// @cdktf/provider-yandex/lib/cdn-origin-group.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CdnOriginGroup = exports.CdnOriginGroupTimeoutsOutputReference = exports.cdnOriginGroupTimeoutsToTerraform = exports.CdnOriginGroupOriginList = exports.CdnOriginGroupOriginOutputReference = exports.cdnOriginGroupOriginToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function cdnOriginGroupOriginToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        backup: cdktf.booleanToTerraform(struct.backup),
        enabled: cdktf.booleanToTerraform(struct.enabled),
        source: cdktf.stringToTerraform(struct.source),
    };
}
exports.cdnOriginGroupOriginToTerraform = cdnOriginGroupOriginToTerraform;
class CdnOriginGroupOriginOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._backup !== undefined) {
            hasAnyValues = true;
            internalValueResult.backup = this._backup;
        }
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._source !== undefined) {
            hasAnyValues = true;
            internalValueResult.source = this._source;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._backup = undefined;
            this._enabled = undefined;
            this._source = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._backup = value.backup;
            this._enabled = value.enabled;
            this._source = value.source;
        }
    }
    get backup() {
        return this.getBooleanAttribute('backup');
    }
    set backup(value) {
        this._backup = value;
    }
    resetBackup() {
        this._backup = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get backupInput() {
        return this._backup;
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    // origin_group_id - computed: true, optional: false, required: false
    get originGroupId() {
        return this.getNumberAttribute('origin_group_id');
    }
    get source() {
        return this.getStringAttribute('source');
    }
    set source(value) {
        this._source = value;
    }
    // Temporarily expose input value. Use with caution.
    get sourceInput() {
        return this._source;
    }
}
exports.CdnOriginGroupOriginOutputReference = CdnOriginGroupOriginOutputReference;
_a = JSII_RTTI_SYMBOL_1;
CdnOriginGroupOriginOutputReference[_a] = { fqn: "@cdktf/provider-yandex.CdnOriginGroupOriginOutputReference", version: "2.0.32" };
class CdnOriginGroupOriginList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new CdnOriginGroupOriginOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.CdnOriginGroupOriginList = CdnOriginGroupOriginList;
_b = JSII_RTTI_SYMBOL_1;
CdnOriginGroupOriginList[_b] = { fqn: "@cdktf/provider-yandex.CdnOriginGroupOriginList", version: "2.0.32" };
function cdnOriginGroupTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.cdnOriginGroupTimeoutsToTerraform = cdnOriginGroupTimeoutsToTerraform;
class CdnOriginGroupTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.CdnOriginGroupTimeoutsOutputReference = CdnOriginGroupTimeoutsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
CdnOriginGroupTimeoutsOutputReference[_c] = { fqn: "@cdktf/provider-yandex.CdnOriginGroupTimeoutsOutputReference", version: "2.0.32" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/r/cdn_origin_group yandex_cdn_origin_group}
*/
class CdnOriginGroup extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/r/cdn_origin_group yandex_cdn_origin_group} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options CdnOriginGroupConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'yandex_cdn_origin_group',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // origin - computed: false, optional: false, required: true
        this._origin = new CdnOriginGroupOriginList(this, "origin", true);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new CdnOriginGroupTimeoutsOutputReference(this, "timeouts");
        this._folderId = config.folderId;
        this._id = config.id;
        this._name = config.name;
        this._useNext = config.useNext;
        this._origin.internalValue = config.origin;
        this._timeouts.internalValue = config.timeouts;
    }
    get folderId() {
        return this.getStringAttribute('folder_id');
    }
    set folderId(value) {
        this._folderId = value;
    }
    resetFolderId() {
        this._folderId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get folderIdInput() {
        return this._folderId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get useNext() {
        return this.getBooleanAttribute('use_next');
    }
    set useNext(value) {
        this._useNext = value;
    }
    resetUseNext() {
        this._useNext = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get useNextInput() {
        return this._useNext;
    }
    get origin() {
        return this._origin;
    }
    putOrigin(value) {
        this._origin.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get originInput() {
        return this._origin.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            folder_id: cdktf.stringToTerraform(this._folderId),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            use_next: cdktf.booleanToTerraform(this._useNext),
            origin: cdktf.listMapper(cdnOriginGroupOriginToTerraform, true)(this._origin.internalValue),
            timeouts: cdnOriginGroupTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.CdnOriginGroup = CdnOriginGroup;
_d = JSII_RTTI_SYMBOL_1;
CdnOriginGroup[_d] = { fqn: "@cdktf/provider-yandex.CdnOriginGroup", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
CdnOriginGroup.tfResourceType = "yandex_cdn_origin_group";
