// @cdktf/provider-yandex/lib/compute-instance.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ComputeInstance = exports.ComputeInstanceTimeoutsOutputReference = exports.computeInstanceTimeoutsToTerraform = exports.ComputeInstanceSecondaryDiskList = exports.ComputeInstanceSecondaryDiskOutputReference = exports.computeInstanceSecondaryDiskToTerraform = exports.ComputeInstanceSchedulingPolicyOutputReference = exports.computeInstanceSchedulingPolicyToTerraform = exports.ComputeInstanceResourcesOutputReference = exports.computeInstanceResourcesToTerraform = exports.ComputeInstancePlacementPolicyOutputReference = exports.computeInstancePlacementPolicyToTerraform = exports.ComputeInstancePlacementPolicyHostAffinityRulesList = exports.ComputeInstancePlacementPolicyHostAffinityRulesOutputReference = exports.computeInstancePlacementPolicyHostAffinityRulesToTerraform = exports.ComputeInstanceNetworkInterfaceList = exports.ComputeInstanceNetworkInterfaceOutputReference = exports.computeInstanceNetworkInterfaceToTerraform = exports.ComputeInstanceNetworkInterfaceNatDnsRecordList = exports.ComputeInstanceNetworkInterfaceNatDnsRecordOutputReference = exports.computeInstanceNetworkInterfaceNatDnsRecordToTerraform = exports.ComputeInstanceNetworkInterfaceIpv6DnsRecordList = exports.ComputeInstanceNetworkInterfaceIpv6DnsRecordOutputReference = exports.computeInstanceNetworkInterfaceIpv6DnsRecordToTerraform = exports.ComputeInstanceNetworkInterfaceDnsRecordList = exports.ComputeInstanceNetworkInterfaceDnsRecordOutputReference = exports.computeInstanceNetworkInterfaceDnsRecordToTerraform = exports.ComputeInstanceLocalDiskList = exports.ComputeInstanceLocalDiskOutputReference = exports.computeInstanceLocalDiskToTerraform = exports.ComputeInstanceBootDiskOutputReference = exports.computeInstanceBootDiskToTerraform = exports.ComputeInstanceBootDiskInitializeParamsOutputReference = exports.computeInstanceBootDiskInitializeParamsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function computeInstanceBootDiskInitializeParamsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        block_size: cdktf.numberToTerraform(struct.blockSize),
        description: cdktf.stringToTerraform(struct.description),
        image_id: cdktf.stringToTerraform(struct.imageId),
        name: cdktf.stringToTerraform(struct.name),
        size: cdktf.numberToTerraform(struct.size),
        snapshot_id: cdktf.stringToTerraform(struct.snapshotId),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.computeInstanceBootDiskInitializeParamsToTerraform = computeInstanceBootDiskInitializeParamsToTerraform;
class ComputeInstanceBootDiskInitializeParamsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._blockSize !== undefined) {
            hasAnyValues = true;
            internalValueResult.blockSize = this._blockSize;
        }
        if (this._description !== undefined) {
            hasAnyValues = true;
            internalValueResult.description = this._description;
        }
        if (this._imageId !== undefined) {
            hasAnyValues = true;
            internalValueResult.imageId = this._imageId;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._size !== undefined) {
            hasAnyValues = true;
            internalValueResult.size = this._size;
        }
        if (this._snapshotId !== undefined) {
            hasAnyValues = true;
            internalValueResult.snapshotId = this._snapshotId;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._blockSize = undefined;
            this._description = undefined;
            this._imageId = undefined;
            this._name = undefined;
            this._size = undefined;
            this._snapshotId = undefined;
            this._type = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._blockSize = value.blockSize;
            this._description = value.description;
            this._imageId = value.imageId;
            this._name = value.name;
            this._size = value.size;
            this._snapshotId = value.snapshotId;
            this._type = value.type;
        }
    }
    get blockSize() {
        return this.getNumberAttribute('block_size');
    }
    set blockSize(value) {
        this._blockSize = value;
    }
    resetBlockSize() {
        this._blockSize = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get blockSizeInput() {
        return this._blockSize;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get imageId() {
        return this.getStringAttribute('image_id');
    }
    set imageId(value) {
        this._imageId = value;
    }
    resetImageId() {
        this._imageId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get imageIdInput() {
        return this._imageId;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get size() {
        return this.getNumberAttribute('size');
    }
    set size(value) {
        this._size = value;
    }
    resetSize() {
        this._size = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sizeInput() {
        return this._size;
    }
    get snapshotId() {
        return this.getStringAttribute('snapshot_id');
    }
    set snapshotId(value) {
        this._snapshotId = value;
    }
    resetSnapshotId() {
        this._snapshotId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get snapshotIdInput() {
        return this._snapshotId;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.ComputeInstanceBootDiskInitializeParamsOutputReference = ComputeInstanceBootDiskInitializeParamsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
ComputeInstanceBootDiskInitializeParamsOutputReference[_a] = { fqn: "@cdktf/provider-yandex.ComputeInstanceBootDiskInitializeParamsOutputReference", version: "2.0.32" };
function computeInstanceBootDiskToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        auto_delete: cdktf.booleanToTerraform(struct.autoDelete),
        device_name: cdktf.stringToTerraform(struct.deviceName),
        disk_id: cdktf.stringToTerraform(struct.diskId),
        mode: cdktf.stringToTerraform(struct.mode),
        initialize_params: computeInstanceBootDiskInitializeParamsToTerraform(struct.initializeParams),
    };
}
exports.computeInstanceBootDiskToTerraform = computeInstanceBootDiskToTerraform;
class ComputeInstanceBootDiskOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // initialize_params - computed: false, optional: true, required: false
        this._initializeParams = new ComputeInstanceBootDiskInitializeParamsOutputReference(this, "initialize_params");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._autoDelete !== undefined) {
            hasAnyValues = true;
            internalValueResult.autoDelete = this._autoDelete;
        }
        if (this._deviceName !== undefined) {
            hasAnyValues = true;
            internalValueResult.deviceName = this._deviceName;
        }
        if (this._diskId !== undefined) {
            hasAnyValues = true;
            internalValueResult.diskId = this._diskId;
        }
        if (this._mode !== undefined) {
            hasAnyValues = true;
            internalValueResult.mode = this._mode;
        }
        if (this._initializeParams?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.initializeParams = this._initializeParams?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._autoDelete = undefined;
            this._deviceName = undefined;
            this._diskId = undefined;
            this._mode = undefined;
            this._initializeParams.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._autoDelete = value.autoDelete;
            this._deviceName = value.deviceName;
            this._diskId = value.diskId;
            this._mode = value.mode;
            this._initializeParams.internalValue = value.initializeParams;
        }
    }
    get autoDelete() {
        return this.getBooleanAttribute('auto_delete');
    }
    set autoDelete(value) {
        this._autoDelete = value;
    }
    resetAutoDelete() {
        this._autoDelete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get autoDeleteInput() {
        return this._autoDelete;
    }
    get deviceName() {
        return this.getStringAttribute('device_name');
    }
    set deviceName(value) {
        this._deviceName = value;
    }
    resetDeviceName() {
        this._deviceName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deviceNameInput() {
        return this._deviceName;
    }
    get diskId() {
        return this.getStringAttribute('disk_id');
    }
    set diskId(value) {
        this._diskId = value;
    }
    resetDiskId() {
        this._diskId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get diskIdInput() {
        return this._diskId;
    }
    get mode() {
        return this.getStringAttribute('mode');
    }
    set mode(value) {
        this._mode = value;
    }
    resetMode() {
        this._mode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get modeInput() {
        return this._mode;
    }
    get initializeParams() {
        return this._initializeParams;
    }
    putInitializeParams(value) {
        this._initializeParams.internalValue = value;
    }
    resetInitializeParams() {
        this._initializeParams.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get initializeParamsInput() {
        return this._initializeParams.internalValue;
    }
}
exports.ComputeInstanceBootDiskOutputReference = ComputeInstanceBootDiskOutputReference;
_b = JSII_RTTI_SYMBOL_1;
ComputeInstanceBootDiskOutputReference[_b] = { fqn: "@cdktf/provider-yandex.ComputeInstanceBootDiskOutputReference", version: "2.0.32" };
function computeInstanceLocalDiskToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        size_bytes: cdktf.numberToTerraform(struct.sizeBytes),
    };
}
exports.computeInstanceLocalDiskToTerraform = computeInstanceLocalDiskToTerraform;
class ComputeInstanceLocalDiskOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._sizeBytes !== undefined) {
            hasAnyValues = true;
            internalValueResult.sizeBytes = this._sizeBytes;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._sizeBytes = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._sizeBytes = value.sizeBytes;
        }
    }
    // device_name - computed: true, optional: false, required: false
    get deviceName() {
        return this.getStringAttribute('device_name');
    }
    get sizeBytes() {
        return this.getNumberAttribute('size_bytes');
    }
    set sizeBytes(value) {
        this._sizeBytes = value;
    }
    // Temporarily expose input value. Use with caution.
    get sizeBytesInput() {
        return this._sizeBytes;
    }
}
exports.ComputeInstanceLocalDiskOutputReference = ComputeInstanceLocalDiskOutputReference;
_c = JSII_RTTI_SYMBOL_1;
ComputeInstanceLocalDiskOutputReference[_c] = { fqn: "@cdktf/provider-yandex.ComputeInstanceLocalDiskOutputReference", version: "2.0.32" };
class ComputeInstanceLocalDiskList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeInstanceLocalDiskOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeInstanceLocalDiskList = ComputeInstanceLocalDiskList;
_d = JSII_RTTI_SYMBOL_1;
ComputeInstanceLocalDiskList[_d] = { fqn: "@cdktf/provider-yandex.ComputeInstanceLocalDiskList", version: "2.0.32" };
function computeInstanceNetworkInterfaceDnsRecordToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        dns_zone_id: cdktf.stringToTerraform(struct.dnsZoneId),
        fqdn: cdktf.stringToTerraform(struct.fqdn),
        ptr: cdktf.booleanToTerraform(struct.ptr),
        ttl: cdktf.numberToTerraform(struct.ttl),
    };
}
exports.computeInstanceNetworkInterfaceDnsRecordToTerraform = computeInstanceNetworkInterfaceDnsRecordToTerraform;
class ComputeInstanceNetworkInterfaceDnsRecordOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._dnsZoneId !== undefined) {
            hasAnyValues = true;
            internalValueResult.dnsZoneId = this._dnsZoneId;
        }
        if (this._fqdn !== undefined) {
            hasAnyValues = true;
            internalValueResult.fqdn = this._fqdn;
        }
        if (this._ptr !== undefined) {
            hasAnyValues = true;
            internalValueResult.ptr = this._ptr;
        }
        if (this._ttl !== undefined) {
            hasAnyValues = true;
            internalValueResult.ttl = this._ttl;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._dnsZoneId = undefined;
            this._fqdn = undefined;
            this._ptr = undefined;
            this._ttl = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._dnsZoneId = value.dnsZoneId;
            this._fqdn = value.fqdn;
            this._ptr = value.ptr;
            this._ttl = value.ttl;
        }
    }
    get dnsZoneId() {
        return this.getStringAttribute('dns_zone_id');
    }
    set dnsZoneId(value) {
        this._dnsZoneId = value;
    }
    resetDnsZoneId() {
        this._dnsZoneId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dnsZoneIdInput() {
        return this._dnsZoneId;
    }
    get fqdn() {
        return this.getStringAttribute('fqdn');
    }
    set fqdn(value) {
        this._fqdn = value;
    }
    // Temporarily expose input value. Use with caution.
    get fqdnInput() {
        return this._fqdn;
    }
    get ptr() {
        return this.getBooleanAttribute('ptr');
    }
    set ptr(value) {
        this._ptr = value;
    }
    resetPtr() {
        this._ptr = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ptrInput() {
        return this._ptr;
    }
    get ttl() {
        return this.getNumberAttribute('ttl');
    }
    set ttl(value) {
        this._ttl = value;
    }
    resetTtl() {
        this._ttl = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ttlInput() {
        return this._ttl;
    }
}
exports.ComputeInstanceNetworkInterfaceDnsRecordOutputReference = ComputeInstanceNetworkInterfaceDnsRecordOutputReference;
_e = JSII_RTTI_SYMBOL_1;
ComputeInstanceNetworkInterfaceDnsRecordOutputReference[_e] = { fqn: "@cdktf/provider-yandex.ComputeInstanceNetworkInterfaceDnsRecordOutputReference", version: "2.0.32" };
class ComputeInstanceNetworkInterfaceDnsRecordList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeInstanceNetworkInterfaceDnsRecordOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeInstanceNetworkInterfaceDnsRecordList = ComputeInstanceNetworkInterfaceDnsRecordList;
_f = JSII_RTTI_SYMBOL_1;
ComputeInstanceNetworkInterfaceDnsRecordList[_f] = { fqn: "@cdktf/provider-yandex.ComputeInstanceNetworkInterfaceDnsRecordList", version: "2.0.32" };
function computeInstanceNetworkInterfaceIpv6DnsRecordToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        dns_zone_id: cdktf.stringToTerraform(struct.dnsZoneId),
        fqdn: cdktf.stringToTerraform(struct.fqdn),
        ptr: cdktf.booleanToTerraform(struct.ptr),
        ttl: cdktf.numberToTerraform(struct.ttl),
    };
}
exports.computeInstanceNetworkInterfaceIpv6DnsRecordToTerraform = computeInstanceNetworkInterfaceIpv6DnsRecordToTerraform;
class ComputeInstanceNetworkInterfaceIpv6DnsRecordOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._dnsZoneId !== undefined) {
            hasAnyValues = true;
            internalValueResult.dnsZoneId = this._dnsZoneId;
        }
        if (this._fqdn !== undefined) {
            hasAnyValues = true;
            internalValueResult.fqdn = this._fqdn;
        }
        if (this._ptr !== undefined) {
            hasAnyValues = true;
            internalValueResult.ptr = this._ptr;
        }
        if (this._ttl !== undefined) {
            hasAnyValues = true;
            internalValueResult.ttl = this._ttl;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._dnsZoneId = undefined;
            this._fqdn = undefined;
            this._ptr = undefined;
            this._ttl = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._dnsZoneId = value.dnsZoneId;
            this._fqdn = value.fqdn;
            this._ptr = value.ptr;
            this._ttl = value.ttl;
        }
    }
    get dnsZoneId() {
        return this.getStringAttribute('dns_zone_id');
    }
    set dnsZoneId(value) {
        this._dnsZoneId = value;
    }
    resetDnsZoneId() {
        this._dnsZoneId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dnsZoneIdInput() {
        return this._dnsZoneId;
    }
    get fqdn() {
        return this.getStringAttribute('fqdn');
    }
    set fqdn(value) {
        this._fqdn = value;
    }
    // Temporarily expose input value. Use with caution.
    get fqdnInput() {
        return this._fqdn;
    }
    get ptr() {
        return this.getBooleanAttribute('ptr');
    }
    set ptr(value) {
        this._ptr = value;
    }
    resetPtr() {
        this._ptr = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ptrInput() {
        return this._ptr;
    }
    get ttl() {
        return this.getNumberAttribute('ttl');
    }
    set ttl(value) {
        this._ttl = value;
    }
    resetTtl() {
        this._ttl = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ttlInput() {
        return this._ttl;
    }
}
exports.ComputeInstanceNetworkInterfaceIpv6DnsRecordOutputReference = ComputeInstanceNetworkInterfaceIpv6DnsRecordOutputReference;
_g = JSII_RTTI_SYMBOL_1;
ComputeInstanceNetworkInterfaceIpv6DnsRecordOutputReference[_g] = { fqn: "@cdktf/provider-yandex.ComputeInstanceNetworkInterfaceIpv6DnsRecordOutputReference", version: "2.0.32" };
class ComputeInstanceNetworkInterfaceIpv6DnsRecordList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeInstanceNetworkInterfaceIpv6DnsRecordOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeInstanceNetworkInterfaceIpv6DnsRecordList = ComputeInstanceNetworkInterfaceIpv6DnsRecordList;
_h = JSII_RTTI_SYMBOL_1;
ComputeInstanceNetworkInterfaceIpv6DnsRecordList[_h] = { fqn: "@cdktf/provider-yandex.ComputeInstanceNetworkInterfaceIpv6DnsRecordList", version: "2.0.32" };
function computeInstanceNetworkInterfaceNatDnsRecordToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        dns_zone_id: cdktf.stringToTerraform(struct.dnsZoneId),
        fqdn: cdktf.stringToTerraform(struct.fqdn),
        ptr: cdktf.booleanToTerraform(struct.ptr),
        ttl: cdktf.numberToTerraform(struct.ttl),
    };
}
exports.computeInstanceNetworkInterfaceNatDnsRecordToTerraform = computeInstanceNetworkInterfaceNatDnsRecordToTerraform;
class ComputeInstanceNetworkInterfaceNatDnsRecordOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._dnsZoneId !== undefined) {
            hasAnyValues = true;
            internalValueResult.dnsZoneId = this._dnsZoneId;
        }
        if (this._fqdn !== undefined) {
            hasAnyValues = true;
            internalValueResult.fqdn = this._fqdn;
        }
        if (this._ptr !== undefined) {
            hasAnyValues = true;
            internalValueResult.ptr = this._ptr;
        }
        if (this._ttl !== undefined) {
            hasAnyValues = true;
            internalValueResult.ttl = this._ttl;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._dnsZoneId = undefined;
            this._fqdn = undefined;
            this._ptr = undefined;
            this._ttl = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._dnsZoneId = value.dnsZoneId;
            this._fqdn = value.fqdn;
            this._ptr = value.ptr;
            this._ttl = value.ttl;
        }
    }
    get dnsZoneId() {
        return this.getStringAttribute('dns_zone_id');
    }
    set dnsZoneId(value) {
        this._dnsZoneId = value;
    }
    resetDnsZoneId() {
        this._dnsZoneId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dnsZoneIdInput() {
        return this._dnsZoneId;
    }
    get fqdn() {
        return this.getStringAttribute('fqdn');
    }
    set fqdn(value) {
        this._fqdn = value;
    }
    // Temporarily expose input value. Use with caution.
    get fqdnInput() {
        return this._fqdn;
    }
    get ptr() {
        return this.getBooleanAttribute('ptr');
    }
    set ptr(value) {
        this._ptr = value;
    }
    resetPtr() {
        this._ptr = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ptrInput() {
        return this._ptr;
    }
    get ttl() {
        return this.getNumberAttribute('ttl');
    }
    set ttl(value) {
        this._ttl = value;
    }
    resetTtl() {
        this._ttl = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ttlInput() {
        return this._ttl;
    }
}
exports.ComputeInstanceNetworkInterfaceNatDnsRecordOutputReference = ComputeInstanceNetworkInterfaceNatDnsRecordOutputReference;
_j = JSII_RTTI_SYMBOL_1;
ComputeInstanceNetworkInterfaceNatDnsRecordOutputReference[_j] = { fqn: "@cdktf/provider-yandex.ComputeInstanceNetworkInterfaceNatDnsRecordOutputReference", version: "2.0.32" };
class ComputeInstanceNetworkInterfaceNatDnsRecordList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeInstanceNetworkInterfaceNatDnsRecordOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeInstanceNetworkInterfaceNatDnsRecordList = ComputeInstanceNetworkInterfaceNatDnsRecordList;
_k = JSII_RTTI_SYMBOL_1;
ComputeInstanceNetworkInterfaceNatDnsRecordList[_k] = { fqn: "@cdktf/provider-yandex.ComputeInstanceNetworkInterfaceNatDnsRecordList", version: "2.0.32" };
function computeInstanceNetworkInterfaceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        ip_address: cdktf.stringToTerraform(struct.ipAddress),
        ipv4: cdktf.booleanToTerraform(struct.ipv4),
        ipv6: cdktf.booleanToTerraform(struct.ipv6),
        ipv6_address: cdktf.stringToTerraform(struct.ipv6Address),
        nat: cdktf.booleanToTerraform(struct.nat),
        nat_ip_address: cdktf.stringToTerraform(struct.natIpAddress),
        security_group_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.securityGroupIds),
        subnet_id: cdktf.stringToTerraform(struct.subnetId),
        dns_record: cdktf.listMapper(computeInstanceNetworkInterfaceDnsRecordToTerraform, true)(struct.dnsRecord),
        ipv6_dns_record: cdktf.listMapper(computeInstanceNetworkInterfaceIpv6DnsRecordToTerraform, true)(struct.ipv6DnsRecord),
        nat_dns_record: cdktf.listMapper(computeInstanceNetworkInterfaceNatDnsRecordToTerraform, true)(struct.natDnsRecord),
    };
}
exports.computeInstanceNetworkInterfaceToTerraform = computeInstanceNetworkInterfaceToTerraform;
class ComputeInstanceNetworkInterfaceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // dns_record - computed: false, optional: true, required: false
        this._dnsRecord = new ComputeInstanceNetworkInterfaceDnsRecordList(this, "dns_record", false);
        // ipv6_dns_record - computed: false, optional: true, required: false
        this._ipv6DnsRecord = new ComputeInstanceNetworkInterfaceIpv6DnsRecordList(this, "ipv6_dns_record", false);
        // nat_dns_record - computed: false, optional: true, required: false
        this._natDnsRecord = new ComputeInstanceNetworkInterfaceNatDnsRecordList(this, "nat_dns_record", false);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._ipAddress !== undefined) {
            hasAnyValues = true;
            internalValueResult.ipAddress = this._ipAddress;
        }
        if (this._ipv4 !== undefined) {
            hasAnyValues = true;
            internalValueResult.ipv4 = this._ipv4;
        }
        if (this._ipv6 !== undefined) {
            hasAnyValues = true;
            internalValueResult.ipv6 = this._ipv6;
        }
        if (this._ipv6Address !== undefined) {
            hasAnyValues = true;
            internalValueResult.ipv6Address = this._ipv6Address;
        }
        if (this._nat !== undefined) {
            hasAnyValues = true;
            internalValueResult.nat = this._nat;
        }
        if (this._natIpAddress !== undefined) {
            hasAnyValues = true;
            internalValueResult.natIpAddress = this._natIpAddress;
        }
        if (this._securityGroupIds !== undefined) {
            hasAnyValues = true;
            internalValueResult.securityGroupIds = this._securityGroupIds;
        }
        if (this._subnetId !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnetId = this._subnetId;
        }
        if (this._dnsRecord?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.dnsRecord = this._dnsRecord?.internalValue;
        }
        if (this._ipv6DnsRecord?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.ipv6DnsRecord = this._ipv6DnsRecord?.internalValue;
        }
        if (this._natDnsRecord?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.natDnsRecord = this._natDnsRecord?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._ipAddress = undefined;
            this._ipv4 = undefined;
            this._ipv6 = undefined;
            this._ipv6Address = undefined;
            this._nat = undefined;
            this._natIpAddress = undefined;
            this._securityGroupIds = undefined;
            this._subnetId = undefined;
            this._dnsRecord.internalValue = undefined;
            this._ipv6DnsRecord.internalValue = undefined;
            this._natDnsRecord.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._ipAddress = value.ipAddress;
            this._ipv4 = value.ipv4;
            this._ipv6 = value.ipv6;
            this._ipv6Address = value.ipv6Address;
            this._nat = value.nat;
            this._natIpAddress = value.natIpAddress;
            this._securityGroupIds = value.securityGroupIds;
            this._subnetId = value.subnetId;
            this._dnsRecord.internalValue = value.dnsRecord;
            this._ipv6DnsRecord.internalValue = value.ipv6DnsRecord;
            this._natDnsRecord.internalValue = value.natDnsRecord;
        }
    }
    // index - computed: true, optional: false, required: false
    get index() {
        return this.getNumberAttribute('index');
    }
    get ipAddress() {
        return this.getStringAttribute('ip_address');
    }
    set ipAddress(value) {
        this._ipAddress = value;
    }
    resetIpAddress() {
        this._ipAddress = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ipAddressInput() {
        return this._ipAddress;
    }
    get ipv4() {
        return this.getBooleanAttribute('ipv4');
    }
    set ipv4(value) {
        this._ipv4 = value;
    }
    resetIpv4() {
        this._ipv4 = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ipv4Input() {
        return this._ipv4;
    }
    get ipv6() {
        return this.getBooleanAttribute('ipv6');
    }
    set ipv6(value) {
        this._ipv6 = value;
    }
    resetIpv6() {
        this._ipv6 = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ipv6Input() {
        return this._ipv6;
    }
    get ipv6Address() {
        return this.getStringAttribute('ipv6_address');
    }
    set ipv6Address(value) {
        this._ipv6Address = value;
    }
    resetIpv6Address() {
        this._ipv6Address = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ipv6AddressInput() {
        return this._ipv6Address;
    }
    // mac_address - computed: true, optional: false, required: false
    get macAddress() {
        return this.getStringAttribute('mac_address');
    }
    get nat() {
        return this.getBooleanAttribute('nat');
    }
    set nat(value) {
        this._nat = value;
    }
    resetNat() {
        this._nat = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get natInput() {
        return this._nat;
    }
    get natIpAddress() {
        return this.getStringAttribute('nat_ip_address');
    }
    set natIpAddress(value) {
        this._natIpAddress = value;
    }
    resetNatIpAddress() {
        this._natIpAddress = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get natIpAddressInput() {
        return this._natIpAddress;
    }
    // nat_ip_version - computed: true, optional: false, required: false
    get natIpVersion() {
        return this.getStringAttribute('nat_ip_version');
    }
    get securityGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('security_group_ids'));
    }
    set securityGroupIds(value) {
        this._securityGroupIds = value;
    }
    resetSecurityGroupIds() {
        this._securityGroupIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupIdsInput() {
        return this._securityGroupIds;
    }
    get subnetId() {
        return this.getStringAttribute('subnet_id');
    }
    set subnetId(value) {
        this._subnetId = value;
    }
    // Temporarily expose input value. Use with caution.
    get subnetIdInput() {
        return this._subnetId;
    }
    get dnsRecord() {
        return this._dnsRecord;
    }
    putDnsRecord(value) {
        this._dnsRecord.internalValue = value;
    }
    resetDnsRecord() {
        this._dnsRecord.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dnsRecordInput() {
        return this._dnsRecord.internalValue;
    }
    get ipv6DnsRecord() {
        return this._ipv6DnsRecord;
    }
    putIpv6DnsRecord(value) {
        this._ipv6DnsRecord.internalValue = value;
    }
    resetIpv6DnsRecord() {
        this._ipv6DnsRecord.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ipv6DnsRecordInput() {
        return this._ipv6DnsRecord.internalValue;
    }
    get natDnsRecord() {
        return this._natDnsRecord;
    }
    putNatDnsRecord(value) {
        this._natDnsRecord.internalValue = value;
    }
    resetNatDnsRecord() {
        this._natDnsRecord.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get natDnsRecordInput() {
        return this._natDnsRecord.internalValue;
    }
}
exports.ComputeInstanceNetworkInterfaceOutputReference = ComputeInstanceNetworkInterfaceOutputReference;
_l = JSII_RTTI_SYMBOL_1;
ComputeInstanceNetworkInterfaceOutputReference[_l] = { fqn: "@cdktf/provider-yandex.ComputeInstanceNetworkInterfaceOutputReference", version: "2.0.32" };
class ComputeInstanceNetworkInterfaceList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeInstanceNetworkInterfaceOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeInstanceNetworkInterfaceList = ComputeInstanceNetworkInterfaceList;
_m = JSII_RTTI_SYMBOL_1;
ComputeInstanceNetworkInterfaceList[_m] = { fqn: "@cdktf/provider-yandex.ComputeInstanceNetworkInterfaceList", version: "2.0.32" };
function computeInstancePlacementPolicyHostAffinityRulesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key: cdktf.stringToTerraform(struct.key),
        op: cdktf.stringToTerraform(struct.op),
        values: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.values),
    };
}
exports.computeInstancePlacementPolicyHostAffinityRulesToTerraform = computeInstancePlacementPolicyHostAffinityRulesToTerraform;
class ComputeInstancePlacementPolicyHostAffinityRulesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._key !== undefined) {
            hasAnyValues = true;
            internalValueResult.key = this._key;
        }
        if (this._op !== undefined) {
            hasAnyValues = true;
            internalValueResult.op = this._op;
        }
        if (this._values !== undefined) {
            hasAnyValues = true;
            internalValueResult.values = this._values;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._key = undefined;
            this._op = undefined;
            this._values = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._key = value.key;
            this._op = value.op;
            this._values = value.values;
        }
    }
    get key() {
        return this.getStringAttribute('key');
    }
    set key(value) {
        this._key = value;
    }
    resetKey() {
        this._key = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keyInput() {
        return this._key;
    }
    get op() {
        return this.getStringAttribute('op');
    }
    set op(value) {
        this._op = value;
    }
    resetOp() {
        this._op = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get opInput() {
        return this._op;
    }
    get values() {
        return this.getListAttribute('values');
    }
    set values(value) {
        this._values = value;
    }
    resetValues() {
        this._values = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get valuesInput() {
        return this._values;
    }
}
exports.ComputeInstancePlacementPolicyHostAffinityRulesOutputReference = ComputeInstancePlacementPolicyHostAffinityRulesOutputReference;
_o = JSII_RTTI_SYMBOL_1;
ComputeInstancePlacementPolicyHostAffinityRulesOutputReference[_o] = { fqn: "@cdktf/provider-yandex.ComputeInstancePlacementPolicyHostAffinityRulesOutputReference", version: "2.0.32" };
class ComputeInstancePlacementPolicyHostAffinityRulesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeInstancePlacementPolicyHostAffinityRulesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeInstancePlacementPolicyHostAffinityRulesList = ComputeInstancePlacementPolicyHostAffinityRulesList;
_p = JSII_RTTI_SYMBOL_1;
ComputeInstancePlacementPolicyHostAffinityRulesList[_p] = { fqn: "@cdktf/provider-yandex.ComputeInstancePlacementPolicyHostAffinityRulesList", version: "2.0.32" };
function computeInstancePlacementPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        host_affinity_rules: cdktf.listMapper(computeInstancePlacementPolicyHostAffinityRulesToTerraform, false)(struct.hostAffinityRules),
        placement_group_id: cdktf.stringToTerraform(struct.placementGroupId),
    };
}
exports.computeInstancePlacementPolicyToTerraform = computeInstancePlacementPolicyToTerraform;
class ComputeInstancePlacementPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // host_affinity_rules - computed: true, optional: true, required: false
        this._hostAffinityRules = new ComputeInstancePlacementPolicyHostAffinityRulesList(this, "host_affinity_rules", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._hostAffinityRules?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.hostAffinityRules = this._hostAffinityRules?.internalValue;
        }
        if (this._placementGroupId !== undefined) {
            hasAnyValues = true;
            internalValueResult.placementGroupId = this._placementGroupId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._hostAffinityRules.internalValue = undefined;
            this._placementGroupId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._hostAffinityRules.internalValue = value.hostAffinityRules;
            this._placementGroupId = value.placementGroupId;
        }
    }
    get hostAffinityRules() {
        return this._hostAffinityRules;
    }
    putHostAffinityRules(value) {
        this._hostAffinityRules.internalValue = value;
    }
    resetHostAffinityRules() {
        this._hostAffinityRules.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get hostAffinityRulesInput() {
        return this._hostAffinityRules.internalValue;
    }
    get placementGroupId() {
        return this.getStringAttribute('placement_group_id');
    }
    set placementGroupId(value) {
        this._placementGroupId = value;
    }
    resetPlacementGroupId() {
        this._placementGroupId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get placementGroupIdInput() {
        return this._placementGroupId;
    }
}
exports.ComputeInstancePlacementPolicyOutputReference = ComputeInstancePlacementPolicyOutputReference;
_q = JSII_RTTI_SYMBOL_1;
ComputeInstancePlacementPolicyOutputReference[_q] = { fqn: "@cdktf/provider-yandex.ComputeInstancePlacementPolicyOutputReference", version: "2.0.32" };
function computeInstanceResourcesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        core_fraction: cdktf.numberToTerraform(struct.coreFraction),
        cores: cdktf.numberToTerraform(struct.cores),
        gpus: cdktf.numberToTerraform(struct.gpus),
        memory: cdktf.numberToTerraform(struct.memory),
    };
}
exports.computeInstanceResourcesToTerraform = computeInstanceResourcesToTerraform;
class ComputeInstanceResourcesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._coreFraction !== undefined) {
            hasAnyValues = true;
            internalValueResult.coreFraction = this._coreFraction;
        }
        if (this._cores !== undefined) {
            hasAnyValues = true;
            internalValueResult.cores = this._cores;
        }
        if (this._gpus !== undefined) {
            hasAnyValues = true;
            internalValueResult.gpus = this._gpus;
        }
        if (this._memory !== undefined) {
            hasAnyValues = true;
            internalValueResult.memory = this._memory;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._coreFraction = undefined;
            this._cores = undefined;
            this._gpus = undefined;
            this._memory = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._coreFraction = value.coreFraction;
            this._cores = value.cores;
            this._gpus = value.gpus;
            this._memory = value.memory;
        }
    }
    get coreFraction() {
        return this.getNumberAttribute('core_fraction');
    }
    set coreFraction(value) {
        this._coreFraction = value;
    }
    resetCoreFraction() {
        this._coreFraction = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get coreFractionInput() {
        return this._coreFraction;
    }
    get cores() {
        return this.getNumberAttribute('cores');
    }
    set cores(value) {
        this._cores = value;
    }
    // Temporarily expose input value. Use with caution.
    get coresInput() {
        return this._cores;
    }
    get gpus() {
        return this.getNumberAttribute('gpus');
    }
    set gpus(value) {
        this._gpus = value;
    }
    resetGpus() {
        this._gpus = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get gpusInput() {
        return this._gpus;
    }
    get memory() {
        return this.getNumberAttribute('memory');
    }
    set memory(value) {
        this._memory = value;
    }
    // Temporarily expose input value. Use with caution.
    get memoryInput() {
        return this._memory;
    }
}
exports.ComputeInstanceResourcesOutputReference = ComputeInstanceResourcesOutputReference;
_r = JSII_RTTI_SYMBOL_1;
ComputeInstanceResourcesOutputReference[_r] = { fqn: "@cdktf/provider-yandex.ComputeInstanceResourcesOutputReference", version: "2.0.32" };
function computeInstanceSchedulingPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        preemptible: cdktf.booleanToTerraform(struct.preemptible),
    };
}
exports.computeInstanceSchedulingPolicyToTerraform = computeInstanceSchedulingPolicyToTerraform;
class ComputeInstanceSchedulingPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._preemptible !== undefined) {
            hasAnyValues = true;
            internalValueResult.preemptible = this._preemptible;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._preemptible = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._preemptible = value.preemptible;
        }
    }
    get preemptible() {
        return this.getBooleanAttribute('preemptible');
    }
    set preemptible(value) {
        this._preemptible = value;
    }
    resetPreemptible() {
        this._preemptible = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get preemptibleInput() {
        return this._preemptible;
    }
}
exports.ComputeInstanceSchedulingPolicyOutputReference = ComputeInstanceSchedulingPolicyOutputReference;
_s = JSII_RTTI_SYMBOL_1;
ComputeInstanceSchedulingPolicyOutputReference[_s] = { fqn: "@cdktf/provider-yandex.ComputeInstanceSchedulingPolicyOutputReference", version: "2.0.32" };
function computeInstanceSecondaryDiskToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        auto_delete: cdktf.booleanToTerraform(struct.autoDelete),
        device_name: cdktf.stringToTerraform(struct.deviceName),
        disk_id: cdktf.stringToTerraform(struct.diskId),
        mode: cdktf.stringToTerraform(struct.mode),
    };
}
exports.computeInstanceSecondaryDiskToTerraform = computeInstanceSecondaryDiskToTerraform;
class ComputeInstanceSecondaryDiskOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._autoDelete !== undefined) {
            hasAnyValues = true;
            internalValueResult.autoDelete = this._autoDelete;
        }
        if (this._deviceName !== undefined) {
            hasAnyValues = true;
            internalValueResult.deviceName = this._deviceName;
        }
        if (this._diskId !== undefined) {
            hasAnyValues = true;
            internalValueResult.diskId = this._diskId;
        }
        if (this._mode !== undefined) {
            hasAnyValues = true;
            internalValueResult.mode = this._mode;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._autoDelete = undefined;
            this._deviceName = undefined;
            this._diskId = undefined;
            this._mode = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._autoDelete = value.autoDelete;
            this._deviceName = value.deviceName;
            this._diskId = value.diskId;
            this._mode = value.mode;
        }
    }
    get autoDelete() {
        return this.getBooleanAttribute('auto_delete');
    }
    set autoDelete(value) {
        this._autoDelete = value;
    }
    resetAutoDelete() {
        this._autoDelete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get autoDeleteInput() {
        return this._autoDelete;
    }
    get deviceName() {
        return this.getStringAttribute('device_name');
    }
    set deviceName(value) {
        this._deviceName = value;
    }
    resetDeviceName() {
        this._deviceName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deviceNameInput() {
        return this._deviceName;
    }
    get diskId() {
        return this.getStringAttribute('disk_id');
    }
    set diskId(value) {
        this._diskId = value;
    }
    // Temporarily expose input value. Use with caution.
    get diskIdInput() {
        return this._diskId;
    }
    get mode() {
        return this.getStringAttribute('mode');
    }
    set mode(value) {
        this._mode = value;
    }
    resetMode() {
        this._mode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get modeInput() {
        return this._mode;
    }
}
exports.ComputeInstanceSecondaryDiskOutputReference = ComputeInstanceSecondaryDiskOutputReference;
_t = JSII_RTTI_SYMBOL_1;
ComputeInstanceSecondaryDiskOutputReference[_t] = { fqn: "@cdktf/provider-yandex.ComputeInstanceSecondaryDiskOutputReference", version: "2.0.32" };
class ComputeInstanceSecondaryDiskList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new ComputeInstanceSecondaryDiskOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.ComputeInstanceSecondaryDiskList = ComputeInstanceSecondaryDiskList;
_u = JSII_RTTI_SYMBOL_1;
ComputeInstanceSecondaryDiskList[_u] = { fqn: "@cdktf/provider-yandex.ComputeInstanceSecondaryDiskList", version: "2.0.32" };
function computeInstanceTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.computeInstanceTimeoutsToTerraform = computeInstanceTimeoutsToTerraform;
class ComputeInstanceTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.ComputeInstanceTimeoutsOutputReference = ComputeInstanceTimeoutsOutputReference;
_v = JSII_RTTI_SYMBOL_1;
ComputeInstanceTimeoutsOutputReference[_v] = { fqn: "@cdktf/provider-yandex.ComputeInstanceTimeoutsOutputReference", version: "2.0.32" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/r/compute_instance yandex_compute_instance}
*/
class ComputeInstance extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/r/compute_instance yandex_compute_instance} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ComputeInstanceConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'yandex_compute_instance',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // boot_disk - computed: false, optional: false, required: true
        this._bootDisk = new ComputeInstanceBootDiskOutputReference(this, "boot_disk");
        // local_disk - computed: false, optional: true, required: false
        this._localDisk = new ComputeInstanceLocalDiskList(this, "local_disk", false);
        // network_interface - computed: false, optional: false, required: true
        this._networkInterface = new ComputeInstanceNetworkInterfaceList(this, "network_interface", false);
        // placement_policy - computed: false, optional: true, required: false
        this._placementPolicy = new ComputeInstancePlacementPolicyOutputReference(this, "placement_policy");
        // resources - computed: false, optional: false, required: true
        this._resources = new ComputeInstanceResourcesOutputReference(this, "resources");
        // scheduling_policy - computed: false, optional: true, required: false
        this._schedulingPolicy = new ComputeInstanceSchedulingPolicyOutputReference(this, "scheduling_policy");
        // secondary_disk - computed: false, optional: true, required: false
        this._secondaryDisk = new ComputeInstanceSecondaryDiskList(this, "secondary_disk", false);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new ComputeInstanceTimeoutsOutputReference(this, "timeouts");
        this._allowRecreate = config.allowRecreate;
        this._allowStoppingForUpdate = config.allowStoppingForUpdate;
        this._description = config.description;
        this._folderId = config.folderId;
        this._hostname = config.hostname;
        this._id = config.id;
        this._labels = config.labels;
        this._metadata = config.metadata;
        this._name = config.name;
        this._networkAccelerationType = config.networkAccelerationType;
        this._platformId = config.platformId;
        this._serviceAccountId = config.serviceAccountId;
        this._zone = config.zone;
        this._bootDisk.internalValue = config.bootDisk;
        this._localDisk.internalValue = config.localDisk;
        this._networkInterface.internalValue = config.networkInterface;
        this._placementPolicy.internalValue = config.placementPolicy;
        this._resources.internalValue = config.resources;
        this._schedulingPolicy.internalValue = config.schedulingPolicy;
        this._secondaryDisk.internalValue = config.secondaryDisk;
        this._timeouts.internalValue = config.timeouts;
    }
    get allowRecreate() {
        return this.getBooleanAttribute('allow_recreate');
    }
    set allowRecreate(value) {
        this._allowRecreate = value;
    }
    resetAllowRecreate() {
        this._allowRecreate = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get allowRecreateInput() {
        return this._allowRecreate;
    }
    get allowStoppingForUpdate() {
        return this.getBooleanAttribute('allow_stopping_for_update');
    }
    set allowStoppingForUpdate(value) {
        this._allowStoppingForUpdate = value;
    }
    resetAllowStoppingForUpdate() {
        this._allowStoppingForUpdate = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get allowStoppingForUpdateInput() {
        return this._allowStoppingForUpdate;
    }
    // created_at - computed: true, optional: false, required: false
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get folderId() {
        return this.getStringAttribute('folder_id');
    }
    set folderId(value) {
        this._folderId = value;
    }
    resetFolderId() {
        this._folderId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get folderIdInput() {
        return this._folderId;
    }
    // fqdn - computed: true, optional: false, required: false
    get fqdn() {
        return this.getStringAttribute('fqdn');
    }
    get hostname() {
        return this.getStringAttribute('hostname');
    }
    set hostname(value) {
        this._hostname = value;
    }
    resetHostname() {
        this._hostname = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get hostnameInput() {
        return this._hostname;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get metadata() {
        return this.getStringMapAttribute('metadata');
    }
    set metadata(value) {
        this._metadata = value;
    }
    resetMetadata() {
        this._metadata = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get metadataInput() {
        return this._metadata;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get networkAccelerationType() {
        return this.getStringAttribute('network_acceleration_type');
    }
    set networkAccelerationType(value) {
        this._networkAccelerationType = value;
    }
    resetNetworkAccelerationType() {
        this._networkAccelerationType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get networkAccelerationTypeInput() {
        return this._networkAccelerationType;
    }
    get platformId() {
        return this.getStringAttribute('platform_id');
    }
    set platformId(value) {
        this._platformId = value;
    }
    resetPlatformId() {
        this._platformId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get platformIdInput() {
        return this._platformId;
    }
    get serviceAccountId() {
        return this.getStringAttribute('service_account_id');
    }
    set serviceAccountId(value) {
        this._serviceAccountId = value;
    }
    resetServiceAccountId() {
        this._serviceAccountId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get serviceAccountIdInput() {
        return this._serviceAccountId;
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    get zone() {
        return this.getStringAttribute('zone');
    }
    set zone(value) {
        this._zone = value;
    }
    resetZone() {
        this._zone = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get zoneInput() {
        return this._zone;
    }
    get bootDisk() {
        return this._bootDisk;
    }
    putBootDisk(value) {
        this._bootDisk.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get bootDiskInput() {
        return this._bootDisk.internalValue;
    }
    get localDisk() {
        return this._localDisk;
    }
    putLocalDisk(value) {
        this._localDisk.internalValue = value;
    }
    resetLocalDisk() {
        this._localDisk.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get localDiskInput() {
        return this._localDisk.internalValue;
    }
    get networkInterface() {
        return this._networkInterface;
    }
    putNetworkInterface(value) {
        this._networkInterface.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get networkInterfaceInput() {
        return this._networkInterface.internalValue;
    }
    get placementPolicy() {
        return this._placementPolicy;
    }
    putPlacementPolicy(value) {
        this._placementPolicy.internalValue = value;
    }
    resetPlacementPolicy() {
        this._placementPolicy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get placementPolicyInput() {
        return this._placementPolicy.internalValue;
    }
    get resources() {
        return this._resources;
    }
    putResources(value) {
        this._resources.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get resourcesInput() {
        return this._resources.internalValue;
    }
    get schedulingPolicy() {
        return this._schedulingPolicy;
    }
    putSchedulingPolicy(value) {
        this._schedulingPolicy.internalValue = value;
    }
    resetSchedulingPolicy() {
        this._schedulingPolicy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get schedulingPolicyInput() {
        return this._schedulingPolicy.internalValue;
    }
    get secondaryDisk() {
        return this._secondaryDisk;
    }
    putSecondaryDisk(value) {
        this._secondaryDisk.internalValue = value;
    }
    resetSecondaryDisk() {
        this._secondaryDisk.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get secondaryDiskInput() {
        return this._secondaryDisk.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            allow_recreate: cdktf.booleanToTerraform(this._allowRecreate),
            allow_stopping_for_update: cdktf.booleanToTerraform(this._allowStoppingForUpdate),
            description: cdktf.stringToTerraform(this._description),
            folder_id: cdktf.stringToTerraform(this._folderId),
            hostname: cdktf.stringToTerraform(this._hostname),
            id: cdktf.stringToTerraform(this._id),
            labels: cdktf.hashMapper(cdktf.stringToTerraform)(this._labels),
            metadata: cdktf.hashMapper(cdktf.stringToTerraform)(this._metadata),
            name: cdktf.stringToTerraform(this._name),
            network_acceleration_type: cdktf.stringToTerraform(this._networkAccelerationType),
            platform_id: cdktf.stringToTerraform(this._platformId),
            service_account_id: cdktf.stringToTerraform(this._serviceAccountId),
            zone: cdktf.stringToTerraform(this._zone),
            boot_disk: computeInstanceBootDiskToTerraform(this._bootDisk.internalValue),
            local_disk: cdktf.listMapper(computeInstanceLocalDiskToTerraform, true)(this._localDisk.internalValue),
            network_interface: cdktf.listMapper(computeInstanceNetworkInterfaceToTerraform, true)(this._networkInterface.internalValue),
            placement_policy: computeInstancePlacementPolicyToTerraform(this._placementPolicy.internalValue),
            resources: computeInstanceResourcesToTerraform(this._resources.internalValue),
            scheduling_policy: computeInstanceSchedulingPolicyToTerraform(this._schedulingPolicy.internalValue),
            secondary_disk: cdktf.listMapper(computeInstanceSecondaryDiskToTerraform, true)(this._secondaryDisk.internalValue),
            timeouts: computeInstanceTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.ComputeInstance = ComputeInstance;
_w = JSII_RTTI_SYMBOL_1;
ComputeInstance[_w] = { fqn: "@cdktf/provider-yandex.ComputeInstance", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
ComputeInstance.tfResourceType = "yandex_compute_instance";
