// @cdktf/provider-yandex/lib/container-registry-iam-binding.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContainerRegistryIamBinding = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/r/container_registry_iam_binding yandex_container_registry_iam_binding}
*/
class ContainerRegistryIamBinding extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/r/container_registry_iam_binding yandex_container_registry_iam_binding} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ContainerRegistryIamBindingConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'yandex_container_registry_iam_binding',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._id = config.id;
        this._members = config.members;
        this._registryId = config.registryId;
        this._role = config.role;
        this._sleepAfter = config.sleepAfter;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get members() {
        return cdktf.Fn.tolist(this.getListAttribute('members'));
    }
    set members(value) {
        this._members = value;
    }
    // Temporarily expose input value. Use with caution.
    get membersInput() {
        return this._members;
    }
    get registryId() {
        return this.getStringAttribute('registry_id');
    }
    set registryId(value) {
        this._registryId = value;
    }
    // Temporarily expose input value. Use with caution.
    get registryIdInput() {
        return this._registryId;
    }
    get role() {
        return this.getStringAttribute('role');
    }
    set role(value) {
        this._role = value;
    }
    // Temporarily expose input value. Use with caution.
    get roleInput() {
        return this._role;
    }
    get sleepAfter() {
        return this.getNumberAttribute('sleep_after');
    }
    set sleepAfter(value) {
        this._sleepAfter = value;
    }
    resetSleepAfter() {
        this._sleepAfter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sleepAfterInput() {
        return this._sleepAfter;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            members: cdktf.listMapper(cdktf.stringToTerraform, false)(this._members),
            registry_id: cdktf.stringToTerraform(this._registryId),
            role: cdktf.stringToTerraform(this._role),
            sleep_after: cdktf.numberToTerraform(this._sleepAfter),
        };
    }
}
exports.ContainerRegistryIamBinding = ContainerRegistryIamBinding;
_a = JSII_RTTI_SYMBOL_1;
ContainerRegistryIamBinding[_a] = { fqn: "@cdktf/provider-yandex.ContainerRegistryIamBinding", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
ContainerRegistryIamBinding.tfResourceType = "yandex_container_registry_iam_binding";
