// @cdktf/provider-yandex/lib/data-yandex-mdb-postgresql-user.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataYandexMdbPostgresqlUser = exports.DataYandexMdbPostgresqlUserPermissionList = exports.DataYandexMdbPostgresqlUserPermissionOutputReference = exports.dataYandexMdbPostgresqlUserPermissionToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataYandexMdbPostgresqlUserPermissionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexMdbPostgresqlUserPermissionToTerraform = dataYandexMdbPostgresqlUserPermissionToTerraform;
class DataYandexMdbPostgresqlUserPermissionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // database_name - computed: true, optional: false, required: false
    get databaseName() {
        return this.getStringAttribute('database_name');
    }
}
exports.DataYandexMdbPostgresqlUserPermissionOutputReference = DataYandexMdbPostgresqlUserPermissionOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataYandexMdbPostgresqlUserPermissionOutputReference[_a] = { fqn: "@cdktf/provider-yandex.DataYandexMdbPostgresqlUserPermissionOutputReference", version: "2.0.32" };
class DataYandexMdbPostgresqlUserPermissionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbPostgresqlUserPermissionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbPostgresqlUserPermissionList = DataYandexMdbPostgresqlUserPermissionList;
_b = JSII_RTTI_SYMBOL_1;
DataYandexMdbPostgresqlUserPermissionList[_b] = { fqn: "@cdktf/provider-yandex.DataYandexMdbPostgresqlUserPermissionList", version: "2.0.32" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/d/mdb_postgresql_user yandex_mdb_postgresql_user}
*/
class DataYandexMdbPostgresqlUser extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/d/mdb_postgresql_user yandex_mdb_postgresql_user} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataYandexMdbPostgresqlUserConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'yandex_mdb_postgresql_user',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // permission - computed: true, optional: false, required: false
        this._permission = new DataYandexMdbPostgresqlUserPermissionList(this, "permission", true);
        this._clusterId = config.clusterId;
        this._connLimit = config.connLimit;
        this._id = config.id;
        this._login = config.login;
        this._name = config.name;
        this._settings = config.settings;
    }
    get clusterId() {
        return this.getStringAttribute('cluster_id');
    }
    set clusterId(value) {
        this._clusterId = value;
    }
    // Temporarily expose input value. Use with caution.
    get clusterIdInput() {
        return this._clusterId;
    }
    get connLimit() {
        return this.getNumberAttribute('conn_limit');
    }
    set connLimit(value) {
        this._connLimit = value;
    }
    resetConnLimit() {
        this._connLimit = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get connLimitInput() {
        return this._connLimit;
    }
    // grants - computed: true, optional: false, required: false
    get grants() {
        return this.getListAttribute('grants');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get login() {
        return this.getBooleanAttribute('login');
    }
    set login(value) {
        this._login = value;
    }
    resetLogin() {
        this._login = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get loginInput() {
        return this._login;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // password - computed: true, optional: false, required: false
    get password() {
        return this.getStringAttribute('password');
    }
    get permission() {
        return this._permission;
    }
    get settings() {
        return this.getStringMapAttribute('settings');
    }
    set settings(value) {
        this._settings = value;
    }
    resetSettings() {
        this._settings = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get settingsInput() {
        return this._settings;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            cluster_id: cdktf.stringToTerraform(this._clusterId),
            conn_limit: cdktf.numberToTerraform(this._connLimit),
            id: cdktf.stringToTerraform(this._id),
            login: cdktf.booleanToTerraform(this._login),
            name: cdktf.stringToTerraform(this._name),
            settings: cdktf.hashMapper(cdktf.stringToTerraform)(this._settings),
        };
    }
}
exports.DataYandexMdbPostgresqlUser = DataYandexMdbPostgresqlUser;
_c = JSII_RTTI_SYMBOL_1;
DataYandexMdbPostgresqlUser[_c] = { fqn: "@cdktf/provider-yandex.DataYandexMdbPostgresqlUser", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
DataYandexMdbPostgresqlUser.tfResourceType = "yandex_mdb_postgresql_user";
