// @cdktf/provider-yandex/lib/data-yandex-iot-core-device.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataYandexIotCoreDevice = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/d/iot_core_device yandex_iot_core_device}
*/
class DataYandexIotCoreDevice extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/d/iot_core_device yandex_iot_core_device} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataYandexIotCoreDeviceConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'yandex_iot_core_device',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ==========
        // ATTRIBUTES
        // ==========
        // aliases - computed: true, optional: false, required: false
        this._aliases = new cdktf.StringMap(this, "aliases");
        this._deviceId = config.deviceId;
        this._id = config.id;
        this._name = config.name;
    }
    get aliases() {
        return this._aliases;
    }
    // certificates - computed: true, optional: false, required: false
    get certificates() {
        return cdktf.Fn.tolist(this.getListAttribute('certificates'));
    }
    // created_at - computed: true, optional: false, required: false
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    get deviceId() {
        return this.getStringAttribute('device_id');
    }
    set deviceId(value) {
        this._deviceId = value;
    }
    resetDeviceId() {
        this._deviceId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deviceIdInput() {
        return this._deviceId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // passwords - computed: true, optional: false, required: false
    get passwords() {
        return cdktf.Fn.tolist(this.getListAttribute('passwords'));
    }
    // registry_id - computed: true, optional: false, required: false
    get registryId() {
        return this.getStringAttribute('registry_id');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            device_id: cdktf.stringToTerraform(this._deviceId),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
        };
    }
}
exports.DataYandexIotCoreDevice = DataYandexIotCoreDevice;
_a = JSII_RTTI_SYMBOL_1;
DataYandexIotCoreDevice[_a] = { fqn: "@cdktf/provider-yandex.DataYandexIotCoreDevice", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
DataYandexIotCoreDevice.tfResourceType = "yandex_iot_core_device";
