// @cdktf/provider-yandex/lib/data-yandex-vpc-address.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataYandexVpcAddress = exports.DataYandexVpcAddressExternalIpv4AddressList = exports.DataYandexVpcAddressExternalIpv4AddressOutputReference = exports.dataYandexVpcAddressExternalIpv4AddressToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataYandexVpcAddressExternalIpv4AddressToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexVpcAddressExternalIpv4AddressToTerraform = dataYandexVpcAddressExternalIpv4AddressToTerraform;
class DataYandexVpcAddressExternalIpv4AddressOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // address - computed: true, optional: false, required: false
    get address() {
        return this.getStringAttribute('address');
    }
    // ddos_protection_provider - computed: true, optional: false, required: false
    get ddosProtectionProvider() {
        return this.getStringAttribute('ddos_protection_provider');
    }
    // outgoing_smtp_capability - computed: true, optional: false, required: false
    get outgoingSmtpCapability() {
        return this.getStringAttribute('outgoing_smtp_capability');
    }
    // zone_id - computed: true, optional: false, required: false
    get zoneId() {
        return this.getStringAttribute('zone_id');
    }
}
exports.DataYandexVpcAddressExternalIpv4AddressOutputReference = DataYandexVpcAddressExternalIpv4AddressOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataYandexVpcAddressExternalIpv4AddressOutputReference[_a] = { fqn: "@cdktf/provider-yandex.DataYandexVpcAddressExternalIpv4AddressOutputReference", version: "2.0.32" };
class DataYandexVpcAddressExternalIpv4AddressList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexVpcAddressExternalIpv4AddressOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexVpcAddressExternalIpv4AddressList = DataYandexVpcAddressExternalIpv4AddressList;
_b = JSII_RTTI_SYMBOL_1;
DataYandexVpcAddressExternalIpv4AddressList[_b] = { fqn: "@cdktf/provider-yandex.DataYandexVpcAddressExternalIpv4AddressList", version: "2.0.32" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/d/vpc_address yandex_vpc_address}
*/
class DataYandexVpcAddress extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/d/vpc_address yandex_vpc_address} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataYandexVpcAddressConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'yandex_vpc_address',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // external_ipv4_address - computed: true, optional: false, required: false
        this._externalIpv4Address = new DataYandexVpcAddressExternalIpv4AddressList(this, "external_ipv4_address", false);
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        this._addressId = config.addressId;
        this._folderId = config.folderId;
        this._id = config.id;
        this._name = config.name;
    }
    get addressId() {
        return this.getStringAttribute('address_id');
    }
    set addressId(value) {
        this._addressId = value;
    }
    resetAddressId() {
        this._addressId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get addressIdInput() {
        return this._addressId;
    }
    // created_at - computed: true, optional: false, required: false
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    get externalIpv4Address() {
        return this._externalIpv4Address;
    }
    get folderId() {
        return this.getStringAttribute('folder_id');
    }
    set folderId(value) {
        this._folderId = value;
    }
    resetFolderId() {
        this._folderId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get folderIdInput() {
        return this._folderId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this._labels;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // reserved - computed: true, optional: false, required: false
    get reserved() {
        return this.getBooleanAttribute('reserved');
    }
    // used - computed: true, optional: false, required: false
    get used() {
        return this.getBooleanAttribute('used');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            address_id: cdktf.stringToTerraform(this._addressId),
            folder_id: cdktf.stringToTerraform(this._folderId),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
        };
    }
}
exports.DataYandexVpcAddress = DataYandexVpcAddress;
_c = JSII_RTTI_SYMBOL_1;
DataYandexVpcAddress[_c] = { fqn: "@cdktf/provider-yandex.DataYandexVpcAddress", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
DataYandexVpcAddress.tfResourceType = "yandex_vpc_address";
