// @cdktf/provider-yandex/lib/data-yandex-api-gateway.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataYandexApiGateway = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/d/api_gateway yandex_api_gateway}
*/
class DataYandexApiGateway extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/d/api_gateway yandex_api_gateway} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataYandexApiGatewayConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'yandex_api_gateway',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        this._apiGatewayId = config.apiGatewayId;
        this._folderId = config.folderId;
        this._id = config.id;
        this._name = config.name;
    }
    get apiGatewayId() {
        return this.getStringAttribute('api_gateway_id');
    }
    set apiGatewayId(value) {
        this._apiGatewayId = value;
    }
    resetApiGatewayId() {
        this._apiGatewayId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get apiGatewayIdInput() {
        return this._apiGatewayId;
    }
    // created_at - computed: true, optional: false, required: false
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    // domain - computed: true, optional: false, required: false
    get domain() {
        return this.getStringAttribute('domain');
    }
    get folderId() {
        return this.getStringAttribute('folder_id');
    }
    set folderId(value) {
        this._folderId = value;
    }
    resetFolderId() {
        this._folderId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get folderIdInput() {
        return this._folderId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this._labels;
    }
    // log_group_id - computed: true, optional: false, required: false
    get logGroupId() {
        return this.getStringAttribute('log_group_id');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    // user_domains - computed: true, optional: false, required: false
    get userDomains() {
        return cdktf.Fn.tolist(this.getListAttribute('user_domains'));
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            api_gateway_id: cdktf.stringToTerraform(this._apiGatewayId),
            folder_id: cdktf.stringToTerraform(this._folderId),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
        };
    }
}
exports.DataYandexApiGateway = DataYandexApiGateway;
_a = JSII_RTTI_SYMBOL_1;
DataYandexApiGateway[_a] = { fqn: "@cdktf/provider-yandex.DataYandexApiGateway", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
DataYandexApiGateway.tfResourceType = "yandex_api_gateway";
