// @cdktf/provider-yandex/lib/dataproc-cluster.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataprocCluster = exports.DataprocClusterTimeoutsOutputReference = exports.dataprocClusterTimeoutsToTerraform = exports.DataprocClusterClusterConfigOutputReference = exports.dataprocClusterClusterConfigToTerraform = exports.DataprocClusterClusterConfigSubclusterSpecList = exports.DataprocClusterClusterConfigSubclusterSpecOutputReference = exports.dataprocClusterClusterConfigSubclusterSpecToTerraform = exports.DataprocClusterClusterConfigSubclusterSpecResourcesOutputReference = exports.dataprocClusterClusterConfigSubclusterSpecResourcesToTerraform = exports.DataprocClusterClusterConfigSubclusterSpecAutoscalingConfigOutputReference = exports.dataprocClusterClusterConfigSubclusterSpecAutoscalingConfigToTerraform = exports.DataprocClusterClusterConfigHadoopOutputReference = exports.dataprocClusterClusterConfigHadoopToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataprocClusterClusterConfigHadoopToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        properties: cdktf.hashMapper(cdktf.stringToTerraform)(struct.properties),
        services: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.services),
        ssh_public_keys: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.sshPublicKeys),
    };
}
exports.dataprocClusterClusterConfigHadoopToTerraform = dataprocClusterClusterConfigHadoopToTerraform;
class DataprocClusterClusterConfigHadoopOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._properties !== undefined) {
            hasAnyValues = true;
            internalValueResult.properties = this._properties;
        }
        if (this._services !== undefined) {
            hasAnyValues = true;
            internalValueResult.services = this._services;
        }
        if (this._sshPublicKeys !== undefined) {
            hasAnyValues = true;
            internalValueResult.sshPublicKeys = this._sshPublicKeys;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._properties = undefined;
            this._services = undefined;
            this._sshPublicKeys = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._properties = value.properties;
            this._services = value.services;
            this._sshPublicKeys = value.sshPublicKeys;
        }
    }
    get properties() {
        return this.getStringMapAttribute('properties');
    }
    set properties(value) {
        this._properties = value;
    }
    resetProperties() {
        this._properties = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get propertiesInput() {
        return this._properties;
    }
    get services() {
        return cdktf.Fn.tolist(this.getListAttribute('services'));
    }
    set services(value) {
        this._services = value;
    }
    resetServices() {
        this._services = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get servicesInput() {
        return this._services;
    }
    get sshPublicKeys() {
        return cdktf.Fn.tolist(this.getListAttribute('ssh_public_keys'));
    }
    set sshPublicKeys(value) {
        this._sshPublicKeys = value;
    }
    resetSshPublicKeys() {
        this._sshPublicKeys = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sshPublicKeysInput() {
        return this._sshPublicKeys;
    }
}
exports.DataprocClusterClusterConfigHadoopOutputReference = DataprocClusterClusterConfigHadoopOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataprocClusterClusterConfigHadoopOutputReference[_a] = { fqn: "@cdktf/provider-yandex.DataprocClusterClusterConfigHadoopOutputReference", version: "2.0.32" };
function dataprocClusterClusterConfigSubclusterSpecAutoscalingConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cpu_utilization_target: cdktf.numberToTerraform(struct.cpuUtilizationTarget),
        decommission_timeout: cdktf.numberToTerraform(struct.decommissionTimeout),
        max_hosts_count: cdktf.numberToTerraform(struct.maxHostsCount),
        measurement_duration: cdktf.numberToTerraform(struct.measurementDuration),
        preemptible: cdktf.booleanToTerraform(struct.preemptible),
        stabilization_duration: cdktf.numberToTerraform(struct.stabilizationDuration),
        warmup_duration: cdktf.numberToTerraform(struct.warmupDuration),
    };
}
exports.dataprocClusterClusterConfigSubclusterSpecAutoscalingConfigToTerraform = dataprocClusterClusterConfigSubclusterSpecAutoscalingConfigToTerraform;
class DataprocClusterClusterConfigSubclusterSpecAutoscalingConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._cpuUtilizationTarget !== undefined) {
            hasAnyValues = true;
            internalValueResult.cpuUtilizationTarget = this._cpuUtilizationTarget;
        }
        if (this._decommissionTimeout !== undefined) {
            hasAnyValues = true;
            internalValueResult.decommissionTimeout = this._decommissionTimeout;
        }
        if (this._maxHostsCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxHostsCount = this._maxHostsCount;
        }
        if (this._measurementDuration !== undefined) {
            hasAnyValues = true;
            internalValueResult.measurementDuration = this._measurementDuration;
        }
        if (this._preemptible !== undefined) {
            hasAnyValues = true;
            internalValueResult.preemptible = this._preemptible;
        }
        if (this._stabilizationDuration !== undefined) {
            hasAnyValues = true;
            internalValueResult.stabilizationDuration = this._stabilizationDuration;
        }
        if (this._warmupDuration !== undefined) {
            hasAnyValues = true;
            internalValueResult.warmupDuration = this._warmupDuration;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._cpuUtilizationTarget = undefined;
            this._decommissionTimeout = undefined;
            this._maxHostsCount = undefined;
            this._measurementDuration = undefined;
            this._preemptible = undefined;
            this._stabilizationDuration = undefined;
            this._warmupDuration = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._cpuUtilizationTarget = value.cpuUtilizationTarget;
            this._decommissionTimeout = value.decommissionTimeout;
            this._maxHostsCount = value.maxHostsCount;
            this._measurementDuration = value.measurementDuration;
            this._preemptible = value.preemptible;
            this._stabilizationDuration = value.stabilizationDuration;
            this._warmupDuration = value.warmupDuration;
        }
    }
    get cpuUtilizationTarget() {
        return this.getNumberAttribute('cpu_utilization_target');
    }
    set cpuUtilizationTarget(value) {
        this._cpuUtilizationTarget = value;
    }
    resetCpuUtilizationTarget() {
        this._cpuUtilizationTarget = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cpuUtilizationTargetInput() {
        return this._cpuUtilizationTarget;
    }
    get decommissionTimeout() {
        return this.getNumberAttribute('decommission_timeout');
    }
    set decommissionTimeout(value) {
        this._decommissionTimeout = value;
    }
    resetDecommissionTimeout() {
        this._decommissionTimeout = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get decommissionTimeoutInput() {
        return this._decommissionTimeout;
    }
    get maxHostsCount() {
        return this.getNumberAttribute('max_hosts_count');
    }
    set maxHostsCount(value) {
        this._maxHostsCount = value;
    }
    // Temporarily expose input value. Use with caution.
    get maxHostsCountInput() {
        return this._maxHostsCount;
    }
    get measurementDuration() {
        return this.getNumberAttribute('measurement_duration');
    }
    set measurementDuration(value) {
        this._measurementDuration = value;
    }
    resetMeasurementDuration() {
        this._measurementDuration = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get measurementDurationInput() {
        return this._measurementDuration;
    }
    get preemptible() {
        return this.getBooleanAttribute('preemptible');
    }
    set preemptible(value) {
        this._preemptible = value;
    }
    resetPreemptible() {
        this._preemptible = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get preemptibleInput() {
        return this._preemptible;
    }
    get stabilizationDuration() {
        return this.getNumberAttribute('stabilization_duration');
    }
    set stabilizationDuration(value) {
        this._stabilizationDuration = value;
    }
    resetStabilizationDuration() {
        this._stabilizationDuration = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get stabilizationDurationInput() {
        return this._stabilizationDuration;
    }
    get warmupDuration() {
        return this.getNumberAttribute('warmup_duration');
    }
    set warmupDuration(value) {
        this._warmupDuration = value;
    }
    resetWarmupDuration() {
        this._warmupDuration = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get warmupDurationInput() {
        return this._warmupDuration;
    }
}
exports.DataprocClusterClusterConfigSubclusterSpecAutoscalingConfigOutputReference = DataprocClusterClusterConfigSubclusterSpecAutoscalingConfigOutputReference;
_b = JSII_RTTI_SYMBOL_1;
DataprocClusterClusterConfigSubclusterSpecAutoscalingConfigOutputReference[_b] = { fqn: "@cdktf/provider-yandex.DataprocClusterClusterConfigSubclusterSpecAutoscalingConfigOutputReference", version: "2.0.32" };
function dataprocClusterClusterConfigSubclusterSpecResourcesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        disk_size: cdktf.numberToTerraform(struct.diskSize),
        disk_type_id: cdktf.stringToTerraform(struct.diskTypeId),
        resource_preset_id: cdktf.stringToTerraform(struct.resourcePresetId),
    };
}
exports.dataprocClusterClusterConfigSubclusterSpecResourcesToTerraform = dataprocClusterClusterConfigSubclusterSpecResourcesToTerraform;
class DataprocClusterClusterConfigSubclusterSpecResourcesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._diskSize !== undefined) {
            hasAnyValues = true;
            internalValueResult.diskSize = this._diskSize;
        }
        if (this._diskTypeId !== undefined) {
            hasAnyValues = true;
            internalValueResult.diskTypeId = this._diskTypeId;
        }
        if (this._resourcePresetId !== undefined) {
            hasAnyValues = true;
            internalValueResult.resourcePresetId = this._resourcePresetId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._diskSize = undefined;
            this._diskTypeId = undefined;
            this._resourcePresetId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._diskSize = value.diskSize;
            this._diskTypeId = value.diskTypeId;
            this._resourcePresetId = value.resourcePresetId;
        }
    }
    get diskSize() {
        return this.getNumberAttribute('disk_size');
    }
    set diskSize(value) {
        this._diskSize = value;
    }
    // Temporarily expose input value. Use with caution.
    get diskSizeInput() {
        return this._diskSize;
    }
    get diskTypeId() {
        return this.getStringAttribute('disk_type_id');
    }
    set diskTypeId(value) {
        this._diskTypeId = value;
    }
    resetDiskTypeId() {
        this._diskTypeId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get diskTypeIdInput() {
        return this._diskTypeId;
    }
    get resourcePresetId() {
        return this.getStringAttribute('resource_preset_id');
    }
    set resourcePresetId(value) {
        this._resourcePresetId = value;
    }
    // Temporarily expose input value. Use with caution.
    get resourcePresetIdInput() {
        return this._resourcePresetId;
    }
}
exports.DataprocClusterClusterConfigSubclusterSpecResourcesOutputReference = DataprocClusterClusterConfigSubclusterSpecResourcesOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataprocClusterClusterConfigSubclusterSpecResourcesOutputReference[_c] = { fqn: "@cdktf/provider-yandex.DataprocClusterClusterConfigSubclusterSpecResourcesOutputReference", version: "2.0.32" };
function dataprocClusterClusterConfigSubclusterSpecToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        assign_public_ip: cdktf.booleanToTerraform(struct.assignPublicIp),
        hosts_count: cdktf.numberToTerraform(struct.hostsCount),
        name: cdktf.stringToTerraform(struct.name),
        role: cdktf.stringToTerraform(struct.role),
        subnet_id: cdktf.stringToTerraform(struct.subnetId),
        autoscaling_config: dataprocClusterClusterConfigSubclusterSpecAutoscalingConfigToTerraform(struct.autoscalingConfig),
        resources: dataprocClusterClusterConfigSubclusterSpecResourcesToTerraform(struct.resources),
    };
}
exports.dataprocClusterClusterConfigSubclusterSpecToTerraform = dataprocClusterClusterConfigSubclusterSpecToTerraform;
class DataprocClusterClusterConfigSubclusterSpecOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // autoscaling_config - computed: false, optional: true, required: false
        this._autoscalingConfig = new DataprocClusterClusterConfigSubclusterSpecAutoscalingConfigOutputReference(this, "autoscaling_config");
        // resources - computed: false, optional: false, required: true
        this._resources = new DataprocClusterClusterConfigSubclusterSpecResourcesOutputReference(this, "resources");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._assignPublicIp !== undefined) {
            hasAnyValues = true;
            internalValueResult.assignPublicIp = this._assignPublicIp;
        }
        if (this._hostsCount !== undefined) {
            hasAnyValues = true;
            internalValueResult.hostsCount = this._hostsCount;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._role !== undefined) {
            hasAnyValues = true;
            internalValueResult.role = this._role;
        }
        if (this._subnetId !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnetId = this._subnetId;
        }
        if (this._autoscalingConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.autoscalingConfig = this._autoscalingConfig?.internalValue;
        }
        if (this._resources?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.resources = this._resources?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._assignPublicIp = undefined;
            this._hostsCount = undefined;
            this._name = undefined;
            this._role = undefined;
            this._subnetId = undefined;
            this._autoscalingConfig.internalValue = undefined;
            this._resources.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._assignPublicIp = value.assignPublicIp;
            this._hostsCount = value.hostsCount;
            this._name = value.name;
            this._role = value.role;
            this._subnetId = value.subnetId;
            this._autoscalingConfig.internalValue = value.autoscalingConfig;
            this._resources.internalValue = value.resources;
        }
    }
    get assignPublicIp() {
        return this.getBooleanAttribute('assign_public_ip');
    }
    set assignPublicIp(value) {
        this._assignPublicIp = value;
    }
    resetAssignPublicIp() {
        this._assignPublicIp = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get assignPublicIpInput() {
        return this._assignPublicIp;
    }
    get hostsCount() {
        return this.getNumberAttribute('hosts_count');
    }
    set hostsCount(value) {
        this._hostsCount = value;
    }
    // Temporarily expose input value. Use with caution.
    get hostsCountInput() {
        return this._hostsCount;
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get role() {
        return this.getStringAttribute('role');
    }
    set role(value) {
        this._role = value;
    }
    // Temporarily expose input value. Use with caution.
    get roleInput() {
        return this._role;
    }
    get subnetId() {
        return this.getStringAttribute('subnet_id');
    }
    set subnetId(value) {
        this._subnetId = value;
    }
    // Temporarily expose input value. Use with caution.
    get subnetIdInput() {
        return this._subnetId;
    }
    get autoscalingConfig() {
        return this._autoscalingConfig;
    }
    putAutoscalingConfig(value) {
        this._autoscalingConfig.internalValue = value;
    }
    resetAutoscalingConfig() {
        this._autoscalingConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get autoscalingConfigInput() {
        return this._autoscalingConfig.internalValue;
    }
    get resources() {
        return this._resources;
    }
    putResources(value) {
        this._resources.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get resourcesInput() {
        return this._resources.internalValue;
    }
}
exports.DataprocClusterClusterConfigSubclusterSpecOutputReference = DataprocClusterClusterConfigSubclusterSpecOutputReference;
_d = JSII_RTTI_SYMBOL_1;
DataprocClusterClusterConfigSubclusterSpecOutputReference[_d] = { fqn: "@cdktf/provider-yandex.DataprocClusterClusterConfigSubclusterSpecOutputReference", version: "2.0.32" };
class DataprocClusterClusterConfigSubclusterSpecList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataprocClusterClusterConfigSubclusterSpecOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataprocClusterClusterConfigSubclusterSpecList = DataprocClusterClusterConfigSubclusterSpecList;
_e = JSII_RTTI_SYMBOL_1;
DataprocClusterClusterConfigSubclusterSpecList[_e] = { fqn: "@cdktf/provider-yandex.DataprocClusterClusterConfigSubclusterSpecList", version: "2.0.32" };
function dataprocClusterClusterConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        version_id: cdktf.stringToTerraform(struct.versionId),
        hadoop: dataprocClusterClusterConfigHadoopToTerraform(struct.hadoop),
        subcluster_spec: cdktf.listMapper(dataprocClusterClusterConfigSubclusterSpecToTerraform, true)(struct.subclusterSpec),
    };
}
exports.dataprocClusterClusterConfigToTerraform = dataprocClusterClusterConfigToTerraform;
class DataprocClusterClusterConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // hadoop - computed: false, optional: true, required: false
        this._hadoop = new DataprocClusterClusterConfigHadoopOutputReference(this, "hadoop");
        // subcluster_spec - computed: false, optional: false, required: true
        this._subclusterSpec = new DataprocClusterClusterConfigSubclusterSpecList(this, "subcluster_spec", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._versionId !== undefined) {
            hasAnyValues = true;
            internalValueResult.versionId = this._versionId;
        }
        if (this._hadoop?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.hadoop = this._hadoop?.internalValue;
        }
        if (this._subclusterSpec?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.subclusterSpec = this._subclusterSpec?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._versionId = undefined;
            this._hadoop.internalValue = undefined;
            this._subclusterSpec.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._versionId = value.versionId;
            this._hadoop.internalValue = value.hadoop;
            this._subclusterSpec.internalValue = value.subclusterSpec;
        }
    }
    get versionId() {
        return this.getStringAttribute('version_id');
    }
    set versionId(value) {
        this._versionId = value;
    }
    resetVersionId() {
        this._versionId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get versionIdInput() {
        return this._versionId;
    }
    get hadoop() {
        return this._hadoop;
    }
    putHadoop(value) {
        this._hadoop.internalValue = value;
    }
    resetHadoop() {
        this._hadoop.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get hadoopInput() {
        return this._hadoop.internalValue;
    }
    get subclusterSpec() {
        return this._subclusterSpec;
    }
    putSubclusterSpec(value) {
        this._subclusterSpec.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get subclusterSpecInput() {
        return this._subclusterSpec.internalValue;
    }
}
exports.DataprocClusterClusterConfigOutputReference = DataprocClusterClusterConfigOutputReference;
_f = JSII_RTTI_SYMBOL_1;
DataprocClusterClusterConfigOutputReference[_f] = { fqn: "@cdktf/provider-yandex.DataprocClusterClusterConfigOutputReference", version: "2.0.32" };
function dataprocClusterTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.dataprocClusterTimeoutsToTerraform = dataprocClusterTimeoutsToTerraform;
class DataprocClusterTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.DataprocClusterTimeoutsOutputReference = DataprocClusterTimeoutsOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataprocClusterTimeoutsOutputReference[_g] = { fqn: "@cdktf/provider-yandex.DataprocClusterTimeoutsOutputReference", version: "2.0.32" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/r/dataproc_cluster yandex_dataproc_cluster}
*/
class DataprocCluster extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/r/dataproc_cluster yandex_dataproc_cluster} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataprocClusterConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'yandex_dataproc_cluster',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // cluster_config - computed: false, optional: false, required: true
        this._clusterConfig = new DataprocClusterClusterConfigOutputReference(this, "cluster_config");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new DataprocClusterTimeoutsOutputReference(this, "timeouts");
        this._bucket = config.bucket;
        this._deletionProtection = config.deletionProtection;
        this._description = config.description;
        this._folderId = config.folderId;
        this._hostGroupIds = config.hostGroupIds;
        this._id = config.id;
        this._labels = config.labels;
        this._name = config.name;
        this._securityGroupIds = config.securityGroupIds;
        this._serviceAccountId = config.serviceAccountId;
        this._uiProxy = config.uiProxy;
        this._zoneId = config.zoneId;
        this._clusterConfig.internalValue = config.clusterConfig;
        this._timeouts.internalValue = config.timeouts;
    }
    get bucket() {
        return this.getStringAttribute('bucket');
    }
    set bucket(value) {
        this._bucket = value;
    }
    resetBucket() {
        this._bucket = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get bucketInput() {
        return this._bucket;
    }
    // created_at - computed: true, optional: false, required: false
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    get deletionProtection() {
        return this.getBooleanAttribute('deletion_protection');
    }
    set deletionProtection(value) {
        this._deletionProtection = value;
    }
    resetDeletionProtection() {
        this._deletionProtection = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deletionProtectionInput() {
        return this._deletionProtection;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get folderId() {
        return this.getStringAttribute('folder_id');
    }
    set folderId(value) {
        this._folderId = value;
    }
    resetFolderId() {
        this._folderId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get folderIdInput() {
        return this._folderId;
    }
    get hostGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('host_group_ids'));
    }
    set hostGroupIds(value) {
        this._hostGroupIds = value;
    }
    resetHostGroupIds() {
        this._hostGroupIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get hostGroupIdsInput() {
        return this._hostGroupIds;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get securityGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('security_group_ids'));
    }
    set securityGroupIds(value) {
        this._securityGroupIds = value;
    }
    resetSecurityGroupIds() {
        this._securityGroupIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupIdsInput() {
        return this._securityGroupIds;
    }
    get serviceAccountId() {
        return this.getStringAttribute('service_account_id');
    }
    set serviceAccountId(value) {
        this._serviceAccountId = value;
    }
    // Temporarily expose input value. Use with caution.
    get serviceAccountIdInput() {
        return this._serviceAccountId;
    }
    get uiProxy() {
        return this.getBooleanAttribute('ui_proxy');
    }
    set uiProxy(value) {
        this._uiProxy = value;
    }
    resetUiProxy() {
        this._uiProxy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get uiProxyInput() {
        return this._uiProxy;
    }
    get zoneId() {
        return this.getStringAttribute('zone_id');
    }
    set zoneId(value) {
        this._zoneId = value;
    }
    resetZoneId() {
        this._zoneId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get zoneIdInput() {
        return this._zoneId;
    }
    get clusterConfig() {
        return this._clusterConfig;
    }
    putClusterConfig(value) {
        this._clusterConfig.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get clusterConfigInput() {
        return this._clusterConfig.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            bucket: cdktf.stringToTerraform(this._bucket),
            deletion_protection: cdktf.booleanToTerraform(this._deletionProtection),
            description: cdktf.stringToTerraform(this._description),
            folder_id: cdktf.stringToTerraform(this._folderId),
            host_group_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(this._hostGroupIds),
            id: cdktf.stringToTerraform(this._id),
            labels: cdktf.hashMapper(cdktf.stringToTerraform)(this._labels),
            name: cdktf.stringToTerraform(this._name),
            security_group_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(this._securityGroupIds),
            service_account_id: cdktf.stringToTerraform(this._serviceAccountId),
            ui_proxy: cdktf.booleanToTerraform(this._uiProxy),
            zone_id: cdktf.stringToTerraform(this._zoneId),
            cluster_config: dataprocClusterClusterConfigToTerraform(this._clusterConfig.internalValue),
            timeouts: dataprocClusterTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.DataprocCluster = DataprocCluster;
_h = JSII_RTTI_SYMBOL_1;
DataprocCluster[_h] = { fqn: "@cdktf/provider-yandex.DataprocCluster", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
DataprocCluster.tfResourceType = "yandex_dataproc_cluster";
