// @cdktf/provider-yandex/lib/data-yandex-compute-instance-group.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _0, _1, _2, _3, _4, _5, _6, _7, _8, _9, _10, _11, _12, _13, _14, _15, _16, _17, _18, _19, _20, _21, _22, _23, _24, _25, _26, _27, _28, _29, _30, _31, _32, _33, _34, _35, _36;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataYandexComputeInstanceGroup = exports.DataYandexComputeInstanceGroupScalePolicyList = exports.DataYandexComputeInstanceGroupScalePolicyOutputReference = exports.dataYandexComputeInstanceGroupScalePolicyToTerraform = exports.DataYandexComputeInstanceGroupScalePolicyTestAutoScaleList = exports.DataYandexComputeInstanceGroupScalePolicyTestAutoScaleOutputReference = exports.dataYandexComputeInstanceGroupScalePolicyTestAutoScaleToTerraform = exports.DataYandexComputeInstanceGroupScalePolicyTestAutoScaleCustomRuleList = exports.DataYandexComputeInstanceGroupScalePolicyTestAutoScaleCustomRuleOutputReference = exports.dataYandexComputeInstanceGroupScalePolicyTestAutoScaleCustomRuleToTerraform = exports.DataYandexComputeInstanceGroupScalePolicyFixedScaleList = exports.DataYandexComputeInstanceGroupScalePolicyFixedScaleOutputReference = exports.dataYandexComputeInstanceGroupScalePolicyFixedScaleToTerraform = exports.DataYandexComputeInstanceGroupScalePolicyAutoScaleList = exports.DataYandexComputeInstanceGroupScalePolicyAutoScaleOutputReference = exports.dataYandexComputeInstanceGroupScalePolicyAutoScaleToTerraform = exports.DataYandexComputeInstanceGroupScalePolicyAutoScaleCustomRuleList = exports.DataYandexComputeInstanceGroupScalePolicyAutoScaleCustomRuleOutputReference = exports.dataYandexComputeInstanceGroupScalePolicyAutoScaleCustomRuleToTerraform = exports.DataYandexComputeInstanceGroupLoadBalancerStateList = exports.DataYandexComputeInstanceGroupLoadBalancerStateOutputReference = exports.dataYandexComputeInstanceGroupLoadBalancerStateToTerraform = exports.DataYandexComputeInstanceGroupLoadBalancerList = exports.DataYandexComputeInstanceGroupLoadBalancerOutputReference = exports.dataYandexComputeInstanceGroupLoadBalancerToTerraform = exports.DataYandexComputeInstanceGroupInstancesList = exports.DataYandexComputeInstanceGroupInstancesOutputReference = exports.dataYandexComputeInstanceGroupInstancesToTerraform = exports.DataYandexComputeInstanceGroupInstancesNetworkInterfaceList = exports.DataYandexComputeInstanceGroupInstancesNetworkInterfaceOutputReference = exports.dataYandexComputeInstanceGroupInstancesNetworkInterfaceToTerraform = exports.DataYandexComputeInstanceGroupInstanceTemplateList = exports.DataYandexComputeInstanceGroupInstanceTemplateOutputReference = exports.dataYandexComputeInstanceGroupInstanceTemplateToTerraform = exports.DataYandexComputeInstanceGroupInstanceTemplateSecondaryDiskList = exports.DataYandexComputeInstanceGroupInstanceTemplateSecondaryDiskOutputReference = exports.dataYandexComputeInstanceGroupInstanceTemplateSecondaryDiskToTerraform = exports.DataYandexComputeInstanceGroupInstanceTemplateSecondaryDiskInitializeParamsList = exports.DataYandexComputeInstanceGroupInstanceTemplateSecondaryDiskInitializeParamsOutputReference = exports.dataYandexComputeInstanceGroupInstanceTemplateSecondaryDiskInitializeParamsToTerraform = exports.DataYandexComputeInstanceGroupInstanceTemplateSchedulingPolicyList = exports.DataYandexComputeInstanceGroupInstanceTemplateSchedulingPolicyOutputReference = exports.dataYandexComputeInstanceGroupInstanceTemplateSchedulingPolicyToTerraform = exports.DataYandexComputeInstanceGroupInstanceTemplateResourcesList = exports.DataYandexComputeInstanceGroupInstanceTemplateResourcesOutputReference = exports.dataYandexComputeInstanceGroupInstanceTemplateResourcesToTerraform = exports.DataYandexComputeInstanceGroupInstanceTemplatePlacementPolicyList = exports.DataYandexComputeInstanceGroupInstanceTemplatePlacementPolicyOutputReference = exports.dataYandexComputeInstanceGroupInstanceTemplatePlacementPolicyToTerraform = exports.DataYandexComputeInstanceGroupInstanceTemplateNetworkSettingsList = exports.DataYandexComputeInstanceGroupInstanceTemplateNetworkSettingsOutputReference = exports.dataYandexComputeInstanceGroupInstanceTemplateNetworkSettingsToTerraform = exports.DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceList = exports.DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceOutputReference = exports.dataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceToTerraform = exports.DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceNatDnsRecordList = exports.DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceNatDnsRecordOutputReference = exports.dataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceNatDnsRecordToTerraform = exports.DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceIpv6DnsRecordList = exports.DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceIpv6DnsRecordOutputReference = exports.dataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceIpv6DnsRecordToTerraform = exports.DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceDnsRecordList = exports.DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceDnsRecordOutputReference = exports.dataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceDnsRecordToTerraform = exports.DataYandexComputeInstanceGroupInstanceTemplateBootDiskList = exports.DataYandexComputeInstanceGroupInstanceTemplateBootDiskOutputReference = exports.dataYandexComputeInstanceGroupInstanceTemplateBootDiskToTerraform = exports.DataYandexComputeInstanceGroupInstanceTemplateBootDiskInitializeParamsList = exports.DataYandexComputeInstanceGroupInstanceTemplateBootDiskInitializeParamsOutputReference = exports.dataYandexComputeInstanceGroupInstanceTemplateBootDiskInitializeParamsToTerraform = exports.DataYandexComputeInstanceGroupHealthCheckList = exports.DataYandexComputeInstanceGroupHealthCheckOutputReference = exports.dataYandexComputeInstanceGroupHealthCheckToTerraform = exports.DataYandexComputeInstanceGroupHealthCheckTcpOptionsList = exports.DataYandexComputeInstanceGroupHealthCheckTcpOptionsOutputReference = exports.dataYandexComputeInstanceGroupHealthCheckTcpOptionsToTerraform = exports.DataYandexComputeInstanceGroupHealthCheckHttpOptionsList = exports.DataYandexComputeInstanceGroupHealthCheckHttpOptionsOutputReference = exports.dataYandexComputeInstanceGroupHealthCheckHttpOptionsToTerraform = exports.DataYandexComputeInstanceGroupDeployPolicyList = exports.DataYandexComputeInstanceGroupDeployPolicyOutputReference = exports.dataYandexComputeInstanceGroupDeployPolicyToTerraform = exports.DataYandexComputeInstanceGroupApplicationLoadBalancerList = exports.DataYandexComputeInstanceGroupApplicationLoadBalancerOutputReference = exports.dataYandexComputeInstanceGroupApplicationLoadBalancerToTerraform = exports.DataYandexComputeInstanceGroupApplicationBalancerStateList = exports.DataYandexComputeInstanceGroupApplicationBalancerStateOutputReference = exports.dataYandexComputeInstanceGroupApplicationBalancerStateToTerraform = exports.DataYandexComputeInstanceGroupAllocationPolicyList = exports.DataYandexComputeInstanceGroupAllocationPolicyOutputReference = exports.dataYandexComputeInstanceGroupAllocationPolicyToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataYandexComputeInstanceGroupAllocationPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexComputeInstanceGroupAllocationPolicyToTerraform = dataYandexComputeInstanceGroupAllocationPolicyToTerraform;
class DataYandexComputeInstanceGroupAllocationPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // zones - computed: true, optional: false, required: false
    get zones() {
        return cdktf.Fn.tolist(this.getListAttribute('zones'));
    }
}
exports.DataYandexComputeInstanceGroupAllocationPolicyOutputReference = DataYandexComputeInstanceGroupAllocationPolicyOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupAllocationPolicyOutputReference[_a] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupAllocationPolicyOutputReference", version: "2.0.32" };
class DataYandexComputeInstanceGroupAllocationPolicyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexComputeInstanceGroupAllocationPolicyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexComputeInstanceGroupAllocationPolicyList = DataYandexComputeInstanceGroupAllocationPolicyList;
_b = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupAllocationPolicyList[_b] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupAllocationPolicyList", version: "2.0.32" };
function dataYandexComputeInstanceGroupApplicationBalancerStateToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexComputeInstanceGroupApplicationBalancerStateToTerraform = dataYandexComputeInstanceGroupApplicationBalancerStateToTerraform;
class DataYandexComputeInstanceGroupApplicationBalancerStateOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // status_message - computed: true, optional: false, required: false
    get statusMessage() {
        return this.getStringAttribute('status_message');
    }
    // target_group_id - computed: true, optional: false, required: false
    get targetGroupId() {
        return this.getStringAttribute('target_group_id');
    }
}
exports.DataYandexComputeInstanceGroupApplicationBalancerStateOutputReference = DataYandexComputeInstanceGroupApplicationBalancerStateOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupApplicationBalancerStateOutputReference[_c] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupApplicationBalancerStateOutputReference", version: "2.0.32" };
class DataYandexComputeInstanceGroupApplicationBalancerStateList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexComputeInstanceGroupApplicationBalancerStateOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexComputeInstanceGroupApplicationBalancerStateList = DataYandexComputeInstanceGroupApplicationBalancerStateList;
_d = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupApplicationBalancerStateList[_d] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupApplicationBalancerStateList", version: "2.0.32" };
function dataYandexComputeInstanceGroupApplicationLoadBalancerToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexComputeInstanceGroupApplicationLoadBalancerToTerraform = dataYandexComputeInstanceGroupApplicationLoadBalancerToTerraform;
class DataYandexComputeInstanceGroupApplicationLoadBalancerOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // target_group_labels - computed: true, optional: false, required: false
        this._targetGroupLabels = new cdktf.StringMap(this, "target_group_labels");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // max_opening_traffic_duration - computed: true, optional: false, required: false
    get maxOpeningTrafficDuration() {
        return this.getNumberAttribute('max_opening_traffic_duration');
    }
    // status_message - computed: true, optional: false, required: false
    get statusMessage() {
        return this.getStringAttribute('status_message');
    }
    // target_group_description - computed: true, optional: false, required: false
    get targetGroupDescription() {
        return this.getStringAttribute('target_group_description');
    }
    // target_group_id - computed: true, optional: false, required: false
    get targetGroupId() {
        return this.getStringAttribute('target_group_id');
    }
    get targetGroupLabels() {
        return this._targetGroupLabels;
    }
    // target_group_name - computed: true, optional: false, required: false
    get targetGroupName() {
        return this.getStringAttribute('target_group_name');
    }
}
exports.DataYandexComputeInstanceGroupApplicationLoadBalancerOutputReference = DataYandexComputeInstanceGroupApplicationLoadBalancerOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupApplicationLoadBalancerOutputReference[_e] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupApplicationLoadBalancerOutputReference", version: "2.0.32" };
class DataYandexComputeInstanceGroupApplicationLoadBalancerList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexComputeInstanceGroupApplicationLoadBalancerOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexComputeInstanceGroupApplicationLoadBalancerList = DataYandexComputeInstanceGroupApplicationLoadBalancerList;
_f = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupApplicationLoadBalancerList[_f] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupApplicationLoadBalancerList", version: "2.0.32" };
function dataYandexComputeInstanceGroupDeployPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexComputeInstanceGroupDeployPolicyToTerraform = dataYandexComputeInstanceGroupDeployPolicyToTerraform;
class DataYandexComputeInstanceGroupDeployPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // max_creating - computed: true, optional: false, required: false
    get maxCreating() {
        return this.getNumberAttribute('max_creating');
    }
    // max_deleting - computed: true, optional: false, required: false
    get maxDeleting() {
        return this.getNumberAttribute('max_deleting');
    }
    // max_expansion - computed: true, optional: false, required: false
    get maxExpansion() {
        return this.getNumberAttribute('max_expansion');
    }
    // max_unavailable - computed: true, optional: false, required: false
    get maxUnavailable() {
        return this.getNumberAttribute('max_unavailable');
    }
    // startup_duration - computed: true, optional: false, required: false
    get startupDuration() {
        return this.getNumberAttribute('startup_duration');
    }
    // strategy - computed: true, optional: false, required: false
    get strategy() {
        return this.getStringAttribute('strategy');
    }
}
exports.DataYandexComputeInstanceGroupDeployPolicyOutputReference = DataYandexComputeInstanceGroupDeployPolicyOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupDeployPolicyOutputReference[_g] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupDeployPolicyOutputReference", version: "2.0.32" };
class DataYandexComputeInstanceGroupDeployPolicyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexComputeInstanceGroupDeployPolicyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexComputeInstanceGroupDeployPolicyList = DataYandexComputeInstanceGroupDeployPolicyList;
_h = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupDeployPolicyList[_h] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupDeployPolicyList", version: "2.0.32" };
function dataYandexComputeInstanceGroupHealthCheckHttpOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexComputeInstanceGroupHealthCheckHttpOptionsToTerraform = dataYandexComputeInstanceGroupHealthCheckHttpOptionsToTerraform;
class DataYandexComputeInstanceGroupHealthCheckHttpOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // path - computed: true, optional: false, required: false
    get path() {
        return this.getStringAttribute('path');
    }
    // port - computed: true, optional: false, required: false
    get port() {
        return this.getNumberAttribute('port');
    }
}
exports.DataYandexComputeInstanceGroupHealthCheckHttpOptionsOutputReference = DataYandexComputeInstanceGroupHealthCheckHttpOptionsOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupHealthCheckHttpOptionsOutputReference[_j] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupHealthCheckHttpOptionsOutputReference", version: "2.0.32" };
class DataYandexComputeInstanceGroupHealthCheckHttpOptionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexComputeInstanceGroupHealthCheckHttpOptionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexComputeInstanceGroupHealthCheckHttpOptionsList = DataYandexComputeInstanceGroupHealthCheckHttpOptionsList;
_k = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupHealthCheckHttpOptionsList[_k] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupHealthCheckHttpOptionsList", version: "2.0.32" };
function dataYandexComputeInstanceGroupHealthCheckTcpOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexComputeInstanceGroupHealthCheckTcpOptionsToTerraform = dataYandexComputeInstanceGroupHealthCheckTcpOptionsToTerraform;
class DataYandexComputeInstanceGroupHealthCheckTcpOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // port - computed: true, optional: false, required: false
    get port() {
        return this.getNumberAttribute('port');
    }
}
exports.DataYandexComputeInstanceGroupHealthCheckTcpOptionsOutputReference = DataYandexComputeInstanceGroupHealthCheckTcpOptionsOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupHealthCheckTcpOptionsOutputReference[_l] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupHealthCheckTcpOptionsOutputReference", version: "2.0.32" };
class DataYandexComputeInstanceGroupHealthCheckTcpOptionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexComputeInstanceGroupHealthCheckTcpOptionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexComputeInstanceGroupHealthCheckTcpOptionsList = DataYandexComputeInstanceGroupHealthCheckTcpOptionsList;
_m = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupHealthCheckTcpOptionsList[_m] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupHealthCheckTcpOptionsList", version: "2.0.32" };
function dataYandexComputeInstanceGroupHealthCheckToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexComputeInstanceGroupHealthCheckToTerraform = dataYandexComputeInstanceGroupHealthCheckToTerraform;
class DataYandexComputeInstanceGroupHealthCheckOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // http_options - computed: true, optional: false, required: false
        this._httpOptions = new DataYandexComputeInstanceGroupHealthCheckHttpOptionsList(this, "http_options", false);
        // tcp_options - computed: true, optional: false, required: false
        this._tcpOptions = new DataYandexComputeInstanceGroupHealthCheckTcpOptionsList(this, "tcp_options", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // healthy_threshold - computed: true, optional: false, required: false
    get healthyThreshold() {
        return this.getNumberAttribute('healthy_threshold');
    }
    get httpOptions() {
        return this._httpOptions;
    }
    // interval - computed: true, optional: false, required: false
    get interval() {
        return this.getNumberAttribute('interval');
    }
    get tcpOptions() {
        return this._tcpOptions;
    }
    // timeout - computed: true, optional: false, required: false
    get timeout() {
        return this.getNumberAttribute('timeout');
    }
    // unhealthy_threshold - computed: true, optional: false, required: false
    get unhealthyThreshold() {
        return this.getNumberAttribute('unhealthy_threshold');
    }
}
exports.DataYandexComputeInstanceGroupHealthCheckOutputReference = DataYandexComputeInstanceGroupHealthCheckOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupHealthCheckOutputReference[_o] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupHealthCheckOutputReference", version: "2.0.32" };
class DataYandexComputeInstanceGroupHealthCheckList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexComputeInstanceGroupHealthCheckOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexComputeInstanceGroupHealthCheckList = DataYandexComputeInstanceGroupHealthCheckList;
_p = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupHealthCheckList[_p] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupHealthCheckList", version: "2.0.32" };
function dataYandexComputeInstanceGroupInstanceTemplateBootDiskInitializeParamsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexComputeInstanceGroupInstanceTemplateBootDiskInitializeParamsToTerraform = dataYandexComputeInstanceGroupInstanceTemplateBootDiskInitializeParamsToTerraform;
class DataYandexComputeInstanceGroupInstanceTemplateBootDiskInitializeParamsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    // image_id - computed: true, optional: false, required: false
    get imageId() {
        return this.getStringAttribute('image_id');
    }
    // size - computed: true, optional: false, required: false
    get size() {
        return this.getNumberAttribute('size');
    }
    // snapshot_id - computed: true, optional: false, required: false
    get snapshotId() {
        return this.getStringAttribute('snapshot_id');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataYandexComputeInstanceGroupInstanceTemplateBootDiskInitializeParamsOutputReference = DataYandexComputeInstanceGroupInstanceTemplateBootDiskInitializeParamsOutputReference;
_q = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupInstanceTemplateBootDiskInitializeParamsOutputReference[_q] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupInstanceTemplateBootDiskInitializeParamsOutputReference", version: "2.0.32" };
class DataYandexComputeInstanceGroupInstanceTemplateBootDiskInitializeParamsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexComputeInstanceGroupInstanceTemplateBootDiskInitializeParamsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexComputeInstanceGroupInstanceTemplateBootDiskInitializeParamsList = DataYandexComputeInstanceGroupInstanceTemplateBootDiskInitializeParamsList;
_r = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupInstanceTemplateBootDiskInitializeParamsList[_r] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupInstanceTemplateBootDiskInitializeParamsList", version: "2.0.32" };
function dataYandexComputeInstanceGroupInstanceTemplateBootDiskToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexComputeInstanceGroupInstanceTemplateBootDiskToTerraform = dataYandexComputeInstanceGroupInstanceTemplateBootDiskToTerraform;
class DataYandexComputeInstanceGroupInstanceTemplateBootDiskOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // initialize_params - computed: true, optional: false, required: false
        this._initializeParams = new DataYandexComputeInstanceGroupInstanceTemplateBootDiskInitializeParamsList(this, "initialize_params", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // device_name - computed: true, optional: false, required: false
    get deviceName() {
        return this.getStringAttribute('device_name');
    }
    // disk_id - computed: true, optional: false, required: false
    get diskId() {
        return this.getStringAttribute('disk_id');
    }
    get initializeParams() {
        return this._initializeParams;
    }
    // mode - computed: true, optional: false, required: false
    get mode() {
        return this.getStringAttribute('mode');
    }
}
exports.DataYandexComputeInstanceGroupInstanceTemplateBootDiskOutputReference = DataYandexComputeInstanceGroupInstanceTemplateBootDiskOutputReference;
_s = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupInstanceTemplateBootDiskOutputReference[_s] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupInstanceTemplateBootDiskOutputReference", version: "2.0.32" };
class DataYandexComputeInstanceGroupInstanceTemplateBootDiskList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexComputeInstanceGroupInstanceTemplateBootDiskOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexComputeInstanceGroupInstanceTemplateBootDiskList = DataYandexComputeInstanceGroupInstanceTemplateBootDiskList;
_t = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupInstanceTemplateBootDiskList[_t] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupInstanceTemplateBootDiskList", version: "2.0.32" };
function dataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceDnsRecordToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceDnsRecordToTerraform = dataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceDnsRecordToTerraform;
class DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceDnsRecordOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // dns_zone_id - computed: true, optional: false, required: false
    get dnsZoneId() {
        return this.getStringAttribute('dns_zone_id');
    }
    // fqdn - computed: true, optional: false, required: false
    get fqdn() {
        return this.getStringAttribute('fqdn');
    }
    // ptr - computed: true, optional: false, required: false
    get ptr() {
        return this.getBooleanAttribute('ptr');
    }
    // ttl - computed: true, optional: false, required: false
    get ttl() {
        return this.getNumberAttribute('ttl');
    }
}
exports.DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceDnsRecordOutputReference = DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceDnsRecordOutputReference;
_u = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceDnsRecordOutputReference[_u] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceDnsRecordOutputReference", version: "2.0.32" };
class DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceDnsRecordList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceDnsRecordOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceDnsRecordList = DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceDnsRecordList;
_v = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceDnsRecordList[_v] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceDnsRecordList", version: "2.0.32" };
function dataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceIpv6DnsRecordToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceIpv6DnsRecordToTerraform = dataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceIpv6DnsRecordToTerraform;
class DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceIpv6DnsRecordOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // dns_zone_id - computed: true, optional: false, required: false
    get dnsZoneId() {
        return this.getStringAttribute('dns_zone_id');
    }
    // fqdn - computed: true, optional: false, required: false
    get fqdn() {
        return this.getStringAttribute('fqdn');
    }
    // ptr - computed: true, optional: false, required: false
    get ptr() {
        return this.getBooleanAttribute('ptr');
    }
    // ttl - computed: true, optional: false, required: false
    get ttl() {
        return this.getNumberAttribute('ttl');
    }
}
exports.DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceIpv6DnsRecordOutputReference = DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceIpv6DnsRecordOutputReference;
_w = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceIpv6DnsRecordOutputReference[_w] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceIpv6DnsRecordOutputReference", version: "2.0.32" };
class DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceIpv6DnsRecordList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceIpv6DnsRecordOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceIpv6DnsRecordList = DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceIpv6DnsRecordList;
_x = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceIpv6DnsRecordList[_x] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceIpv6DnsRecordList", version: "2.0.32" };
function dataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceNatDnsRecordToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceNatDnsRecordToTerraform = dataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceNatDnsRecordToTerraform;
class DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceNatDnsRecordOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // dns_zone_id - computed: true, optional: false, required: false
    get dnsZoneId() {
        return this.getStringAttribute('dns_zone_id');
    }
    // fqdn - computed: true, optional: false, required: false
    get fqdn() {
        return this.getStringAttribute('fqdn');
    }
    // ptr - computed: true, optional: false, required: false
    get ptr() {
        return this.getBooleanAttribute('ptr');
    }
    // ttl - computed: true, optional: false, required: false
    get ttl() {
        return this.getNumberAttribute('ttl');
    }
}
exports.DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceNatDnsRecordOutputReference = DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceNatDnsRecordOutputReference;
_y = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceNatDnsRecordOutputReference[_y] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceNatDnsRecordOutputReference", version: "2.0.32" };
class DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceNatDnsRecordList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceNatDnsRecordOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceNatDnsRecordList = DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceNatDnsRecordList;
_z = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceNatDnsRecordList[_z] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceNatDnsRecordList", version: "2.0.32" };
function dataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceToTerraform = dataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceToTerraform;
class DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // dns_record - computed: true, optional: false, required: false
        this._dnsRecord = new DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceDnsRecordList(this, "dns_record", false);
        // ipv6_dns_record - computed: true, optional: false, required: false
        this._ipv6DnsRecord = new DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceIpv6DnsRecordList(this, "ipv6_dns_record", false);
        // nat_dns_record - computed: true, optional: false, required: false
        this._natDnsRecord = new DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceNatDnsRecordList(this, "nat_dns_record", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get dnsRecord() {
        return this._dnsRecord;
    }
    // ip_address - computed: true, optional: false, required: false
    get ipAddress() {
        return this.getStringAttribute('ip_address');
    }
    // ipv4 - computed: true, optional: false, required: false
    get ipv4() {
        return this.getBooleanAttribute('ipv4');
    }
    // ipv6 - computed: true, optional: false, required: false
    get ipv6() {
        return this.getBooleanAttribute('ipv6');
    }
    // ipv6_address - computed: true, optional: false, required: false
    get ipv6Address() {
        return this.getStringAttribute('ipv6_address');
    }
    get ipv6DnsRecord() {
        return this._ipv6DnsRecord;
    }
    // nat - computed: true, optional: false, required: false
    get nat() {
        return this.getBooleanAttribute('nat');
    }
    get natDnsRecord() {
        return this._natDnsRecord;
    }
    // nat_ip_address - computed: true, optional: false, required: false
    get natIpAddress() {
        return this.getStringAttribute('nat_ip_address');
    }
    // network_id - computed: true, optional: false, required: false
    get networkId() {
        return this.getStringAttribute('network_id');
    }
    // security_group_ids - computed: true, optional: false, required: false
    get securityGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('security_group_ids'));
    }
    // subnet_ids - computed: true, optional: false, required: false
    get subnetIds() {
        return cdktf.Fn.tolist(this.getListAttribute('subnet_ids'));
    }
}
exports.DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceOutputReference = DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceOutputReference;
_0 = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceOutputReference[_0] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceOutputReference", version: "2.0.32" };
class DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceList = DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceList;
_1 = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceList[_1] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceList", version: "2.0.32" };
function dataYandexComputeInstanceGroupInstanceTemplateNetworkSettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexComputeInstanceGroupInstanceTemplateNetworkSettingsToTerraform = dataYandexComputeInstanceGroupInstanceTemplateNetworkSettingsToTerraform;
class DataYandexComputeInstanceGroupInstanceTemplateNetworkSettingsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataYandexComputeInstanceGroupInstanceTemplateNetworkSettingsOutputReference = DataYandexComputeInstanceGroupInstanceTemplateNetworkSettingsOutputReference;
_2 = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupInstanceTemplateNetworkSettingsOutputReference[_2] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupInstanceTemplateNetworkSettingsOutputReference", version: "2.0.32" };
class DataYandexComputeInstanceGroupInstanceTemplateNetworkSettingsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexComputeInstanceGroupInstanceTemplateNetworkSettingsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexComputeInstanceGroupInstanceTemplateNetworkSettingsList = DataYandexComputeInstanceGroupInstanceTemplateNetworkSettingsList;
_3 = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupInstanceTemplateNetworkSettingsList[_3] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupInstanceTemplateNetworkSettingsList", version: "2.0.32" };
function dataYandexComputeInstanceGroupInstanceTemplatePlacementPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexComputeInstanceGroupInstanceTemplatePlacementPolicyToTerraform = dataYandexComputeInstanceGroupInstanceTemplatePlacementPolicyToTerraform;
class DataYandexComputeInstanceGroupInstanceTemplatePlacementPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // placement_group_id - computed: true, optional: false, required: false
    get placementGroupId() {
        return this.getStringAttribute('placement_group_id');
    }
}
exports.DataYandexComputeInstanceGroupInstanceTemplatePlacementPolicyOutputReference = DataYandexComputeInstanceGroupInstanceTemplatePlacementPolicyOutputReference;
_4 = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupInstanceTemplatePlacementPolicyOutputReference[_4] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupInstanceTemplatePlacementPolicyOutputReference", version: "2.0.32" };
class DataYandexComputeInstanceGroupInstanceTemplatePlacementPolicyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexComputeInstanceGroupInstanceTemplatePlacementPolicyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexComputeInstanceGroupInstanceTemplatePlacementPolicyList = DataYandexComputeInstanceGroupInstanceTemplatePlacementPolicyList;
_5 = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupInstanceTemplatePlacementPolicyList[_5] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupInstanceTemplatePlacementPolicyList", version: "2.0.32" };
function dataYandexComputeInstanceGroupInstanceTemplateResourcesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexComputeInstanceGroupInstanceTemplateResourcesToTerraform = dataYandexComputeInstanceGroupInstanceTemplateResourcesToTerraform;
class DataYandexComputeInstanceGroupInstanceTemplateResourcesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // core_fraction - computed: true, optional: false, required: false
    get coreFraction() {
        return this.getNumberAttribute('core_fraction');
    }
    // cores - computed: true, optional: false, required: false
    get cores() {
        return this.getNumberAttribute('cores');
    }
    // gpus - computed: true, optional: false, required: false
    get gpus() {
        return this.getNumberAttribute('gpus');
    }
    // memory - computed: true, optional: false, required: false
    get memory() {
        return this.getNumberAttribute('memory');
    }
}
exports.DataYandexComputeInstanceGroupInstanceTemplateResourcesOutputReference = DataYandexComputeInstanceGroupInstanceTemplateResourcesOutputReference;
_6 = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupInstanceTemplateResourcesOutputReference[_6] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupInstanceTemplateResourcesOutputReference", version: "2.0.32" };
class DataYandexComputeInstanceGroupInstanceTemplateResourcesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexComputeInstanceGroupInstanceTemplateResourcesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexComputeInstanceGroupInstanceTemplateResourcesList = DataYandexComputeInstanceGroupInstanceTemplateResourcesList;
_7 = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupInstanceTemplateResourcesList[_7] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupInstanceTemplateResourcesList", version: "2.0.32" };
function dataYandexComputeInstanceGroupInstanceTemplateSchedulingPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexComputeInstanceGroupInstanceTemplateSchedulingPolicyToTerraform = dataYandexComputeInstanceGroupInstanceTemplateSchedulingPolicyToTerraform;
class DataYandexComputeInstanceGroupInstanceTemplateSchedulingPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // preemptible - computed: true, optional: false, required: false
    get preemptible() {
        return this.getBooleanAttribute('preemptible');
    }
}
exports.DataYandexComputeInstanceGroupInstanceTemplateSchedulingPolicyOutputReference = DataYandexComputeInstanceGroupInstanceTemplateSchedulingPolicyOutputReference;
_8 = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupInstanceTemplateSchedulingPolicyOutputReference[_8] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupInstanceTemplateSchedulingPolicyOutputReference", version: "2.0.32" };
class DataYandexComputeInstanceGroupInstanceTemplateSchedulingPolicyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexComputeInstanceGroupInstanceTemplateSchedulingPolicyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexComputeInstanceGroupInstanceTemplateSchedulingPolicyList = DataYandexComputeInstanceGroupInstanceTemplateSchedulingPolicyList;
_9 = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupInstanceTemplateSchedulingPolicyList[_9] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupInstanceTemplateSchedulingPolicyList", version: "2.0.32" };
function dataYandexComputeInstanceGroupInstanceTemplateSecondaryDiskInitializeParamsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexComputeInstanceGroupInstanceTemplateSecondaryDiskInitializeParamsToTerraform = dataYandexComputeInstanceGroupInstanceTemplateSecondaryDiskInitializeParamsToTerraform;
class DataYandexComputeInstanceGroupInstanceTemplateSecondaryDiskInitializeParamsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    // image_id - computed: true, optional: false, required: false
    get imageId() {
        return this.getStringAttribute('image_id');
    }
    // size - computed: true, optional: false, required: false
    get size() {
        return this.getNumberAttribute('size');
    }
    // snapshot_id - computed: true, optional: false, required: false
    get snapshotId() {
        return this.getStringAttribute('snapshot_id');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataYandexComputeInstanceGroupInstanceTemplateSecondaryDiskInitializeParamsOutputReference = DataYandexComputeInstanceGroupInstanceTemplateSecondaryDiskInitializeParamsOutputReference;
_10 = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupInstanceTemplateSecondaryDiskInitializeParamsOutputReference[_10] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupInstanceTemplateSecondaryDiskInitializeParamsOutputReference", version: "2.0.32" };
class DataYandexComputeInstanceGroupInstanceTemplateSecondaryDiskInitializeParamsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexComputeInstanceGroupInstanceTemplateSecondaryDiskInitializeParamsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexComputeInstanceGroupInstanceTemplateSecondaryDiskInitializeParamsList = DataYandexComputeInstanceGroupInstanceTemplateSecondaryDiskInitializeParamsList;
_11 = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupInstanceTemplateSecondaryDiskInitializeParamsList[_11] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupInstanceTemplateSecondaryDiskInitializeParamsList", version: "2.0.32" };
function dataYandexComputeInstanceGroupInstanceTemplateSecondaryDiskToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexComputeInstanceGroupInstanceTemplateSecondaryDiskToTerraform = dataYandexComputeInstanceGroupInstanceTemplateSecondaryDiskToTerraform;
class DataYandexComputeInstanceGroupInstanceTemplateSecondaryDiskOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // initialize_params - computed: true, optional: false, required: false
        this._initializeParams = new DataYandexComputeInstanceGroupInstanceTemplateSecondaryDiskInitializeParamsList(this, "initialize_params", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // device_name - computed: true, optional: false, required: false
    get deviceName() {
        return this.getStringAttribute('device_name');
    }
    // disk_id - computed: true, optional: false, required: false
    get diskId() {
        return this.getStringAttribute('disk_id');
    }
    get initializeParams() {
        return this._initializeParams;
    }
    // mode - computed: true, optional: false, required: false
    get mode() {
        return this.getStringAttribute('mode');
    }
}
exports.DataYandexComputeInstanceGroupInstanceTemplateSecondaryDiskOutputReference = DataYandexComputeInstanceGroupInstanceTemplateSecondaryDiskOutputReference;
_12 = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupInstanceTemplateSecondaryDiskOutputReference[_12] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupInstanceTemplateSecondaryDiskOutputReference", version: "2.0.32" };
class DataYandexComputeInstanceGroupInstanceTemplateSecondaryDiskList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexComputeInstanceGroupInstanceTemplateSecondaryDiskOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexComputeInstanceGroupInstanceTemplateSecondaryDiskList = DataYandexComputeInstanceGroupInstanceTemplateSecondaryDiskList;
_13 = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupInstanceTemplateSecondaryDiskList[_13] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupInstanceTemplateSecondaryDiskList", version: "2.0.32" };
function dataYandexComputeInstanceGroupInstanceTemplateToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexComputeInstanceGroupInstanceTemplateToTerraform = dataYandexComputeInstanceGroupInstanceTemplateToTerraform;
class DataYandexComputeInstanceGroupInstanceTemplateOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // boot_disk - computed: true, optional: false, required: false
        this._bootDisk = new DataYandexComputeInstanceGroupInstanceTemplateBootDiskList(this, "boot_disk", false);
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        // metadata - computed: true, optional: false, required: false
        this._metadata = new cdktf.StringMap(this, "metadata");
        // network_interface - computed: true, optional: false, required: false
        this._networkInterface = new DataYandexComputeInstanceGroupInstanceTemplateNetworkInterfaceList(this, "network_interface", false);
        // network_settings - computed: true, optional: false, required: false
        this._networkSettings = new DataYandexComputeInstanceGroupInstanceTemplateNetworkSettingsList(this, "network_settings", false);
        // placement_policy - computed: true, optional: false, required: false
        this._placementPolicy = new DataYandexComputeInstanceGroupInstanceTemplatePlacementPolicyList(this, "placement_policy", false);
        // resources - computed: true, optional: false, required: false
        this._resources = new DataYandexComputeInstanceGroupInstanceTemplateResourcesList(this, "resources", false);
        // scheduling_policy - computed: true, optional: false, required: false
        this._schedulingPolicy = new DataYandexComputeInstanceGroupInstanceTemplateSchedulingPolicyList(this, "scheduling_policy", false);
        // secondary_disk - computed: true, optional: false, required: false
        this._secondaryDisk = new DataYandexComputeInstanceGroupInstanceTemplateSecondaryDiskList(this, "secondary_disk", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get bootDisk() {
        return this._bootDisk;
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    // hostname - computed: true, optional: false, required: false
    get hostname() {
        return this.getStringAttribute('hostname');
    }
    get labels() {
        return this._labels;
    }
    get metadata() {
        return this._metadata;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get networkInterface() {
        return this._networkInterface;
    }
    get networkSettings() {
        return this._networkSettings;
    }
    get placementPolicy() {
        return this._placementPolicy;
    }
    // platform_id - computed: true, optional: false, required: false
    get platformId() {
        return this.getStringAttribute('platform_id');
    }
    get resources() {
        return this._resources;
    }
    get schedulingPolicy() {
        return this._schedulingPolicy;
    }
    get secondaryDisk() {
        return this._secondaryDisk;
    }
    // service_account_id - computed: true, optional: false, required: false
    get serviceAccountId() {
        return this.getStringAttribute('service_account_id');
    }
}
exports.DataYandexComputeInstanceGroupInstanceTemplateOutputReference = DataYandexComputeInstanceGroupInstanceTemplateOutputReference;
_14 = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupInstanceTemplateOutputReference[_14] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupInstanceTemplateOutputReference", version: "2.0.32" };
class DataYandexComputeInstanceGroupInstanceTemplateList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexComputeInstanceGroupInstanceTemplateOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexComputeInstanceGroupInstanceTemplateList = DataYandexComputeInstanceGroupInstanceTemplateList;
_15 = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupInstanceTemplateList[_15] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupInstanceTemplateList", version: "2.0.32" };
function dataYandexComputeInstanceGroupInstancesNetworkInterfaceToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexComputeInstanceGroupInstancesNetworkInterfaceToTerraform = dataYandexComputeInstanceGroupInstancesNetworkInterfaceToTerraform;
class DataYandexComputeInstanceGroupInstancesNetworkInterfaceOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // index - computed: true, optional: false, required: false
    get index() {
        return this.getNumberAttribute('index');
    }
    // ip_address - computed: true, optional: false, required: false
    get ipAddress() {
        return this.getStringAttribute('ip_address');
    }
    // ipv4 - computed: true, optional: false, required: false
    get ipv4() {
        return this.getBooleanAttribute('ipv4');
    }
    // ipv6 - computed: true, optional: false, required: false
    get ipv6() {
        return this.getBooleanAttribute('ipv6');
    }
    // ipv6_address - computed: true, optional: false, required: false
    get ipv6Address() {
        return this.getStringAttribute('ipv6_address');
    }
    // mac_address - computed: true, optional: false, required: false
    get macAddress() {
        return this.getStringAttribute('mac_address');
    }
    // nat - computed: true, optional: false, required: false
    get nat() {
        return this.getBooleanAttribute('nat');
    }
    // nat_ip_address - computed: true, optional: false, required: false
    get natIpAddress() {
        return this.getStringAttribute('nat_ip_address');
    }
    // nat_ip_version - computed: true, optional: false, required: false
    get natIpVersion() {
        return this.getStringAttribute('nat_ip_version');
    }
    // subnet_id - computed: true, optional: false, required: false
    get subnetId() {
        return this.getStringAttribute('subnet_id');
    }
}
exports.DataYandexComputeInstanceGroupInstancesNetworkInterfaceOutputReference = DataYandexComputeInstanceGroupInstancesNetworkInterfaceOutputReference;
_16 = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupInstancesNetworkInterfaceOutputReference[_16] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupInstancesNetworkInterfaceOutputReference", version: "2.0.32" };
class DataYandexComputeInstanceGroupInstancesNetworkInterfaceList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexComputeInstanceGroupInstancesNetworkInterfaceOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexComputeInstanceGroupInstancesNetworkInterfaceList = DataYandexComputeInstanceGroupInstancesNetworkInterfaceList;
_17 = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupInstancesNetworkInterfaceList[_17] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupInstancesNetworkInterfaceList", version: "2.0.32" };
function dataYandexComputeInstanceGroupInstancesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexComputeInstanceGroupInstancesToTerraform = dataYandexComputeInstanceGroupInstancesToTerraform;
class DataYandexComputeInstanceGroupInstancesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // network_interface - computed: true, optional: false, required: false
        this._networkInterface = new DataYandexComputeInstanceGroupInstancesNetworkInterfaceList(this, "network_interface", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // fqdn - computed: true, optional: false, required: false
    get fqdn() {
        return this.getStringAttribute('fqdn');
    }
    // instance_id - computed: true, optional: false, required: false
    get instanceId() {
        return this.getStringAttribute('instance_id');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get networkInterface() {
        return this._networkInterface;
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    // status_changed_at - computed: true, optional: false, required: false
    get statusChangedAt() {
        return this.getStringAttribute('status_changed_at');
    }
    // status_message - computed: true, optional: false, required: false
    get statusMessage() {
        return this.getStringAttribute('status_message');
    }
    // zone_id - computed: true, optional: false, required: false
    get zoneId() {
        return this.getStringAttribute('zone_id');
    }
}
exports.DataYandexComputeInstanceGroupInstancesOutputReference = DataYandexComputeInstanceGroupInstancesOutputReference;
_18 = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupInstancesOutputReference[_18] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupInstancesOutputReference", version: "2.0.32" };
class DataYandexComputeInstanceGroupInstancesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexComputeInstanceGroupInstancesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexComputeInstanceGroupInstancesList = DataYandexComputeInstanceGroupInstancesList;
_19 = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupInstancesList[_19] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupInstancesList", version: "2.0.32" };
function dataYandexComputeInstanceGroupLoadBalancerToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexComputeInstanceGroupLoadBalancerToTerraform = dataYandexComputeInstanceGroupLoadBalancerToTerraform;
class DataYandexComputeInstanceGroupLoadBalancerOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // target_group_labels - computed: true, optional: false, required: false
        this._targetGroupLabels = new cdktf.StringMap(this, "target_group_labels");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // max_opening_traffic_duration - computed: true, optional: false, required: false
    get maxOpeningTrafficDuration() {
        return this.getNumberAttribute('max_opening_traffic_duration');
    }
    // status_message - computed: true, optional: false, required: false
    get statusMessage() {
        return this.getStringAttribute('status_message');
    }
    // target_group_description - computed: true, optional: false, required: false
    get targetGroupDescription() {
        return this.getStringAttribute('target_group_description');
    }
    // target_group_id - computed: true, optional: false, required: false
    get targetGroupId() {
        return this.getStringAttribute('target_group_id');
    }
    get targetGroupLabels() {
        return this._targetGroupLabels;
    }
    // target_group_name - computed: true, optional: false, required: false
    get targetGroupName() {
        return this.getStringAttribute('target_group_name');
    }
}
exports.DataYandexComputeInstanceGroupLoadBalancerOutputReference = DataYandexComputeInstanceGroupLoadBalancerOutputReference;
_20 = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupLoadBalancerOutputReference[_20] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupLoadBalancerOutputReference", version: "2.0.32" };
class DataYandexComputeInstanceGroupLoadBalancerList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexComputeInstanceGroupLoadBalancerOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexComputeInstanceGroupLoadBalancerList = DataYandexComputeInstanceGroupLoadBalancerList;
_21 = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupLoadBalancerList[_21] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupLoadBalancerList", version: "2.0.32" };
function dataYandexComputeInstanceGroupLoadBalancerStateToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexComputeInstanceGroupLoadBalancerStateToTerraform = dataYandexComputeInstanceGroupLoadBalancerStateToTerraform;
class DataYandexComputeInstanceGroupLoadBalancerStateOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // status_message - computed: true, optional: false, required: false
    get statusMessage() {
        return this.getStringAttribute('status_message');
    }
    // target_group_id - computed: true, optional: false, required: false
    get targetGroupId() {
        return this.getStringAttribute('target_group_id');
    }
}
exports.DataYandexComputeInstanceGroupLoadBalancerStateOutputReference = DataYandexComputeInstanceGroupLoadBalancerStateOutputReference;
_22 = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupLoadBalancerStateOutputReference[_22] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupLoadBalancerStateOutputReference", version: "2.0.32" };
class DataYandexComputeInstanceGroupLoadBalancerStateList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexComputeInstanceGroupLoadBalancerStateOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexComputeInstanceGroupLoadBalancerStateList = DataYandexComputeInstanceGroupLoadBalancerStateList;
_23 = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupLoadBalancerStateList[_23] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupLoadBalancerStateList", version: "2.0.32" };
function dataYandexComputeInstanceGroupScalePolicyAutoScaleCustomRuleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexComputeInstanceGroupScalePolicyAutoScaleCustomRuleToTerraform = dataYandexComputeInstanceGroupScalePolicyAutoScaleCustomRuleToTerraform;
class DataYandexComputeInstanceGroupScalePolicyAutoScaleCustomRuleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // folder_id - computed: true, optional: false, required: false
    get folderId() {
        return this.getStringAttribute('folder_id');
    }
    get labels() {
        return this._labels;
    }
    // metric_name - computed: true, optional: false, required: false
    get metricName() {
        return this.getStringAttribute('metric_name');
    }
    // metric_type - computed: true, optional: false, required: false
    get metricType() {
        return this.getStringAttribute('metric_type');
    }
    // rule_type - computed: true, optional: false, required: false
    get ruleType() {
        return this.getStringAttribute('rule_type');
    }
    // service - computed: true, optional: false, required: false
    get service() {
        return this.getStringAttribute('service');
    }
    // target - computed: true, optional: false, required: false
    get target() {
        return this.getNumberAttribute('target');
    }
}
exports.DataYandexComputeInstanceGroupScalePolicyAutoScaleCustomRuleOutputReference = DataYandexComputeInstanceGroupScalePolicyAutoScaleCustomRuleOutputReference;
_24 = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupScalePolicyAutoScaleCustomRuleOutputReference[_24] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupScalePolicyAutoScaleCustomRuleOutputReference", version: "2.0.32" };
class DataYandexComputeInstanceGroupScalePolicyAutoScaleCustomRuleList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexComputeInstanceGroupScalePolicyAutoScaleCustomRuleOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexComputeInstanceGroupScalePolicyAutoScaleCustomRuleList = DataYandexComputeInstanceGroupScalePolicyAutoScaleCustomRuleList;
_25 = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupScalePolicyAutoScaleCustomRuleList[_25] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupScalePolicyAutoScaleCustomRuleList", version: "2.0.32" };
function dataYandexComputeInstanceGroupScalePolicyAutoScaleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexComputeInstanceGroupScalePolicyAutoScaleToTerraform = dataYandexComputeInstanceGroupScalePolicyAutoScaleToTerraform;
class DataYandexComputeInstanceGroupScalePolicyAutoScaleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // custom_rule - computed: true, optional: false, required: false
        this._customRule = new DataYandexComputeInstanceGroupScalePolicyAutoScaleCustomRuleList(this, "custom_rule", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // cpu_utilization_target - computed: true, optional: false, required: false
    get cpuUtilizationTarget() {
        return this.getNumberAttribute('cpu_utilization_target');
    }
    get customRule() {
        return this._customRule;
    }
    // initial_size - computed: true, optional: false, required: false
    get initialSize() {
        return this.getNumberAttribute('initial_size');
    }
    // max_size - computed: true, optional: false, required: false
    get maxSize() {
        return this.getNumberAttribute('max_size');
    }
    // measurement_duration - computed: true, optional: false, required: false
    get measurementDuration() {
        return this.getNumberAttribute('measurement_duration');
    }
    // min_zone_size - computed: true, optional: false, required: false
    get minZoneSize() {
        return this.getNumberAttribute('min_zone_size');
    }
    // stabilization_duration - computed: true, optional: false, required: false
    get stabilizationDuration() {
        return this.getNumberAttribute('stabilization_duration');
    }
    // warmup_duration - computed: true, optional: false, required: false
    get warmupDuration() {
        return this.getNumberAttribute('warmup_duration');
    }
}
exports.DataYandexComputeInstanceGroupScalePolicyAutoScaleOutputReference = DataYandexComputeInstanceGroupScalePolicyAutoScaleOutputReference;
_26 = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupScalePolicyAutoScaleOutputReference[_26] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupScalePolicyAutoScaleOutputReference", version: "2.0.32" };
class DataYandexComputeInstanceGroupScalePolicyAutoScaleList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexComputeInstanceGroupScalePolicyAutoScaleOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexComputeInstanceGroupScalePolicyAutoScaleList = DataYandexComputeInstanceGroupScalePolicyAutoScaleList;
_27 = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupScalePolicyAutoScaleList[_27] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupScalePolicyAutoScaleList", version: "2.0.32" };
function dataYandexComputeInstanceGroupScalePolicyFixedScaleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexComputeInstanceGroupScalePolicyFixedScaleToTerraform = dataYandexComputeInstanceGroupScalePolicyFixedScaleToTerraform;
class DataYandexComputeInstanceGroupScalePolicyFixedScaleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // size - computed: true, optional: false, required: false
    get size() {
        return this.getNumberAttribute('size');
    }
}
exports.DataYandexComputeInstanceGroupScalePolicyFixedScaleOutputReference = DataYandexComputeInstanceGroupScalePolicyFixedScaleOutputReference;
_28 = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupScalePolicyFixedScaleOutputReference[_28] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupScalePolicyFixedScaleOutputReference", version: "2.0.32" };
class DataYandexComputeInstanceGroupScalePolicyFixedScaleList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexComputeInstanceGroupScalePolicyFixedScaleOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexComputeInstanceGroupScalePolicyFixedScaleList = DataYandexComputeInstanceGroupScalePolicyFixedScaleList;
_29 = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupScalePolicyFixedScaleList[_29] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupScalePolicyFixedScaleList", version: "2.0.32" };
function dataYandexComputeInstanceGroupScalePolicyTestAutoScaleCustomRuleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexComputeInstanceGroupScalePolicyTestAutoScaleCustomRuleToTerraform = dataYandexComputeInstanceGroupScalePolicyTestAutoScaleCustomRuleToTerraform;
class DataYandexComputeInstanceGroupScalePolicyTestAutoScaleCustomRuleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // folder_id - computed: true, optional: false, required: false
    get folderId() {
        return this.getStringAttribute('folder_id');
    }
    get labels() {
        return this._labels;
    }
    // metric_name - computed: true, optional: false, required: false
    get metricName() {
        return this.getStringAttribute('metric_name');
    }
    // metric_type - computed: true, optional: false, required: false
    get metricType() {
        return this.getStringAttribute('metric_type');
    }
    // rule_type - computed: true, optional: false, required: false
    get ruleType() {
        return this.getStringAttribute('rule_type');
    }
    // service - computed: true, optional: false, required: false
    get service() {
        return this.getStringAttribute('service');
    }
    // target - computed: true, optional: false, required: false
    get target() {
        return this.getNumberAttribute('target');
    }
}
exports.DataYandexComputeInstanceGroupScalePolicyTestAutoScaleCustomRuleOutputReference = DataYandexComputeInstanceGroupScalePolicyTestAutoScaleCustomRuleOutputReference;
_30 = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupScalePolicyTestAutoScaleCustomRuleOutputReference[_30] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupScalePolicyTestAutoScaleCustomRuleOutputReference", version: "2.0.32" };
class DataYandexComputeInstanceGroupScalePolicyTestAutoScaleCustomRuleList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexComputeInstanceGroupScalePolicyTestAutoScaleCustomRuleOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexComputeInstanceGroupScalePolicyTestAutoScaleCustomRuleList = DataYandexComputeInstanceGroupScalePolicyTestAutoScaleCustomRuleList;
_31 = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupScalePolicyTestAutoScaleCustomRuleList[_31] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupScalePolicyTestAutoScaleCustomRuleList", version: "2.0.32" };
function dataYandexComputeInstanceGroupScalePolicyTestAutoScaleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexComputeInstanceGroupScalePolicyTestAutoScaleToTerraform = dataYandexComputeInstanceGroupScalePolicyTestAutoScaleToTerraform;
class DataYandexComputeInstanceGroupScalePolicyTestAutoScaleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // custom_rule - computed: true, optional: false, required: false
        this._customRule = new DataYandexComputeInstanceGroupScalePolicyTestAutoScaleCustomRuleList(this, "custom_rule", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // cpu_utilization_target - computed: true, optional: false, required: false
    get cpuUtilizationTarget() {
        return this.getNumberAttribute('cpu_utilization_target');
    }
    get customRule() {
        return this._customRule;
    }
    // initial_size - computed: true, optional: false, required: false
    get initialSize() {
        return this.getNumberAttribute('initial_size');
    }
    // max_size - computed: true, optional: false, required: false
    get maxSize() {
        return this.getNumberAttribute('max_size');
    }
    // measurement_duration - computed: true, optional: false, required: false
    get measurementDuration() {
        return this.getNumberAttribute('measurement_duration');
    }
    // min_zone_size - computed: true, optional: false, required: false
    get minZoneSize() {
        return this.getNumberAttribute('min_zone_size');
    }
    // stabilization_duration - computed: true, optional: false, required: false
    get stabilizationDuration() {
        return this.getNumberAttribute('stabilization_duration');
    }
    // warmup_duration - computed: true, optional: false, required: false
    get warmupDuration() {
        return this.getNumberAttribute('warmup_duration');
    }
}
exports.DataYandexComputeInstanceGroupScalePolicyTestAutoScaleOutputReference = DataYandexComputeInstanceGroupScalePolicyTestAutoScaleOutputReference;
_32 = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupScalePolicyTestAutoScaleOutputReference[_32] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupScalePolicyTestAutoScaleOutputReference", version: "2.0.32" };
class DataYandexComputeInstanceGroupScalePolicyTestAutoScaleList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexComputeInstanceGroupScalePolicyTestAutoScaleOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexComputeInstanceGroupScalePolicyTestAutoScaleList = DataYandexComputeInstanceGroupScalePolicyTestAutoScaleList;
_33 = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupScalePolicyTestAutoScaleList[_33] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupScalePolicyTestAutoScaleList", version: "2.0.32" };
function dataYandexComputeInstanceGroupScalePolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexComputeInstanceGroupScalePolicyToTerraform = dataYandexComputeInstanceGroupScalePolicyToTerraform;
class DataYandexComputeInstanceGroupScalePolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // auto_scale - computed: true, optional: false, required: false
        this._autoScale = new DataYandexComputeInstanceGroupScalePolicyAutoScaleList(this, "auto_scale", false);
        // fixed_scale - computed: true, optional: false, required: false
        this._fixedScale = new DataYandexComputeInstanceGroupScalePolicyFixedScaleList(this, "fixed_scale", false);
        // test_auto_scale - computed: true, optional: false, required: false
        this._testAutoScale = new DataYandexComputeInstanceGroupScalePolicyTestAutoScaleList(this, "test_auto_scale", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get autoScale() {
        return this._autoScale;
    }
    get fixedScale() {
        return this._fixedScale;
    }
    get testAutoScale() {
        return this._testAutoScale;
    }
}
exports.DataYandexComputeInstanceGroupScalePolicyOutputReference = DataYandexComputeInstanceGroupScalePolicyOutputReference;
_34 = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupScalePolicyOutputReference[_34] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupScalePolicyOutputReference", version: "2.0.32" };
class DataYandexComputeInstanceGroupScalePolicyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexComputeInstanceGroupScalePolicyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexComputeInstanceGroupScalePolicyList = DataYandexComputeInstanceGroupScalePolicyList;
_35 = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroupScalePolicyList[_35] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroupScalePolicyList", version: "2.0.32" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/d/compute_instance_group yandex_compute_instance_group}
*/
class DataYandexComputeInstanceGroup extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/d/compute_instance_group yandex_compute_instance_group} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataYandexComputeInstanceGroupConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'yandex_compute_instance_group',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ==========
        // ATTRIBUTES
        // ==========
        // allocation_policy - computed: true, optional: false, required: false
        this._allocationPolicy = new DataYandexComputeInstanceGroupAllocationPolicyList(this, "allocation_policy", false);
        // application_balancer_state - computed: true, optional: false, required: false
        this._applicationBalancerState = new DataYandexComputeInstanceGroupApplicationBalancerStateList(this, "application_balancer_state", false);
        // application_load_balancer - computed: true, optional: false, required: false
        this._applicationLoadBalancer = new DataYandexComputeInstanceGroupApplicationLoadBalancerList(this, "application_load_balancer", false);
        // deploy_policy - computed: true, optional: false, required: false
        this._deployPolicy = new DataYandexComputeInstanceGroupDeployPolicyList(this, "deploy_policy", false);
        // health_check - computed: true, optional: false, required: false
        this._healthCheck = new DataYandexComputeInstanceGroupHealthCheckList(this, "health_check", false);
        // instance_template - computed: true, optional: false, required: false
        this._instanceTemplate = new DataYandexComputeInstanceGroupInstanceTemplateList(this, "instance_template", false);
        // instances - computed: true, optional: false, required: false
        this._instances = new DataYandexComputeInstanceGroupInstancesList(this, "instances", false);
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        // load_balancer - computed: true, optional: false, required: false
        this._loadBalancer = new DataYandexComputeInstanceGroupLoadBalancerList(this, "load_balancer", false);
        // load_balancer_state - computed: true, optional: false, required: false
        this._loadBalancerState = new DataYandexComputeInstanceGroupLoadBalancerStateList(this, "load_balancer_state", false);
        // scale_policy - computed: true, optional: false, required: false
        this._scalePolicy = new DataYandexComputeInstanceGroupScalePolicyList(this, "scale_policy", false);
        // variables - computed: true, optional: false, required: false
        this._variables = new cdktf.StringMap(this, "variables");
        this._id = config.id;
        this._instanceGroupId = config.instanceGroupId;
    }
    get allocationPolicy() {
        return this._allocationPolicy;
    }
    get applicationBalancerState() {
        return this._applicationBalancerState;
    }
    get applicationLoadBalancer() {
        return this._applicationLoadBalancer;
    }
    // created_at - computed: true, optional: false, required: false
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    // deletion_protection - computed: true, optional: false, required: false
    get deletionProtection() {
        return this.getBooleanAttribute('deletion_protection');
    }
    get deployPolicy() {
        return this._deployPolicy;
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    // folder_id - computed: true, optional: false, required: false
    get folderId() {
        return this.getStringAttribute('folder_id');
    }
    get healthCheck() {
        return this._healthCheck;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get instanceGroupId() {
        return this.getStringAttribute('instance_group_id');
    }
    set instanceGroupId(value) {
        this._instanceGroupId = value;
    }
    // Temporarily expose input value. Use with caution.
    get instanceGroupIdInput() {
        return this._instanceGroupId;
    }
    get instanceTemplate() {
        return this._instanceTemplate;
    }
    get instances() {
        return this._instances;
    }
    get labels() {
        return this._labels;
    }
    get loadBalancer() {
        return this._loadBalancer;
    }
    get loadBalancerState() {
        return this._loadBalancerState;
    }
    // max_checking_health_duration - computed: true, optional: false, required: false
    get maxCheckingHealthDuration() {
        return this.getNumberAttribute('max_checking_health_duration');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get scalePolicy() {
        return this._scalePolicy;
    }
    // service_account_id - computed: true, optional: false, required: false
    get serviceAccountId() {
        return this.getStringAttribute('service_account_id');
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    get variables() {
        return this._variables;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
            instance_group_id: cdktf.stringToTerraform(this._instanceGroupId),
        };
    }
}
exports.DataYandexComputeInstanceGroup = DataYandexComputeInstanceGroup;
_36 = JSII_RTTI_SYMBOL_1;
DataYandexComputeInstanceGroup[_36] = { fqn: "@cdktf/provider-yandex.DataYandexComputeInstanceGroup", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
DataYandexComputeInstanceGroup.tfResourceType = "yandex_compute_instance_group";
