// @cdktf/provider-yandex/lib/data-yandex-cdn-origin-group.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataYandexCdnOriginGroup = exports.DataYandexCdnOriginGroupOriginList = exports.DataYandexCdnOriginGroupOriginOutputReference = exports.dataYandexCdnOriginGroupOriginToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataYandexCdnOriginGroupOriginToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexCdnOriginGroupOriginToTerraform = dataYandexCdnOriginGroupOriginToTerraform;
class DataYandexCdnOriginGroupOriginOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // backup - computed: true, optional: false, required: false
    get backup() {
        return this.getBooleanAttribute('backup');
    }
    // enabled - computed: true, optional: false, required: false
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    // origin_group_id - computed: true, optional: false, required: false
    get originGroupId() {
        return this.getNumberAttribute('origin_group_id');
    }
    // source - computed: true, optional: false, required: false
    get source() {
        return this.getStringAttribute('source');
    }
}
exports.DataYandexCdnOriginGroupOriginOutputReference = DataYandexCdnOriginGroupOriginOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataYandexCdnOriginGroupOriginOutputReference[_a] = { fqn: "@cdktf/provider-yandex.DataYandexCdnOriginGroupOriginOutputReference", version: "2.0.32" };
class DataYandexCdnOriginGroupOriginList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexCdnOriginGroupOriginOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexCdnOriginGroupOriginList = DataYandexCdnOriginGroupOriginList;
_b = JSII_RTTI_SYMBOL_1;
DataYandexCdnOriginGroupOriginList[_b] = { fqn: "@cdktf/provider-yandex.DataYandexCdnOriginGroupOriginList", version: "2.0.32" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/d/cdn_origin_group yandex_cdn_origin_group}
*/
class DataYandexCdnOriginGroup extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/d/cdn_origin_group yandex_cdn_origin_group} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataYandexCdnOriginGroupConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'yandex_cdn_origin_group',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // origin - computed: true, optional: false, required: false
        this._origin = new DataYandexCdnOriginGroupOriginList(this, "origin", true);
        this._folderId = config.folderId;
        this._id = config.id;
        this._name = config.name;
        this._originGroupId = config.originGroupId;
    }
    get folderId() {
        return this.getStringAttribute('folder_id');
    }
    set folderId(value) {
        this._folderId = value;
    }
    resetFolderId() {
        this._folderId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get folderIdInput() {
        return this._folderId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get origin() {
        return this._origin;
    }
    get originGroupId() {
        return this.getNumberAttribute('origin_group_id');
    }
    set originGroupId(value) {
        this._originGroupId = value;
    }
    resetOriginGroupId() {
        this._originGroupId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get originGroupIdInput() {
        return this._originGroupId;
    }
    // use_next - computed: true, optional: false, required: false
    get useNext() {
        return this.getBooleanAttribute('use_next');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            folder_id: cdktf.stringToTerraform(this._folderId),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            origin_group_id: cdktf.numberToTerraform(this._originGroupId),
        };
    }
}
exports.DataYandexCdnOriginGroup = DataYandexCdnOriginGroup;
_c = JSII_RTTI_SYMBOL_1;
DataYandexCdnOriginGroup[_c] = { fqn: "@cdktf/provider-yandex.DataYandexCdnOriginGroup", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
DataYandexCdnOriginGroup.tfResourceType = "yandex_cdn_origin_group";
