// @cdktf/provider-yandex/lib/data-yandex-ydb-database-serverless.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataYandexYdbDatabaseServerless = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/d/ydb_database_serverless yandex_ydb_database_serverless}
*/
class DataYandexYdbDatabaseServerless extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/d/ydb_database_serverless yandex_ydb_database_serverless} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataYandexYdbDatabaseServerlessConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'yandex_ydb_database_serverless',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        this._databaseId = config.databaseId;
        this._folderId = config.folderId;
        this._id = config.id;
        this._name = config.name;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // created_at - computed: true, optional: false, required: false
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    get databaseId() {
        return this.getStringAttribute('database_id');
    }
    set databaseId(value) {
        this._databaseId = value;
    }
    resetDatabaseId() {
        this._databaseId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get databaseIdInput() {
        return this._databaseId;
    }
    // database_path - computed: true, optional: false, required: false
    get databasePath() {
        return this.getStringAttribute('database_path');
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    // document_api_endpoint - computed: true, optional: false, required: false
    get documentApiEndpoint() {
        return this.getStringAttribute('document_api_endpoint');
    }
    get folderId() {
        return this.getStringAttribute('folder_id');
    }
    set folderId(value) {
        this._folderId = value;
    }
    resetFolderId() {
        this._folderId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get folderIdInput() {
        return this._folderId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this._labels;
    }
    // location_id - computed: true, optional: false, required: false
    get locationId() {
        return this.getStringAttribute('location_id');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    // tls_enabled - computed: true, optional: false, required: false
    get tlsEnabled() {
        return this.getBooleanAttribute('tls_enabled');
    }
    // ydb_api_endpoint - computed: true, optional: false, required: false
    get ydbApiEndpoint() {
        return this.getStringAttribute('ydb_api_endpoint');
    }
    // ydb_full_endpoint - computed: true, optional: false, required: false
    get ydbFullEndpoint() {
        return this.getStringAttribute('ydb_full_endpoint');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            database_id: cdktf.stringToTerraform(this._databaseId),
            folder_id: cdktf.stringToTerraform(this._folderId),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
        };
    }
}
exports.DataYandexYdbDatabaseServerless = DataYandexYdbDatabaseServerless;
_a = JSII_RTTI_SYMBOL_1;
DataYandexYdbDatabaseServerless[_a] = { fqn: "@cdktf/provider-yandex.DataYandexYdbDatabaseServerless", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
DataYandexYdbDatabaseServerless.tfResourceType = "yandex_ydb_database_serverless";
