// @cdktf/provider-yandex/lib/data-yandex-mdb-kafka-topic.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataYandexMdbKafkaTopic = exports.DataYandexMdbKafkaTopicTopicConfigList = exports.DataYandexMdbKafkaTopicTopicConfigOutputReference = exports.dataYandexMdbKafkaTopicTopicConfigToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataYandexMdbKafkaTopicTopicConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexMdbKafkaTopicTopicConfigToTerraform = dataYandexMdbKafkaTopicTopicConfigToTerraform;
class DataYandexMdbKafkaTopicTopicConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // cleanup_policy - computed: true, optional: false, required: false
    get cleanupPolicy() {
        return this.getStringAttribute('cleanup_policy');
    }
    // compression_type - computed: true, optional: false, required: false
    get compressionType() {
        return this.getStringAttribute('compression_type');
    }
    // delete_retention_ms - computed: true, optional: false, required: false
    get deleteRetentionMs() {
        return this.getStringAttribute('delete_retention_ms');
    }
    // file_delete_delay_ms - computed: true, optional: false, required: false
    get fileDeleteDelayMs() {
        return this.getStringAttribute('file_delete_delay_ms');
    }
    // flush_messages - computed: true, optional: false, required: false
    get flushMessages() {
        return this.getStringAttribute('flush_messages');
    }
    // flush_ms - computed: true, optional: false, required: false
    get flushMs() {
        return this.getStringAttribute('flush_ms');
    }
    // max_message_bytes - computed: true, optional: false, required: false
    get maxMessageBytes() {
        return this.getStringAttribute('max_message_bytes');
    }
    // min_compaction_lag_ms - computed: true, optional: false, required: false
    get minCompactionLagMs() {
        return this.getStringAttribute('min_compaction_lag_ms');
    }
    // min_insync_replicas - computed: true, optional: false, required: false
    get minInsyncReplicas() {
        return this.getStringAttribute('min_insync_replicas');
    }
    // preallocate - computed: true, optional: false, required: false
    get preallocate() {
        return this.getBooleanAttribute('preallocate');
    }
    // retention_bytes - computed: true, optional: false, required: false
    get retentionBytes() {
        return this.getStringAttribute('retention_bytes');
    }
    // retention_ms - computed: true, optional: false, required: false
    get retentionMs() {
        return this.getStringAttribute('retention_ms');
    }
    // segment_bytes - computed: true, optional: false, required: false
    get segmentBytes() {
        return this.getStringAttribute('segment_bytes');
    }
}
exports.DataYandexMdbKafkaTopicTopicConfigOutputReference = DataYandexMdbKafkaTopicTopicConfigOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataYandexMdbKafkaTopicTopicConfigOutputReference[_a] = { fqn: "@cdktf/provider-yandex.DataYandexMdbKafkaTopicTopicConfigOutputReference", version: "2.0.32" };
class DataYandexMdbKafkaTopicTopicConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbKafkaTopicTopicConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbKafkaTopicTopicConfigList = DataYandexMdbKafkaTopicTopicConfigList;
_b = JSII_RTTI_SYMBOL_1;
DataYandexMdbKafkaTopicTopicConfigList[_b] = { fqn: "@cdktf/provider-yandex.DataYandexMdbKafkaTopicTopicConfigList", version: "2.0.32" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/d/mdb_kafka_topic yandex_mdb_kafka_topic}
*/
class DataYandexMdbKafkaTopic extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/d/mdb_kafka_topic yandex_mdb_kafka_topic} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataYandexMdbKafkaTopicConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'yandex_mdb_kafka_topic',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // topic_config - computed: true, optional: false, required: false
        this._topicConfig = new DataYandexMdbKafkaTopicTopicConfigList(this, "topic_config", false);
        this._clusterId = config.clusterId;
        this._id = config.id;
        this._name = config.name;
    }
    get clusterId() {
        return this.getStringAttribute('cluster_id');
    }
    set clusterId(value) {
        this._clusterId = value;
    }
    // Temporarily expose input value. Use with caution.
    get clusterIdInput() {
        return this._clusterId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // partitions - computed: true, optional: false, required: false
    get partitions() {
        return this.getNumberAttribute('partitions');
    }
    // replication_factor - computed: true, optional: false, required: false
    get replicationFactor() {
        return this.getNumberAttribute('replication_factor');
    }
    get topicConfig() {
        return this._topicConfig;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            cluster_id: cdktf.stringToTerraform(this._clusterId),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
        };
    }
}
exports.DataYandexMdbKafkaTopic = DataYandexMdbKafkaTopic;
_c = JSII_RTTI_SYMBOL_1;
DataYandexMdbKafkaTopic[_c] = { fqn: "@cdktf/provider-yandex.DataYandexMdbKafkaTopic", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
DataYandexMdbKafkaTopic.tfResourceType = "yandex_mdb_kafka_topic";
