// @cdktf/provider-yandex/lib/data-yandex-mdb-mysql-user.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataYandexMdbMysqlUser = exports.DataYandexMdbMysqlUserPermissionList = exports.DataYandexMdbMysqlUserPermissionOutputReference = exports.dataYandexMdbMysqlUserPermissionToTerraform = exports.DataYandexMdbMysqlUserConnectionLimitsList = exports.DataYandexMdbMysqlUserConnectionLimitsOutputReference = exports.dataYandexMdbMysqlUserConnectionLimitsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataYandexMdbMysqlUserConnectionLimitsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexMdbMysqlUserConnectionLimitsToTerraform = dataYandexMdbMysqlUserConnectionLimitsToTerraform;
class DataYandexMdbMysqlUserConnectionLimitsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // max_connections_per_hour - computed: true, optional: false, required: false
    get maxConnectionsPerHour() {
        return this.getNumberAttribute('max_connections_per_hour');
    }
    // max_questions_per_hour - computed: true, optional: false, required: false
    get maxQuestionsPerHour() {
        return this.getNumberAttribute('max_questions_per_hour');
    }
    // max_updates_per_hour - computed: true, optional: false, required: false
    get maxUpdatesPerHour() {
        return this.getNumberAttribute('max_updates_per_hour');
    }
    // max_user_connections - computed: true, optional: false, required: false
    get maxUserConnections() {
        return this.getNumberAttribute('max_user_connections');
    }
}
exports.DataYandexMdbMysqlUserConnectionLimitsOutputReference = DataYandexMdbMysqlUserConnectionLimitsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataYandexMdbMysqlUserConnectionLimitsOutputReference[_a] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMysqlUserConnectionLimitsOutputReference", version: "2.0.32" };
class DataYandexMdbMysqlUserConnectionLimitsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbMysqlUserConnectionLimitsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbMysqlUserConnectionLimitsList = DataYandexMdbMysqlUserConnectionLimitsList;
_b = JSII_RTTI_SYMBOL_1;
DataYandexMdbMysqlUserConnectionLimitsList[_b] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMysqlUserConnectionLimitsList", version: "2.0.32" };
function dataYandexMdbMysqlUserPermissionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        roles: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.roles),
    };
}
exports.dataYandexMdbMysqlUserPermissionToTerraform = dataYandexMdbMysqlUserPermissionToTerraform;
class DataYandexMdbMysqlUserPermissionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._roles !== undefined) {
            hasAnyValues = true;
            internalValueResult.roles = this._roles;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._roles = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._roles = value.roles;
        }
    }
    // database_name - computed: true, optional: false, required: false
    get databaseName() {
        return this.getStringAttribute('database_name');
    }
    get roles() {
        return this.getListAttribute('roles');
    }
    set roles(value) {
        this._roles = value;
    }
    resetRoles() {
        this._roles = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get rolesInput() {
        return this._roles;
    }
}
exports.DataYandexMdbMysqlUserPermissionOutputReference = DataYandexMdbMysqlUserPermissionOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataYandexMdbMysqlUserPermissionOutputReference[_c] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMysqlUserPermissionOutputReference", version: "2.0.32" };
class DataYandexMdbMysqlUserPermissionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbMysqlUserPermissionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbMysqlUserPermissionList = DataYandexMdbMysqlUserPermissionList;
_d = JSII_RTTI_SYMBOL_1;
DataYandexMdbMysqlUserPermissionList[_d] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMysqlUserPermissionList", version: "2.0.32" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/d/mdb_mysql_user yandex_mdb_mysql_user}
*/
class DataYandexMdbMysqlUser extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/d/mdb_mysql_user yandex_mdb_mysql_user} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataYandexMdbMysqlUserConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'yandex_mdb_mysql_user',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // connection_limits - computed: true, optional: false, required: false
        this._connectionLimits = new DataYandexMdbMysqlUserConnectionLimitsList(this, "connection_limits", false);
        // permission - computed: false, optional: true, required: false
        this._permission = new DataYandexMdbMysqlUserPermissionList(this, "permission", true);
        this._clusterId = config.clusterId;
        this._id = config.id;
        this._name = config.name;
        this._permission.internalValue = config.permission;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // authentication_plugin - computed: true, optional: false, required: false
    get authenticationPlugin() {
        return this.getStringAttribute('authentication_plugin');
    }
    get clusterId() {
        return this.getStringAttribute('cluster_id');
    }
    set clusterId(value) {
        this._clusterId = value;
    }
    // Temporarily expose input value. Use with caution.
    get clusterIdInput() {
        return this._clusterId;
    }
    get connectionLimits() {
        return this._connectionLimits;
    }
    // global_permissions - computed: true, optional: false, required: false
    get globalPermissions() {
        return cdktf.Fn.tolist(this.getListAttribute('global_permissions'));
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get permission() {
        return this._permission;
    }
    putPermission(value) {
        this._permission.internalValue = value;
    }
    resetPermission() {
        this._permission.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get permissionInput() {
        return this._permission.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            cluster_id: cdktf.stringToTerraform(this._clusterId),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            permission: cdktf.listMapper(dataYandexMdbMysqlUserPermissionToTerraform, true)(this._permission.internalValue),
        };
    }
}
exports.DataYandexMdbMysqlUser = DataYandexMdbMysqlUser;
_e = JSII_RTTI_SYMBOL_1;
DataYandexMdbMysqlUser[_e] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMysqlUser", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
DataYandexMdbMysqlUser.tfResourceType = "yandex_mdb_mysql_user";
