// @cdktf/provider-yandex/lib/data-yandex-vpc-subnet.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataYandexVpcSubnet = exports.DataYandexVpcSubnetDhcpOptionsList = exports.DataYandexVpcSubnetDhcpOptionsOutputReference = exports.dataYandexVpcSubnetDhcpOptionsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataYandexVpcSubnetDhcpOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexVpcSubnetDhcpOptionsToTerraform = dataYandexVpcSubnetDhcpOptionsToTerraform;
class DataYandexVpcSubnetDhcpOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // domain_name - computed: true, optional: false, required: false
    get domainName() {
        return this.getStringAttribute('domain_name');
    }
    // domain_name_servers - computed: true, optional: false, required: false
    get domainNameServers() {
        return this.getListAttribute('domain_name_servers');
    }
    // ntp_servers - computed: true, optional: false, required: false
    get ntpServers() {
        return this.getListAttribute('ntp_servers');
    }
}
exports.DataYandexVpcSubnetDhcpOptionsOutputReference = DataYandexVpcSubnetDhcpOptionsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataYandexVpcSubnetDhcpOptionsOutputReference[_a] = { fqn: "@cdktf/provider-yandex.DataYandexVpcSubnetDhcpOptionsOutputReference", version: "2.0.32" };
class DataYandexVpcSubnetDhcpOptionsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexVpcSubnetDhcpOptionsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexVpcSubnetDhcpOptionsList = DataYandexVpcSubnetDhcpOptionsList;
_b = JSII_RTTI_SYMBOL_1;
DataYandexVpcSubnetDhcpOptionsList[_b] = { fqn: "@cdktf/provider-yandex.DataYandexVpcSubnetDhcpOptionsList", version: "2.0.32" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/d/vpc_subnet yandex_vpc_subnet}
*/
class DataYandexVpcSubnet extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/d/vpc_subnet yandex_vpc_subnet} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataYandexVpcSubnetConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'yandex_vpc_subnet',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // dhcp_options - computed: true, optional: false, required: false
        this._dhcpOptions = new DataYandexVpcSubnetDhcpOptionsList(this, "dhcp_options", false);
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        this._folderId = config.folderId;
        this._id = config.id;
        this._name = config.name;
        this._subnetId = config.subnetId;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // created_at - computed: true, optional: false, required: false
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    get dhcpOptions() {
        return this._dhcpOptions;
    }
    get folderId() {
        return this.getStringAttribute('folder_id');
    }
    set folderId(value) {
        this._folderId = value;
    }
    resetFolderId() {
        this._folderId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get folderIdInput() {
        return this._folderId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this._labels;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // network_id - computed: true, optional: false, required: false
    get networkId() {
        return this.getStringAttribute('network_id');
    }
    // route_table_id - computed: true, optional: false, required: false
    get routeTableId() {
        return this.getStringAttribute('route_table_id');
    }
    get subnetId() {
        return this.getStringAttribute('subnet_id');
    }
    set subnetId(value) {
        this._subnetId = value;
    }
    resetSubnetId() {
        this._subnetId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get subnetIdInput() {
        return this._subnetId;
    }
    // v4_cidr_blocks - computed: true, optional: false, required: false
    get v4CidrBlocks() {
        return this.getListAttribute('v4_cidr_blocks');
    }
    // v6_cidr_blocks - computed: true, optional: false, required: false
    get v6CidrBlocks() {
        return this.getListAttribute('v6_cidr_blocks');
    }
    // zone - computed: true, optional: false, required: false
    get zone() {
        return this.getStringAttribute('zone');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            folder_id: cdktf.stringToTerraform(this._folderId),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            subnet_id: cdktf.stringToTerraform(this._subnetId),
        };
    }
}
exports.DataYandexVpcSubnet = DataYandexVpcSubnet;
_c = JSII_RTTI_SYMBOL_1;
DataYandexVpcSubnet[_c] = { fqn: "@cdktf/provider-yandex.DataYandexVpcSubnet", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
DataYandexVpcSubnet.tfResourceType = "yandex_vpc_subnet";
