// @cdktf/provider-yandex/lib/data-yandex-mdb-postgresql-cluster.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataYandexMdbPostgresqlCluster = exports.DataYandexMdbPostgresqlClusterMaintenanceWindowList = exports.DataYandexMdbPostgresqlClusterMaintenanceWindowOutputReference = exports.dataYandexMdbPostgresqlClusterMaintenanceWindowToTerraform = exports.DataYandexMdbPostgresqlClusterHostList = exports.DataYandexMdbPostgresqlClusterHostOutputReference = exports.dataYandexMdbPostgresqlClusterHostToTerraform = exports.DataYandexMdbPostgresqlClusterConfigAList = exports.DataYandexMdbPostgresqlClusterConfigAOutputReference = exports.dataYandexMdbPostgresqlClusterConfigAToTerraform = exports.DataYandexMdbPostgresqlClusterConfigResourcesList = exports.DataYandexMdbPostgresqlClusterConfigResourcesOutputReference = exports.dataYandexMdbPostgresqlClusterConfigResourcesToTerraform = exports.DataYandexMdbPostgresqlClusterConfigPoolerConfigList = exports.DataYandexMdbPostgresqlClusterConfigPoolerConfigOutputReference = exports.dataYandexMdbPostgresqlClusterConfigPoolerConfigToTerraform = exports.DataYandexMdbPostgresqlClusterConfigPerformanceDiagnosticsList = exports.DataYandexMdbPostgresqlClusterConfigPerformanceDiagnosticsOutputReference = exports.dataYandexMdbPostgresqlClusterConfigPerformanceDiagnosticsToTerraform = exports.DataYandexMdbPostgresqlClusterConfigBackupWindowStartList = exports.DataYandexMdbPostgresqlClusterConfigBackupWindowStartOutputReference = exports.dataYandexMdbPostgresqlClusterConfigBackupWindowStartToTerraform = exports.DataYandexMdbPostgresqlClusterConfigAccessList = exports.DataYandexMdbPostgresqlClusterConfigAccessOutputReference = exports.dataYandexMdbPostgresqlClusterConfigAccessToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataYandexMdbPostgresqlClusterConfigAccessToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexMdbPostgresqlClusterConfigAccessToTerraform = dataYandexMdbPostgresqlClusterConfigAccessToTerraform;
class DataYandexMdbPostgresqlClusterConfigAccessOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // data_lens - computed: true, optional: false, required: false
    get dataLens() {
        return this.getBooleanAttribute('data_lens');
    }
    // serverless - computed: true, optional: false, required: false
    get serverless() {
        return this.getBooleanAttribute('serverless');
    }
    // web_sql - computed: true, optional: false, required: false
    get webSql() {
        return this.getBooleanAttribute('web_sql');
    }
}
exports.DataYandexMdbPostgresqlClusterConfigAccessOutputReference = DataYandexMdbPostgresqlClusterConfigAccessOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataYandexMdbPostgresqlClusterConfigAccessOutputReference[_a] = { fqn: "@cdktf/provider-yandex.DataYandexMdbPostgresqlClusterConfigAccessOutputReference", version: "2.0.32" };
class DataYandexMdbPostgresqlClusterConfigAccessList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbPostgresqlClusterConfigAccessOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbPostgresqlClusterConfigAccessList = DataYandexMdbPostgresqlClusterConfigAccessList;
_b = JSII_RTTI_SYMBOL_1;
DataYandexMdbPostgresqlClusterConfigAccessList[_b] = { fqn: "@cdktf/provider-yandex.DataYandexMdbPostgresqlClusterConfigAccessList", version: "2.0.32" };
function dataYandexMdbPostgresqlClusterConfigBackupWindowStartToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexMdbPostgresqlClusterConfigBackupWindowStartToTerraform = dataYandexMdbPostgresqlClusterConfigBackupWindowStartToTerraform;
class DataYandexMdbPostgresqlClusterConfigBackupWindowStartOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // hours - computed: true, optional: false, required: false
    get hours() {
        return this.getNumberAttribute('hours');
    }
    // minutes - computed: true, optional: false, required: false
    get minutes() {
        return this.getNumberAttribute('minutes');
    }
}
exports.DataYandexMdbPostgresqlClusterConfigBackupWindowStartOutputReference = DataYandexMdbPostgresqlClusterConfigBackupWindowStartOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataYandexMdbPostgresqlClusterConfigBackupWindowStartOutputReference[_c] = { fqn: "@cdktf/provider-yandex.DataYandexMdbPostgresqlClusterConfigBackupWindowStartOutputReference", version: "2.0.32" };
class DataYandexMdbPostgresqlClusterConfigBackupWindowStartList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbPostgresqlClusterConfigBackupWindowStartOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbPostgresqlClusterConfigBackupWindowStartList = DataYandexMdbPostgresqlClusterConfigBackupWindowStartList;
_d = JSII_RTTI_SYMBOL_1;
DataYandexMdbPostgresqlClusterConfigBackupWindowStartList[_d] = { fqn: "@cdktf/provider-yandex.DataYandexMdbPostgresqlClusterConfigBackupWindowStartList", version: "2.0.32" };
function dataYandexMdbPostgresqlClusterConfigPerformanceDiagnosticsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexMdbPostgresqlClusterConfigPerformanceDiagnosticsToTerraform = dataYandexMdbPostgresqlClusterConfigPerformanceDiagnosticsToTerraform;
class DataYandexMdbPostgresqlClusterConfigPerformanceDiagnosticsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // enabled - computed: true, optional: false, required: false
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    // sessions_sampling_interval - computed: true, optional: false, required: false
    get sessionsSamplingInterval() {
        return this.getNumberAttribute('sessions_sampling_interval');
    }
    // statements_sampling_interval - computed: true, optional: false, required: false
    get statementsSamplingInterval() {
        return this.getNumberAttribute('statements_sampling_interval');
    }
}
exports.DataYandexMdbPostgresqlClusterConfigPerformanceDiagnosticsOutputReference = DataYandexMdbPostgresqlClusterConfigPerformanceDiagnosticsOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataYandexMdbPostgresqlClusterConfigPerformanceDiagnosticsOutputReference[_e] = { fqn: "@cdktf/provider-yandex.DataYandexMdbPostgresqlClusterConfigPerformanceDiagnosticsOutputReference", version: "2.0.32" };
class DataYandexMdbPostgresqlClusterConfigPerformanceDiagnosticsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbPostgresqlClusterConfigPerformanceDiagnosticsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbPostgresqlClusterConfigPerformanceDiagnosticsList = DataYandexMdbPostgresqlClusterConfigPerformanceDiagnosticsList;
_f = JSII_RTTI_SYMBOL_1;
DataYandexMdbPostgresqlClusterConfigPerformanceDiagnosticsList[_f] = { fqn: "@cdktf/provider-yandex.DataYandexMdbPostgresqlClusterConfigPerformanceDiagnosticsList", version: "2.0.32" };
function dataYandexMdbPostgresqlClusterConfigPoolerConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexMdbPostgresqlClusterConfigPoolerConfigToTerraform = dataYandexMdbPostgresqlClusterConfigPoolerConfigToTerraform;
class DataYandexMdbPostgresqlClusterConfigPoolerConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // pool_discard - computed: true, optional: false, required: false
    get poolDiscard() {
        return this.getBooleanAttribute('pool_discard');
    }
    // pooling_mode - computed: true, optional: false, required: false
    get poolingMode() {
        return this.getStringAttribute('pooling_mode');
    }
}
exports.DataYandexMdbPostgresqlClusterConfigPoolerConfigOutputReference = DataYandexMdbPostgresqlClusterConfigPoolerConfigOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataYandexMdbPostgresqlClusterConfigPoolerConfigOutputReference[_g] = { fqn: "@cdktf/provider-yandex.DataYandexMdbPostgresqlClusterConfigPoolerConfigOutputReference", version: "2.0.32" };
class DataYandexMdbPostgresqlClusterConfigPoolerConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbPostgresqlClusterConfigPoolerConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbPostgresqlClusterConfigPoolerConfigList = DataYandexMdbPostgresqlClusterConfigPoolerConfigList;
_h = JSII_RTTI_SYMBOL_1;
DataYandexMdbPostgresqlClusterConfigPoolerConfigList[_h] = { fqn: "@cdktf/provider-yandex.DataYandexMdbPostgresqlClusterConfigPoolerConfigList", version: "2.0.32" };
function dataYandexMdbPostgresqlClusterConfigResourcesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexMdbPostgresqlClusterConfigResourcesToTerraform = dataYandexMdbPostgresqlClusterConfigResourcesToTerraform;
class DataYandexMdbPostgresqlClusterConfigResourcesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // disk_size - computed: true, optional: false, required: false
    get diskSize() {
        return this.getNumberAttribute('disk_size');
    }
    // disk_type_id - computed: true, optional: false, required: false
    get diskTypeId() {
        return this.getStringAttribute('disk_type_id');
    }
    // resource_preset_id - computed: true, optional: false, required: false
    get resourcePresetId() {
        return this.getStringAttribute('resource_preset_id');
    }
}
exports.DataYandexMdbPostgresqlClusterConfigResourcesOutputReference = DataYandexMdbPostgresqlClusterConfigResourcesOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataYandexMdbPostgresqlClusterConfigResourcesOutputReference[_j] = { fqn: "@cdktf/provider-yandex.DataYandexMdbPostgresqlClusterConfigResourcesOutputReference", version: "2.0.32" };
class DataYandexMdbPostgresqlClusterConfigResourcesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbPostgresqlClusterConfigResourcesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbPostgresqlClusterConfigResourcesList = DataYandexMdbPostgresqlClusterConfigResourcesList;
_k = JSII_RTTI_SYMBOL_1;
DataYandexMdbPostgresqlClusterConfigResourcesList[_k] = { fqn: "@cdktf/provider-yandex.DataYandexMdbPostgresqlClusterConfigResourcesList", version: "2.0.32" };
function dataYandexMdbPostgresqlClusterConfigAToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexMdbPostgresqlClusterConfigAToTerraform = dataYandexMdbPostgresqlClusterConfigAToTerraform;
class DataYandexMdbPostgresqlClusterConfigAOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // access - computed: true, optional: false, required: false
        this._access = new DataYandexMdbPostgresqlClusterConfigAccessList(this, "access", false);
        // backup_window_start - computed: true, optional: false, required: false
        this._backupWindowStart = new DataYandexMdbPostgresqlClusterConfigBackupWindowStartList(this, "backup_window_start", false);
        // performance_diagnostics - computed: true, optional: false, required: false
        this._performanceDiagnostics = new DataYandexMdbPostgresqlClusterConfigPerformanceDiagnosticsList(this, "performance_diagnostics", false);
        // pooler_config - computed: true, optional: false, required: false
        this._poolerConfig = new DataYandexMdbPostgresqlClusterConfigPoolerConfigList(this, "pooler_config", false);
        // postgresql_config - computed: true, optional: false, required: false
        this._postgresqlConfig = new cdktf.StringMap(this, "postgresql_config");
        // resources - computed: true, optional: false, required: false
        this._resources = new DataYandexMdbPostgresqlClusterConfigResourcesList(this, "resources", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get access() {
        return this._access;
    }
    // autofailover - computed: true, optional: false, required: false
    get autofailover() {
        return this.getBooleanAttribute('autofailover');
    }
    // backup_retain_period_days - computed: true, optional: false, required: false
    get backupRetainPeriodDays() {
        return this.getNumberAttribute('backup_retain_period_days');
    }
    get backupWindowStart() {
        return this._backupWindowStart;
    }
    get performanceDiagnostics() {
        return this._performanceDiagnostics;
    }
    get poolerConfig() {
        return this._poolerConfig;
    }
    get postgresqlConfig() {
        return this._postgresqlConfig;
    }
    get resources() {
        return this._resources;
    }
    // version - computed: true, optional: false, required: false
    get version() {
        return this.getStringAttribute('version');
    }
}
exports.DataYandexMdbPostgresqlClusterConfigAOutputReference = DataYandexMdbPostgresqlClusterConfigAOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataYandexMdbPostgresqlClusterConfigAOutputReference[_l] = { fqn: "@cdktf/provider-yandex.DataYandexMdbPostgresqlClusterConfigAOutputReference", version: "2.0.32" };
class DataYandexMdbPostgresqlClusterConfigAList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbPostgresqlClusterConfigAOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbPostgresqlClusterConfigAList = DataYandexMdbPostgresqlClusterConfigAList;
_m = JSII_RTTI_SYMBOL_1;
DataYandexMdbPostgresqlClusterConfigAList[_m] = { fqn: "@cdktf/provider-yandex.DataYandexMdbPostgresqlClusterConfigAList", version: "2.0.32" };
function dataYandexMdbPostgresqlClusterHostToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexMdbPostgresqlClusterHostToTerraform = dataYandexMdbPostgresqlClusterHostToTerraform;
class DataYandexMdbPostgresqlClusterHostOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // assign_public_ip - computed: true, optional: false, required: false
    get assignPublicIp() {
        return this.getBooleanAttribute('assign_public_ip');
    }
    // fqdn - computed: true, optional: false, required: false
    get fqdn() {
        return this.getStringAttribute('fqdn');
    }
    // priority - computed: true, optional: false, required: false
    get priority() {
        return this.getNumberAttribute('priority');
    }
    // replication_source - computed: true, optional: false, required: false
    get replicationSource() {
        return this.getStringAttribute('replication_source');
    }
    // role - computed: true, optional: false, required: false
    get role() {
        return this.getStringAttribute('role');
    }
    // subnet_id - computed: true, optional: false, required: false
    get subnetId() {
        return this.getStringAttribute('subnet_id');
    }
    // zone - computed: true, optional: false, required: false
    get zone() {
        return this.getStringAttribute('zone');
    }
}
exports.DataYandexMdbPostgresqlClusterHostOutputReference = DataYandexMdbPostgresqlClusterHostOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataYandexMdbPostgresqlClusterHostOutputReference[_o] = { fqn: "@cdktf/provider-yandex.DataYandexMdbPostgresqlClusterHostOutputReference", version: "2.0.32" };
class DataYandexMdbPostgresqlClusterHostList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbPostgresqlClusterHostOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbPostgresqlClusterHostList = DataYandexMdbPostgresqlClusterHostList;
_p = JSII_RTTI_SYMBOL_1;
DataYandexMdbPostgresqlClusterHostList[_p] = { fqn: "@cdktf/provider-yandex.DataYandexMdbPostgresqlClusterHostList", version: "2.0.32" };
function dataYandexMdbPostgresqlClusterMaintenanceWindowToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexMdbPostgresqlClusterMaintenanceWindowToTerraform = dataYandexMdbPostgresqlClusterMaintenanceWindowToTerraform;
class DataYandexMdbPostgresqlClusterMaintenanceWindowOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // day - computed: true, optional: false, required: false
    get day() {
        return this.getStringAttribute('day');
    }
    // hour - computed: true, optional: false, required: false
    get hour() {
        return this.getNumberAttribute('hour');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataYandexMdbPostgresqlClusterMaintenanceWindowOutputReference = DataYandexMdbPostgresqlClusterMaintenanceWindowOutputReference;
_q = JSII_RTTI_SYMBOL_1;
DataYandexMdbPostgresqlClusterMaintenanceWindowOutputReference[_q] = { fqn: "@cdktf/provider-yandex.DataYandexMdbPostgresqlClusterMaintenanceWindowOutputReference", version: "2.0.32" };
class DataYandexMdbPostgresqlClusterMaintenanceWindowList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbPostgresqlClusterMaintenanceWindowOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbPostgresqlClusterMaintenanceWindowList = DataYandexMdbPostgresqlClusterMaintenanceWindowList;
_r = JSII_RTTI_SYMBOL_1;
DataYandexMdbPostgresqlClusterMaintenanceWindowList[_r] = { fqn: "@cdktf/provider-yandex.DataYandexMdbPostgresqlClusterMaintenanceWindowList", version: "2.0.32" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/d/mdb_postgresql_cluster yandex_mdb_postgresql_cluster}
*/
class DataYandexMdbPostgresqlCluster extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/d/mdb_postgresql_cluster yandex_mdb_postgresql_cluster} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataYandexMdbPostgresqlClusterConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'yandex_mdb_postgresql_cluster',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // config - computed: true, optional: false, required: false
        this._config = new DataYandexMdbPostgresqlClusterConfigAList(this, "config", false);
        // host - computed: true, optional: false, required: false
        this._host = new DataYandexMdbPostgresqlClusterHostList(this, "host", false);
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        // maintenance_window - computed: true, optional: false, required: false
        this._maintenanceWindow = new DataYandexMdbPostgresqlClusterMaintenanceWindowList(this, "maintenance_window", false);
        this._clusterId = config.clusterId;
        this._deletionProtection = config.deletionProtection;
        this._description = config.description;
        this._folderId = config.folderId;
        this._id = config.id;
        this._name = config.name;
    }
    get clusterId() {
        return this.getStringAttribute('cluster_id');
    }
    set clusterId(value) {
        this._clusterId = value;
    }
    resetClusterId() {
        this._clusterId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clusterIdInput() {
        return this._clusterId;
    }
    get config() {
        return this._config;
    }
    // created_at - computed: true, optional: false, required: false
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    get deletionProtection() {
        return this.getBooleanAttribute('deletion_protection');
    }
    set deletionProtection(value) {
        this._deletionProtection = value;
    }
    resetDeletionProtection() {
        this._deletionProtection = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deletionProtectionInput() {
        return this._deletionProtection;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    // environment - computed: true, optional: false, required: false
    get environment() {
        return this.getStringAttribute('environment');
    }
    get folderId() {
        return this.getStringAttribute('folder_id');
    }
    set folderId(value) {
        this._folderId = value;
    }
    resetFolderId() {
        this._folderId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get folderIdInput() {
        return this._folderId;
    }
    // health - computed: true, optional: false, required: false
    get health() {
        return this.getStringAttribute('health');
    }
    get host() {
        return this._host;
    }
    // host_group_ids - computed: true, optional: false, required: false
    get hostGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('host_group_ids'));
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this._labels;
    }
    get maintenanceWindow() {
        return this._maintenanceWindow;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // network_id - computed: true, optional: false, required: false
    get networkId() {
        return this.getStringAttribute('network_id');
    }
    // security_group_ids - computed: true, optional: false, required: false
    get securityGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('security_group_ids'));
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            cluster_id: cdktf.stringToTerraform(this._clusterId),
            deletion_protection: cdktf.booleanToTerraform(this._deletionProtection),
            description: cdktf.stringToTerraform(this._description),
            folder_id: cdktf.stringToTerraform(this._folderId),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
        };
    }
}
exports.DataYandexMdbPostgresqlCluster = DataYandexMdbPostgresqlCluster;
_s = JSII_RTTI_SYMBOL_1;
DataYandexMdbPostgresqlCluster[_s] = { fqn: "@cdktf/provider-yandex.DataYandexMdbPostgresqlCluster", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
DataYandexMdbPostgresqlCluster.tfResourceType = "yandex_mdb_postgresql_cluster";
