// @cdktf/provider-yandex/lib/data-yandex-ydb-database-dedicated.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataYandexYdbDatabaseDedicated = exports.DataYandexYdbDatabaseDedicatedStorageConfigList = exports.DataYandexYdbDatabaseDedicatedStorageConfigOutputReference = exports.dataYandexYdbDatabaseDedicatedStorageConfigToTerraform = exports.DataYandexYdbDatabaseDedicatedScalePolicyList = exports.DataYandexYdbDatabaseDedicatedScalePolicyOutputReference = exports.dataYandexYdbDatabaseDedicatedScalePolicyToTerraform = exports.DataYandexYdbDatabaseDedicatedScalePolicyFixedScaleList = exports.DataYandexYdbDatabaseDedicatedScalePolicyFixedScaleOutputReference = exports.dataYandexYdbDatabaseDedicatedScalePolicyFixedScaleToTerraform = exports.DataYandexYdbDatabaseDedicatedLocationList = exports.DataYandexYdbDatabaseDedicatedLocationOutputReference = exports.dataYandexYdbDatabaseDedicatedLocationToTerraform = exports.DataYandexYdbDatabaseDedicatedLocationZoneList = exports.DataYandexYdbDatabaseDedicatedLocationZoneOutputReference = exports.dataYandexYdbDatabaseDedicatedLocationZoneToTerraform = exports.DataYandexYdbDatabaseDedicatedLocationRegionList = exports.DataYandexYdbDatabaseDedicatedLocationRegionOutputReference = exports.dataYandexYdbDatabaseDedicatedLocationRegionToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataYandexYdbDatabaseDedicatedLocationRegionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexYdbDatabaseDedicatedLocationRegionToTerraform = dataYandexYdbDatabaseDedicatedLocationRegionToTerraform;
class DataYandexYdbDatabaseDedicatedLocationRegionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
}
exports.DataYandexYdbDatabaseDedicatedLocationRegionOutputReference = DataYandexYdbDatabaseDedicatedLocationRegionOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataYandexYdbDatabaseDedicatedLocationRegionOutputReference[_a] = { fqn: "@cdktf/provider-yandex.DataYandexYdbDatabaseDedicatedLocationRegionOutputReference", version: "2.0.32" };
class DataYandexYdbDatabaseDedicatedLocationRegionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexYdbDatabaseDedicatedLocationRegionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexYdbDatabaseDedicatedLocationRegionList = DataYandexYdbDatabaseDedicatedLocationRegionList;
_b = JSII_RTTI_SYMBOL_1;
DataYandexYdbDatabaseDedicatedLocationRegionList[_b] = { fqn: "@cdktf/provider-yandex.DataYandexYdbDatabaseDedicatedLocationRegionList", version: "2.0.32" };
function dataYandexYdbDatabaseDedicatedLocationZoneToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexYdbDatabaseDedicatedLocationZoneToTerraform = dataYandexYdbDatabaseDedicatedLocationZoneToTerraform;
class DataYandexYdbDatabaseDedicatedLocationZoneOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
}
exports.DataYandexYdbDatabaseDedicatedLocationZoneOutputReference = DataYandexYdbDatabaseDedicatedLocationZoneOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataYandexYdbDatabaseDedicatedLocationZoneOutputReference[_c] = { fqn: "@cdktf/provider-yandex.DataYandexYdbDatabaseDedicatedLocationZoneOutputReference", version: "2.0.32" };
class DataYandexYdbDatabaseDedicatedLocationZoneList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexYdbDatabaseDedicatedLocationZoneOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexYdbDatabaseDedicatedLocationZoneList = DataYandexYdbDatabaseDedicatedLocationZoneList;
_d = JSII_RTTI_SYMBOL_1;
DataYandexYdbDatabaseDedicatedLocationZoneList[_d] = { fqn: "@cdktf/provider-yandex.DataYandexYdbDatabaseDedicatedLocationZoneList", version: "2.0.32" };
function dataYandexYdbDatabaseDedicatedLocationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexYdbDatabaseDedicatedLocationToTerraform = dataYandexYdbDatabaseDedicatedLocationToTerraform;
class DataYandexYdbDatabaseDedicatedLocationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // region - computed: true, optional: false, required: false
        this._region = new DataYandexYdbDatabaseDedicatedLocationRegionList(this, "region", false);
        // zone - computed: true, optional: false, required: false
        this._zone = new DataYandexYdbDatabaseDedicatedLocationZoneList(this, "zone", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get region() {
        return this._region;
    }
    get zone() {
        return this._zone;
    }
}
exports.DataYandexYdbDatabaseDedicatedLocationOutputReference = DataYandexYdbDatabaseDedicatedLocationOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataYandexYdbDatabaseDedicatedLocationOutputReference[_e] = { fqn: "@cdktf/provider-yandex.DataYandexYdbDatabaseDedicatedLocationOutputReference", version: "2.0.32" };
class DataYandexYdbDatabaseDedicatedLocationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexYdbDatabaseDedicatedLocationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexYdbDatabaseDedicatedLocationList = DataYandexYdbDatabaseDedicatedLocationList;
_f = JSII_RTTI_SYMBOL_1;
DataYandexYdbDatabaseDedicatedLocationList[_f] = { fqn: "@cdktf/provider-yandex.DataYandexYdbDatabaseDedicatedLocationList", version: "2.0.32" };
function dataYandexYdbDatabaseDedicatedScalePolicyFixedScaleToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexYdbDatabaseDedicatedScalePolicyFixedScaleToTerraform = dataYandexYdbDatabaseDedicatedScalePolicyFixedScaleToTerraform;
class DataYandexYdbDatabaseDedicatedScalePolicyFixedScaleOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // size - computed: true, optional: false, required: false
    get size() {
        return this.getNumberAttribute('size');
    }
}
exports.DataYandexYdbDatabaseDedicatedScalePolicyFixedScaleOutputReference = DataYandexYdbDatabaseDedicatedScalePolicyFixedScaleOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataYandexYdbDatabaseDedicatedScalePolicyFixedScaleOutputReference[_g] = { fqn: "@cdktf/provider-yandex.DataYandexYdbDatabaseDedicatedScalePolicyFixedScaleOutputReference", version: "2.0.32" };
class DataYandexYdbDatabaseDedicatedScalePolicyFixedScaleList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexYdbDatabaseDedicatedScalePolicyFixedScaleOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexYdbDatabaseDedicatedScalePolicyFixedScaleList = DataYandexYdbDatabaseDedicatedScalePolicyFixedScaleList;
_h = JSII_RTTI_SYMBOL_1;
DataYandexYdbDatabaseDedicatedScalePolicyFixedScaleList[_h] = { fqn: "@cdktf/provider-yandex.DataYandexYdbDatabaseDedicatedScalePolicyFixedScaleList", version: "2.0.32" };
function dataYandexYdbDatabaseDedicatedScalePolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexYdbDatabaseDedicatedScalePolicyToTerraform = dataYandexYdbDatabaseDedicatedScalePolicyToTerraform;
class DataYandexYdbDatabaseDedicatedScalePolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // fixed_scale - computed: true, optional: false, required: false
        this._fixedScale = new DataYandexYdbDatabaseDedicatedScalePolicyFixedScaleList(this, "fixed_scale", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get fixedScale() {
        return this._fixedScale;
    }
}
exports.DataYandexYdbDatabaseDedicatedScalePolicyOutputReference = DataYandexYdbDatabaseDedicatedScalePolicyOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataYandexYdbDatabaseDedicatedScalePolicyOutputReference[_j] = { fqn: "@cdktf/provider-yandex.DataYandexYdbDatabaseDedicatedScalePolicyOutputReference", version: "2.0.32" };
class DataYandexYdbDatabaseDedicatedScalePolicyList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexYdbDatabaseDedicatedScalePolicyOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexYdbDatabaseDedicatedScalePolicyList = DataYandexYdbDatabaseDedicatedScalePolicyList;
_k = JSII_RTTI_SYMBOL_1;
DataYandexYdbDatabaseDedicatedScalePolicyList[_k] = { fqn: "@cdktf/provider-yandex.DataYandexYdbDatabaseDedicatedScalePolicyList", version: "2.0.32" };
function dataYandexYdbDatabaseDedicatedStorageConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexYdbDatabaseDedicatedStorageConfigToTerraform = dataYandexYdbDatabaseDedicatedStorageConfigToTerraform;
class DataYandexYdbDatabaseDedicatedStorageConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // group_count - computed: true, optional: false, required: false
    get groupCount() {
        return this.getNumberAttribute('group_count');
    }
    // storage_type_id - computed: true, optional: false, required: false
    get storageTypeId() {
        return this.getStringAttribute('storage_type_id');
    }
}
exports.DataYandexYdbDatabaseDedicatedStorageConfigOutputReference = DataYandexYdbDatabaseDedicatedStorageConfigOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataYandexYdbDatabaseDedicatedStorageConfigOutputReference[_l] = { fqn: "@cdktf/provider-yandex.DataYandexYdbDatabaseDedicatedStorageConfigOutputReference", version: "2.0.32" };
class DataYandexYdbDatabaseDedicatedStorageConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexYdbDatabaseDedicatedStorageConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexYdbDatabaseDedicatedStorageConfigList = DataYandexYdbDatabaseDedicatedStorageConfigList;
_m = JSII_RTTI_SYMBOL_1;
DataYandexYdbDatabaseDedicatedStorageConfigList[_m] = { fqn: "@cdktf/provider-yandex.DataYandexYdbDatabaseDedicatedStorageConfigList", version: "2.0.32" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/d/ydb_database_dedicated yandex_ydb_database_dedicated}
*/
class DataYandexYdbDatabaseDedicated extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/d/ydb_database_dedicated yandex_ydb_database_dedicated} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataYandexYdbDatabaseDedicatedConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'yandex_ydb_database_dedicated',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        // location - computed: true, optional: false, required: false
        this._location = new DataYandexYdbDatabaseDedicatedLocationList(this, "location", false);
        // scale_policy - computed: true, optional: false, required: false
        this._scalePolicy = new DataYandexYdbDatabaseDedicatedScalePolicyList(this, "scale_policy", false);
        // storage_config - computed: true, optional: false, required: false
        this._storageConfig = new DataYandexYdbDatabaseDedicatedStorageConfigList(this, "storage_config", false);
        this._databaseId = config.databaseId;
        this._folderId = config.folderId;
        this._id = config.id;
        this._name = config.name;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // assign_public_ips - computed: true, optional: false, required: false
    get assignPublicIps() {
        return this.getBooleanAttribute('assign_public_ips');
    }
    // created_at - computed: true, optional: false, required: false
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    get databaseId() {
        return this.getStringAttribute('database_id');
    }
    set databaseId(value) {
        this._databaseId = value;
    }
    resetDatabaseId() {
        this._databaseId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get databaseIdInput() {
        return this._databaseId;
    }
    // database_path - computed: true, optional: false, required: false
    get databasePath() {
        return this.getStringAttribute('database_path');
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    get folderId() {
        return this.getStringAttribute('folder_id');
    }
    set folderId(value) {
        this._folderId = value;
    }
    resetFolderId() {
        this._folderId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get folderIdInput() {
        return this._folderId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this._labels;
    }
    get location() {
        return this._location;
    }
    // location_id - computed: true, optional: false, required: false
    get locationId() {
        return this.getStringAttribute('location_id');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // network_id - computed: true, optional: false, required: false
    get networkId() {
        return this.getStringAttribute('network_id');
    }
    // resource_preset_id - computed: true, optional: false, required: false
    get resourcePresetId() {
        return this.getStringAttribute('resource_preset_id');
    }
    get scalePolicy() {
        return this._scalePolicy;
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    get storageConfig() {
        return this._storageConfig;
    }
    // subnet_ids - computed: true, optional: false, required: false
    get subnetIds() {
        return cdktf.Fn.tolist(this.getListAttribute('subnet_ids'));
    }
    // tls_enabled - computed: true, optional: false, required: false
    get tlsEnabled() {
        return this.getBooleanAttribute('tls_enabled');
    }
    // ydb_api_endpoint - computed: true, optional: false, required: false
    get ydbApiEndpoint() {
        return this.getStringAttribute('ydb_api_endpoint');
    }
    // ydb_full_endpoint - computed: true, optional: false, required: false
    get ydbFullEndpoint() {
        return this.getStringAttribute('ydb_full_endpoint');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            database_id: cdktf.stringToTerraform(this._databaseId),
            folder_id: cdktf.stringToTerraform(this._folderId),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
        };
    }
}
exports.DataYandexYdbDatabaseDedicated = DataYandexYdbDatabaseDedicated;
_o = JSII_RTTI_SYMBOL_1;
DataYandexYdbDatabaseDedicated[_o] = { fqn: "@cdktf/provider-yandex.DataYandexYdbDatabaseDedicated", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
DataYandexYdbDatabaseDedicated.tfResourceType = "yandex_ydb_database_dedicated";
