// @cdktf/provider-yandex/lib/data-yandex-lb-target-group.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataYandexLbTargetGroup = exports.DataYandexLbTargetGroupTargetList = exports.DataYandexLbTargetGroupTargetOutputReference = exports.dataYandexLbTargetGroupTargetToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataYandexLbTargetGroupTargetToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexLbTargetGroupTargetToTerraform = dataYandexLbTargetGroupTargetToTerraform;
class DataYandexLbTargetGroupTargetOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // address - computed: true, optional: false, required: false
    get address() {
        return this.getStringAttribute('address');
    }
    // subnet_id - computed: true, optional: false, required: false
    get subnetId() {
        return this.getStringAttribute('subnet_id');
    }
}
exports.DataYandexLbTargetGroupTargetOutputReference = DataYandexLbTargetGroupTargetOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataYandexLbTargetGroupTargetOutputReference[_a] = { fqn: "@cdktf/provider-yandex.DataYandexLbTargetGroupTargetOutputReference", version: "2.0.32" };
class DataYandexLbTargetGroupTargetList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexLbTargetGroupTargetOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexLbTargetGroupTargetList = DataYandexLbTargetGroupTargetList;
_b = JSII_RTTI_SYMBOL_1;
DataYandexLbTargetGroupTargetList[_b] = { fqn: "@cdktf/provider-yandex.DataYandexLbTargetGroupTargetList", version: "2.0.32" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/d/lb_target_group yandex_lb_target_group}
*/
class DataYandexLbTargetGroup extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/d/lb_target_group yandex_lb_target_group} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataYandexLbTargetGroupConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'yandex_lb_target_group',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        // target - computed: true, optional: false, required: false
        this._target = new DataYandexLbTargetGroupTargetList(this, "target", true);
        this._folderId = config.folderId;
        this._id = config.id;
        this._name = config.name;
        this._targetGroupId = config.targetGroupId;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // created_at - computed: true, optional: false, required: false
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    get folderId() {
        return this.getStringAttribute('folder_id');
    }
    set folderId(value) {
        this._folderId = value;
    }
    resetFolderId() {
        this._folderId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get folderIdInput() {
        return this._folderId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this._labels;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get target() {
        return this._target;
    }
    get targetGroupId() {
        return this.getStringAttribute('target_group_id');
    }
    set targetGroupId(value) {
        this._targetGroupId = value;
    }
    resetTargetGroupId() {
        this._targetGroupId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetGroupIdInput() {
        return this._targetGroupId;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            folder_id: cdktf.stringToTerraform(this._folderId),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            target_group_id: cdktf.stringToTerraform(this._targetGroupId),
        };
    }
}
exports.DataYandexLbTargetGroup = DataYandexLbTargetGroup;
_c = JSII_RTTI_SYMBOL_1;
DataYandexLbTargetGroup[_c] = { fqn: "@cdktf/provider-yandex.DataYandexLbTargetGroup", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
DataYandexLbTargetGroup.tfResourceType = "yandex_lb_target_group";
