// @cdktf/provider-yandex/lib/data-yandex-mdb-sqlserver-cluster.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataYandexMdbSqlserverCluster = exports.DataYandexMdbSqlserverClusterUserList = exports.DataYandexMdbSqlserverClusterUserOutputReference = exports.dataYandexMdbSqlserverClusterUserToTerraform = exports.DataYandexMdbSqlserverClusterUserPermissionList = exports.DataYandexMdbSqlserverClusterUserPermissionOutputReference = exports.dataYandexMdbSqlserverClusterUserPermissionToTerraform = exports.DataYandexMdbSqlserverClusterResourcesList = exports.DataYandexMdbSqlserverClusterResourcesOutputReference = exports.dataYandexMdbSqlserverClusterResourcesToTerraform = exports.DataYandexMdbSqlserverClusterHostList = exports.DataYandexMdbSqlserverClusterHostOutputReference = exports.dataYandexMdbSqlserverClusterHostToTerraform = exports.DataYandexMdbSqlserverClusterDatabaseList = exports.DataYandexMdbSqlserverClusterDatabaseOutputReference = exports.dataYandexMdbSqlserverClusterDatabaseToTerraform = exports.DataYandexMdbSqlserverClusterBackupWindowStartList = exports.DataYandexMdbSqlserverClusterBackupWindowStartOutputReference = exports.dataYandexMdbSqlserverClusterBackupWindowStartToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataYandexMdbSqlserverClusterBackupWindowStartToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexMdbSqlserverClusterBackupWindowStartToTerraform = dataYandexMdbSqlserverClusterBackupWindowStartToTerraform;
class DataYandexMdbSqlserverClusterBackupWindowStartOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // hours - computed: true, optional: false, required: false
    get hours() {
        return this.getNumberAttribute('hours');
    }
    // minutes - computed: true, optional: false, required: false
    get minutes() {
        return this.getNumberAttribute('minutes');
    }
}
exports.DataYandexMdbSqlserverClusterBackupWindowStartOutputReference = DataYandexMdbSqlserverClusterBackupWindowStartOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataYandexMdbSqlserverClusterBackupWindowStartOutputReference[_a] = { fqn: "@cdktf/provider-yandex.DataYandexMdbSqlserverClusterBackupWindowStartOutputReference", version: "2.0.32" };
class DataYandexMdbSqlserverClusterBackupWindowStartList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbSqlserverClusterBackupWindowStartOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbSqlserverClusterBackupWindowStartList = DataYandexMdbSqlserverClusterBackupWindowStartList;
_b = JSII_RTTI_SYMBOL_1;
DataYandexMdbSqlserverClusterBackupWindowStartList[_b] = { fqn: "@cdktf/provider-yandex.DataYandexMdbSqlserverClusterBackupWindowStartList", version: "2.0.32" };
function dataYandexMdbSqlserverClusterDatabaseToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexMdbSqlserverClusterDatabaseToTerraform = dataYandexMdbSqlserverClusterDatabaseToTerraform;
class DataYandexMdbSqlserverClusterDatabaseOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
}
exports.DataYandexMdbSqlserverClusterDatabaseOutputReference = DataYandexMdbSqlserverClusterDatabaseOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataYandexMdbSqlserverClusterDatabaseOutputReference[_c] = { fqn: "@cdktf/provider-yandex.DataYandexMdbSqlserverClusterDatabaseOutputReference", version: "2.0.32" };
class DataYandexMdbSqlserverClusterDatabaseList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbSqlserverClusterDatabaseOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbSqlserverClusterDatabaseList = DataYandexMdbSqlserverClusterDatabaseList;
_d = JSII_RTTI_SYMBOL_1;
DataYandexMdbSqlserverClusterDatabaseList[_d] = { fqn: "@cdktf/provider-yandex.DataYandexMdbSqlserverClusterDatabaseList", version: "2.0.32" };
function dataYandexMdbSqlserverClusterHostToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexMdbSqlserverClusterHostToTerraform = dataYandexMdbSqlserverClusterHostToTerraform;
class DataYandexMdbSqlserverClusterHostOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // assign_public_ip - computed: true, optional: false, required: false
    get assignPublicIp() {
        return this.getBooleanAttribute('assign_public_ip');
    }
    // fqdn - computed: true, optional: false, required: false
    get fqdn() {
        return this.getStringAttribute('fqdn');
    }
    // subnet_id - computed: true, optional: false, required: false
    get subnetId() {
        return this.getStringAttribute('subnet_id');
    }
    // zone - computed: true, optional: false, required: false
    get zone() {
        return this.getStringAttribute('zone');
    }
}
exports.DataYandexMdbSqlserverClusterHostOutputReference = DataYandexMdbSqlserverClusterHostOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataYandexMdbSqlserverClusterHostOutputReference[_e] = { fqn: "@cdktf/provider-yandex.DataYandexMdbSqlserverClusterHostOutputReference", version: "2.0.32" };
class DataYandexMdbSqlserverClusterHostList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbSqlserverClusterHostOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbSqlserverClusterHostList = DataYandexMdbSqlserverClusterHostList;
_f = JSII_RTTI_SYMBOL_1;
DataYandexMdbSqlserverClusterHostList[_f] = { fqn: "@cdktf/provider-yandex.DataYandexMdbSqlserverClusterHostList", version: "2.0.32" };
function dataYandexMdbSqlserverClusterResourcesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexMdbSqlserverClusterResourcesToTerraform = dataYandexMdbSqlserverClusterResourcesToTerraform;
class DataYandexMdbSqlserverClusterResourcesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // disk_size - computed: true, optional: false, required: false
    get diskSize() {
        return this.getNumberAttribute('disk_size');
    }
    // disk_type_id - computed: true, optional: false, required: false
    get diskTypeId() {
        return this.getStringAttribute('disk_type_id');
    }
    // resource_preset_id - computed: true, optional: false, required: false
    get resourcePresetId() {
        return this.getStringAttribute('resource_preset_id');
    }
}
exports.DataYandexMdbSqlserverClusterResourcesOutputReference = DataYandexMdbSqlserverClusterResourcesOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataYandexMdbSqlserverClusterResourcesOutputReference[_g] = { fqn: "@cdktf/provider-yandex.DataYandexMdbSqlserverClusterResourcesOutputReference", version: "2.0.32" };
class DataYandexMdbSqlserverClusterResourcesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbSqlserverClusterResourcesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbSqlserverClusterResourcesList = DataYandexMdbSqlserverClusterResourcesList;
_h = JSII_RTTI_SYMBOL_1;
DataYandexMdbSqlserverClusterResourcesList[_h] = { fqn: "@cdktf/provider-yandex.DataYandexMdbSqlserverClusterResourcesList", version: "2.0.32" };
function dataYandexMdbSqlserverClusterUserPermissionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexMdbSqlserverClusterUserPermissionToTerraform = dataYandexMdbSqlserverClusterUserPermissionToTerraform;
class DataYandexMdbSqlserverClusterUserPermissionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // database_name - computed: true, optional: false, required: false
    get databaseName() {
        return this.getStringAttribute('database_name');
    }
    // roles - computed: true, optional: false, required: false
    get roles() {
        return cdktf.Fn.tolist(this.getListAttribute('roles'));
    }
}
exports.DataYandexMdbSqlserverClusterUserPermissionOutputReference = DataYandexMdbSqlserverClusterUserPermissionOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataYandexMdbSqlserverClusterUserPermissionOutputReference[_j] = { fqn: "@cdktf/provider-yandex.DataYandexMdbSqlserverClusterUserPermissionOutputReference", version: "2.0.32" };
class DataYandexMdbSqlserverClusterUserPermissionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbSqlserverClusterUserPermissionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbSqlserverClusterUserPermissionList = DataYandexMdbSqlserverClusterUserPermissionList;
_k = JSII_RTTI_SYMBOL_1;
DataYandexMdbSqlserverClusterUserPermissionList[_k] = { fqn: "@cdktf/provider-yandex.DataYandexMdbSqlserverClusterUserPermissionList", version: "2.0.32" };
function dataYandexMdbSqlserverClusterUserToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexMdbSqlserverClusterUserToTerraform = dataYandexMdbSqlserverClusterUserToTerraform;
class DataYandexMdbSqlserverClusterUserOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // permission - computed: true, optional: false, required: false
        this._permission = new DataYandexMdbSqlserverClusterUserPermissionList(this, "permission", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // password - computed: true, optional: false, required: false
    get password() {
        return this.getStringAttribute('password');
    }
    get permission() {
        return this._permission;
    }
}
exports.DataYandexMdbSqlserverClusterUserOutputReference = DataYandexMdbSqlserverClusterUserOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataYandexMdbSqlserverClusterUserOutputReference[_l] = { fqn: "@cdktf/provider-yandex.DataYandexMdbSqlserverClusterUserOutputReference", version: "2.0.32" };
class DataYandexMdbSqlserverClusterUserList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbSqlserverClusterUserOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbSqlserverClusterUserList = DataYandexMdbSqlserverClusterUserList;
_m = JSII_RTTI_SYMBOL_1;
DataYandexMdbSqlserverClusterUserList[_m] = { fqn: "@cdktf/provider-yandex.DataYandexMdbSqlserverClusterUserList", version: "2.0.32" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/d/mdb_sqlserver_cluster yandex_mdb_sqlserver_cluster}
*/
class DataYandexMdbSqlserverCluster extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/d/mdb_sqlserver_cluster yandex_mdb_sqlserver_cluster} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataYandexMdbSqlserverClusterConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'yandex_mdb_sqlserver_cluster',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ==========
        // ATTRIBUTES
        // ==========
        // backup_window_start - computed: true, optional: false, required: false
        this._backupWindowStart = new DataYandexMdbSqlserverClusterBackupWindowStartList(this, "backup_window_start", false);
        // database - computed: true, optional: false, required: false
        this._database = new DataYandexMdbSqlserverClusterDatabaseList(this, "database", false);
        // host - computed: true, optional: false, required: false
        this._host = new DataYandexMdbSqlserverClusterHostList(this, "host", false);
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        // resources - computed: true, optional: false, required: false
        this._resources = new DataYandexMdbSqlserverClusterResourcesList(this, "resources", false);
        // user - computed: true, optional: false, required: false
        this._user = new DataYandexMdbSqlserverClusterUserList(this, "user", false);
        this._clusterId = config.clusterId;
        this._deletionProtection = config.deletionProtection;
        this._folderId = config.folderId;
        this._id = config.id;
        this._name = config.name;
        this._sqlcollation = config.sqlcollation;
        this._sqlserverConfig = config.sqlserverConfig;
    }
    get backupWindowStart() {
        return this._backupWindowStart;
    }
    get clusterId() {
        return this.getStringAttribute('cluster_id');
    }
    set clusterId(value) {
        this._clusterId = value;
    }
    resetClusterId() {
        this._clusterId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clusterIdInput() {
        return this._clusterId;
    }
    // created_at - computed: true, optional: false, required: false
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    get database() {
        return this._database;
    }
    get deletionProtection() {
        return this.getBooleanAttribute('deletion_protection');
    }
    set deletionProtection(value) {
        this._deletionProtection = value;
    }
    resetDeletionProtection() {
        this._deletionProtection = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deletionProtectionInput() {
        return this._deletionProtection;
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    // environment - computed: true, optional: false, required: false
    get environment() {
        return this.getStringAttribute('environment');
    }
    get folderId() {
        return this.getStringAttribute('folder_id');
    }
    set folderId(value) {
        this._folderId = value;
    }
    resetFolderId() {
        this._folderId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get folderIdInput() {
        return this._folderId;
    }
    // health - computed: true, optional: false, required: false
    get health() {
        return this.getStringAttribute('health');
    }
    get host() {
        return this._host;
    }
    // host_group_ids - computed: true, optional: false, required: false
    get hostGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('host_group_ids'));
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this._labels;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // network_id - computed: true, optional: false, required: false
    get networkId() {
        return this.getStringAttribute('network_id');
    }
    get resources() {
        return this._resources;
    }
    // security_group_ids - computed: true, optional: false, required: false
    get securityGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('security_group_ids'));
    }
    get sqlcollation() {
        return this.getStringAttribute('sqlcollation');
    }
    set sqlcollation(value) {
        this._sqlcollation = value;
    }
    resetSqlcollation() {
        this._sqlcollation = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sqlcollationInput() {
        return this._sqlcollation;
    }
    get sqlserverConfig() {
        return this.getStringMapAttribute('sqlserver_config');
    }
    set sqlserverConfig(value) {
        this._sqlserverConfig = value;
    }
    resetSqlserverConfig() {
        this._sqlserverConfig = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sqlserverConfigInput() {
        return this._sqlserverConfig;
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    get user() {
        return this._user;
    }
    // version - computed: true, optional: false, required: false
    get version() {
        return this.getStringAttribute('version');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            cluster_id: cdktf.stringToTerraform(this._clusterId),
            deletion_protection: cdktf.booleanToTerraform(this._deletionProtection),
            folder_id: cdktf.stringToTerraform(this._folderId),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            sqlcollation: cdktf.stringToTerraform(this._sqlcollation),
            sqlserver_config: cdktf.hashMapper(cdktf.stringToTerraform)(this._sqlserverConfig),
        };
    }
}
exports.DataYandexMdbSqlserverCluster = DataYandexMdbSqlserverCluster;
_o = JSII_RTTI_SYMBOL_1;
DataYandexMdbSqlserverCluster[_o] = { fqn: "@cdktf/provider-yandex.DataYandexMdbSqlserverCluster", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
DataYandexMdbSqlserverCluster.tfResourceType = "yandex_mdb_sqlserver_cluster";
