// @cdktf/provider-yandex/lib/data-yandex-mdb-mysql-cluster.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataYandexMdbMysqlCluster = exports.DataYandexMdbMysqlClusterAccessOutputReference = exports.dataYandexMdbMysqlClusterAccessToTerraform = exports.DataYandexMdbMysqlClusterUserList = exports.DataYandexMdbMysqlClusterUserOutputReference = exports.dataYandexMdbMysqlClusterUserToTerraform = exports.DataYandexMdbMysqlClusterUserPermissionList = exports.DataYandexMdbMysqlClusterUserPermissionOutputReference = exports.dataYandexMdbMysqlClusterUserPermissionToTerraform = exports.DataYandexMdbMysqlClusterUserConnectionLimitsList = exports.DataYandexMdbMysqlClusterUserConnectionLimitsOutputReference = exports.dataYandexMdbMysqlClusterUserConnectionLimitsToTerraform = exports.DataYandexMdbMysqlClusterResourcesList = exports.DataYandexMdbMysqlClusterResourcesOutputReference = exports.dataYandexMdbMysqlClusterResourcesToTerraform = exports.DataYandexMdbMysqlClusterPerformanceDiagnosticsList = exports.DataYandexMdbMysqlClusterPerformanceDiagnosticsOutputReference = exports.dataYandexMdbMysqlClusterPerformanceDiagnosticsToTerraform = exports.DataYandexMdbMysqlClusterMaintenanceWindowList = exports.DataYandexMdbMysqlClusterMaintenanceWindowOutputReference = exports.dataYandexMdbMysqlClusterMaintenanceWindowToTerraform = exports.DataYandexMdbMysqlClusterHostList = exports.DataYandexMdbMysqlClusterHostOutputReference = exports.dataYandexMdbMysqlClusterHostToTerraform = exports.DataYandexMdbMysqlClusterDatabaseList = exports.DataYandexMdbMysqlClusterDatabaseOutputReference = exports.dataYandexMdbMysqlClusterDatabaseToTerraform = exports.DataYandexMdbMysqlClusterBackupWindowStartList = exports.DataYandexMdbMysqlClusterBackupWindowStartOutputReference = exports.dataYandexMdbMysqlClusterBackupWindowStartToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataYandexMdbMysqlClusterBackupWindowStartToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexMdbMysqlClusterBackupWindowStartToTerraform = dataYandexMdbMysqlClusterBackupWindowStartToTerraform;
class DataYandexMdbMysqlClusterBackupWindowStartOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // hours - computed: true, optional: false, required: false
    get hours() {
        return this.getNumberAttribute('hours');
    }
    // minutes - computed: true, optional: false, required: false
    get minutes() {
        return this.getNumberAttribute('minutes');
    }
}
exports.DataYandexMdbMysqlClusterBackupWindowStartOutputReference = DataYandexMdbMysqlClusterBackupWindowStartOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataYandexMdbMysqlClusterBackupWindowStartOutputReference[_a] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMysqlClusterBackupWindowStartOutputReference", version: "2.0.32" };
class DataYandexMdbMysqlClusterBackupWindowStartList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbMysqlClusterBackupWindowStartOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbMysqlClusterBackupWindowStartList = DataYandexMdbMysqlClusterBackupWindowStartList;
_b = JSII_RTTI_SYMBOL_1;
DataYandexMdbMysqlClusterBackupWindowStartList[_b] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMysqlClusterBackupWindowStartList", version: "2.0.32" };
function dataYandexMdbMysqlClusterDatabaseToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexMdbMysqlClusterDatabaseToTerraform = dataYandexMdbMysqlClusterDatabaseToTerraform;
class DataYandexMdbMysqlClusterDatabaseOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
}
exports.DataYandexMdbMysqlClusterDatabaseOutputReference = DataYandexMdbMysqlClusterDatabaseOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataYandexMdbMysqlClusterDatabaseOutputReference[_c] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMysqlClusterDatabaseOutputReference", version: "2.0.32" };
class DataYandexMdbMysqlClusterDatabaseList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbMysqlClusterDatabaseOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbMysqlClusterDatabaseList = DataYandexMdbMysqlClusterDatabaseList;
_d = JSII_RTTI_SYMBOL_1;
DataYandexMdbMysqlClusterDatabaseList[_d] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMysqlClusterDatabaseList", version: "2.0.32" };
function dataYandexMdbMysqlClusterHostToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexMdbMysqlClusterHostToTerraform = dataYandexMdbMysqlClusterHostToTerraform;
class DataYandexMdbMysqlClusterHostOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // assign_public_ip - computed: true, optional: false, required: false
    get assignPublicIp() {
        return this.getBooleanAttribute('assign_public_ip');
    }
    // backup_priority - computed: true, optional: false, required: false
    get backupPriority() {
        return this.getNumberAttribute('backup_priority');
    }
    // fqdn - computed: true, optional: false, required: false
    get fqdn() {
        return this.getStringAttribute('fqdn');
    }
    // priority - computed: true, optional: false, required: false
    get priority() {
        return this.getNumberAttribute('priority');
    }
    // replication_source - computed: true, optional: false, required: false
    get replicationSource() {
        return this.getStringAttribute('replication_source');
    }
    // subnet_id - computed: true, optional: false, required: false
    get subnetId() {
        return this.getStringAttribute('subnet_id');
    }
    // zone - computed: true, optional: false, required: false
    get zone() {
        return this.getStringAttribute('zone');
    }
}
exports.DataYandexMdbMysqlClusterHostOutputReference = DataYandexMdbMysqlClusterHostOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataYandexMdbMysqlClusterHostOutputReference[_e] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMysqlClusterHostOutputReference", version: "2.0.32" };
class DataYandexMdbMysqlClusterHostList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbMysqlClusterHostOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbMysqlClusterHostList = DataYandexMdbMysqlClusterHostList;
_f = JSII_RTTI_SYMBOL_1;
DataYandexMdbMysqlClusterHostList[_f] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMysqlClusterHostList", version: "2.0.32" };
function dataYandexMdbMysqlClusterMaintenanceWindowToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexMdbMysqlClusterMaintenanceWindowToTerraform = dataYandexMdbMysqlClusterMaintenanceWindowToTerraform;
class DataYandexMdbMysqlClusterMaintenanceWindowOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // day - computed: true, optional: false, required: false
    get day() {
        return this.getStringAttribute('day');
    }
    // hour - computed: true, optional: false, required: false
    get hour() {
        return this.getNumberAttribute('hour');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataYandexMdbMysqlClusterMaintenanceWindowOutputReference = DataYandexMdbMysqlClusterMaintenanceWindowOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataYandexMdbMysqlClusterMaintenanceWindowOutputReference[_g] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMysqlClusterMaintenanceWindowOutputReference", version: "2.0.32" };
class DataYandexMdbMysqlClusterMaintenanceWindowList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbMysqlClusterMaintenanceWindowOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbMysqlClusterMaintenanceWindowList = DataYandexMdbMysqlClusterMaintenanceWindowList;
_h = JSII_RTTI_SYMBOL_1;
DataYandexMdbMysqlClusterMaintenanceWindowList[_h] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMysqlClusterMaintenanceWindowList", version: "2.0.32" };
function dataYandexMdbMysqlClusterPerformanceDiagnosticsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexMdbMysqlClusterPerformanceDiagnosticsToTerraform = dataYandexMdbMysqlClusterPerformanceDiagnosticsToTerraform;
class DataYandexMdbMysqlClusterPerformanceDiagnosticsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // enabled - computed: true, optional: false, required: false
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    // sessions_sampling_interval - computed: true, optional: false, required: false
    get sessionsSamplingInterval() {
        return this.getNumberAttribute('sessions_sampling_interval');
    }
    // statements_sampling_interval - computed: true, optional: false, required: false
    get statementsSamplingInterval() {
        return this.getNumberAttribute('statements_sampling_interval');
    }
}
exports.DataYandexMdbMysqlClusterPerformanceDiagnosticsOutputReference = DataYandexMdbMysqlClusterPerformanceDiagnosticsOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataYandexMdbMysqlClusterPerformanceDiagnosticsOutputReference[_j] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMysqlClusterPerformanceDiagnosticsOutputReference", version: "2.0.32" };
class DataYandexMdbMysqlClusterPerformanceDiagnosticsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbMysqlClusterPerformanceDiagnosticsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbMysqlClusterPerformanceDiagnosticsList = DataYandexMdbMysqlClusterPerformanceDiagnosticsList;
_k = JSII_RTTI_SYMBOL_1;
DataYandexMdbMysqlClusterPerformanceDiagnosticsList[_k] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMysqlClusterPerformanceDiagnosticsList", version: "2.0.32" };
function dataYandexMdbMysqlClusterResourcesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexMdbMysqlClusterResourcesToTerraform = dataYandexMdbMysqlClusterResourcesToTerraform;
class DataYandexMdbMysqlClusterResourcesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // disk_size - computed: true, optional: false, required: false
    get diskSize() {
        return this.getNumberAttribute('disk_size');
    }
    // disk_type_id - computed: true, optional: false, required: false
    get diskTypeId() {
        return this.getStringAttribute('disk_type_id');
    }
    // resource_preset_id - computed: true, optional: false, required: false
    get resourcePresetId() {
        return this.getStringAttribute('resource_preset_id');
    }
}
exports.DataYandexMdbMysqlClusterResourcesOutputReference = DataYandexMdbMysqlClusterResourcesOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataYandexMdbMysqlClusterResourcesOutputReference[_l] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMysqlClusterResourcesOutputReference", version: "2.0.32" };
class DataYandexMdbMysqlClusterResourcesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbMysqlClusterResourcesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbMysqlClusterResourcesList = DataYandexMdbMysqlClusterResourcesList;
_m = JSII_RTTI_SYMBOL_1;
DataYandexMdbMysqlClusterResourcesList[_m] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMysqlClusterResourcesList", version: "2.0.32" };
function dataYandexMdbMysqlClusterUserConnectionLimitsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexMdbMysqlClusterUserConnectionLimitsToTerraform = dataYandexMdbMysqlClusterUserConnectionLimitsToTerraform;
class DataYandexMdbMysqlClusterUserConnectionLimitsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // max_connections_per_hour - computed: true, optional: false, required: false
    get maxConnectionsPerHour() {
        return this.getNumberAttribute('max_connections_per_hour');
    }
    // max_questions_per_hour - computed: true, optional: false, required: false
    get maxQuestionsPerHour() {
        return this.getNumberAttribute('max_questions_per_hour');
    }
    // max_updates_per_hour - computed: true, optional: false, required: false
    get maxUpdatesPerHour() {
        return this.getNumberAttribute('max_updates_per_hour');
    }
    // max_user_connections - computed: true, optional: false, required: false
    get maxUserConnections() {
        return this.getNumberAttribute('max_user_connections');
    }
}
exports.DataYandexMdbMysqlClusterUserConnectionLimitsOutputReference = DataYandexMdbMysqlClusterUserConnectionLimitsOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataYandexMdbMysqlClusterUserConnectionLimitsOutputReference[_o] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMysqlClusterUserConnectionLimitsOutputReference", version: "2.0.32" };
class DataYandexMdbMysqlClusterUserConnectionLimitsList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbMysqlClusterUserConnectionLimitsOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbMysqlClusterUserConnectionLimitsList = DataYandexMdbMysqlClusterUserConnectionLimitsList;
_p = JSII_RTTI_SYMBOL_1;
DataYandexMdbMysqlClusterUserConnectionLimitsList[_p] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMysqlClusterUserConnectionLimitsList", version: "2.0.32" };
function dataYandexMdbMysqlClusterUserPermissionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexMdbMysqlClusterUserPermissionToTerraform = dataYandexMdbMysqlClusterUserPermissionToTerraform;
class DataYandexMdbMysqlClusterUserPermissionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // database_name - computed: true, optional: false, required: false
    get databaseName() {
        return this.getStringAttribute('database_name');
    }
    // roles - computed: true, optional: false, required: false
    get roles() {
        return this.getListAttribute('roles');
    }
}
exports.DataYandexMdbMysqlClusterUserPermissionOutputReference = DataYandexMdbMysqlClusterUserPermissionOutputReference;
_q = JSII_RTTI_SYMBOL_1;
DataYandexMdbMysqlClusterUserPermissionOutputReference[_q] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMysqlClusterUserPermissionOutputReference", version: "2.0.32" };
class DataYandexMdbMysqlClusterUserPermissionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbMysqlClusterUserPermissionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbMysqlClusterUserPermissionList = DataYandexMdbMysqlClusterUserPermissionList;
_r = JSII_RTTI_SYMBOL_1;
DataYandexMdbMysqlClusterUserPermissionList[_r] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMysqlClusterUserPermissionList", version: "2.0.32" };
function dataYandexMdbMysqlClusterUserToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexMdbMysqlClusterUserToTerraform = dataYandexMdbMysqlClusterUserToTerraform;
class DataYandexMdbMysqlClusterUserOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // connection_limits - computed: true, optional: false, required: false
        this._connectionLimits = new DataYandexMdbMysqlClusterUserConnectionLimitsList(this, "connection_limits", false);
        // permission - computed: true, optional: false, required: false
        this._permission = new DataYandexMdbMysqlClusterUserPermissionList(this, "permission", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // authentication_plugin - computed: true, optional: false, required: false
    get authenticationPlugin() {
        return this.getStringAttribute('authentication_plugin');
    }
    get connectionLimits() {
        return this._connectionLimits;
    }
    // global_permissions - computed: true, optional: false, required: false
    get globalPermissions() {
        return cdktf.Fn.tolist(this.getListAttribute('global_permissions'));
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    // password - computed: true, optional: false, required: false
    get password() {
        return this.getStringAttribute('password');
    }
    get permission() {
        return this._permission;
    }
}
exports.DataYandexMdbMysqlClusterUserOutputReference = DataYandexMdbMysqlClusterUserOutputReference;
_s = JSII_RTTI_SYMBOL_1;
DataYandexMdbMysqlClusterUserOutputReference[_s] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMysqlClusterUserOutputReference", version: "2.0.32" };
class DataYandexMdbMysqlClusterUserList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbMysqlClusterUserOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbMysqlClusterUserList = DataYandexMdbMysqlClusterUserList;
_t = JSII_RTTI_SYMBOL_1;
DataYandexMdbMysqlClusterUserList[_t] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMysqlClusterUserList", version: "2.0.32" };
function dataYandexMdbMysqlClusterAccessToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexMdbMysqlClusterAccessToTerraform = dataYandexMdbMysqlClusterAccessToTerraform;
class DataYandexMdbMysqlClusterAccessOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // data_lens - computed: true, optional: false, required: false
    get dataLens() {
        return this.getBooleanAttribute('data_lens');
    }
    // web_sql - computed: true, optional: false, required: false
    get webSql() {
        return this.getBooleanAttribute('web_sql');
    }
}
exports.DataYandexMdbMysqlClusterAccessOutputReference = DataYandexMdbMysqlClusterAccessOutputReference;
_u = JSII_RTTI_SYMBOL_1;
DataYandexMdbMysqlClusterAccessOutputReference[_u] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMysqlClusterAccessOutputReference", version: "2.0.32" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/d/mdb_mysql_cluster yandex_mdb_mysql_cluster}
*/
class DataYandexMdbMysqlCluster extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/d/mdb_mysql_cluster yandex_mdb_mysql_cluster} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataYandexMdbMysqlClusterConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'yandex_mdb_mysql_cluster',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // ==========
        // ATTRIBUTES
        // ==========
        // backup_window_start - computed: true, optional: false, required: false
        this._backupWindowStart = new DataYandexMdbMysqlClusterBackupWindowStartList(this, "backup_window_start", false);
        // database - computed: true, optional: false, required: false
        this._database = new DataYandexMdbMysqlClusterDatabaseList(this, "database", true);
        // host - computed: true, optional: false, required: false
        this._host = new DataYandexMdbMysqlClusterHostList(this, "host", false);
        // maintenance_window - computed: true, optional: false, required: false
        this._maintenanceWindow = new DataYandexMdbMysqlClusterMaintenanceWindowList(this, "maintenance_window", false);
        // performance_diagnostics - computed: true, optional: false, required: false
        this._performanceDiagnostics = new DataYandexMdbMysqlClusterPerformanceDiagnosticsList(this, "performance_diagnostics", false);
        // resources - computed: true, optional: false, required: false
        this._resources = new DataYandexMdbMysqlClusterResourcesList(this, "resources", false);
        // user - computed: true, optional: false, required: false
        this._user = new DataYandexMdbMysqlClusterUserList(this, "user", false);
        // access - computed: false, optional: true, required: false
        this._access = new DataYandexMdbMysqlClusterAccessOutputReference(this, "access");
        this._clusterId = config.clusterId;
        this._deletionProtection = config.deletionProtection;
        this._description = config.description;
        this._folderId = config.folderId;
        this._id = config.id;
        this._labels = config.labels;
        this._mysqlConfig = config.mysqlConfig;
        this._name = config.name;
        this._access.internalValue = config.access;
    }
    get backupWindowStart() {
        return this._backupWindowStart;
    }
    get clusterId() {
        return this.getStringAttribute('cluster_id');
    }
    set clusterId(value) {
        this._clusterId = value;
    }
    resetClusterId() {
        this._clusterId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clusterIdInput() {
        return this._clusterId;
    }
    // created_at - computed: true, optional: false, required: false
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    get database() {
        return this._database;
    }
    get deletionProtection() {
        return this.getBooleanAttribute('deletion_protection');
    }
    set deletionProtection(value) {
        this._deletionProtection = value;
    }
    resetDeletionProtection() {
        this._deletionProtection = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deletionProtectionInput() {
        return this._deletionProtection;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    // environment - computed: true, optional: false, required: false
    get environment() {
        return this.getStringAttribute('environment');
    }
    get folderId() {
        return this.getStringAttribute('folder_id');
    }
    set folderId(value) {
        this._folderId = value;
    }
    resetFolderId() {
        this._folderId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get folderIdInput() {
        return this._folderId;
    }
    // health - computed: true, optional: false, required: false
    get health() {
        return this.getStringAttribute('health');
    }
    get host() {
        return this._host;
    }
    // host_group_ids - computed: true, optional: false, required: false
    get hostGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('host_group_ids'));
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get maintenanceWindow() {
        return this._maintenanceWindow;
    }
    get mysqlConfig() {
        return this.getStringMapAttribute('mysql_config');
    }
    set mysqlConfig(value) {
        this._mysqlConfig = value;
    }
    resetMysqlConfig() {
        this._mysqlConfig = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get mysqlConfigInput() {
        return this._mysqlConfig;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // network_id - computed: true, optional: false, required: false
    get networkId() {
        return this.getStringAttribute('network_id');
    }
    get performanceDiagnostics() {
        return this._performanceDiagnostics;
    }
    get resources() {
        return this._resources;
    }
    // security_group_ids - computed: true, optional: false, required: false
    get securityGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('security_group_ids'));
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    get user() {
        return this._user;
    }
    // version - computed: true, optional: false, required: false
    get version() {
        return this.getStringAttribute('version');
    }
    get access() {
        return this._access;
    }
    putAccess(value) {
        this._access.internalValue = value;
    }
    resetAccess() {
        this._access.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get accessInput() {
        return this._access.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            cluster_id: cdktf.stringToTerraform(this._clusterId),
            deletion_protection: cdktf.booleanToTerraform(this._deletionProtection),
            description: cdktf.stringToTerraform(this._description),
            folder_id: cdktf.stringToTerraform(this._folderId),
            id: cdktf.stringToTerraform(this._id),
            labels: cdktf.hashMapper(cdktf.stringToTerraform)(this._labels),
            mysql_config: cdktf.hashMapper(cdktf.stringToTerraform)(this._mysqlConfig),
            name: cdktf.stringToTerraform(this._name),
            access: dataYandexMdbMysqlClusterAccessToTerraform(this._access.internalValue),
        };
    }
}
exports.DataYandexMdbMysqlCluster = DataYandexMdbMysqlCluster;
_v = JSII_RTTI_SYMBOL_1;
DataYandexMdbMysqlCluster[_v] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMysqlCluster", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
DataYandexMdbMysqlCluster.tfResourceType = "yandex_mdb_mysql_cluster";
