// @cdktf/provider-yandex/lib/data-yandex-compute-disk.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataYandexComputeDisk = exports.DataYandexComputeDiskDiskPlacementPolicyOutputReference = exports.dataYandexComputeDiskDiskPlacementPolicyToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataYandexComputeDiskDiskPlacementPolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        disk_placement_group_id: cdktf.stringToTerraform(struct.diskPlacementGroupId),
    };
}
exports.dataYandexComputeDiskDiskPlacementPolicyToTerraform = dataYandexComputeDiskDiskPlacementPolicyToTerraform;
class DataYandexComputeDiskDiskPlacementPolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._diskPlacementGroupId !== undefined) {
            hasAnyValues = true;
            internalValueResult.diskPlacementGroupId = this._diskPlacementGroupId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._diskPlacementGroupId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._diskPlacementGroupId = value.diskPlacementGroupId;
        }
    }
    get diskPlacementGroupId() {
        return this.getStringAttribute('disk_placement_group_id');
    }
    set diskPlacementGroupId(value) {
        this._diskPlacementGroupId = value;
    }
    // Temporarily expose input value. Use with caution.
    get diskPlacementGroupIdInput() {
        return this._diskPlacementGroupId;
    }
}
exports.DataYandexComputeDiskDiskPlacementPolicyOutputReference = DataYandexComputeDiskDiskPlacementPolicyOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataYandexComputeDiskDiskPlacementPolicyOutputReference[_a] = { fqn: "@cdktf/provider-yandex.DataYandexComputeDiskDiskPlacementPolicyOutputReference", version: "2.0.32" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/d/compute_disk yandex_compute_disk}
*/
class DataYandexComputeDisk extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/d/compute_disk yandex_compute_disk} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataYandexComputeDiskConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'yandex_compute_disk',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        // disk_placement_policy - computed: false, optional: true, required: false
        this._diskPlacementPolicy = new DataYandexComputeDiskDiskPlacementPolicyOutputReference(this, "disk_placement_policy");
        this._diskId = config.diskId;
        this._folderId = config.folderId;
        this._id = config.id;
        this._name = config.name;
        this._diskPlacementPolicy.internalValue = config.diskPlacementPolicy;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // block_size - computed: true, optional: false, required: false
    get blockSize() {
        return this.getNumberAttribute('block_size');
    }
    // created_at - computed: true, optional: false, required: false
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    get diskId() {
        return this.getStringAttribute('disk_id');
    }
    set diskId(value) {
        this._diskId = value;
    }
    resetDiskId() {
        this._diskId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get diskIdInput() {
        return this._diskId;
    }
    get folderId() {
        return this.getStringAttribute('folder_id');
    }
    set folderId(value) {
        this._folderId = value;
    }
    resetFolderId() {
        this._folderId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get folderIdInput() {
        return this._folderId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // image_id - computed: true, optional: false, required: false
    get imageId() {
        return this.getStringAttribute('image_id');
    }
    // instance_ids - computed: true, optional: false, required: false
    get instanceIds() {
        return this.getListAttribute('instance_ids');
    }
    get labels() {
        return this._labels;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // product_ids - computed: true, optional: false, required: false
    get productIds() {
        return this.getListAttribute('product_ids');
    }
    // size - computed: true, optional: false, required: false
    get size() {
        return this.getNumberAttribute('size');
    }
    // snapshot_id - computed: true, optional: false, required: false
    get snapshotId() {
        return this.getStringAttribute('snapshot_id');
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
    // zone - computed: true, optional: false, required: false
    get zone() {
        return this.getStringAttribute('zone');
    }
    get diskPlacementPolicy() {
        return this._diskPlacementPolicy;
    }
    putDiskPlacementPolicy(value) {
        this._diskPlacementPolicy.internalValue = value;
    }
    resetDiskPlacementPolicy() {
        this._diskPlacementPolicy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get diskPlacementPolicyInput() {
        return this._diskPlacementPolicy.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            disk_id: cdktf.stringToTerraform(this._diskId),
            folder_id: cdktf.stringToTerraform(this._folderId),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            disk_placement_policy: dataYandexComputeDiskDiskPlacementPolicyToTerraform(this._diskPlacementPolicy.internalValue),
        };
    }
}
exports.DataYandexComputeDisk = DataYandexComputeDisk;
_b = JSII_RTTI_SYMBOL_1;
DataYandexComputeDisk[_b] = { fqn: "@cdktf/provider-yandex.DataYandexComputeDisk", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
DataYandexComputeDisk.tfResourceType = "yandex_compute_disk";
