// @cdktf/provider-yandex/lib/data-yandex-client-config.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataYandexClientConfig = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/d/client_config yandex_client_config}
*/
class DataYandexClientConfig extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/d/client_config yandex_client_config} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataYandexClientConfigConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'yandex_client_config',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._id = config.id;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // cloud_id - computed: true, optional: false, required: false
    get cloudId() {
        return this.getStringAttribute('cloud_id');
    }
    // folder_id - computed: true, optional: false, required: false
    get folderId() {
        return this.getStringAttribute('folder_id');
    }
    // iam_token - computed: true, optional: false, required: false
    get iamToken() {
        return this.getStringAttribute('iam_token');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // zone - computed: true, optional: false, required: false
    get zone() {
        return this.getStringAttribute('zone');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            id: cdktf.stringToTerraform(this._id),
        };
    }
}
exports.DataYandexClientConfig = DataYandexClientConfig;
_a = JSII_RTTI_SYMBOL_1;
DataYandexClientConfig[_a] = { fqn: "@cdktf/provider-yandex.DataYandexClientConfig", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
DataYandexClientConfig.tfResourceType = "yandex_client_config";
