// @cdktf/provider-yandex/lib/data-yandex-mdb-postgresql-database.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataYandexMdbPostgresqlDatabase = exports.DataYandexMdbPostgresqlDatabaseExtensionList = exports.DataYandexMdbPostgresqlDatabaseExtensionOutputReference = exports.dataYandexMdbPostgresqlDatabaseExtensionToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataYandexMdbPostgresqlDatabaseExtensionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        version: cdktf.stringToTerraform(struct.version),
    };
}
exports.dataYandexMdbPostgresqlDatabaseExtensionToTerraform = dataYandexMdbPostgresqlDatabaseExtensionToTerraform;
class DataYandexMdbPostgresqlDatabaseExtensionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._version !== undefined) {
            hasAnyValues = true;
            internalValueResult.version = this._version;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._version = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._version = value.version;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get version() {
        return this.getStringAttribute('version');
    }
    set version(value) {
        this._version = value;
    }
    resetVersion() {
        this._version = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get versionInput() {
        return this._version;
    }
}
exports.DataYandexMdbPostgresqlDatabaseExtensionOutputReference = DataYandexMdbPostgresqlDatabaseExtensionOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataYandexMdbPostgresqlDatabaseExtensionOutputReference[_a] = { fqn: "@cdktf/provider-yandex.DataYandexMdbPostgresqlDatabaseExtensionOutputReference", version: "2.0.32" };
class DataYandexMdbPostgresqlDatabaseExtensionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbPostgresqlDatabaseExtensionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbPostgresqlDatabaseExtensionList = DataYandexMdbPostgresqlDatabaseExtensionList;
_b = JSII_RTTI_SYMBOL_1;
DataYandexMdbPostgresqlDatabaseExtensionList[_b] = { fqn: "@cdktf/provider-yandex.DataYandexMdbPostgresqlDatabaseExtensionList", version: "2.0.32" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/d/mdb_postgresql_database yandex_mdb_postgresql_database}
*/
class DataYandexMdbPostgresqlDatabase extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/d/mdb_postgresql_database yandex_mdb_postgresql_database} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataYandexMdbPostgresqlDatabaseConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'yandex_mdb_postgresql_database',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // extension - computed: false, optional: true, required: false
        this._extension = new DataYandexMdbPostgresqlDatabaseExtensionList(this, "extension", true);
        this._clusterId = config.clusterId;
        this._id = config.id;
        this._lcCollate = config.lcCollate;
        this._lcType = config.lcType;
        this._name = config.name;
        this._owner = config.owner;
        this._extension.internalValue = config.extension;
    }
    get clusterId() {
        return this.getStringAttribute('cluster_id');
    }
    set clusterId(value) {
        this._clusterId = value;
    }
    // Temporarily expose input value. Use with caution.
    get clusterIdInput() {
        return this._clusterId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get lcCollate() {
        return this.getStringAttribute('lc_collate');
    }
    set lcCollate(value) {
        this._lcCollate = value;
    }
    resetLcCollate() {
        this._lcCollate = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get lcCollateInput() {
        return this._lcCollate;
    }
    get lcType() {
        return this.getStringAttribute('lc_type');
    }
    set lcType(value) {
        this._lcType = value;
    }
    resetLcType() {
        this._lcType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get lcTypeInput() {
        return this._lcType;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get owner() {
        return this.getStringAttribute('owner');
    }
    set owner(value) {
        this._owner = value;
    }
    resetOwner() {
        this._owner = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ownerInput() {
        return this._owner;
    }
    get extension() {
        return this._extension;
    }
    putExtension(value) {
        this._extension.internalValue = value;
    }
    resetExtension() {
        this._extension.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get extensionInput() {
        return this._extension.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            cluster_id: cdktf.stringToTerraform(this._clusterId),
            id: cdktf.stringToTerraform(this._id),
            lc_collate: cdktf.stringToTerraform(this._lcCollate),
            lc_type: cdktf.stringToTerraform(this._lcType),
            name: cdktf.stringToTerraform(this._name),
            owner: cdktf.stringToTerraform(this._owner),
            extension: cdktf.listMapper(dataYandexMdbPostgresqlDatabaseExtensionToTerraform, true)(this._extension.internalValue),
        };
    }
}
exports.DataYandexMdbPostgresqlDatabase = DataYandexMdbPostgresqlDatabase;
_c = JSII_RTTI_SYMBOL_1;
DataYandexMdbPostgresqlDatabase[_c] = { fqn: "@cdktf/provider-yandex.DataYandexMdbPostgresqlDatabase", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
DataYandexMdbPostgresqlDatabase.tfResourceType = "yandex_mdb_postgresql_database";
