// @cdktf/provider-yandex/lib/data-yandex-mdb-kafka-connector.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataYandexMdbKafkaConnector = exports.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerList = exports.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerOutputReference = exports.dataYandexMdbKafkaConnectorConnectorConfigMirrormakerToTerraform = exports.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterList = exports.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterOutputReference = exports.dataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterToTerraform = exports.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterThisClusterList = exports.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterThisClusterOutputReference = exports.dataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterThisClusterToTerraform = exports.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterExternalClusterList = exports.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterExternalClusterOutputReference = exports.dataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterExternalClusterToTerraform = exports.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterList = exports.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterOutputReference = exports.dataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterToTerraform = exports.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterThisClusterList = exports.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterThisClusterOutputReference = exports.dataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterThisClusterToTerraform = exports.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterExternalClusterList = exports.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterExternalClusterOutputReference = exports.dataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterExternalClusterToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterExternalClusterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterExternalClusterToTerraform = dataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterExternalClusterToTerraform;
class DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterExternalClusterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // bootstrap_servers - computed: true, optional: false, required: false
    get bootstrapServers() {
        return this.getStringAttribute('bootstrap_servers');
    }
    // sasl_mechanism - computed: true, optional: false, required: false
    get saslMechanism() {
        return this.getStringAttribute('sasl_mechanism');
    }
    // sasl_password - computed: true, optional: false, required: false
    get saslPassword() {
        return this.getStringAttribute('sasl_password');
    }
    // sasl_username - computed: true, optional: false, required: false
    get saslUsername() {
        return this.getStringAttribute('sasl_username');
    }
    // security_protocol - computed: true, optional: false, required: false
    get securityProtocol() {
        return this.getStringAttribute('security_protocol');
    }
}
exports.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterExternalClusterOutputReference = DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterExternalClusterOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterExternalClusterOutputReference[_a] = { fqn: "@cdktf/provider-yandex.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterExternalClusterOutputReference", version: "2.0.32" };
class DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterExternalClusterList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterExternalClusterOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterExternalClusterList = DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterExternalClusterList;
_b = JSII_RTTI_SYMBOL_1;
DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterExternalClusterList[_b] = { fqn: "@cdktf/provider-yandex.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterExternalClusterList", version: "2.0.32" };
function dataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterThisClusterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterThisClusterToTerraform = dataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterThisClusterToTerraform;
class DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterThisClusterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
}
exports.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterThisClusterOutputReference = DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterThisClusterOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterThisClusterOutputReference[_c] = { fqn: "@cdktf/provider-yandex.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterThisClusterOutputReference", version: "2.0.32" };
class DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterThisClusterList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterThisClusterOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterThisClusterList = DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterThisClusterList;
_d = JSII_RTTI_SYMBOL_1;
DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterThisClusterList[_d] = { fqn: "@cdktf/provider-yandex.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterThisClusterList", version: "2.0.32" };
function dataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterToTerraform = dataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterToTerraform;
class DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // external_cluster - computed: true, optional: false, required: false
        this._externalCluster = new DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterExternalClusterList(this, "external_cluster", false);
        // this_cluster - computed: true, optional: false, required: false
        this._thisCluster = new DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterThisClusterList(this, "this_cluster", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // alias - computed: true, optional: false, required: false
    get alias() {
        return this.getStringAttribute('alias');
    }
    get externalCluster() {
        return this._externalCluster;
    }
    get thisCluster() {
        return this._thisCluster;
    }
}
exports.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterOutputReference = DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterOutputReference[_e] = { fqn: "@cdktf/provider-yandex.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterOutputReference", version: "2.0.32" };
class DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterList = DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterList;
_f = JSII_RTTI_SYMBOL_1;
DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterList[_f] = { fqn: "@cdktf/provider-yandex.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterList", version: "2.0.32" };
function dataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterExternalClusterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterExternalClusterToTerraform = dataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterExternalClusterToTerraform;
class DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterExternalClusterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // bootstrap_servers - computed: true, optional: false, required: false
    get bootstrapServers() {
        return this.getStringAttribute('bootstrap_servers');
    }
    // sasl_mechanism - computed: true, optional: false, required: false
    get saslMechanism() {
        return this.getStringAttribute('sasl_mechanism');
    }
    // sasl_password - computed: true, optional: false, required: false
    get saslPassword() {
        return this.getStringAttribute('sasl_password');
    }
    // sasl_username - computed: true, optional: false, required: false
    get saslUsername() {
        return this.getStringAttribute('sasl_username');
    }
    // security_protocol - computed: true, optional: false, required: false
    get securityProtocol() {
        return this.getStringAttribute('security_protocol');
    }
}
exports.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterExternalClusterOutputReference = DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterExternalClusterOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterExternalClusterOutputReference[_g] = { fqn: "@cdktf/provider-yandex.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterExternalClusterOutputReference", version: "2.0.32" };
class DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterExternalClusterList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterExternalClusterOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterExternalClusterList = DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterExternalClusterList;
_h = JSII_RTTI_SYMBOL_1;
DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterExternalClusterList[_h] = { fqn: "@cdktf/provider-yandex.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterExternalClusterList", version: "2.0.32" };
function dataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterThisClusterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterThisClusterToTerraform = dataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterThisClusterToTerraform;
class DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterThisClusterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
}
exports.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterThisClusterOutputReference = DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterThisClusterOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterThisClusterOutputReference[_j] = { fqn: "@cdktf/provider-yandex.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterThisClusterOutputReference", version: "2.0.32" };
class DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterThisClusterList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterThisClusterOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterThisClusterList = DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterThisClusterList;
_k = JSII_RTTI_SYMBOL_1;
DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterThisClusterList[_k] = { fqn: "@cdktf/provider-yandex.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterThisClusterList", version: "2.0.32" };
function dataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterToTerraform = dataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterToTerraform;
class DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // external_cluster - computed: true, optional: false, required: false
        this._externalCluster = new DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterExternalClusterList(this, "external_cluster", false);
        // this_cluster - computed: true, optional: false, required: false
        this._thisCluster = new DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterThisClusterList(this, "this_cluster", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // alias - computed: true, optional: false, required: false
    get alias() {
        return this.getStringAttribute('alias');
    }
    get externalCluster() {
        return this._externalCluster;
    }
    get thisCluster() {
        return this._thisCluster;
    }
}
exports.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterOutputReference = DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterOutputReference[_l] = { fqn: "@cdktf/provider-yandex.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterOutputReference", version: "2.0.32" };
class DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterList = DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterList;
_m = JSII_RTTI_SYMBOL_1;
DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterList[_m] = { fqn: "@cdktf/provider-yandex.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterList", version: "2.0.32" };
function dataYandexMdbKafkaConnectorConnectorConfigMirrormakerToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexMdbKafkaConnectorConnectorConfigMirrormakerToTerraform = dataYandexMdbKafkaConnectorConnectorConfigMirrormakerToTerraform;
class DataYandexMdbKafkaConnectorConnectorConfigMirrormakerOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // source_cluster - computed: true, optional: false, required: false
        this._sourceCluster = new DataYandexMdbKafkaConnectorConnectorConfigMirrormakerSourceClusterList(this, "source_cluster", false);
        // target_cluster - computed: true, optional: false, required: false
        this._targetCluster = new DataYandexMdbKafkaConnectorConnectorConfigMirrormakerTargetClusterList(this, "target_cluster", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // replication_factor - computed: true, optional: false, required: false
    get replicationFactor() {
        return this.getNumberAttribute('replication_factor');
    }
    get sourceCluster() {
        return this._sourceCluster;
    }
    get targetCluster() {
        return this._targetCluster;
    }
    // topics - computed: true, optional: false, required: false
    get topics() {
        return this.getStringAttribute('topics');
    }
}
exports.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerOutputReference = DataYandexMdbKafkaConnectorConnectorConfigMirrormakerOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataYandexMdbKafkaConnectorConnectorConfigMirrormakerOutputReference[_o] = { fqn: "@cdktf/provider-yandex.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerOutputReference", version: "2.0.32" };
class DataYandexMdbKafkaConnectorConnectorConfigMirrormakerList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbKafkaConnectorConnectorConfigMirrormakerOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerList = DataYandexMdbKafkaConnectorConnectorConfigMirrormakerList;
_p = JSII_RTTI_SYMBOL_1;
DataYandexMdbKafkaConnectorConnectorConfigMirrormakerList[_p] = { fqn: "@cdktf/provider-yandex.DataYandexMdbKafkaConnectorConnectorConfigMirrormakerList", version: "2.0.32" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/d/mdb_kafka_connector yandex_mdb_kafka_connector}
*/
class DataYandexMdbKafkaConnector extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/d/mdb_kafka_connector yandex_mdb_kafka_connector} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataYandexMdbKafkaConnectorConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'yandex_mdb_kafka_connector',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // connector_config_mirrormaker - computed: true, optional: false, required: false
        this._connectorConfigMirrormaker = new DataYandexMdbKafkaConnectorConnectorConfigMirrormakerList(this, "connector_config_mirrormaker", false);
        // properties - computed: true, optional: false, required: false
        this._properties = new cdktf.StringMap(this, "properties");
        this._clusterId = config.clusterId;
        this._id = config.id;
        this._name = config.name;
    }
    get clusterId() {
        return this.getStringAttribute('cluster_id');
    }
    set clusterId(value) {
        this._clusterId = value;
    }
    // Temporarily expose input value. Use with caution.
    get clusterIdInput() {
        return this._clusterId;
    }
    get connectorConfigMirrormaker() {
        return this._connectorConfigMirrormaker;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get properties() {
        return this._properties;
    }
    // tasks_max - computed: true, optional: false, required: false
    get tasksMax() {
        return this.getNumberAttribute('tasks_max');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            cluster_id: cdktf.stringToTerraform(this._clusterId),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
        };
    }
}
exports.DataYandexMdbKafkaConnector = DataYandexMdbKafkaConnector;
_q = JSII_RTTI_SYMBOL_1;
DataYandexMdbKafkaConnector[_q] = { fqn: "@cdktf/provider-yandex.DataYandexMdbKafkaConnector", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
DataYandexMdbKafkaConnector.tfResourceType = "yandex_mdb_kafka_connector";
