// @cdktf/provider-yandex/lib/data-yandex-mdb-redis-cluster.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataYandexMdbRedisCluster = exports.DataYandexMdbRedisClusterResourcesList = exports.DataYandexMdbRedisClusterResourcesOutputReference = exports.dataYandexMdbRedisClusterResourcesToTerraform = exports.DataYandexMdbRedisClusterMaintenanceWindowList = exports.DataYandexMdbRedisClusterMaintenanceWindowOutputReference = exports.dataYandexMdbRedisClusterMaintenanceWindowToTerraform = exports.DataYandexMdbRedisClusterHostList = exports.DataYandexMdbRedisClusterHostOutputReference = exports.dataYandexMdbRedisClusterHostToTerraform = exports.DataYandexMdbRedisClusterConfigAList = exports.DataYandexMdbRedisClusterConfigAOutputReference = exports.dataYandexMdbRedisClusterConfigAToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataYandexMdbRedisClusterConfigAToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexMdbRedisClusterConfigAToTerraform = dataYandexMdbRedisClusterConfigAToTerraform;
class DataYandexMdbRedisClusterConfigAOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // client_output_buffer_limit_normal - computed: true, optional: false, required: false
    get clientOutputBufferLimitNormal() {
        return this.getStringAttribute('client_output_buffer_limit_normal');
    }
    // client_output_buffer_limit_pubsub - computed: true, optional: false, required: false
    get clientOutputBufferLimitPubsub() {
        return this.getStringAttribute('client_output_buffer_limit_pubsub');
    }
    // databases - computed: true, optional: false, required: false
    get databases() {
        return this.getNumberAttribute('databases');
    }
    // maxmemory_policy - computed: true, optional: false, required: false
    get maxmemoryPolicy() {
        return this.getStringAttribute('maxmemory_policy');
    }
    // notify_keyspace_events - computed: true, optional: false, required: false
    get notifyKeyspaceEvents() {
        return this.getStringAttribute('notify_keyspace_events');
    }
    // slowlog_log_slower_than - computed: true, optional: false, required: false
    get slowlogLogSlowerThan() {
        return this.getNumberAttribute('slowlog_log_slower_than');
    }
    // slowlog_max_len - computed: true, optional: false, required: false
    get slowlogMaxLen() {
        return this.getNumberAttribute('slowlog_max_len');
    }
    // timeout - computed: true, optional: false, required: false
    get timeout() {
        return this.getNumberAttribute('timeout');
    }
    // version - computed: true, optional: false, required: false
    get version() {
        return this.getStringAttribute('version');
    }
}
exports.DataYandexMdbRedisClusterConfigAOutputReference = DataYandexMdbRedisClusterConfigAOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataYandexMdbRedisClusterConfigAOutputReference[_a] = { fqn: "@cdktf/provider-yandex.DataYandexMdbRedisClusterConfigAOutputReference", version: "2.0.32" };
class DataYandexMdbRedisClusterConfigAList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbRedisClusterConfigAOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbRedisClusterConfigAList = DataYandexMdbRedisClusterConfigAList;
_b = JSII_RTTI_SYMBOL_1;
DataYandexMdbRedisClusterConfigAList[_b] = { fqn: "@cdktf/provider-yandex.DataYandexMdbRedisClusterConfigAList", version: "2.0.32" };
function dataYandexMdbRedisClusterHostToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexMdbRedisClusterHostToTerraform = dataYandexMdbRedisClusterHostToTerraform;
class DataYandexMdbRedisClusterHostOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // assign_public_ip - computed: true, optional: false, required: false
    get assignPublicIp() {
        return this.getBooleanAttribute('assign_public_ip');
    }
    // fqdn - computed: true, optional: false, required: false
    get fqdn() {
        return this.getStringAttribute('fqdn');
    }
    // replica_priority - computed: true, optional: false, required: false
    get replicaPriority() {
        return this.getNumberAttribute('replica_priority');
    }
    // shard_name - computed: true, optional: false, required: false
    get shardName() {
        return this.getStringAttribute('shard_name');
    }
    // subnet_id - computed: true, optional: false, required: false
    get subnetId() {
        return this.getStringAttribute('subnet_id');
    }
    // zone - computed: true, optional: false, required: false
    get zone() {
        return this.getStringAttribute('zone');
    }
}
exports.DataYandexMdbRedisClusterHostOutputReference = DataYandexMdbRedisClusterHostOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataYandexMdbRedisClusterHostOutputReference[_c] = { fqn: "@cdktf/provider-yandex.DataYandexMdbRedisClusterHostOutputReference", version: "2.0.32" };
class DataYandexMdbRedisClusterHostList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbRedisClusterHostOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbRedisClusterHostList = DataYandexMdbRedisClusterHostList;
_d = JSII_RTTI_SYMBOL_1;
DataYandexMdbRedisClusterHostList[_d] = { fqn: "@cdktf/provider-yandex.DataYandexMdbRedisClusterHostList", version: "2.0.32" };
function dataYandexMdbRedisClusterMaintenanceWindowToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexMdbRedisClusterMaintenanceWindowToTerraform = dataYandexMdbRedisClusterMaintenanceWindowToTerraform;
class DataYandexMdbRedisClusterMaintenanceWindowOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // day - computed: true, optional: false, required: false
    get day() {
        return this.getStringAttribute('day');
    }
    // hour - computed: true, optional: false, required: false
    get hour() {
        return this.getNumberAttribute('hour');
    }
    // type - computed: true, optional: false, required: false
    get type() {
        return this.getStringAttribute('type');
    }
}
exports.DataYandexMdbRedisClusterMaintenanceWindowOutputReference = DataYandexMdbRedisClusterMaintenanceWindowOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataYandexMdbRedisClusterMaintenanceWindowOutputReference[_e] = { fqn: "@cdktf/provider-yandex.DataYandexMdbRedisClusterMaintenanceWindowOutputReference", version: "2.0.32" };
class DataYandexMdbRedisClusterMaintenanceWindowList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbRedisClusterMaintenanceWindowOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbRedisClusterMaintenanceWindowList = DataYandexMdbRedisClusterMaintenanceWindowList;
_f = JSII_RTTI_SYMBOL_1;
DataYandexMdbRedisClusterMaintenanceWindowList[_f] = { fqn: "@cdktf/provider-yandex.DataYandexMdbRedisClusterMaintenanceWindowList", version: "2.0.32" };
function dataYandexMdbRedisClusterResourcesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexMdbRedisClusterResourcesToTerraform = dataYandexMdbRedisClusterResourcesToTerraform;
class DataYandexMdbRedisClusterResourcesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // disk_size - computed: true, optional: false, required: false
    get diskSize() {
        return this.getNumberAttribute('disk_size');
    }
    // disk_type_id - computed: true, optional: false, required: false
    get diskTypeId() {
        return this.getStringAttribute('disk_type_id');
    }
    // resource_preset_id - computed: true, optional: false, required: false
    get resourcePresetId() {
        return this.getStringAttribute('resource_preset_id');
    }
}
exports.DataYandexMdbRedisClusterResourcesOutputReference = DataYandexMdbRedisClusterResourcesOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataYandexMdbRedisClusterResourcesOutputReference[_g] = { fqn: "@cdktf/provider-yandex.DataYandexMdbRedisClusterResourcesOutputReference", version: "2.0.32" };
class DataYandexMdbRedisClusterResourcesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbRedisClusterResourcesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbRedisClusterResourcesList = DataYandexMdbRedisClusterResourcesList;
_h = JSII_RTTI_SYMBOL_1;
DataYandexMdbRedisClusterResourcesList[_h] = { fqn: "@cdktf/provider-yandex.DataYandexMdbRedisClusterResourcesList", version: "2.0.32" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/d/mdb_redis_cluster yandex_mdb_redis_cluster}
*/
class DataYandexMdbRedisCluster extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/d/mdb_redis_cluster yandex_mdb_redis_cluster} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataYandexMdbRedisClusterConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'yandex_mdb_redis_cluster',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // config - computed: true, optional: false, required: false
        this._config = new DataYandexMdbRedisClusterConfigAList(this, "config", false);
        // host - computed: true, optional: false, required: false
        this._host = new DataYandexMdbRedisClusterHostList(this, "host", false);
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        // maintenance_window - computed: true, optional: false, required: false
        this._maintenanceWindow = new DataYandexMdbRedisClusterMaintenanceWindowList(this, "maintenance_window", false);
        // resources - computed: true, optional: false, required: false
        this._resources = new DataYandexMdbRedisClusterResourcesList(this, "resources", false);
        this._clusterId = config.clusterId;
        this._deletionProtection = config.deletionProtection;
        this._folderId = config.folderId;
        this._id = config.id;
        this._name = config.name;
    }
    get clusterId() {
        return this.getStringAttribute('cluster_id');
    }
    set clusterId(value) {
        this._clusterId = value;
    }
    resetClusterId() {
        this._clusterId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clusterIdInput() {
        return this._clusterId;
    }
    get config() {
        return this._config;
    }
    // created_at - computed: true, optional: false, required: false
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    get deletionProtection() {
        return this.getBooleanAttribute('deletion_protection');
    }
    set deletionProtection(value) {
        this._deletionProtection = value;
    }
    resetDeletionProtection() {
        this._deletionProtection = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deletionProtectionInput() {
        return this._deletionProtection;
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    // environment - computed: true, optional: false, required: false
    get environment() {
        return this.getStringAttribute('environment');
    }
    get folderId() {
        return this.getStringAttribute('folder_id');
    }
    set folderId(value) {
        this._folderId = value;
    }
    resetFolderId() {
        this._folderId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get folderIdInput() {
        return this._folderId;
    }
    // health - computed: true, optional: false, required: false
    get health() {
        return this.getStringAttribute('health');
    }
    get host() {
        return this._host;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this._labels;
    }
    get maintenanceWindow() {
        return this._maintenanceWindow;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // network_id - computed: true, optional: false, required: false
    get networkId() {
        return this.getStringAttribute('network_id');
    }
    // persistence_mode - computed: true, optional: false, required: false
    get persistenceMode() {
        return this.getStringAttribute('persistence_mode');
    }
    get resources() {
        return this._resources;
    }
    // security_group_ids - computed: true, optional: false, required: false
    get securityGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('security_group_ids'));
    }
    // sharded - computed: true, optional: false, required: false
    get sharded() {
        return this.getBooleanAttribute('sharded');
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    // tls_enabled - computed: true, optional: false, required: false
    get tlsEnabled() {
        return this.getBooleanAttribute('tls_enabled');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            cluster_id: cdktf.stringToTerraform(this._clusterId),
            deletion_protection: cdktf.booleanToTerraform(this._deletionProtection),
            folder_id: cdktf.stringToTerraform(this._folderId),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
        };
    }
}
exports.DataYandexMdbRedisCluster = DataYandexMdbRedisCluster;
_j = JSII_RTTI_SYMBOL_1;
DataYandexMdbRedisCluster[_j] = { fqn: "@cdktf/provider-yandex.DataYandexMdbRedisCluster", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
DataYandexMdbRedisCluster.tfResourceType = "yandex_mdb_redis_cluster";
