// @cdktf/provider-yandex/lib/data-yandex-function-trigger.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataYandexFunctionTrigger = exports.DataYandexFunctionTriggerTimerList = exports.DataYandexFunctionTriggerTimerOutputReference = exports.dataYandexFunctionTriggerTimerToTerraform = exports.DataYandexFunctionTriggerObjectStorageList = exports.DataYandexFunctionTriggerObjectStorageOutputReference = exports.dataYandexFunctionTriggerObjectStorageToTerraform = exports.DataYandexFunctionTriggerMessageQueueList = exports.DataYandexFunctionTriggerMessageQueueOutputReference = exports.dataYandexFunctionTriggerMessageQueueToTerraform = exports.DataYandexFunctionTriggerLoggingList = exports.DataYandexFunctionTriggerLoggingOutputReference = exports.dataYandexFunctionTriggerLoggingToTerraform = exports.DataYandexFunctionTriggerLogGroupList = exports.DataYandexFunctionTriggerLogGroupOutputReference = exports.dataYandexFunctionTriggerLogGroupToTerraform = exports.DataYandexFunctionTriggerIotList = exports.DataYandexFunctionTriggerIotOutputReference = exports.dataYandexFunctionTriggerIotToTerraform = exports.DataYandexFunctionTriggerFunctionList = exports.DataYandexFunctionTriggerFunctionOutputReference = exports.dataYandexFunctionTriggerFunctionToTerraform = exports.DataYandexFunctionTriggerDlqList = exports.DataYandexFunctionTriggerDlqOutputReference = exports.dataYandexFunctionTriggerDlqToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataYandexFunctionTriggerDlqToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexFunctionTriggerDlqToTerraform = dataYandexFunctionTriggerDlqToTerraform;
class DataYandexFunctionTriggerDlqOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // queue_id - computed: true, optional: false, required: false
    get queueId() {
        return this.getStringAttribute('queue_id');
    }
    // service_account_id - computed: true, optional: false, required: false
    get serviceAccountId() {
        return this.getStringAttribute('service_account_id');
    }
}
exports.DataYandexFunctionTriggerDlqOutputReference = DataYandexFunctionTriggerDlqOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataYandexFunctionTriggerDlqOutputReference[_a] = { fqn: "@cdktf/provider-yandex.DataYandexFunctionTriggerDlqOutputReference", version: "2.0.32" };
class DataYandexFunctionTriggerDlqList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexFunctionTriggerDlqOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexFunctionTriggerDlqList = DataYandexFunctionTriggerDlqList;
_b = JSII_RTTI_SYMBOL_1;
DataYandexFunctionTriggerDlqList[_b] = { fqn: "@cdktf/provider-yandex.DataYandexFunctionTriggerDlqList", version: "2.0.32" };
function dataYandexFunctionTriggerFunctionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexFunctionTriggerFunctionToTerraform = dataYandexFunctionTriggerFunctionToTerraform;
class DataYandexFunctionTriggerFunctionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    // retry_attempts - computed: true, optional: false, required: false
    get retryAttempts() {
        return this.getStringAttribute('retry_attempts');
    }
    // retry_interval - computed: true, optional: false, required: false
    get retryInterval() {
        return this.getStringAttribute('retry_interval');
    }
    // service_account_id - computed: true, optional: false, required: false
    get serviceAccountId() {
        return this.getStringAttribute('service_account_id');
    }
    // tag - computed: true, optional: false, required: false
    get tag() {
        return this.getStringAttribute('tag');
    }
}
exports.DataYandexFunctionTriggerFunctionOutputReference = DataYandexFunctionTriggerFunctionOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataYandexFunctionTriggerFunctionOutputReference[_c] = { fqn: "@cdktf/provider-yandex.DataYandexFunctionTriggerFunctionOutputReference", version: "2.0.32" };
class DataYandexFunctionTriggerFunctionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexFunctionTriggerFunctionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexFunctionTriggerFunctionList = DataYandexFunctionTriggerFunctionList;
_d = JSII_RTTI_SYMBOL_1;
DataYandexFunctionTriggerFunctionList[_d] = { fqn: "@cdktf/provider-yandex.DataYandexFunctionTriggerFunctionList", version: "2.0.32" };
function dataYandexFunctionTriggerIotToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexFunctionTriggerIotToTerraform = dataYandexFunctionTriggerIotToTerraform;
class DataYandexFunctionTriggerIotOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // device_id - computed: true, optional: false, required: false
    get deviceId() {
        return this.getStringAttribute('device_id');
    }
    // registry_id - computed: true, optional: false, required: false
    get registryId() {
        return this.getStringAttribute('registry_id');
    }
    // topic - computed: true, optional: false, required: false
    get topic() {
        return this.getStringAttribute('topic');
    }
}
exports.DataYandexFunctionTriggerIotOutputReference = DataYandexFunctionTriggerIotOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataYandexFunctionTriggerIotOutputReference[_e] = { fqn: "@cdktf/provider-yandex.DataYandexFunctionTriggerIotOutputReference", version: "2.0.32" };
class DataYandexFunctionTriggerIotList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexFunctionTriggerIotOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexFunctionTriggerIotList = DataYandexFunctionTriggerIotList;
_f = JSII_RTTI_SYMBOL_1;
DataYandexFunctionTriggerIotList[_f] = { fqn: "@cdktf/provider-yandex.DataYandexFunctionTriggerIotList", version: "2.0.32" };
function dataYandexFunctionTriggerLogGroupToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexFunctionTriggerLogGroupToTerraform = dataYandexFunctionTriggerLogGroupToTerraform;
class DataYandexFunctionTriggerLogGroupOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // batch_cutoff - computed: true, optional: false, required: false
    get batchCutoff() {
        return this.getStringAttribute('batch_cutoff');
    }
    // batch_size - computed: true, optional: false, required: false
    get batchSize() {
        return this.getStringAttribute('batch_size');
    }
    // log_group_ids - computed: true, optional: false, required: false
    get logGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('log_group_ids'));
    }
}
exports.DataYandexFunctionTriggerLogGroupOutputReference = DataYandexFunctionTriggerLogGroupOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataYandexFunctionTriggerLogGroupOutputReference[_g] = { fqn: "@cdktf/provider-yandex.DataYandexFunctionTriggerLogGroupOutputReference", version: "2.0.32" };
class DataYandexFunctionTriggerLogGroupList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexFunctionTriggerLogGroupOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexFunctionTriggerLogGroupList = DataYandexFunctionTriggerLogGroupList;
_h = JSII_RTTI_SYMBOL_1;
DataYandexFunctionTriggerLogGroupList[_h] = { fqn: "@cdktf/provider-yandex.DataYandexFunctionTriggerLogGroupList", version: "2.0.32" };
function dataYandexFunctionTriggerLoggingToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexFunctionTriggerLoggingToTerraform = dataYandexFunctionTriggerLoggingToTerraform;
class DataYandexFunctionTriggerLoggingOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // batch_cutoff - computed: true, optional: false, required: false
    get batchCutoff() {
        return this.getStringAttribute('batch_cutoff');
    }
    // batch_size - computed: true, optional: false, required: false
    get batchSize() {
        return this.getStringAttribute('batch_size');
    }
    // group_id - computed: true, optional: false, required: false
    get groupId() {
        return this.getStringAttribute('group_id');
    }
    // levels - computed: true, optional: false, required: false
    get levels() {
        return cdktf.Fn.tolist(this.getListAttribute('levels'));
    }
    // resource_ids - computed: true, optional: false, required: false
    get resourceIds() {
        return cdktf.Fn.tolist(this.getListAttribute('resource_ids'));
    }
    // resource_types - computed: true, optional: false, required: false
    get resourceTypes() {
        return cdktf.Fn.tolist(this.getListAttribute('resource_types'));
    }
}
exports.DataYandexFunctionTriggerLoggingOutputReference = DataYandexFunctionTriggerLoggingOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataYandexFunctionTriggerLoggingOutputReference[_j] = { fqn: "@cdktf/provider-yandex.DataYandexFunctionTriggerLoggingOutputReference", version: "2.0.32" };
class DataYandexFunctionTriggerLoggingList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexFunctionTriggerLoggingOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexFunctionTriggerLoggingList = DataYandexFunctionTriggerLoggingList;
_k = JSII_RTTI_SYMBOL_1;
DataYandexFunctionTriggerLoggingList[_k] = { fqn: "@cdktf/provider-yandex.DataYandexFunctionTriggerLoggingList", version: "2.0.32" };
function dataYandexFunctionTriggerMessageQueueToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexFunctionTriggerMessageQueueToTerraform = dataYandexFunctionTriggerMessageQueueToTerraform;
class DataYandexFunctionTriggerMessageQueueOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // batch_cutoff - computed: true, optional: false, required: false
    get batchCutoff() {
        return this.getStringAttribute('batch_cutoff');
    }
    // batch_size - computed: true, optional: false, required: false
    get batchSize() {
        return this.getStringAttribute('batch_size');
    }
    // queue_id - computed: true, optional: false, required: false
    get queueId() {
        return this.getStringAttribute('queue_id');
    }
    // service_account_id - computed: true, optional: false, required: false
    get serviceAccountId() {
        return this.getStringAttribute('service_account_id');
    }
    // visibility_timeout - computed: true, optional: false, required: false
    get visibilityTimeout() {
        return this.getStringAttribute('visibility_timeout');
    }
}
exports.DataYandexFunctionTriggerMessageQueueOutputReference = DataYandexFunctionTriggerMessageQueueOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataYandexFunctionTriggerMessageQueueOutputReference[_l] = { fqn: "@cdktf/provider-yandex.DataYandexFunctionTriggerMessageQueueOutputReference", version: "2.0.32" };
class DataYandexFunctionTriggerMessageQueueList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexFunctionTriggerMessageQueueOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexFunctionTriggerMessageQueueList = DataYandexFunctionTriggerMessageQueueList;
_m = JSII_RTTI_SYMBOL_1;
DataYandexFunctionTriggerMessageQueueList[_m] = { fqn: "@cdktf/provider-yandex.DataYandexFunctionTriggerMessageQueueList", version: "2.0.32" };
function dataYandexFunctionTriggerObjectStorageToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexFunctionTriggerObjectStorageToTerraform = dataYandexFunctionTriggerObjectStorageToTerraform;
class DataYandexFunctionTriggerObjectStorageOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // bucket_id - computed: true, optional: false, required: false
    get bucketId() {
        return this.getStringAttribute('bucket_id');
    }
    // create - computed: true, optional: false, required: false
    get create() {
        return this.getBooleanAttribute('create');
    }
    // delete - computed: true, optional: false, required: false
    get delete() {
        return this.getBooleanAttribute('delete');
    }
    // prefix - computed: true, optional: false, required: false
    get prefix() {
        return this.getStringAttribute('prefix');
    }
    // suffix - computed: true, optional: false, required: false
    get suffix() {
        return this.getStringAttribute('suffix');
    }
    // update - computed: true, optional: false, required: false
    get update() {
        return this.getBooleanAttribute('update');
    }
}
exports.DataYandexFunctionTriggerObjectStorageOutputReference = DataYandexFunctionTriggerObjectStorageOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataYandexFunctionTriggerObjectStorageOutputReference[_o] = { fqn: "@cdktf/provider-yandex.DataYandexFunctionTriggerObjectStorageOutputReference", version: "2.0.32" };
class DataYandexFunctionTriggerObjectStorageList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexFunctionTriggerObjectStorageOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexFunctionTriggerObjectStorageList = DataYandexFunctionTriggerObjectStorageList;
_p = JSII_RTTI_SYMBOL_1;
DataYandexFunctionTriggerObjectStorageList[_p] = { fqn: "@cdktf/provider-yandex.DataYandexFunctionTriggerObjectStorageList", version: "2.0.32" };
function dataYandexFunctionTriggerTimerToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexFunctionTriggerTimerToTerraform = dataYandexFunctionTriggerTimerToTerraform;
class DataYandexFunctionTriggerTimerOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // cron_expression - computed: true, optional: false, required: false
    get cronExpression() {
        return this.getStringAttribute('cron_expression');
    }
}
exports.DataYandexFunctionTriggerTimerOutputReference = DataYandexFunctionTriggerTimerOutputReference;
_q = JSII_RTTI_SYMBOL_1;
DataYandexFunctionTriggerTimerOutputReference[_q] = { fqn: "@cdktf/provider-yandex.DataYandexFunctionTriggerTimerOutputReference", version: "2.0.32" };
class DataYandexFunctionTriggerTimerList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexFunctionTriggerTimerOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexFunctionTriggerTimerList = DataYandexFunctionTriggerTimerList;
_r = JSII_RTTI_SYMBOL_1;
DataYandexFunctionTriggerTimerList[_r] = { fqn: "@cdktf/provider-yandex.DataYandexFunctionTriggerTimerList", version: "2.0.32" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/d/function_trigger yandex_function_trigger}
*/
class DataYandexFunctionTrigger extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/d/function_trigger yandex_function_trigger} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataYandexFunctionTriggerConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'yandex_function_trigger',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // dlq - computed: true, optional: false, required: false
        this._dlq = new DataYandexFunctionTriggerDlqList(this, "dlq", false);
        // function - computed: true, optional: false, required: false
        this._function = new DataYandexFunctionTriggerFunctionList(this, "function", false);
        // iot - computed: true, optional: false, required: false
        this._iot = new DataYandexFunctionTriggerIotList(this, "iot", false);
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        // log_group - computed: true, optional: false, required: false
        this._logGroup = new DataYandexFunctionTriggerLogGroupList(this, "log_group", false);
        // logging - computed: true, optional: false, required: false
        this._logging = new DataYandexFunctionTriggerLoggingList(this, "logging", false);
        // message_queue - computed: true, optional: false, required: false
        this._messageQueue = new DataYandexFunctionTriggerMessageQueueList(this, "message_queue", false);
        // object_storage - computed: true, optional: false, required: false
        this._objectStorage = new DataYandexFunctionTriggerObjectStorageList(this, "object_storage", false);
        // timer - computed: true, optional: false, required: false
        this._timer = new DataYandexFunctionTriggerTimerList(this, "timer", false);
        this._folderId = config.folderId;
        this._id = config.id;
        this._name = config.name;
        this._triggerId = config.triggerId;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // created_at - computed: true, optional: false, required: false
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    get dlq() {
        return this._dlq;
    }
    get folderId() {
        return this.getStringAttribute('folder_id');
    }
    set folderId(value) {
        this._folderId = value;
    }
    resetFolderId() {
        this._folderId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get folderIdInput() {
        return this._folderId;
    }
    get function() {
        return this._function;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get iot() {
        return this._iot;
    }
    get labels() {
        return this._labels;
    }
    get logGroup() {
        return this._logGroup;
    }
    get logging() {
        return this._logging;
    }
    get messageQueue() {
        return this._messageQueue;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get objectStorage() {
        return this._objectStorage;
    }
    get timer() {
        return this._timer;
    }
    get triggerId() {
        return this.getStringAttribute('trigger_id');
    }
    set triggerId(value) {
        this._triggerId = value;
    }
    resetTriggerId() {
        this._triggerId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get triggerIdInput() {
        return this._triggerId;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            folder_id: cdktf.stringToTerraform(this._folderId),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            trigger_id: cdktf.stringToTerraform(this._triggerId),
        };
    }
}
exports.DataYandexFunctionTrigger = DataYandexFunctionTrigger;
_s = JSII_RTTI_SYMBOL_1;
DataYandexFunctionTrigger[_s] = { fqn: "@cdktf/provider-yandex.DataYandexFunctionTrigger", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
DataYandexFunctionTrigger.tfResourceType = "yandex_function_trigger";
