// @cdktf/provider-yandex/lib/data-yandex-mdb-mongodb-cluster.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataYandexMdbMongodbCluster = exports.DataYandexMdbMongodbClusterUserList = exports.DataYandexMdbMongodbClusterUserOutputReference = exports.dataYandexMdbMongodbClusterUserToTerraform = exports.DataYandexMdbMongodbClusterUserPermissionList = exports.DataYandexMdbMongodbClusterUserPermissionOutputReference = exports.dataYandexMdbMongodbClusterUserPermissionToTerraform = exports.DataYandexMdbMongodbClusterResourcesOutputReference = exports.dataYandexMdbMongodbClusterResourcesToTerraform = exports.DataYandexMdbMongodbClusterMaintenanceWindowOutputReference = exports.dataYandexMdbMongodbClusterMaintenanceWindowToTerraform = exports.DataYandexMdbMongodbClusterHostList = exports.DataYandexMdbMongodbClusterHostOutputReference = exports.dataYandexMdbMongodbClusterHostToTerraform = exports.DataYandexMdbMongodbClusterDatabaseList = exports.DataYandexMdbMongodbClusterDatabaseOutputReference = exports.dataYandexMdbMongodbClusterDatabaseToTerraform = exports.DataYandexMdbMongodbClusterClusterConfigOutputReference = exports.dataYandexMdbMongodbClusterClusterConfigToTerraform = exports.DataYandexMdbMongodbClusterClusterConfigMongodOutputReference = exports.dataYandexMdbMongodbClusterClusterConfigMongodToTerraform = exports.DataYandexMdbMongodbClusterClusterConfigMongodSetParameterOutputReference = exports.dataYandexMdbMongodbClusterClusterConfigMongodSetParameterToTerraform = exports.DataYandexMdbMongodbClusterClusterConfigMongodSecurityOutputReference = exports.dataYandexMdbMongodbClusterClusterConfigMongodSecurityToTerraform = exports.DataYandexMdbMongodbClusterClusterConfigMongodSecurityKmipOutputReference = exports.dataYandexMdbMongodbClusterClusterConfigMongodSecurityKmipToTerraform = exports.DataYandexMdbMongodbClusterClusterConfigMongodAuditLogOutputReference = exports.dataYandexMdbMongodbClusterClusterConfigMongodAuditLogToTerraform = exports.DataYandexMdbMongodbClusterClusterConfigBackupWindowStartOutputReference = exports.dataYandexMdbMongodbClusterClusterConfigBackupWindowStartToTerraform = exports.DataYandexMdbMongodbClusterClusterConfigAccessOutputReference = exports.dataYandexMdbMongodbClusterClusterConfigAccessToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataYandexMdbMongodbClusterClusterConfigAccessToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        data_lens: cdktf.booleanToTerraform(struct.dataLens),
    };
}
exports.dataYandexMdbMongodbClusterClusterConfigAccessToTerraform = dataYandexMdbMongodbClusterClusterConfigAccessToTerraform;
class DataYandexMdbMongodbClusterClusterConfigAccessOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._dataLens !== undefined) {
            hasAnyValues = true;
            internalValueResult.dataLens = this._dataLens;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._dataLens = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._dataLens = value.dataLens;
        }
    }
    get dataLens() {
        return this.getBooleanAttribute('data_lens');
    }
    set dataLens(value) {
        this._dataLens = value;
    }
    resetDataLens() {
        this._dataLens = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dataLensInput() {
        return this._dataLens;
    }
}
exports.DataYandexMdbMongodbClusterClusterConfigAccessOutputReference = DataYandexMdbMongodbClusterClusterConfigAccessOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataYandexMdbMongodbClusterClusterConfigAccessOutputReference[_a] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMongodbClusterClusterConfigAccessOutputReference", version: "2.0.32" };
function dataYandexMdbMongodbClusterClusterConfigBackupWindowStartToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        hours: cdktf.numberToTerraform(struct.hours),
        minutes: cdktf.numberToTerraform(struct.minutes),
    };
}
exports.dataYandexMdbMongodbClusterClusterConfigBackupWindowStartToTerraform = dataYandexMdbMongodbClusterClusterConfigBackupWindowStartToTerraform;
class DataYandexMdbMongodbClusterClusterConfigBackupWindowStartOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._hours !== undefined) {
            hasAnyValues = true;
            internalValueResult.hours = this._hours;
        }
        if (this._minutes !== undefined) {
            hasAnyValues = true;
            internalValueResult.minutes = this._minutes;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._hours = undefined;
            this._minutes = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._hours = value.hours;
            this._minutes = value.minutes;
        }
    }
    get hours() {
        return this.getNumberAttribute('hours');
    }
    set hours(value) {
        this._hours = value;
    }
    resetHours() {
        this._hours = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get hoursInput() {
        return this._hours;
    }
    get minutes() {
        return this.getNumberAttribute('minutes');
    }
    set minutes(value) {
        this._minutes = value;
    }
    resetMinutes() {
        this._minutes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get minutesInput() {
        return this._minutes;
    }
}
exports.DataYandexMdbMongodbClusterClusterConfigBackupWindowStartOutputReference = DataYandexMdbMongodbClusterClusterConfigBackupWindowStartOutputReference;
_b = JSII_RTTI_SYMBOL_1;
DataYandexMdbMongodbClusterClusterConfigBackupWindowStartOutputReference[_b] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMongodbClusterClusterConfigBackupWindowStartOutputReference", version: "2.0.32" };
function dataYandexMdbMongodbClusterClusterConfigMongodAuditLogToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        filter: cdktf.stringToTerraform(struct.filter),
        runtime_configuration: cdktf.booleanToTerraform(struct.runtimeConfiguration),
    };
}
exports.dataYandexMdbMongodbClusterClusterConfigMongodAuditLogToTerraform = dataYandexMdbMongodbClusterClusterConfigMongodAuditLogToTerraform;
class DataYandexMdbMongodbClusterClusterConfigMongodAuditLogOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._filter !== undefined) {
            hasAnyValues = true;
            internalValueResult.filter = this._filter;
        }
        if (this._runtimeConfiguration !== undefined) {
            hasAnyValues = true;
            internalValueResult.runtimeConfiguration = this._runtimeConfiguration;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._filter = undefined;
            this._runtimeConfiguration = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._filter = value.filter;
            this._runtimeConfiguration = value.runtimeConfiguration;
        }
    }
    get filter() {
        return this.getStringAttribute('filter');
    }
    set filter(value) {
        this._filter = value;
    }
    resetFilter() {
        this._filter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get filterInput() {
        return this._filter;
    }
    get runtimeConfiguration() {
        return this.getBooleanAttribute('runtime_configuration');
    }
    set runtimeConfiguration(value) {
        this._runtimeConfiguration = value;
    }
    resetRuntimeConfiguration() {
        this._runtimeConfiguration = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get runtimeConfigurationInput() {
        return this._runtimeConfiguration;
    }
}
exports.DataYandexMdbMongodbClusterClusterConfigMongodAuditLogOutputReference = DataYandexMdbMongodbClusterClusterConfigMongodAuditLogOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataYandexMdbMongodbClusterClusterConfigMongodAuditLogOutputReference[_c] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMongodbClusterClusterConfigMongodAuditLogOutputReference", version: "2.0.32" };
function dataYandexMdbMongodbClusterClusterConfigMongodSecurityKmipToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        client_certificate: cdktf.stringToTerraform(struct.clientCertificate),
        key_identifier: cdktf.stringToTerraform(struct.keyIdentifier),
        port: cdktf.numberToTerraform(struct.port),
        server_ca: cdktf.stringToTerraform(struct.serverCa),
        server_name: cdktf.stringToTerraform(struct.serverName),
    };
}
exports.dataYandexMdbMongodbClusterClusterConfigMongodSecurityKmipToTerraform = dataYandexMdbMongodbClusterClusterConfigMongodSecurityKmipToTerraform;
class DataYandexMdbMongodbClusterClusterConfigMongodSecurityKmipOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._clientCertificate !== undefined) {
            hasAnyValues = true;
            internalValueResult.clientCertificate = this._clientCertificate;
        }
        if (this._keyIdentifier !== undefined) {
            hasAnyValues = true;
            internalValueResult.keyIdentifier = this._keyIdentifier;
        }
        if (this._port !== undefined) {
            hasAnyValues = true;
            internalValueResult.port = this._port;
        }
        if (this._serverCa !== undefined) {
            hasAnyValues = true;
            internalValueResult.serverCa = this._serverCa;
        }
        if (this._serverName !== undefined) {
            hasAnyValues = true;
            internalValueResult.serverName = this._serverName;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._clientCertificate = undefined;
            this._keyIdentifier = undefined;
            this._port = undefined;
            this._serverCa = undefined;
            this._serverName = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._clientCertificate = value.clientCertificate;
            this._keyIdentifier = value.keyIdentifier;
            this._port = value.port;
            this._serverCa = value.serverCa;
            this._serverName = value.serverName;
        }
    }
    get clientCertificate() {
        return this.getStringAttribute('client_certificate');
    }
    set clientCertificate(value) {
        this._clientCertificate = value;
    }
    resetClientCertificate() {
        this._clientCertificate = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clientCertificateInput() {
        return this._clientCertificate;
    }
    get keyIdentifier() {
        return this.getStringAttribute('key_identifier');
    }
    set keyIdentifier(value) {
        this._keyIdentifier = value;
    }
    resetKeyIdentifier() {
        this._keyIdentifier = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keyIdentifierInput() {
        return this._keyIdentifier;
    }
    get port() {
        return this.getNumberAttribute('port');
    }
    set port(value) {
        this._port = value;
    }
    resetPort() {
        this._port = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get portInput() {
        return this._port;
    }
    get serverCa() {
        return this.getStringAttribute('server_ca');
    }
    set serverCa(value) {
        this._serverCa = value;
    }
    resetServerCa() {
        this._serverCa = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get serverCaInput() {
        return this._serverCa;
    }
    get serverName() {
        return this.getStringAttribute('server_name');
    }
    set serverName(value) {
        this._serverName = value;
    }
    resetServerName() {
        this._serverName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get serverNameInput() {
        return this._serverName;
    }
}
exports.DataYandexMdbMongodbClusterClusterConfigMongodSecurityKmipOutputReference = DataYandexMdbMongodbClusterClusterConfigMongodSecurityKmipOutputReference;
_d = JSII_RTTI_SYMBOL_1;
DataYandexMdbMongodbClusterClusterConfigMongodSecurityKmipOutputReference[_d] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMongodbClusterClusterConfigMongodSecurityKmipOutputReference", version: "2.0.32" };
function dataYandexMdbMongodbClusterClusterConfigMongodSecurityToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enable_encryption: cdktf.booleanToTerraform(struct.enableEncryption),
        kmip: dataYandexMdbMongodbClusterClusterConfigMongodSecurityKmipToTerraform(struct.kmip),
    };
}
exports.dataYandexMdbMongodbClusterClusterConfigMongodSecurityToTerraform = dataYandexMdbMongodbClusterClusterConfigMongodSecurityToTerraform;
class DataYandexMdbMongodbClusterClusterConfigMongodSecurityOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // kmip - computed: false, optional: true, required: false
        this._kmip = new DataYandexMdbMongodbClusterClusterConfigMongodSecurityKmipOutputReference(this, "kmip");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enableEncryption !== undefined) {
            hasAnyValues = true;
            internalValueResult.enableEncryption = this._enableEncryption;
        }
        if (this._kmip?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.kmip = this._kmip?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enableEncryption = undefined;
            this._kmip.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enableEncryption = value.enableEncryption;
            this._kmip.internalValue = value.kmip;
        }
    }
    get enableEncryption() {
        return this.getBooleanAttribute('enable_encryption');
    }
    set enableEncryption(value) {
        this._enableEncryption = value;
    }
    resetEnableEncryption() {
        this._enableEncryption = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableEncryptionInput() {
        return this._enableEncryption;
    }
    get kmip() {
        return this._kmip;
    }
    putKmip(value) {
        this._kmip.internalValue = value;
    }
    resetKmip() {
        this._kmip.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kmipInput() {
        return this._kmip.internalValue;
    }
}
exports.DataYandexMdbMongodbClusterClusterConfigMongodSecurityOutputReference = DataYandexMdbMongodbClusterClusterConfigMongodSecurityOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataYandexMdbMongodbClusterClusterConfigMongodSecurityOutputReference[_e] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMongodbClusterClusterConfigMongodSecurityOutputReference", version: "2.0.32" };
function dataYandexMdbMongodbClusterClusterConfigMongodSetParameterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        audit_authorization_success: cdktf.booleanToTerraform(struct.auditAuthorizationSuccess),
    };
}
exports.dataYandexMdbMongodbClusterClusterConfigMongodSetParameterToTerraform = dataYandexMdbMongodbClusterClusterConfigMongodSetParameterToTerraform;
class DataYandexMdbMongodbClusterClusterConfigMongodSetParameterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._auditAuthorizationSuccess !== undefined) {
            hasAnyValues = true;
            internalValueResult.auditAuthorizationSuccess = this._auditAuthorizationSuccess;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._auditAuthorizationSuccess = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._auditAuthorizationSuccess = value.auditAuthorizationSuccess;
        }
    }
    get auditAuthorizationSuccess() {
        return this.getBooleanAttribute('audit_authorization_success');
    }
    set auditAuthorizationSuccess(value) {
        this._auditAuthorizationSuccess = value;
    }
    resetAuditAuthorizationSuccess() {
        this._auditAuthorizationSuccess = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get auditAuthorizationSuccessInput() {
        return this._auditAuthorizationSuccess;
    }
}
exports.DataYandexMdbMongodbClusterClusterConfigMongodSetParameterOutputReference = DataYandexMdbMongodbClusterClusterConfigMongodSetParameterOutputReference;
_f = JSII_RTTI_SYMBOL_1;
DataYandexMdbMongodbClusterClusterConfigMongodSetParameterOutputReference[_f] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMongodbClusterClusterConfigMongodSetParameterOutputReference", version: "2.0.32" };
function dataYandexMdbMongodbClusterClusterConfigMongodToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        audit_log: dataYandexMdbMongodbClusterClusterConfigMongodAuditLogToTerraform(struct.auditLog),
        security: dataYandexMdbMongodbClusterClusterConfigMongodSecurityToTerraform(struct.security),
        set_parameter: dataYandexMdbMongodbClusterClusterConfigMongodSetParameterToTerraform(struct.setParameter),
    };
}
exports.dataYandexMdbMongodbClusterClusterConfigMongodToTerraform = dataYandexMdbMongodbClusterClusterConfigMongodToTerraform;
class DataYandexMdbMongodbClusterClusterConfigMongodOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // audit_log - computed: false, optional: true, required: false
        this._auditLog = new DataYandexMdbMongodbClusterClusterConfigMongodAuditLogOutputReference(this, "audit_log");
        // security - computed: false, optional: true, required: false
        this._security = new DataYandexMdbMongodbClusterClusterConfigMongodSecurityOutputReference(this, "security");
        // set_parameter - computed: false, optional: true, required: false
        this._setParameter = new DataYandexMdbMongodbClusterClusterConfigMongodSetParameterOutputReference(this, "set_parameter");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._auditLog?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.auditLog = this._auditLog?.internalValue;
        }
        if (this._security?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.security = this._security?.internalValue;
        }
        if (this._setParameter?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.setParameter = this._setParameter?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._auditLog.internalValue = undefined;
            this._security.internalValue = undefined;
            this._setParameter.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._auditLog.internalValue = value.auditLog;
            this._security.internalValue = value.security;
            this._setParameter.internalValue = value.setParameter;
        }
    }
    get auditLog() {
        return this._auditLog;
    }
    putAuditLog(value) {
        this._auditLog.internalValue = value;
    }
    resetAuditLog() {
        this._auditLog.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get auditLogInput() {
        return this._auditLog.internalValue;
    }
    get security() {
        return this._security;
    }
    putSecurity(value) {
        this._security.internalValue = value;
    }
    resetSecurity() {
        this._security.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityInput() {
        return this._security.internalValue;
    }
    get setParameter() {
        return this._setParameter;
    }
    putSetParameter(value) {
        this._setParameter.internalValue = value;
    }
    resetSetParameter() {
        this._setParameter.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get setParameterInput() {
        return this._setParameter.internalValue;
    }
}
exports.DataYandexMdbMongodbClusterClusterConfigMongodOutputReference = DataYandexMdbMongodbClusterClusterConfigMongodOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataYandexMdbMongodbClusterClusterConfigMongodOutputReference[_g] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMongodbClusterClusterConfigMongodOutputReference", version: "2.0.32" };
function dataYandexMdbMongodbClusterClusterConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        feature_compatibility_version: cdktf.stringToTerraform(struct.featureCompatibilityVersion),
        version: cdktf.stringToTerraform(struct.version),
        access: dataYandexMdbMongodbClusterClusterConfigAccessToTerraform(struct.access),
        backup_window_start: dataYandexMdbMongodbClusterClusterConfigBackupWindowStartToTerraform(struct.backupWindowStart),
        mongod: dataYandexMdbMongodbClusterClusterConfigMongodToTerraform(struct.mongod),
    };
}
exports.dataYandexMdbMongodbClusterClusterConfigToTerraform = dataYandexMdbMongodbClusterClusterConfigToTerraform;
class DataYandexMdbMongodbClusterClusterConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // access - computed: false, optional: true, required: false
        this._access = new DataYandexMdbMongodbClusterClusterConfigAccessOutputReference(this, "access");
        // backup_window_start - computed: false, optional: true, required: false
        this._backupWindowStart = new DataYandexMdbMongodbClusterClusterConfigBackupWindowStartOutputReference(this, "backup_window_start");
        // mongod - computed: false, optional: true, required: false
        this._mongod = new DataYandexMdbMongodbClusterClusterConfigMongodOutputReference(this, "mongod");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._featureCompatibilityVersion !== undefined) {
            hasAnyValues = true;
            internalValueResult.featureCompatibilityVersion = this._featureCompatibilityVersion;
        }
        if (this._version !== undefined) {
            hasAnyValues = true;
            internalValueResult.version = this._version;
        }
        if (this._access?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.access = this._access?.internalValue;
        }
        if (this._backupWindowStart?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.backupWindowStart = this._backupWindowStart?.internalValue;
        }
        if (this._mongod?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.mongod = this._mongod?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._featureCompatibilityVersion = undefined;
            this._version = undefined;
            this._access.internalValue = undefined;
            this._backupWindowStart.internalValue = undefined;
            this._mongod.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._featureCompatibilityVersion = value.featureCompatibilityVersion;
            this._version = value.version;
            this._access.internalValue = value.access;
            this._backupWindowStart.internalValue = value.backupWindowStart;
            this._mongod.internalValue = value.mongod;
        }
    }
    get featureCompatibilityVersion() {
        return this.getStringAttribute('feature_compatibility_version');
    }
    set featureCompatibilityVersion(value) {
        this._featureCompatibilityVersion = value;
    }
    resetFeatureCompatibilityVersion() {
        this._featureCompatibilityVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get featureCompatibilityVersionInput() {
        return this._featureCompatibilityVersion;
    }
    get version() {
        return this.getStringAttribute('version');
    }
    set version(value) {
        this._version = value;
    }
    resetVersion() {
        this._version = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get versionInput() {
        return this._version;
    }
    get access() {
        return this._access;
    }
    putAccess(value) {
        this._access.internalValue = value;
    }
    resetAccess() {
        this._access.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get accessInput() {
        return this._access.internalValue;
    }
    get backupWindowStart() {
        return this._backupWindowStart;
    }
    putBackupWindowStart(value) {
        this._backupWindowStart.internalValue = value;
    }
    resetBackupWindowStart() {
        this._backupWindowStart.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get backupWindowStartInput() {
        return this._backupWindowStart.internalValue;
    }
    get mongod() {
        return this._mongod;
    }
    putMongod(value) {
        this._mongod.internalValue = value;
    }
    resetMongod() {
        this._mongod.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get mongodInput() {
        return this._mongod.internalValue;
    }
}
exports.DataYandexMdbMongodbClusterClusterConfigOutputReference = DataYandexMdbMongodbClusterClusterConfigOutputReference;
_h = JSII_RTTI_SYMBOL_1;
DataYandexMdbMongodbClusterClusterConfigOutputReference[_h] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMongodbClusterClusterConfigOutputReference", version: "2.0.32" };
function dataYandexMdbMongodbClusterDatabaseToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
    };
}
exports.dataYandexMdbMongodbClusterDatabaseToTerraform = dataYandexMdbMongodbClusterDatabaseToTerraform;
class DataYandexMdbMongodbClusterDatabaseOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
}
exports.DataYandexMdbMongodbClusterDatabaseOutputReference = DataYandexMdbMongodbClusterDatabaseOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataYandexMdbMongodbClusterDatabaseOutputReference[_j] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMongodbClusterDatabaseOutputReference", version: "2.0.32" };
class DataYandexMdbMongodbClusterDatabaseList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbMongodbClusterDatabaseOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbMongodbClusterDatabaseList = DataYandexMdbMongodbClusterDatabaseList;
_k = JSII_RTTI_SYMBOL_1;
DataYandexMdbMongodbClusterDatabaseList[_k] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMongodbClusterDatabaseList", version: "2.0.32" };
function dataYandexMdbMongodbClusterHostToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        assign_public_ip: cdktf.booleanToTerraform(struct.assignPublicIp),
        health: cdktf.stringToTerraform(struct.health),
        name: cdktf.stringToTerraform(struct.name),
        role: cdktf.stringToTerraform(struct.role),
        shard_name: cdktf.stringToTerraform(struct.shardName),
        subnet_id: cdktf.stringToTerraform(struct.subnetId),
        type: cdktf.stringToTerraform(struct.type),
        zone_id: cdktf.stringToTerraform(struct.zoneId),
    };
}
exports.dataYandexMdbMongodbClusterHostToTerraform = dataYandexMdbMongodbClusterHostToTerraform;
class DataYandexMdbMongodbClusterHostOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._assignPublicIp !== undefined) {
            hasAnyValues = true;
            internalValueResult.assignPublicIp = this._assignPublicIp;
        }
        if (this._health !== undefined) {
            hasAnyValues = true;
            internalValueResult.health = this._health;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._role !== undefined) {
            hasAnyValues = true;
            internalValueResult.role = this._role;
        }
        if (this._shardName !== undefined) {
            hasAnyValues = true;
            internalValueResult.shardName = this._shardName;
        }
        if (this._subnetId !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnetId = this._subnetId;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._zoneId !== undefined) {
            hasAnyValues = true;
            internalValueResult.zoneId = this._zoneId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._assignPublicIp = undefined;
            this._health = undefined;
            this._name = undefined;
            this._role = undefined;
            this._shardName = undefined;
            this._subnetId = undefined;
            this._type = undefined;
            this._zoneId = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._assignPublicIp = value.assignPublicIp;
            this._health = value.health;
            this._name = value.name;
            this._role = value.role;
            this._shardName = value.shardName;
            this._subnetId = value.subnetId;
            this._type = value.type;
            this._zoneId = value.zoneId;
        }
    }
    get assignPublicIp() {
        return this.getBooleanAttribute('assign_public_ip');
    }
    set assignPublicIp(value) {
        this._assignPublicIp = value;
    }
    resetAssignPublicIp() {
        this._assignPublicIp = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get assignPublicIpInput() {
        return this._assignPublicIp;
    }
    get health() {
        return this.getStringAttribute('health');
    }
    set health(value) {
        this._health = value;
    }
    resetHealth() {
        this._health = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get healthInput() {
        return this._health;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get role() {
        return this.getStringAttribute('role');
    }
    set role(value) {
        this._role = value;
    }
    resetRole() {
        this._role = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get roleInput() {
        return this._role;
    }
    get shardName() {
        return this.getStringAttribute('shard_name');
    }
    set shardName(value) {
        this._shardName = value;
    }
    resetShardName() {
        this._shardName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get shardNameInput() {
        return this._shardName;
    }
    get subnetId() {
        return this.getStringAttribute('subnet_id');
    }
    set subnetId(value) {
        this._subnetId = value;
    }
    resetSubnetId() {
        this._subnetId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get subnetIdInput() {
        return this._subnetId;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get zoneId() {
        return this.getStringAttribute('zone_id');
    }
    set zoneId(value) {
        this._zoneId = value;
    }
    resetZoneId() {
        this._zoneId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get zoneIdInput() {
        return this._zoneId;
    }
}
exports.DataYandexMdbMongodbClusterHostOutputReference = DataYandexMdbMongodbClusterHostOutputReference;
_l = JSII_RTTI_SYMBOL_1;
DataYandexMdbMongodbClusterHostOutputReference[_l] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMongodbClusterHostOutputReference", version: "2.0.32" };
class DataYandexMdbMongodbClusterHostList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbMongodbClusterHostOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbMongodbClusterHostList = DataYandexMdbMongodbClusterHostList;
_m = JSII_RTTI_SYMBOL_1;
DataYandexMdbMongodbClusterHostList[_m] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMongodbClusterHostList", version: "2.0.32" };
function dataYandexMdbMongodbClusterMaintenanceWindowToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        day: cdktf.stringToTerraform(struct.day),
        hour: cdktf.numberToTerraform(struct.hour),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.dataYandexMdbMongodbClusterMaintenanceWindowToTerraform = dataYandexMdbMongodbClusterMaintenanceWindowToTerraform;
class DataYandexMdbMongodbClusterMaintenanceWindowOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._day !== undefined) {
            hasAnyValues = true;
            internalValueResult.day = this._day;
        }
        if (this._hour !== undefined) {
            hasAnyValues = true;
            internalValueResult.hour = this._hour;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._day = undefined;
            this._hour = undefined;
            this._type = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._day = value.day;
            this._hour = value.hour;
            this._type = value.type;
        }
    }
    get day() {
        return this.getStringAttribute('day');
    }
    set day(value) {
        this._day = value;
    }
    resetDay() {
        this._day = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dayInput() {
        return this._day;
    }
    get hour() {
        return this.getNumberAttribute('hour');
    }
    set hour(value) {
        this._hour = value;
    }
    resetHour() {
        this._hour = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get hourInput() {
        return this._hour;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.DataYandexMdbMongodbClusterMaintenanceWindowOutputReference = DataYandexMdbMongodbClusterMaintenanceWindowOutputReference;
_o = JSII_RTTI_SYMBOL_1;
DataYandexMdbMongodbClusterMaintenanceWindowOutputReference[_o] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMongodbClusterMaintenanceWindowOutputReference", version: "2.0.32" };
function dataYandexMdbMongodbClusterResourcesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        disk_size: cdktf.numberToTerraform(struct.diskSize),
        disk_type_id: cdktf.stringToTerraform(struct.diskTypeId),
        resource_preset_id: cdktf.stringToTerraform(struct.resourcePresetId),
    };
}
exports.dataYandexMdbMongodbClusterResourcesToTerraform = dataYandexMdbMongodbClusterResourcesToTerraform;
class DataYandexMdbMongodbClusterResourcesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._diskSize !== undefined) {
            hasAnyValues = true;
            internalValueResult.diskSize = this._diskSize;
        }
        if (this._diskTypeId !== undefined) {
            hasAnyValues = true;
            internalValueResult.diskTypeId = this._diskTypeId;
        }
        if (this._resourcePresetId !== undefined) {
            hasAnyValues = true;
            internalValueResult.resourcePresetId = this._resourcePresetId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._diskSize = undefined;
            this._diskTypeId = undefined;
            this._resourcePresetId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._diskSize = value.diskSize;
            this._diskTypeId = value.diskTypeId;
            this._resourcePresetId = value.resourcePresetId;
        }
    }
    get diskSize() {
        return this.getNumberAttribute('disk_size');
    }
    set diskSize(value) {
        this._diskSize = value;
    }
    resetDiskSize() {
        this._diskSize = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get diskSizeInput() {
        return this._diskSize;
    }
    get diskTypeId() {
        return this.getStringAttribute('disk_type_id');
    }
    set diskTypeId(value) {
        this._diskTypeId = value;
    }
    resetDiskTypeId() {
        this._diskTypeId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get diskTypeIdInput() {
        return this._diskTypeId;
    }
    get resourcePresetId() {
        return this.getStringAttribute('resource_preset_id');
    }
    set resourcePresetId(value) {
        this._resourcePresetId = value;
    }
    resetResourcePresetId() {
        this._resourcePresetId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get resourcePresetIdInput() {
        return this._resourcePresetId;
    }
}
exports.DataYandexMdbMongodbClusterResourcesOutputReference = DataYandexMdbMongodbClusterResourcesOutputReference;
_p = JSII_RTTI_SYMBOL_1;
DataYandexMdbMongodbClusterResourcesOutputReference[_p] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMongodbClusterResourcesOutputReference", version: "2.0.32" };
function dataYandexMdbMongodbClusterUserPermissionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        database_name: cdktf.stringToTerraform(struct.databaseName),
        roles: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.roles),
    };
}
exports.dataYandexMdbMongodbClusterUserPermissionToTerraform = dataYandexMdbMongodbClusterUserPermissionToTerraform;
class DataYandexMdbMongodbClusterUserPermissionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._databaseName !== undefined) {
            hasAnyValues = true;
            internalValueResult.databaseName = this._databaseName;
        }
        if (this._roles !== undefined) {
            hasAnyValues = true;
            internalValueResult.roles = this._roles;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._databaseName = undefined;
            this._roles = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._databaseName = value.databaseName;
            this._roles = value.roles;
        }
    }
    get databaseName() {
        return this.getStringAttribute('database_name');
    }
    set databaseName(value) {
        this._databaseName = value;
    }
    resetDatabaseName() {
        this._databaseName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get databaseNameInput() {
        return this._databaseName;
    }
    get roles() {
        return this.getListAttribute('roles');
    }
    set roles(value) {
        this._roles = value;
    }
    resetRoles() {
        this._roles = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get rolesInput() {
        return this._roles;
    }
}
exports.DataYandexMdbMongodbClusterUserPermissionOutputReference = DataYandexMdbMongodbClusterUserPermissionOutputReference;
_q = JSII_RTTI_SYMBOL_1;
DataYandexMdbMongodbClusterUserPermissionOutputReference[_q] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMongodbClusterUserPermissionOutputReference", version: "2.0.32" };
class DataYandexMdbMongodbClusterUserPermissionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbMongodbClusterUserPermissionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbMongodbClusterUserPermissionList = DataYandexMdbMongodbClusterUserPermissionList;
_r = JSII_RTTI_SYMBOL_1;
DataYandexMdbMongodbClusterUserPermissionList[_r] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMongodbClusterUserPermissionList", version: "2.0.32" };
function dataYandexMdbMongodbClusterUserToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        password: cdktf.stringToTerraform(struct.password),
        permission: cdktf.listMapper(dataYandexMdbMongodbClusterUserPermissionToTerraform, true)(struct.permission),
    };
}
exports.dataYandexMdbMongodbClusterUserToTerraform = dataYandexMdbMongodbClusterUserToTerraform;
class DataYandexMdbMongodbClusterUserOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // permission - computed: false, optional: true, required: false
        this._permission = new DataYandexMdbMongodbClusterUserPermissionList(this, "permission", true);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._password !== undefined) {
            hasAnyValues = true;
            internalValueResult.password = this._password;
        }
        if (this._permission?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.permission = this._permission?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._password = undefined;
            this._permission.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._password = value.password;
            this._permission.internalValue = value.permission;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get password() {
        return this.getStringAttribute('password');
    }
    set password(value) {
        this._password = value;
    }
    resetPassword() {
        this._password = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get passwordInput() {
        return this._password;
    }
    get permission() {
        return this._permission;
    }
    putPermission(value) {
        this._permission.internalValue = value;
    }
    resetPermission() {
        this._permission.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get permissionInput() {
        return this._permission.internalValue;
    }
}
exports.DataYandexMdbMongodbClusterUserOutputReference = DataYandexMdbMongodbClusterUserOutputReference;
_s = JSII_RTTI_SYMBOL_1;
DataYandexMdbMongodbClusterUserOutputReference[_s] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMongodbClusterUserOutputReference", version: "2.0.32" };
class DataYandexMdbMongodbClusterUserList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexMdbMongodbClusterUserOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexMdbMongodbClusterUserList = DataYandexMdbMongodbClusterUserList;
_t = JSII_RTTI_SYMBOL_1;
DataYandexMdbMongodbClusterUserList[_t] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMongodbClusterUserList", version: "2.0.32" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/d/mdb_mongodb_cluster yandex_mdb_mongodb_cluster}
*/
class DataYandexMdbMongodbCluster extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/d/mdb_mongodb_cluster yandex_mdb_mongodb_cluster} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataYandexMdbMongodbClusterConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'yandex_mdb_mongodb_cluster',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // cluster_config - computed: false, optional: true, required: false
        this._clusterConfig = new DataYandexMdbMongodbClusterClusterConfigOutputReference(this, "cluster_config");
        // database - computed: false, optional: true, required: false
        this._database = new DataYandexMdbMongodbClusterDatabaseList(this, "database", true);
        // host - computed: false, optional: true, required: false
        this._host = new DataYandexMdbMongodbClusterHostList(this, "host", false);
        // maintenance_window - computed: false, optional: true, required: false
        this._maintenanceWindow = new DataYandexMdbMongodbClusterMaintenanceWindowOutputReference(this, "maintenance_window");
        // resources - computed: false, optional: true, required: false
        this._resources = new DataYandexMdbMongodbClusterResourcesOutputReference(this, "resources");
        // user - computed: false, optional: true, required: false
        this._user = new DataYandexMdbMongodbClusterUserList(this, "user", true);
        this._clusterId = config.clusterId;
        this._createdAt = config.createdAt;
        this._deletionProtection = config.deletionProtection;
        this._description = config.description;
        this._environment = config.environment;
        this._folderId = config.folderId;
        this._health = config.health;
        this._id = config.id;
        this._labels = config.labels;
        this._name = config.name;
        this._networkId = config.networkId;
        this._securityGroupIds = config.securityGroupIds;
        this._sharded = config.sharded;
        this._status = config.status;
        this._clusterConfig.internalValue = config.clusterConfig;
        this._database.internalValue = config.database;
        this._host.internalValue = config.host;
        this._maintenanceWindow.internalValue = config.maintenanceWindow;
        this._resources.internalValue = config.resources;
        this._user.internalValue = config.user;
    }
    get clusterId() {
        return this.getStringAttribute('cluster_id');
    }
    set clusterId(value) {
        this._clusterId = value;
    }
    resetClusterId() {
        this._clusterId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clusterIdInput() {
        return this._clusterId;
    }
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    set createdAt(value) {
        this._createdAt = value;
    }
    resetCreatedAt() {
        this._createdAt = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createdAtInput() {
        return this._createdAt;
    }
    get deletionProtection() {
        return this.getBooleanAttribute('deletion_protection');
    }
    set deletionProtection(value) {
        this._deletionProtection = value;
    }
    resetDeletionProtection() {
        this._deletionProtection = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deletionProtectionInput() {
        return this._deletionProtection;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get environment() {
        return this.getStringAttribute('environment');
    }
    set environment(value) {
        this._environment = value;
    }
    resetEnvironment() {
        this._environment = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get environmentInput() {
        return this._environment;
    }
    get folderId() {
        return this.getStringAttribute('folder_id');
    }
    set folderId(value) {
        this._folderId = value;
    }
    resetFolderId() {
        this._folderId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get folderIdInput() {
        return this._folderId;
    }
    get health() {
        return this.getStringAttribute('health');
    }
    set health(value) {
        this._health = value;
    }
    resetHealth() {
        this._health = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get healthInput() {
        return this._health;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get networkId() {
        return this.getStringAttribute('network_id');
    }
    set networkId(value) {
        this._networkId = value;
    }
    resetNetworkId() {
        this._networkId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get networkIdInput() {
        return this._networkId;
    }
    get securityGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('security_group_ids'));
    }
    set securityGroupIds(value) {
        this._securityGroupIds = value;
    }
    resetSecurityGroupIds() {
        this._securityGroupIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupIdsInput() {
        return this._securityGroupIds;
    }
    get sharded() {
        return this.getBooleanAttribute('sharded');
    }
    set sharded(value) {
        this._sharded = value;
    }
    resetSharded() {
        this._sharded = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get shardedInput() {
        return this._sharded;
    }
    get status() {
        return this.getStringAttribute('status');
    }
    set status(value) {
        this._status = value;
    }
    resetStatus() {
        this._status = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get statusInput() {
        return this._status;
    }
    get clusterConfig() {
        return this._clusterConfig;
    }
    putClusterConfig(value) {
        this._clusterConfig.internalValue = value;
    }
    resetClusterConfig() {
        this._clusterConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clusterConfigInput() {
        return this._clusterConfig.internalValue;
    }
    get database() {
        return this._database;
    }
    putDatabase(value) {
        this._database.internalValue = value;
    }
    resetDatabase() {
        this._database.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get databaseInput() {
        return this._database.internalValue;
    }
    get host() {
        return this._host;
    }
    putHost(value) {
        this._host.internalValue = value;
    }
    resetHost() {
        this._host.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get hostInput() {
        return this._host.internalValue;
    }
    get maintenanceWindow() {
        return this._maintenanceWindow;
    }
    putMaintenanceWindow(value) {
        this._maintenanceWindow.internalValue = value;
    }
    resetMaintenanceWindow() {
        this._maintenanceWindow.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maintenanceWindowInput() {
        return this._maintenanceWindow.internalValue;
    }
    get resources() {
        return this._resources;
    }
    putResources(value) {
        this._resources.internalValue = value;
    }
    resetResources() {
        this._resources.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get resourcesInput() {
        return this._resources.internalValue;
    }
    get user() {
        return this._user;
    }
    putUser(value) {
        this._user.internalValue = value;
    }
    resetUser() {
        this._user.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get userInput() {
        return this._user.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            cluster_id: cdktf.stringToTerraform(this._clusterId),
            created_at: cdktf.stringToTerraform(this._createdAt),
            deletion_protection: cdktf.booleanToTerraform(this._deletionProtection),
            description: cdktf.stringToTerraform(this._description),
            environment: cdktf.stringToTerraform(this._environment),
            folder_id: cdktf.stringToTerraform(this._folderId),
            health: cdktf.stringToTerraform(this._health),
            id: cdktf.stringToTerraform(this._id),
            labels: cdktf.hashMapper(cdktf.stringToTerraform)(this._labels),
            name: cdktf.stringToTerraform(this._name),
            network_id: cdktf.stringToTerraform(this._networkId),
            security_group_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(this._securityGroupIds),
            sharded: cdktf.booleanToTerraform(this._sharded),
            status: cdktf.stringToTerraform(this._status),
            cluster_config: dataYandexMdbMongodbClusterClusterConfigToTerraform(this._clusterConfig.internalValue),
            database: cdktf.listMapper(dataYandexMdbMongodbClusterDatabaseToTerraform, true)(this._database.internalValue),
            host: cdktf.listMapper(dataYandexMdbMongodbClusterHostToTerraform, true)(this._host.internalValue),
            maintenance_window: dataYandexMdbMongodbClusterMaintenanceWindowToTerraform(this._maintenanceWindow.internalValue),
            resources: dataYandexMdbMongodbClusterResourcesToTerraform(this._resources.internalValue),
            user: cdktf.listMapper(dataYandexMdbMongodbClusterUserToTerraform, true)(this._user.internalValue),
        };
    }
}
exports.DataYandexMdbMongodbCluster = DataYandexMdbMongodbCluster;
_u = JSII_RTTI_SYMBOL_1;
DataYandexMdbMongodbCluster[_u] = { fqn: "@cdktf/provider-yandex.DataYandexMdbMongodbCluster", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
DataYandexMdbMongodbCluster.tfResourceType = "yandex_mdb_mongodb_cluster";
