// @cdktf/provider-yandex/lib/data-yandex-dataproc-cluster.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataYandexDataprocCluster = exports.DataYandexDataprocClusterClusterConfigList = exports.DataYandexDataprocClusterClusterConfigOutputReference = exports.dataYandexDataprocClusterClusterConfigToTerraform = exports.DataYandexDataprocClusterClusterConfigSubclusterSpecList = exports.DataYandexDataprocClusterClusterConfigSubclusterSpecOutputReference = exports.dataYandexDataprocClusterClusterConfigSubclusterSpecToTerraform = exports.DataYandexDataprocClusterClusterConfigSubclusterSpecResourcesList = exports.DataYandexDataprocClusterClusterConfigSubclusterSpecResourcesOutputReference = exports.dataYandexDataprocClusterClusterConfigSubclusterSpecResourcesToTerraform = exports.DataYandexDataprocClusterClusterConfigSubclusterSpecAutoscalingConfigList = exports.DataYandexDataprocClusterClusterConfigSubclusterSpecAutoscalingConfigOutputReference = exports.dataYandexDataprocClusterClusterConfigSubclusterSpecAutoscalingConfigToTerraform = exports.DataYandexDataprocClusterClusterConfigHadoopList = exports.DataYandexDataprocClusterClusterConfigHadoopOutputReference = exports.dataYandexDataprocClusterClusterConfigHadoopToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function dataYandexDataprocClusterClusterConfigHadoopToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexDataprocClusterClusterConfigHadoopToTerraform = dataYandexDataprocClusterClusterConfigHadoopToTerraform;
class DataYandexDataprocClusterClusterConfigHadoopOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // properties - computed: true, optional: false, required: false
        this._properties = new cdktf.StringMap(this, "properties");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get properties() {
        return this._properties;
    }
    // services - computed: true, optional: false, required: false
    get services() {
        return cdktf.Fn.tolist(this.getListAttribute('services'));
    }
    // ssh_public_keys - computed: true, optional: false, required: false
    get sshPublicKeys() {
        return cdktf.Fn.tolist(this.getListAttribute('ssh_public_keys'));
    }
}
exports.DataYandexDataprocClusterClusterConfigHadoopOutputReference = DataYandexDataprocClusterClusterConfigHadoopOutputReference;
_a = JSII_RTTI_SYMBOL_1;
DataYandexDataprocClusterClusterConfigHadoopOutputReference[_a] = { fqn: "@cdktf/provider-yandex.DataYandexDataprocClusterClusterConfigHadoopOutputReference", version: "2.0.32" };
class DataYandexDataprocClusterClusterConfigHadoopList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexDataprocClusterClusterConfigHadoopOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexDataprocClusterClusterConfigHadoopList = DataYandexDataprocClusterClusterConfigHadoopList;
_b = JSII_RTTI_SYMBOL_1;
DataYandexDataprocClusterClusterConfigHadoopList[_b] = { fqn: "@cdktf/provider-yandex.DataYandexDataprocClusterClusterConfigHadoopList", version: "2.0.32" };
function dataYandexDataprocClusterClusterConfigSubclusterSpecAutoscalingConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexDataprocClusterClusterConfigSubclusterSpecAutoscalingConfigToTerraform = dataYandexDataprocClusterClusterConfigSubclusterSpecAutoscalingConfigToTerraform;
class DataYandexDataprocClusterClusterConfigSubclusterSpecAutoscalingConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // cpu_utilization_target - computed: true, optional: false, required: false
    get cpuUtilizationTarget() {
        return this.getNumberAttribute('cpu_utilization_target');
    }
    // decommission_timeout - computed: true, optional: false, required: false
    get decommissionTimeout() {
        return this.getNumberAttribute('decommission_timeout');
    }
    // max_hosts_count - computed: true, optional: false, required: false
    get maxHostsCount() {
        return this.getNumberAttribute('max_hosts_count');
    }
    // measurement_duration - computed: true, optional: false, required: false
    get measurementDuration() {
        return this.getNumberAttribute('measurement_duration');
    }
    // preemptible - computed: true, optional: false, required: false
    get preemptible() {
        return this.getBooleanAttribute('preemptible');
    }
    // stabilization_duration - computed: true, optional: false, required: false
    get stabilizationDuration() {
        return this.getNumberAttribute('stabilization_duration');
    }
    // warmup_duration - computed: true, optional: false, required: false
    get warmupDuration() {
        return this.getNumberAttribute('warmup_duration');
    }
}
exports.DataYandexDataprocClusterClusterConfigSubclusterSpecAutoscalingConfigOutputReference = DataYandexDataprocClusterClusterConfigSubclusterSpecAutoscalingConfigOutputReference;
_c = JSII_RTTI_SYMBOL_1;
DataYandexDataprocClusterClusterConfigSubclusterSpecAutoscalingConfigOutputReference[_c] = { fqn: "@cdktf/provider-yandex.DataYandexDataprocClusterClusterConfigSubclusterSpecAutoscalingConfigOutputReference", version: "2.0.32" };
class DataYandexDataprocClusterClusterConfigSubclusterSpecAutoscalingConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexDataprocClusterClusterConfigSubclusterSpecAutoscalingConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexDataprocClusterClusterConfigSubclusterSpecAutoscalingConfigList = DataYandexDataprocClusterClusterConfigSubclusterSpecAutoscalingConfigList;
_d = JSII_RTTI_SYMBOL_1;
DataYandexDataprocClusterClusterConfigSubclusterSpecAutoscalingConfigList[_d] = { fqn: "@cdktf/provider-yandex.DataYandexDataprocClusterClusterConfigSubclusterSpecAutoscalingConfigList", version: "2.0.32" };
function dataYandexDataprocClusterClusterConfigSubclusterSpecResourcesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexDataprocClusterClusterConfigSubclusterSpecResourcesToTerraform = dataYandexDataprocClusterClusterConfigSubclusterSpecResourcesToTerraform;
class DataYandexDataprocClusterClusterConfigSubclusterSpecResourcesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // disk_size - computed: true, optional: false, required: false
    get diskSize() {
        return this.getNumberAttribute('disk_size');
    }
    // disk_type_id - computed: true, optional: false, required: false
    get diskTypeId() {
        return this.getStringAttribute('disk_type_id');
    }
    // resource_preset_id - computed: true, optional: false, required: false
    get resourcePresetId() {
        return this.getStringAttribute('resource_preset_id');
    }
}
exports.DataYandexDataprocClusterClusterConfigSubclusterSpecResourcesOutputReference = DataYandexDataprocClusterClusterConfigSubclusterSpecResourcesOutputReference;
_e = JSII_RTTI_SYMBOL_1;
DataYandexDataprocClusterClusterConfigSubclusterSpecResourcesOutputReference[_e] = { fqn: "@cdktf/provider-yandex.DataYandexDataprocClusterClusterConfigSubclusterSpecResourcesOutputReference", version: "2.0.32" };
class DataYandexDataprocClusterClusterConfigSubclusterSpecResourcesList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexDataprocClusterClusterConfigSubclusterSpecResourcesOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexDataprocClusterClusterConfigSubclusterSpecResourcesList = DataYandexDataprocClusterClusterConfigSubclusterSpecResourcesList;
_f = JSII_RTTI_SYMBOL_1;
DataYandexDataprocClusterClusterConfigSubclusterSpecResourcesList[_f] = { fqn: "@cdktf/provider-yandex.DataYandexDataprocClusterClusterConfigSubclusterSpecResourcesList", version: "2.0.32" };
function dataYandexDataprocClusterClusterConfigSubclusterSpecToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexDataprocClusterClusterConfigSubclusterSpecToTerraform = dataYandexDataprocClusterClusterConfigSubclusterSpecToTerraform;
class DataYandexDataprocClusterClusterConfigSubclusterSpecOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // autoscaling_config - computed: true, optional: false, required: false
        this._autoscalingConfig = new DataYandexDataprocClusterClusterConfigSubclusterSpecAutoscalingConfigList(this, "autoscaling_config", false);
        // resources - computed: true, optional: false, required: false
        this._resources = new DataYandexDataprocClusterClusterConfigSubclusterSpecResourcesList(this, "resources", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // assign_public_ip - computed: true, optional: false, required: false
    get assignPublicIp() {
        return this.getBooleanAttribute('assign_public_ip');
    }
    get autoscalingConfig() {
        return this._autoscalingConfig;
    }
    // hosts_count - computed: true, optional: false, required: false
    get hostsCount() {
        return this.getNumberAttribute('hosts_count');
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get resources() {
        return this._resources;
    }
    // role - computed: true, optional: false, required: false
    get role() {
        return this.getStringAttribute('role');
    }
    // subnet_id - computed: true, optional: false, required: false
    get subnetId() {
        return this.getStringAttribute('subnet_id');
    }
}
exports.DataYandexDataprocClusterClusterConfigSubclusterSpecOutputReference = DataYandexDataprocClusterClusterConfigSubclusterSpecOutputReference;
_g = JSII_RTTI_SYMBOL_1;
DataYandexDataprocClusterClusterConfigSubclusterSpecOutputReference[_g] = { fqn: "@cdktf/provider-yandex.DataYandexDataprocClusterClusterConfigSubclusterSpecOutputReference", version: "2.0.32" };
class DataYandexDataprocClusterClusterConfigSubclusterSpecList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexDataprocClusterClusterConfigSubclusterSpecOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexDataprocClusterClusterConfigSubclusterSpecList = DataYandexDataprocClusterClusterConfigSubclusterSpecList;
_h = JSII_RTTI_SYMBOL_1;
DataYandexDataprocClusterClusterConfigSubclusterSpecList[_h] = { fqn: "@cdktf/provider-yandex.DataYandexDataprocClusterClusterConfigSubclusterSpecList", version: "2.0.32" };
function dataYandexDataprocClusterClusterConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.dataYandexDataprocClusterClusterConfigToTerraform = dataYandexDataprocClusterClusterConfigToTerraform;
class DataYandexDataprocClusterClusterConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // hadoop - computed: true, optional: false, required: false
        this._hadoop = new DataYandexDataprocClusterClusterConfigHadoopList(this, "hadoop", false);
        // subcluster_spec - computed: true, optional: false, required: false
        this._subclusterSpec = new DataYandexDataprocClusterClusterConfigSubclusterSpecList(this, "subcluster_spec", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    get hadoop() {
        return this._hadoop;
    }
    get subclusterSpec() {
        return this._subclusterSpec;
    }
    // version_id - computed: true, optional: false, required: false
    get versionId() {
        return this.getStringAttribute('version_id');
    }
}
exports.DataYandexDataprocClusterClusterConfigOutputReference = DataYandexDataprocClusterClusterConfigOutputReference;
_j = JSII_RTTI_SYMBOL_1;
DataYandexDataprocClusterClusterConfigOutputReference[_j] = { fqn: "@cdktf/provider-yandex.DataYandexDataprocClusterClusterConfigOutputReference", version: "2.0.32" };
class DataYandexDataprocClusterClusterConfigList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new DataYandexDataprocClusterClusterConfigOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.DataYandexDataprocClusterClusterConfigList = DataYandexDataprocClusterClusterConfigList;
_k = JSII_RTTI_SYMBOL_1;
DataYandexDataprocClusterClusterConfigList[_k] = { fqn: "@cdktf/provider-yandex.DataYandexDataprocClusterClusterConfigList", version: "2.0.32" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/d/dataproc_cluster yandex_dataproc_cluster}
*/
class DataYandexDataprocCluster extends cdktf.TerraformDataSource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/d/dataproc_cluster yandex_dataproc_cluster} Data Source
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options DataYandexDataprocClusterConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'yandex_dataproc_cluster',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // cluster_config - computed: true, optional: false, required: false
        this._clusterConfig = new DataYandexDataprocClusterClusterConfigList(this, "cluster_config", false);
        // labels - computed: true, optional: false, required: false
        this._labels = new cdktf.StringMap(this, "labels");
        this._clusterId = config.clusterId;
        this._id = config.id;
        this._name = config.name;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // bucket - computed: true, optional: false, required: false
    get bucket() {
        return this.getStringAttribute('bucket');
    }
    get clusterConfig() {
        return this._clusterConfig;
    }
    get clusterId() {
        return this.getStringAttribute('cluster_id');
    }
    set clusterId(value) {
        this._clusterId = value;
    }
    resetClusterId() {
        this._clusterId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clusterIdInput() {
        return this._clusterId;
    }
    // created_at - computed: true, optional: false, required: false
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    // deletion_protection - computed: true, optional: false, required: false
    get deletionProtection() {
        return this.getBooleanAttribute('deletion_protection');
    }
    // description - computed: true, optional: false, required: false
    get description() {
        return this.getStringAttribute('description');
    }
    // folder_id - computed: true, optional: false, required: false
    get folderId() {
        return this.getStringAttribute('folder_id');
    }
    // host_group_ids - computed: true, optional: false, required: false
    get hostGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('host_group_ids'));
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this._labels;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // security_group_ids - computed: true, optional: false, required: false
    get securityGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('security_group_ids'));
    }
    // service_account_id - computed: true, optional: false, required: false
    get serviceAccountId() {
        return this.getStringAttribute('service_account_id');
    }
    // ui_proxy - computed: true, optional: false, required: false
    get uiProxy() {
        return this.getBooleanAttribute('ui_proxy');
    }
    // zone_id - computed: true, optional: false, required: false
    get zoneId() {
        return this.getStringAttribute('zone_id');
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            cluster_id: cdktf.stringToTerraform(this._clusterId),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
        };
    }
}
exports.DataYandexDataprocCluster = DataYandexDataprocCluster;
_l = JSII_RTTI_SYMBOL_1;
DataYandexDataprocCluster[_l] = { fqn: "@cdktf/provider-yandex.DataYandexDataprocCluster", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
DataYandexDataprocCluster.tfResourceType = "yandex_dataproc_cluster";
