// @cdktf/provider-yandex/lib/iam-service-account-static-access-key.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.IamServiceAccountStaticAccessKey = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/r/iam_service_account_static_access_key yandex_iam_service_account_static_access_key}
*/
class IamServiceAccountStaticAccessKey extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/r/iam_service_account_static_access_key yandex_iam_service_account_static_access_key} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options IamServiceAccountStaticAccessKeyConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'yandex_iam_service_account_static_access_key',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._description = config.description;
        this._id = config.id;
        this._pgpKey = config.pgpKey;
        this._serviceAccountId = config.serviceAccountId;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // access_key - computed: true, optional: false, required: false
    get accessKey() {
        return this.getStringAttribute('access_key');
    }
    // created_at - computed: true, optional: false, required: false
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    // encrypted_secret_key - computed: true, optional: false, required: false
    get encryptedSecretKey() {
        return this.getStringAttribute('encrypted_secret_key');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    // key_fingerprint - computed: true, optional: false, required: false
    get keyFingerprint() {
        return this.getStringAttribute('key_fingerprint');
    }
    get pgpKey() {
        return this.getStringAttribute('pgp_key');
    }
    set pgpKey(value) {
        this._pgpKey = value;
    }
    resetPgpKey() {
        this._pgpKey = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pgpKeyInput() {
        return this._pgpKey;
    }
    // secret_key - computed: true, optional: false, required: false
    get secretKey() {
        return this.getStringAttribute('secret_key');
    }
    get serviceAccountId() {
        return this.getStringAttribute('service_account_id');
    }
    set serviceAccountId(value) {
        this._serviceAccountId = value;
    }
    // Temporarily expose input value. Use with caution.
    get serviceAccountIdInput() {
        return this._serviceAccountId;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            pgp_key: cdktf.stringToTerraform(this._pgpKey),
            service_account_id: cdktf.stringToTerraform(this._serviceAccountId),
        };
    }
}
exports.IamServiceAccountStaticAccessKey = IamServiceAccountStaticAccessKey;
_a = JSII_RTTI_SYMBOL_1;
IamServiceAccountStaticAccessKey[_a] = { fqn: "@cdktf/provider-yandex.IamServiceAccountStaticAccessKey", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
IamServiceAccountStaticAccessKey.tfResourceType = "yandex_iam_service_account_static_access_key";
