// @cdktf/provider-yandex/lib/kubernetes-cluster.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q;
Object.defineProperty(exports, "__esModule", { value: true });
exports.KubernetesCluster = exports.KubernetesClusterTimeoutsOutputReference = exports.kubernetesClusterTimeoutsToTerraform = exports.KubernetesClusterNetworkImplementationOutputReference = exports.kubernetesClusterNetworkImplementationToTerraform = exports.KubernetesClusterNetworkImplementationCiliumOutputReference = exports.kubernetesClusterNetworkImplementationCiliumToTerraform = exports.KubernetesClusterMasterOutputReference = exports.kubernetesClusterMasterToTerraform = exports.KubernetesClusterMasterZonalOutputReference = exports.kubernetesClusterMasterZonalToTerraform = exports.KubernetesClusterMasterRegionalOutputReference = exports.kubernetesClusterMasterRegionalToTerraform = exports.KubernetesClusterMasterRegionalLocationList = exports.KubernetesClusterMasterRegionalLocationOutputReference = exports.kubernetesClusterMasterRegionalLocationToTerraform = exports.KubernetesClusterMasterMaintenancePolicyOutputReference = exports.kubernetesClusterMasterMaintenancePolicyToTerraform = exports.KubernetesClusterMasterMaintenancePolicyMaintenanceWindowList = exports.KubernetesClusterMasterMaintenancePolicyMaintenanceWindowOutputReference = exports.kubernetesClusterMasterMaintenancePolicyMaintenanceWindowToTerraform = exports.KubernetesClusterMasterVersionInfoList = exports.KubernetesClusterMasterVersionInfoOutputReference = exports.kubernetesClusterMasterVersionInfoToTerraform = exports.KubernetesClusterKmsProviderOutputReference = exports.kubernetesClusterKmsProviderToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function kubernetesClusterKmsProviderToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        key_id: cdktf.stringToTerraform(struct.keyId),
    };
}
exports.kubernetesClusterKmsProviderToTerraform = kubernetesClusterKmsProviderToTerraform;
class KubernetesClusterKmsProviderOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._keyId !== undefined) {
            hasAnyValues = true;
            internalValueResult.keyId = this._keyId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._keyId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._keyId = value.keyId;
        }
    }
    get keyId() {
        return this.getStringAttribute('key_id');
    }
    set keyId(value) {
        this._keyId = value;
    }
    resetKeyId() {
        this._keyId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keyIdInput() {
        return this._keyId;
    }
}
exports.KubernetesClusterKmsProviderOutputReference = KubernetesClusterKmsProviderOutputReference;
_a = JSII_RTTI_SYMBOL_1;
KubernetesClusterKmsProviderOutputReference[_a] = { fqn: "@cdktf/provider-yandex.KubernetesClusterKmsProviderOutputReference", version: "2.0.32" };
function kubernetesClusterMasterVersionInfoToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.kubernetesClusterMasterVersionInfoToTerraform = kubernetesClusterMasterVersionInfoToTerraform;
class KubernetesClusterMasterVersionInfoOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // current_version - computed: true, optional: false, required: false
    get currentVersion() {
        return this.getStringAttribute('current_version');
    }
    // new_revision_available - computed: true, optional: false, required: false
    get newRevisionAvailable() {
        return this.getBooleanAttribute('new_revision_available');
    }
    // new_revision_summary - computed: true, optional: false, required: false
    get newRevisionSummary() {
        return this.getStringAttribute('new_revision_summary');
    }
    // version_deprecated - computed: true, optional: false, required: false
    get versionDeprecated() {
        return this.getBooleanAttribute('version_deprecated');
    }
}
exports.KubernetesClusterMasterVersionInfoOutputReference = KubernetesClusterMasterVersionInfoOutputReference;
_b = JSII_RTTI_SYMBOL_1;
KubernetesClusterMasterVersionInfoOutputReference[_b] = { fqn: "@cdktf/provider-yandex.KubernetesClusterMasterVersionInfoOutputReference", version: "2.0.32" };
class KubernetesClusterMasterVersionInfoList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new KubernetesClusterMasterVersionInfoOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.KubernetesClusterMasterVersionInfoList = KubernetesClusterMasterVersionInfoList;
_c = JSII_RTTI_SYMBOL_1;
KubernetesClusterMasterVersionInfoList[_c] = { fqn: "@cdktf/provider-yandex.KubernetesClusterMasterVersionInfoList", version: "2.0.32" };
function kubernetesClusterMasterMaintenancePolicyMaintenanceWindowToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        day: cdktf.stringToTerraform(struct.day),
        duration: cdktf.stringToTerraform(struct.duration),
        start_time: cdktf.stringToTerraform(struct.startTime),
    };
}
exports.kubernetesClusterMasterMaintenancePolicyMaintenanceWindowToTerraform = kubernetesClusterMasterMaintenancePolicyMaintenanceWindowToTerraform;
class KubernetesClusterMasterMaintenancePolicyMaintenanceWindowOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._day !== undefined) {
            hasAnyValues = true;
            internalValueResult.day = this._day;
        }
        if (this._duration !== undefined) {
            hasAnyValues = true;
            internalValueResult.duration = this._duration;
        }
        if (this._startTime !== undefined) {
            hasAnyValues = true;
            internalValueResult.startTime = this._startTime;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._day = undefined;
            this._duration = undefined;
            this._startTime = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._day = value.day;
            this._duration = value.duration;
            this._startTime = value.startTime;
        }
    }
    get day() {
        return this.getStringAttribute('day');
    }
    set day(value) {
        this._day = value;
    }
    resetDay() {
        this._day = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dayInput() {
        return this._day;
    }
    get duration() {
        return this.getStringAttribute('duration');
    }
    set duration(value) {
        this._duration = value;
    }
    // Temporarily expose input value. Use with caution.
    get durationInput() {
        return this._duration;
    }
    get startTime() {
        return this.getStringAttribute('start_time');
    }
    set startTime(value) {
        this._startTime = value;
    }
    // Temporarily expose input value. Use with caution.
    get startTimeInput() {
        return this._startTime;
    }
}
exports.KubernetesClusterMasterMaintenancePolicyMaintenanceWindowOutputReference = KubernetesClusterMasterMaintenancePolicyMaintenanceWindowOutputReference;
_d = JSII_RTTI_SYMBOL_1;
KubernetesClusterMasterMaintenancePolicyMaintenanceWindowOutputReference[_d] = { fqn: "@cdktf/provider-yandex.KubernetesClusterMasterMaintenancePolicyMaintenanceWindowOutputReference", version: "2.0.32" };
class KubernetesClusterMasterMaintenancePolicyMaintenanceWindowList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new KubernetesClusterMasterMaintenancePolicyMaintenanceWindowOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.KubernetesClusterMasterMaintenancePolicyMaintenanceWindowList = KubernetesClusterMasterMaintenancePolicyMaintenanceWindowList;
_e = JSII_RTTI_SYMBOL_1;
KubernetesClusterMasterMaintenancePolicyMaintenanceWindowList[_e] = { fqn: "@cdktf/provider-yandex.KubernetesClusterMasterMaintenancePolicyMaintenanceWindowList", version: "2.0.32" };
function kubernetesClusterMasterMaintenancePolicyToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        auto_upgrade: cdktf.booleanToTerraform(struct.autoUpgrade),
        maintenance_window: cdktf.listMapper(kubernetesClusterMasterMaintenancePolicyMaintenanceWindowToTerraform, true)(struct.maintenanceWindow),
    };
}
exports.kubernetesClusterMasterMaintenancePolicyToTerraform = kubernetesClusterMasterMaintenancePolicyToTerraform;
class KubernetesClusterMasterMaintenancePolicyOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // maintenance_window - computed: false, optional: true, required: false
        this._maintenanceWindow = new KubernetesClusterMasterMaintenancePolicyMaintenanceWindowList(this, "maintenance_window", true);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._autoUpgrade !== undefined) {
            hasAnyValues = true;
            internalValueResult.autoUpgrade = this._autoUpgrade;
        }
        if (this._maintenanceWindow?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.maintenanceWindow = this._maintenanceWindow?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._autoUpgrade = undefined;
            this._maintenanceWindow.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._autoUpgrade = value.autoUpgrade;
            this._maintenanceWindow.internalValue = value.maintenanceWindow;
        }
    }
    get autoUpgrade() {
        return this.getBooleanAttribute('auto_upgrade');
    }
    set autoUpgrade(value) {
        this._autoUpgrade = value;
    }
    // Temporarily expose input value. Use with caution.
    get autoUpgradeInput() {
        return this._autoUpgrade;
    }
    get maintenanceWindow() {
        return this._maintenanceWindow;
    }
    putMaintenanceWindow(value) {
        this._maintenanceWindow.internalValue = value;
    }
    resetMaintenanceWindow() {
        this._maintenanceWindow.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maintenanceWindowInput() {
        return this._maintenanceWindow.internalValue;
    }
}
exports.KubernetesClusterMasterMaintenancePolicyOutputReference = KubernetesClusterMasterMaintenancePolicyOutputReference;
_f = JSII_RTTI_SYMBOL_1;
KubernetesClusterMasterMaintenancePolicyOutputReference[_f] = { fqn: "@cdktf/provider-yandex.KubernetesClusterMasterMaintenancePolicyOutputReference", version: "2.0.32" };
function kubernetesClusterMasterRegionalLocationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        subnet_id: cdktf.stringToTerraform(struct.subnetId),
        zone: cdktf.stringToTerraform(struct.zone),
    };
}
exports.kubernetesClusterMasterRegionalLocationToTerraform = kubernetesClusterMasterRegionalLocationToTerraform;
class KubernetesClusterMasterRegionalLocationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._subnetId !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnetId = this._subnetId;
        }
        if (this._zone !== undefined) {
            hasAnyValues = true;
            internalValueResult.zone = this._zone;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._subnetId = undefined;
            this._zone = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._subnetId = value.subnetId;
            this._zone = value.zone;
        }
    }
    get subnetId() {
        return this.getStringAttribute('subnet_id');
    }
    set subnetId(value) {
        this._subnetId = value;
    }
    resetSubnetId() {
        this._subnetId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get subnetIdInput() {
        return this._subnetId;
    }
    get zone() {
        return this.getStringAttribute('zone');
    }
    set zone(value) {
        this._zone = value;
    }
    resetZone() {
        this._zone = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get zoneInput() {
        return this._zone;
    }
}
exports.KubernetesClusterMasterRegionalLocationOutputReference = KubernetesClusterMasterRegionalLocationOutputReference;
_g = JSII_RTTI_SYMBOL_1;
KubernetesClusterMasterRegionalLocationOutputReference[_g] = { fqn: "@cdktf/provider-yandex.KubernetesClusterMasterRegionalLocationOutputReference", version: "2.0.32" };
class KubernetesClusterMasterRegionalLocationList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new KubernetesClusterMasterRegionalLocationOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.KubernetesClusterMasterRegionalLocationList = KubernetesClusterMasterRegionalLocationList;
_h = JSII_RTTI_SYMBOL_1;
KubernetesClusterMasterRegionalLocationList[_h] = { fqn: "@cdktf/provider-yandex.KubernetesClusterMasterRegionalLocationList", version: "2.0.32" };
function kubernetesClusterMasterRegionalToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        region: cdktf.stringToTerraform(struct.region),
        location: cdktf.listMapper(kubernetesClusterMasterRegionalLocationToTerraform, true)(struct.location),
    };
}
exports.kubernetesClusterMasterRegionalToTerraform = kubernetesClusterMasterRegionalToTerraform;
class KubernetesClusterMasterRegionalOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // location - computed: false, optional: true, required: false
        this._location = new KubernetesClusterMasterRegionalLocationList(this, "location", false);
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._region !== undefined) {
            hasAnyValues = true;
            internalValueResult.region = this._region;
        }
        if (this._location?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.location = this._location?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._region = undefined;
            this._location.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._region = value.region;
            this._location.internalValue = value.location;
        }
    }
    get region() {
        return this.getStringAttribute('region');
    }
    set region(value) {
        this._region = value;
    }
    // Temporarily expose input value. Use with caution.
    get regionInput() {
        return this._region;
    }
    get location() {
        return this._location;
    }
    putLocation(value) {
        this._location.internalValue = value;
    }
    resetLocation() {
        this._location.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get locationInput() {
        return this._location.internalValue;
    }
}
exports.KubernetesClusterMasterRegionalOutputReference = KubernetesClusterMasterRegionalOutputReference;
_j = JSII_RTTI_SYMBOL_1;
KubernetesClusterMasterRegionalOutputReference[_j] = { fqn: "@cdktf/provider-yandex.KubernetesClusterMasterRegionalOutputReference", version: "2.0.32" };
function kubernetesClusterMasterZonalToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        subnet_id: cdktf.stringToTerraform(struct.subnetId),
        zone: cdktf.stringToTerraform(struct.zone),
    };
}
exports.kubernetesClusterMasterZonalToTerraform = kubernetesClusterMasterZonalToTerraform;
class KubernetesClusterMasterZonalOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._subnetId !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnetId = this._subnetId;
        }
        if (this._zone !== undefined) {
            hasAnyValues = true;
            internalValueResult.zone = this._zone;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._subnetId = undefined;
            this._zone = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._subnetId = value.subnetId;
            this._zone = value.zone;
        }
    }
    get subnetId() {
        return this.getStringAttribute('subnet_id');
    }
    set subnetId(value) {
        this._subnetId = value;
    }
    resetSubnetId() {
        this._subnetId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get subnetIdInput() {
        return this._subnetId;
    }
    get zone() {
        return this.getStringAttribute('zone');
    }
    set zone(value) {
        this._zone = value;
    }
    resetZone() {
        this._zone = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get zoneInput() {
        return this._zone;
    }
}
exports.KubernetesClusterMasterZonalOutputReference = KubernetesClusterMasterZonalOutputReference;
_k = JSII_RTTI_SYMBOL_1;
KubernetesClusterMasterZonalOutputReference[_k] = { fqn: "@cdktf/provider-yandex.KubernetesClusterMasterZonalOutputReference", version: "2.0.32" };
function kubernetesClusterMasterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        public_ip: cdktf.booleanToTerraform(struct.publicIp),
        security_group_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.securityGroupIds),
        version: cdktf.stringToTerraform(struct.version),
        maintenance_policy: kubernetesClusterMasterMaintenancePolicyToTerraform(struct.maintenancePolicy),
        regional: kubernetesClusterMasterRegionalToTerraform(struct.regional),
        zonal: kubernetesClusterMasterZonalToTerraform(struct.zonal),
    };
}
exports.kubernetesClusterMasterToTerraform = kubernetesClusterMasterToTerraform;
class KubernetesClusterMasterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // version_info - computed: true, optional: false, required: false
        this._versionInfo = new KubernetesClusterMasterVersionInfoList(this, "version_info", false);
        // maintenance_policy - computed: false, optional: true, required: false
        this._maintenancePolicy = new KubernetesClusterMasterMaintenancePolicyOutputReference(this, "maintenance_policy");
        // regional - computed: false, optional: true, required: false
        this._regional = new KubernetesClusterMasterRegionalOutputReference(this, "regional");
        // zonal - computed: false, optional: true, required: false
        this._zonal = new KubernetesClusterMasterZonalOutputReference(this, "zonal");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._publicIp !== undefined) {
            hasAnyValues = true;
            internalValueResult.publicIp = this._publicIp;
        }
        if (this._securityGroupIds !== undefined) {
            hasAnyValues = true;
            internalValueResult.securityGroupIds = this._securityGroupIds;
        }
        if (this._version !== undefined) {
            hasAnyValues = true;
            internalValueResult.version = this._version;
        }
        if (this._maintenancePolicy?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.maintenancePolicy = this._maintenancePolicy?.internalValue;
        }
        if (this._regional?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.regional = this._regional?.internalValue;
        }
        if (this._zonal?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.zonal = this._zonal?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._publicIp = undefined;
            this._securityGroupIds = undefined;
            this._version = undefined;
            this._maintenancePolicy.internalValue = undefined;
            this._regional.internalValue = undefined;
            this._zonal.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._publicIp = value.publicIp;
            this._securityGroupIds = value.securityGroupIds;
            this._version = value.version;
            this._maintenancePolicy.internalValue = value.maintenancePolicy;
            this._regional.internalValue = value.regional;
            this._zonal.internalValue = value.zonal;
        }
    }
    // cluster_ca_certificate - computed: true, optional: false, required: false
    get clusterCaCertificate() {
        return this.getStringAttribute('cluster_ca_certificate');
    }
    // external_v4_address - computed: true, optional: false, required: false
    get externalV4Address() {
        return this.getStringAttribute('external_v4_address');
    }
    // external_v4_endpoint - computed: true, optional: false, required: false
    get externalV4Endpoint() {
        return this.getStringAttribute('external_v4_endpoint');
    }
    // internal_v4_address - computed: true, optional: false, required: false
    get internalV4Address() {
        return this.getStringAttribute('internal_v4_address');
    }
    // internal_v4_endpoint - computed: true, optional: false, required: false
    get internalV4Endpoint() {
        return this.getStringAttribute('internal_v4_endpoint');
    }
    get publicIp() {
        return this.getBooleanAttribute('public_ip');
    }
    set publicIp(value) {
        this._publicIp = value;
    }
    resetPublicIp() {
        this._publicIp = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get publicIpInput() {
        return this._publicIp;
    }
    get securityGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('security_group_ids'));
    }
    set securityGroupIds(value) {
        this._securityGroupIds = value;
    }
    resetSecurityGroupIds() {
        this._securityGroupIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupIdsInput() {
        return this._securityGroupIds;
    }
    get version() {
        return this.getStringAttribute('version');
    }
    set version(value) {
        this._version = value;
    }
    resetVersion() {
        this._version = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get versionInput() {
        return this._version;
    }
    get versionInfo() {
        return this._versionInfo;
    }
    get maintenancePolicy() {
        return this._maintenancePolicy;
    }
    putMaintenancePolicy(value) {
        this._maintenancePolicy.internalValue = value;
    }
    resetMaintenancePolicy() {
        this._maintenancePolicy.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maintenancePolicyInput() {
        return this._maintenancePolicy.internalValue;
    }
    get regional() {
        return this._regional;
    }
    putRegional(value) {
        this._regional.internalValue = value;
    }
    resetRegional() {
        this._regional.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regionalInput() {
        return this._regional.internalValue;
    }
    get zonal() {
        return this._zonal;
    }
    putZonal(value) {
        this._zonal.internalValue = value;
    }
    resetZonal() {
        this._zonal.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get zonalInput() {
        return this._zonal.internalValue;
    }
}
exports.KubernetesClusterMasterOutputReference = KubernetesClusterMasterOutputReference;
_l = JSII_RTTI_SYMBOL_1;
KubernetesClusterMasterOutputReference[_l] = { fqn: "@cdktf/provider-yandex.KubernetesClusterMasterOutputReference", version: "2.0.32" };
function kubernetesClusterNetworkImplementationCiliumToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.kubernetesClusterNetworkImplementationCiliumToTerraform = kubernetesClusterNetworkImplementationCiliumToTerraform;
class KubernetesClusterNetworkImplementationCiliumOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
}
exports.KubernetesClusterNetworkImplementationCiliumOutputReference = KubernetesClusterNetworkImplementationCiliumOutputReference;
_m = JSII_RTTI_SYMBOL_1;
KubernetesClusterNetworkImplementationCiliumOutputReference[_m] = { fqn: "@cdktf/provider-yandex.KubernetesClusterNetworkImplementationCiliumOutputReference", version: "2.0.32" };
function kubernetesClusterNetworkImplementationToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        cilium: kubernetesClusterNetworkImplementationCiliumToTerraform(struct.cilium),
    };
}
exports.kubernetesClusterNetworkImplementationToTerraform = kubernetesClusterNetworkImplementationToTerraform;
class KubernetesClusterNetworkImplementationOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // cilium - computed: false, optional: true, required: false
        this._cilium = new KubernetesClusterNetworkImplementationCiliumOutputReference(this, "cilium");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._cilium?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.cilium = this._cilium?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._cilium.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._cilium.internalValue = value.cilium;
        }
    }
    get cilium() {
        return this._cilium;
    }
    putCilium(value) {
        this._cilium.internalValue = value;
    }
    resetCilium() {
        this._cilium.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ciliumInput() {
        return this._cilium.internalValue;
    }
}
exports.KubernetesClusterNetworkImplementationOutputReference = KubernetesClusterNetworkImplementationOutputReference;
_o = JSII_RTTI_SYMBOL_1;
KubernetesClusterNetworkImplementationOutputReference[_o] = { fqn: "@cdktf/provider-yandex.KubernetesClusterNetworkImplementationOutputReference", version: "2.0.32" };
function kubernetesClusterTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        read: cdktf.stringToTerraform(struct.read),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.kubernetesClusterTimeoutsToTerraform = kubernetesClusterTimeoutsToTerraform;
class KubernetesClusterTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._read !== undefined) {
            hasAnyValues = true;
            internalValueResult.read = this._read;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._read = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._read = value.read;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get read() {
        return this.getStringAttribute('read');
    }
    set read(value) {
        this._read = value;
    }
    resetRead() {
        this._read = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get readInput() {
        return this._read;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.KubernetesClusterTimeoutsOutputReference = KubernetesClusterTimeoutsOutputReference;
_p = JSII_RTTI_SYMBOL_1;
KubernetesClusterTimeoutsOutputReference[_p] = { fqn: "@cdktf/provider-yandex.KubernetesClusterTimeoutsOutputReference", version: "2.0.32" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/r/kubernetes_cluster yandex_kubernetes_cluster}
*/
class KubernetesCluster extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/r/kubernetes_cluster yandex_kubernetes_cluster} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options KubernetesClusterConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'yandex_kubernetes_cluster',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // kms_provider - computed: false, optional: true, required: false
        this._kmsProvider = new KubernetesClusterKmsProviderOutputReference(this, "kms_provider");
        // master - computed: false, optional: false, required: true
        this._master = new KubernetesClusterMasterOutputReference(this, "master");
        // network_implementation - computed: false, optional: true, required: false
        this._networkImplementation = new KubernetesClusterNetworkImplementationOutputReference(this, "network_implementation");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new KubernetesClusterTimeoutsOutputReference(this, "timeouts");
        this._clusterIpv4Range = config.clusterIpv4Range;
        this._clusterIpv6Range = config.clusterIpv6Range;
        this._description = config.description;
        this._folderId = config.folderId;
        this._id = config.id;
        this._labels = config.labels;
        this._name = config.name;
        this._networkId = config.networkId;
        this._networkPolicyProvider = config.networkPolicyProvider;
        this._nodeIpv4CidrMaskSize = config.nodeIpv4CidrMaskSize;
        this._nodeServiceAccountId = config.nodeServiceAccountId;
        this._releaseChannel = config.releaseChannel;
        this._serviceAccountId = config.serviceAccountId;
        this._serviceIpv4Range = config.serviceIpv4Range;
        this._serviceIpv6Range = config.serviceIpv6Range;
        this._kmsProvider.internalValue = config.kmsProvider;
        this._master.internalValue = config.master;
        this._networkImplementation.internalValue = config.networkImplementation;
        this._timeouts.internalValue = config.timeouts;
    }
    get clusterIpv4Range() {
        return this.getStringAttribute('cluster_ipv4_range');
    }
    set clusterIpv4Range(value) {
        this._clusterIpv4Range = value;
    }
    resetClusterIpv4Range() {
        this._clusterIpv4Range = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clusterIpv4RangeInput() {
        return this._clusterIpv4Range;
    }
    get clusterIpv6Range() {
        return this.getStringAttribute('cluster_ipv6_range');
    }
    set clusterIpv6Range(value) {
        this._clusterIpv6Range = value;
    }
    resetClusterIpv6Range() {
        this._clusterIpv6Range = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clusterIpv6RangeInput() {
        return this._clusterIpv6Range;
    }
    // created_at - computed: true, optional: false, required: false
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get folderId() {
        return this.getStringAttribute('folder_id');
    }
    set folderId(value) {
        this._folderId = value;
    }
    resetFolderId() {
        this._folderId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get folderIdInput() {
        return this._folderId;
    }
    // health - computed: true, optional: false, required: false
    get health() {
        return this.getStringAttribute('health');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    // log_group_id - computed: true, optional: false, required: false
    get logGroupId() {
        return this.getStringAttribute('log_group_id');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get networkId() {
        return this.getStringAttribute('network_id');
    }
    set networkId(value) {
        this._networkId = value;
    }
    // Temporarily expose input value. Use with caution.
    get networkIdInput() {
        return this._networkId;
    }
    get networkPolicyProvider() {
        return this.getStringAttribute('network_policy_provider');
    }
    set networkPolicyProvider(value) {
        this._networkPolicyProvider = value;
    }
    resetNetworkPolicyProvider() {
        this._networkPolicyProvider = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get networkPolicyProviderInput() {
        return this._networkPolicyProvider;
    }
    get nodeIpv4CidrMaskSize() {
        return this.getNumberAttribute('node_ipv4_cidr_mask_size');
    }
    set nodeIpv4CidrMaskSize(value) {
        this._nodeIpv4CidrMaskSize = value;
    }
    resetNodeIpv4CidrMaskSize() {
        this._nodeIpv4CidrMaskSize = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nodeIpv4CidrMaskSizeInput() {
        return this._nodeIpv4CidrMaskSize;
    }
    get nodeServiceAccountId() {
        return this.getStringAttribute('node_service_account_id');
    }
    set nodeServiceAccountId(value) {
        this._nodeServiceAccountId = value;
    }
    // Temporarily expose input value. Use with caution.
    get nodeServiceAccountIdInput() {
        return this._nodeServiceAccountId;
    }
    get releaseChannel() {
        return this.getStringAttribute('release_channel');
    }
    set releaseChannel(value) {
        this._releaseChannel = value;
    }
    resetReleaseChannel() {
        this._releaseChannel = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get releaseChannelInput() {
        return this._releaseChannel;
    }
    get serviceAccountId() {
        return this.getStringAttribute('service_account_id');
    }
    set serviceAccountId(value) {
        this._serviceAccountId = value;
    }
    // Temporarily expose input value. Use with caution.
    get serviceAccountIdInput() {
        return this._serviceAccountId;
    }
    get serviceIpv4Range() {
        return this.getStringAttribute('service_ipv4_range');
    }
    set serviceIpv4Range(value) {
        this._serviceIpv4Range = value;
    }
    resetServiceIpv4Range() {
        this._serviceIpv4Range = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get serviceIpv4RangeInput() {
        return this._serviceIpv4Range;
    }
    get serviceIpv6Range() {
        return this.getStringAttribute('service_ipv6_range');
    }
    set serviceIpv6Range(value) {
        this._serviceIpv6Range = value;
    }
    resetServiceIpv6Range() {
        this._serviceIpv6Range = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get serviceIpv6RangeInput() {
        return this._serviceIpv6Range;
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    get kmsProvider() {
        return this._kmsProvider;
    }
    putKmsProvider(value) {
        this._kmsProvider.internalValue = value;
    }
    resetKmsProvider() {
        this._kmsProvider.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kmsProviderInput() {
        return this._kmsProvider.internalValue;
    }
    get master() {
        return this._master;
    }
    putMaster(value) {
        this._master.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get masterInput() {
        return this._master.internalValue;
    }
    get networkImplementation() {
        return this._networkImplementation;
    }
    putNetworkImplementation(value) {
        this._networkImplementation.internalValue = value;
    }
    resetNetworkImplementation() {
        this._networkImplementation.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get networkImplementationInput() {
        return this._networkImplementation.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            cluster_ipv4_range: cdktf.stringToTerraform(this._clusterIpv4Range),
            cluster_ipv6_range: cdktf.stringToTerraform(this._clusterIpv6Range),
            description: cdktf.stringToTerraform(this._description),
            folder_id: cdktf.stringToTerraform(this._folderId),
            id: cdktf.stringToTerraform(this._id),
            labels: cdktf.hashMapper(cdktf.stringToTerraform)(this._labels),
            name: cdktf.stringToTerraform(this._name),
            network_id: cdktf.stringToTerraform(this._networkId),
            network_policy_provider: cdktf.stringToTerraform(this._networkPolicyProvider),
            node_ipv4_cidr_mask_size: cdktf.numberToTerraform(this._nodeIpv4CidrMaskSize),
            node_service_account_id: cdktf.stringToTerraform(this._nodeServiceAccountId),
            release_channel: cdktf.stringToTerraform(this._releaseChannel),
            service_account_id: cdktf.stringToTerraform(this._serviceAccountId),
            service_ipv4_range: cdktf.stringToTerraform(this._serviceIpv4Range),
            service_ipv6_range: cdktf.stringToTerraform(this._serviceIpv6Range),
            kms_provider: kubernetesClusterKmsProviderToTerraform(this._kmsProvider.internalValue),
            master: kubernetesClusterMasterToTerraform(this._master.internalValue),
            network_implementation: kubernetesClusterNetworkImplementationToTerraform(this._networkImplementation.internalValue),
            timeouts: kubernetesClusterTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.KubernetesCluster = KubernetesCluster;
_q = JSII_RTTI_SYMBOL_1;
KubernetesCluster[_q] = { fqn: "@cdktf/provider-yandex.KubernetesCluster", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
KubernetesCluster.tfResourceType = "yandex_kubernetes_cluster";
