// @cdktf/provider-yandex/lib/lb-network-load-balancer.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LbNetworkLoadBalancer = exports.LbNetworkLoadBalancerTimeoutsOutputReference = exports.lbNetworkLoadBalancerTimeoutsToTerraform = exports.LbNetworkLoadBalancerListenerList = exports.LbNetworkLoadBalancerListenerOutputReference = exports.lbNetworkLoadBalancerListenerToTerraform = exports.LbNetworkLoadBalancerListenerInternalAddressSpecOutputReference = exports.lbNetworkLoadBalancerListenerInternalAddressSpecToTerraform = exports.LbNetworkLoadBalancerListenerExternalAddressSpecOutputReference = exports.lbNetworkLoadBalancerListenerExternalAddressSpecToTerraform = exports.LbNetworkLoadBalancerAttachedTargetGroupList = exports.LbNetworkLoadBalancerAttachedTargetGroupOutputReference = exports.lbNetworkLoadBalancerAttachedTargetGroupToTerraform = exports.LbNetworkLoadBalancerAttachedTargetGroupHealthcheckList = exports.LbNetworkLoadBalancerAttachedTargetGroupHealthcheckOutputReference = exports.lbNetworkLoadBalancerAttachedTargetGroupHealthcheckToTerraform = exports.LbNetworkLoadBalancerAttachedTargetGroupHealthcheckTcpOptionsOutputReference = exports.lbNetworkLoadBalancerAttachedTargetGroupHealthcheckTcpOptionsToTerraform = exports.LbNetworkLoadBalancerAttachedTargetGroupHealthcheckHttpOptionsOutputReference = exports.lbNetworkLoadBalancerAttachedTargetGroupHealthcheckHttpOptionsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function lbNetworkLoadBalancerAttachedTargetGroupHealthcheckHttpOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        path: cdktf.stringToTerraform(struct.path),
        port: cdktf.numberToTerraform(struct.port),
    };
}
exports.lbNetworkLoadBalancerAttachedTargetGroupHealthcheckHttpOptionsToTerraform = lbNetworkLoadBalancerAttachedTargetGroupHealthcheckHttpOptionsToTerraform;
class LbNetworkLoadBalancerAttachedTargetGroupHealthcheckHttpOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._path !== undefined) {
            hasAnyValues = true;
            internalValueResult.path = this._path;
        }
        if (this._port !== undefined) {
            hasAnyValues = true;
            internalValueResult.port = this._port;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._path = undefined;
            this._port = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._path = value.path;
            this._port = value.port;
        }
    }
    get path() {
        return this.getStringAttribute('path');
    }
    set path(value) {
        this._path = value;
    }
    resetPath() {
        this._path = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pathInput() {
        return this._path;
    }
    get port() {
        return this.getNumberAttribute('port');
    }
    set port(value) {
        this._port = value;
    }
    // Temporarily expose input value. Use with caution.
    get portInput() {
        return this._port;
    }
}
exports.LbNetworkLoadBalancerAttachedTargetGroupHealthcheckHttpOptionsOutputReference = LbNetworkLoadBalancerAttachedTargetGroupHealthcheckHttpOptionsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
LbNetworkLoadBalancerAttachedTargetGroupHealthcheckHttpOptionsOutputReference[_a] = { fqn: "@cdktf/provider-yandex.LbNetworkLoadBalancerAttachedTargetGroupHealthcheckHttpOptionsOutputReference", version: "2.0.32" };
function lbNetworkLoadBalancerAttachedTargetGroupHealthcheckTcpOptionsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        port: cdktf.numberToTerraform(struct.port),
    };
}
exports.lbNetworkLoadBalancerAttachedTargetGroupHealthcheckTcpOptionsToTerraform = lbNetworkLoadBalancerAttachedTargetGroupHealthcheckTcpOptionsToTerraform;
class LbNetworkLoadBalancerAttachedTargetGroupHealthcheckTcpOptionsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._port !== undefined) {
            hasAnyValues = true;
            internalValueResult.port = this._port;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._port = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._port = value.port;
        }
    }
    get port() {
        return this.getNumberAttribute('port');
    }
    set port(value) {
        this._port = value;
    }
    // Temporarily expose input value. Use with caution.
    get portInput() {
        return this._port;
    }
}
exports.LbNetworkLoadBalancerAttachedTargetGroupHealthcheckTcpOptionsOutputReference = LbNetworkLoadBalancerAttachedTargetGroupHealthcheckTcpOptionsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
LbNetworkLoadBalancerAttachedTargetGroupHealthcheckTcpOptionsOutputReference[_b] = { fqn: "@cdktf/provider-yandex.LbNetworkLoadBalancerAttachedTargetGroupHealthcheckTcpOptionsOutputReference", version: "2.0.32" };
function lbNetworkLoadBalancerAttachedTargetGroupHealthcheckToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        healthy_threshold: cdktf.numberToTerraform(struct.healthyThreshold),
        interval: cdktf.numberToTerraform(struct.interval),
        name: cdktf.stringToTerraform(struct.name),
        timeout: cdktf.numberToTerraform(struct.timeout),
        unhealthy_threshold: cdktf.numberToTerraform(struct.unhealthyThreshold),
        http_options: lbNetworkLoadBalancerAttachedTargetGroupHealthcheckHttpOptionsToTerraform(struct.httpOptions),
        tcp_options: lbNetworkLoadBalancerAttachedTargetGroupHealthcheckTcpOptionsToTerraform(struct.tcpOptions),
    };
}
exports.lbNetworkLoadBalancerAttachedTargetGroupHealthcheckToTerraform = lbNetworkLoadBalancerAttachedTargetGroupHealthcheckToTerraform;
class LbNetworkLoadBalancerAttachedTargetGroupHealthcheckOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // http_options - computed: false, optional: true, required: false
        this._httpOptions = new LbNetworkLoadBalancerAttachedTargetGroupHealthcheckHttpOptionsOutputReference(this, "http_options");
        // tcp_options - computed: false, optional: true, required: false
        this._tcpOptions = new LbNetworkLoadBalancerAttachedTargetGroupHealthcheckTcpOptionsOutputReference(this, "tcp_options");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._healthyThreshold !== undefined) {
            hasAnyValues = true;
            internalValueResult.healthyThreshold = this._healthyThreshold;
        }
        if (this._interval !== undefined) {
            hasAnyValues = true;
            internalValueResult.interval = this._interval;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._timeout !== undefined) {
            hasAnyValues = true;
            internalValueResult.timeout = this._timeout;
        }
        if (this._unhealthyThreshold !== undefined) {
            hasAnyValues = true;
            internalValueResult.unhealthyThreshold = this._unhealthyThreshold;
        }
        if (this._httpOptions?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.httpOptions = this._httpOptions?.internalValue;
        }
        if (this._tcpOptions?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.tcpOptions = this._tcpOptions?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._healthyThreshold = undefined;
            this._interval = undefined;
            this._name = undefined;
            this._timeout = undefined;
            this._unhealthyThreshold = undefined;
            this._httpOptions.internalValue = undefined;
            this._tcpOptions.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._healthyThreshold = value.healthyThreshold;
            this._interval = value.interval;
            this._name = value.name;
            this._timeout = value.timeout;
            this._unhealthyThreshold = value.unhealthyThreshold;
            this._httpOptions.internalValue = value.httpOptions;
            this._tcpOptions.internalValue = value.tcpOptions;
        }
    }
    get healthyThreshold() {
        return this.getNumberAttribute('healthy_threshold');
    }
    set healthyThreshold(value) {
        this._healthyThreshold = value;
    }
    resetHealthyThreshold() {
        this._healthyThreshold = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get healthyThresholdInput() {
        return this._healthyThreshold;
    }
    get interval() {
        return this.getNumberAttribute('interval');
    }
    set interval(value) {
        this._interval = value;
    }
    resetInterval() {
        this._interval = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get intervalInput() {
        return this._interval;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get timeout() {
        return this.getNumberAttribute('timeout');
    }
    set timeout(value) {
        this._timeout = value;
    }
    resetTimeout() {
        this._timeout = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutInput() {
        return this._timeout;
    }
    get unhealthyThreshold() {
        return this.getNumberAttribute('unhealthy_threshold');
    }
    set unhealthyThreshold(value) {
        this._unhealthyThreshold = value;
    }
    resetUnhealthyThreshold() {
        this._unhealthyThreshold = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get unhealthyThresholdInput() {
        return this._unhealthyThreshold;
    }
    get httpOptions() {
        return this._httpOptions;
    }
    putHttpOptions(value) {
        this._httpOptions.internalValue = value;
    }
    resetHttpOptions() {
        this._httpOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get httpOptionsInput() {
        return this._httpOptions.internalValue;
    }
    get tcpOptions() {
        return this._tcpOptions;
    }
    putTcpOptions(value) {
        this._tcpOptions.internalValue = value;
    }
    resetTcpOptions() {
        this._tcpOptions.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tcpOptionsInput() {
        return this._tcpOptions.internalValue;
    }
}
exports.LbNetworkLoadBalancerAttachedTargetGroupHealthcheckOutputReference = LbNetworkLoadBalancerAttachedTargetGroupHealthcheckOutputReference;
_c = JSII_RTTI_SYMBOL_1;
LbNetworkLoadBalancerAttachedTargetGroupHealthcheckOutputReference[_c] = { fqn: "@cdktf/provider-yandex.LbNetworkLoadBalancerAttachedTargetGroupHealthcheckOutputReference", version: "2.0.32" };
class LbNetworkLoadBalancerAttachedTargetGroupHealthcheckList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new LbNetworkLoadBalancerAttachedTargetGroupHealthcheckOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.LbNetworkLoadBalancerAttachedTargetGroupHealthcheckList = LbNetworkLoadBalancerAttachedTargetGroupHealthcheckList;
_d = JSII_RTTI_SYMBOL_1;
LbNetworkLoadBalancerAttachedTargetGroupHealthcheckList[_d] = { fqn: "@cdktf/provider-yandex.LbNetworkLoadBalancerAttachedTargetGroupHealthcheckList", version: "2.0.32" };
function lbNetworkLoadBalancerAttachedTargetGroupToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        target_group_id: cdktf.stringToTerraform(struct.targetGroupId),
        healthcheck: cdktf.listMapper(lbNetworkLoadBalancerAttachedTargetGroupHealthcheckToTerraform, true)(struct.healthcheck),
    };
}
exports.lbNetworkLoadBalancerAttachedTargetGroupToTerraform = lbNetworkLoadBalancerAttachedTargetGroupToTerraform;
class LbNetworkLoadBalancerAttachedTargetGroupOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // healthcheck - computed: false, optional: false, required: true
        this._healthcheck = new LbNetworkLoadBalancerAttachedTargetGroupHealthcheckList(this, "healthcheck", false);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._targetGroupId !== undefined) {
            hasAnyValues = true;
            internalValueResult.targetGroupId = this._targetGroupId;
        }
        if (this._healthcheck?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.healthcheck = this._healthcheck?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._targetGroupId = undefined;
            this._healthcheck.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._targetGroupId = value.targetGroupId;
            this._healthcheck.internalValue = value.healthcheck;
        }
    }
    get targetGroupId() {
        return this.getStringAttribute('target_group_id');
    }
    set targetGroupId(value) {
        this._targetGroupId = value;
    }
    // Temporarily expose input value. Use with caution.
    get targetGroupIdInput() {
        return this._targetGroupId;
    }
    get healthcheck() {
        return this._healthcheck;
    }
    putHealthcheck(value) {
        this._healthcheck.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get healthcheckInput() {
        return this._healthcheck.internalValue;
    }
}
exports.LbNetworkLoadBalancerAttachedTargetGroupOutputReference = LbNetworkLoadBalancerAttachedTargetGroupOutputReference;
_e = JSII_RTTI_SYMBOL_1;
LbNetworkLoadBalancerAttachedTargetGroupOutputReference[_e] = { fqn: "@cdktf/provider-yandex.LbNetworkLoadBalancerAttachedTargetGroupOutputReference", version: "2.0.32" };
class LbNetworkLoadBalancerAttachedTargetGroupList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new LbNetworkLoadBalancerAttachedTargetGroupOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.LbNetworkLoadBalancerAttachedTargetGroupList = LbNetworkLoadBalancerAttachedTargetGroupList;
_f = JSII_RTTI_SYMBOL_1;
LbNetworkLoadBalancerAttachedTargetGroupList[_f] = { fqn: "@cdktf/provider-yandex.LbNetworkLoadBalancerAttachedTargetGroupList", version: "2.0.32" };
function lbNetworkLoadBalancerListenerExternalAddressSpecToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        address: cdktf.stringToTerraform(struct.address),
        ip_version: cdktf.stringToTerraform(struct.ipVersion),
    };
}
exports.lbNetworkLoadBalancerListenerExternalAddressSpecToTerraform = lbNetworkLoadBalancerListenerExternalAddressSpecToTerraform;
class LbNetworkLoadBalancerListenerExternalAddressSpecOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._address !== undefined) {
            hasAnyValues = true;
            internalValueResult.address = this._address;
        }
        if (this._ipVersion !== undefined) {
            hasAnyValues = true;
            internalValueResult.ipVersion = this._ipVersion;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._address = undefined;
            this._ipVersion = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._address = value.address;
            this._ipVersion = value.ipVersion;
        }
    }
    get address() {
        return this.getStringAttribute('address');
    }
    set address(value) {
        this._address = value;
    }
    resetAddress() {
        this._address = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get addressInput() {
        return this._address;
    }
    get ipVersion() {
        return this.getStringAttribute('ip_version');
    }
    set ipVersion(value) {
        this._ipVersion = value;
    }
    resetIpVersion() {
        this._ipVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ipVersionInput() {
        return this._ipVersion;
    }
}
exports.LbNetworkLoadBalancerListenerExternalAddressSpecOutputReference = LbNetworkLoadBalancerListenerExternalAddressSpecOutputReference;
_g = JSII_RTTI_SYMBOL_1;
LbNetworkLoadBalancerListenerExternalAddressSpecOutputReference[_g] = { fqn: "@cdktf/provider-yandex.LbNetworkLoadBalancerListenerExternalAddressSpecOutputReference", version: "2.0.32" };
function lbNetworkLoadBalancerListenerInternalAddressSpecToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        address: cdktf.stringToTerraform(struct.address),
        ip_version: cdktf.stringToTerraform(struct.ipVersion),
        subnet_id: cdktf.stringToTerraform(struct.subnetId),
    };
}
exports.lbNetworkLoadBalancerListenerInternalAddressSpecToTerraform = lbNetworkLoadBalancerListenerInternalAddressSpecToTerraform;
class LbNetworkLoadBalancerListenerInternalAddressSpecOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._address !== undefined) {
            hasAnyValues = true;
            internalValueResult.address = this._address;
        }
        if (this._ipVersion !== undefined) {
            hasAnyValues = true;
            internalValueResult.ipVersion = this._ipVersion;
        }
        if (this._subnetId !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnetId = this._subnetId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._address = undefined;
            this._ipVersion = undefined;
            this._subnetId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._address = value.address;
            this._ipVersion = value.ipVersion;
            this._subnetId = value.subnetId;
        }
    }
    get address() {
        return this.getStringAttribute('address');
    }
    set address(value) {
        this._address = value;
    }
    resetAddress() {
        this._address = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get addressInput() {
        return this._address;
    }
    get ipVersion() {
        return this.getStringAttribute('ip_version');
    }
    set ipVersion(value) {
        this._ipVersion = value;
    }
    resetIpVersion() {
        this._ipVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ipVersionInput() {
        return this._ipVersion;
    }
    get subnetId() {
        return this.getStringAttribute('subnet_id');
    }
    set subnetId(value) {
        this._subnetId = value;
    }
    // Temporarily expose input value. Use with caution.
    get subnetIdInput() {
        return this._subnetId;
    }
}
exports.LbNetworkLoadBalancerListenerInternalAddressSpecOutputReference = LbNetworkLoadBalancerListenerInternalAddressSpecOutputReference;
_h = JSII_RTTI_SYMBOL_1;
LbNetworkLoadBalancerListenerInternalAddressSpecOutputReference[_h] = { fqn: "@cdktf/provider-yandex.LbNetworkLoadBalancerListenerInternalAddressSpecOutputReference", version: "2.0.32" };
function lbNetworkLoadBalancerListenerToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        port: cdktf.numberToTerraform(struct.port),
        protocol: cdktf.stringToTerraform(struct.protocol),
        target_port: cdktf.numberToTerraform(struct.targetPort),
        external_address_spec: lbNetworkLoadBalancerListenerExternalAddressSpecToTerraform(struct.externalAddressSpec),
        internal_address_spec: lbNetworkLoadBalancerListenerInternalAddressSpecToTerraform(struct.internalAddressSpec),
    };
}
exports.lbNetworkLoadBalancerListenerToTerraform = lbNetworkLoadBalancerListenerToTerraform;
class LbNetworkLoadBalancerListenerOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // external_address_spec - computed: false, optional: true, required: false
        this._externalAddressSpec = new LbNetworkLoadBalancerListenerExternalAddressSpecOutputReference(this, "external_address_spec");
        // internal_address_spec - computed: false, optional: true, required: false
        this._internalAddressSpec = new LbNetworkLoadBalancerListenerInternalAddressSpecOutputReference(this, "internal_address_spec");
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._port !== undefined) {
            hasAnyValues = true;
            internalValueResult.port = this._port;
        }
        if (this._protocol !== undefined) {
            hasAnyValues = true;
            internalValueResult.protocol = this._protocol;
        }
        if (this._targetPort !== undefined) {
            hasAnyValues = true;
            internalValueResult.targetPort = this._targetPort;
        }
        if (this._externalAddressSpec?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.externalAddressSpec = this._externalAddressSpec?.internalValue;
        }
        if (this._internalAddressSpec?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.internalAddressSpec = this._internalAddressSpec?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._port = undefined;
            this._protocol = undefined;
            this._targetPort = undefined;
            this._externalAddressSpec.internalValue = undefined;
            this._internalAddressSpec.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._port = value.port;
            this._protocol = value.protocol;
            this._targetPort = value.targetPort;
            this._externalAddressSpec.internalValue = value.externalAddressSpec;
            this._internalAddressSpec.internalValue = value.internalAddressSpec;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get port() {
        return this.getNumberAttribute('port');
    }
    set port(value) {
        this._port = value;
    }
    // Temporarily expose input value. Use with caution.
    get portInput() {
        return this._port;
    }
    get protocol() {
        return this.getStringAttribute('protocol');
    }
    set protocol(value) {
        this._protocol = value;
    }
    resetProtocol() {
        this._protocol = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get protocolInput() {
        return this._protocol;
    }
    get targetPort() {
        return this.getNumberAttribute('target_port');
    }
    set targetPort(value) {
        this._targetPort = value;
    }
    resetTargetPort() {
        this._targetPort = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get targetPortInput() {
        return this._targetPort;
    }
    get externalAddressSpec() {
        return this._externalAddressSpec;
    }
    putExternalAddressSpec(value) {
        this._externalAddressSpec.internalValue = value;
    }
    resetExternalAddressSpec() {
        this._externalAddressSpec.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get externalAddressSpecInput() {
        return this._externalAddressSpec.internalValue;
    }
    get internalAddressSpec() {
        return this._internalAddressSpec;
    }
    putInternalAddressSpec(value) {
        this._internalAddressSpec.internalValue = value;
    }
    resetInternalAddressSpec() {
        this._internalAddressSpec.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get internalAddressSpecInput() {
        return this._internalAddressSpec.internalValue;
    }
}
exports.LbNetworkLoadBalancerListenerOutputReference = LbNetworkLoadBalancerListenerOutputReference;
_j = JSII_RTTI_SYMBOL_1;
LbNetworkLoadBalancerListenerOutputReference[_j] = { fqn: "@cdktf/provider-yandex.LbNetworkLoadBalancerListenerOutputReference", version: "2.0.32" };
class LbNetworkLoadBalancerListenerList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new LbNetworkLoadBalancerListenerOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.LbNetworkLoadBalancerListenerList = LbNetworkLoadBalancerListenerList;
_k = JSII_RTTI_SYMBOL_1;
LbNetworkLoadBalancerListenerList[_k] = { fqn: "@cdktf/provider-yandex.LbNetworkLoadBalancerListenerList", version: "2.0.32" };
function lbNetworkLoadBalancerTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.lbNetworkLoadBalancerTimeoutsToTerraform = lbNetworkLoadBalancerTimeoutsToTerraform;
class LbNetworkLoadBalancerTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.LbNetworkLoadBalancerTimeoutsOutputReference = LbNetworkLoadBalancerTimeoutsOutputReference;
_l = JSII_RTTI_SYMBOL_1;
LbNetworkLoadBalancerTimeoutsOutputReference[_l] = { fqn: "@cdktf/provider-yandex.LbNetworkLoadBalancerTimeoutsOutputReference", version: "2.0.32" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/r/lb_network_load_balancer yandex_lb_network_load_balancer}
*/
class LbNetworkLoadBalancer extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/r/lb_network_load_balancer yandex_lb_network_load_balancer} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options LbNetworkLoadBalancerConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'yandex_lb_network_load_balancer',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // attached_target_group - computed: false, optional: true, required: false
        this._attachedTargetGroup = new LbNetworkLoadBalancerAttachedTargetGroupList(this, "attached_target_group", true);
        // listener - computed: false, optional: true, required: false
        this._listener = new LbNetworkLoadBalancerListenerList(this, "listener", true);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new LbNetworkLoadBalancerTimeoutsOutputReference(this, "timeouts");
        this._description = config.description;
        this._folderId = config.folderId;
        this._id = config.id;
        this._labels = config.labels;
        this._name = config.name;
        this._regionId = config.regionId;
        this._type = config.type;
        this._attachedTargetGroup.internalValue = config.attachedTargetGroup;
        this._listener.internalValue = config.listener;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // created_at - computed: true, optional: false, required: false
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get folderId() {
        return this.getStringAttribute('folder_id');
    }
    set folderId(value) {
        this._folderId = value;
    }
    resetFolderId() {
        this._folderId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get folderIdInput() {
        return this._folderId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get regionId() {
        return this.getStringAttribute('region_id');
    }
    set regionId(value) {
        this._regionId = value;
    }
    resetRegionId() {
        this._regionId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get regionIdInput() {
        return this._regionId;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get attachedTargetGroup() {
        return this._attachedTargetGroup;
    }
    putAttachedTargetGroup(value) {
        this._attachedTargetGroup.internalValue = value;
    }
    resetAttachedTargetGroup() {
        this._attachedTargetGroup.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get attachedTargetGroupInput() {
        return this._attachedTargetGroup.internalValue;
    }
    get listener() {
        return this._listener;
    }
    putListener(value) {
        this._listener.internalValue = value;
    }
    resetListener() {
        this._listener.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get listenerInput() {
        return this._listener.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            folder_id: cdktf.stringToTerraform(this._folderId),
            id: cdktf.stringToTerraform(this._id),
            labels: cdktf.hashMapper(cdktf.stringToTerraform)(this._labels),
            name: cdktf.stringToTerraform(this._name),
            region_id: cdktf.stringToTerraform(this._regionId),
            type: cdktf.stringToTerraform(this._type),
            attached_target_group: cdktf.listMapper(lbNetworkLoadBalancerAttachedTargetGroupToTerraform, true)(this._attachedTargetGroup.internalValue),
            listener: cdktf.listMapper(lbNetworkLoadBalancerListenerToTerraform, true)(this._listener.internalValue),
            timeouts: lbNetworkLoadBalancerTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.LbNetworkLoadBalancer = LbNetworkLoadBalancer;
_m = JSII_RTTI_SYMBOL_1;
LbNetworkLoadBalancer[_m] = { fqn: "@cdktf/provider-yandex.LbNetworkLoadBalancer", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
LbNetworkLoadBalancer.tfResourceType = "yandex_lb_network_load_balancer";
