// @cdktf/provider-yandex/lib/mdb-redis-cluster.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MdbRedisCluster = exports.MdbRedisClusterTimeoutsOutputReference = exports.mdbRedisClusterTimeoutsToTerraform = exports.MdbRedisClusterResourcesOutputReference = exports.mdbRedisClusterResourcesToTerraform = exports.MdbRedisClusterMaintenanceWindowOutputReference = exports.mdbRedisClusterMaintenanceWindowToTerraform = exports.MdbRedisClusterHostList = exports.MdbRedisClusterHostOutputReference = exports.mdbRedisClusterHostToTerraform = exports.MdbRedisClusterConfigAOutputReference = exports.mdbRedisClusterConfigAToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function mdbRedisClusterConfigAToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        client_output_buffer_limit_normal: cdktf.stringToTerraform(struct.clientOutputBufferLimitNormal),
        client_output_buffer_limit_pubsub: cdktf.stringToTerraform(struct.clientOutputBufferLimitPubsub),
        databases: cdktf.numberToTerraform(struct.databases),
        maxmemory_policy: cdktf.stringToTerraform(struct.maxmemoryPolicy),
        notify_keyspace_events: cdktf.stringToTerraform(struct.notifyKeyspaceEvents),
        password: cdktf.stringToTerraform(struct.password),
        slowlog_log_slower_than: cdktf.numberToTerraform(struct.slowlogLogSlowerThan),
        slowlog_max_len: cdktf.numberToTerraform(struct.slowlogMaxLen),
        timeout: cdktf.numberToTerraform(struct.timeout),
        version: cdktf.stringToTerraform(struct.version),
    };
}
exports.mdbRedisClusterConfigAToTerraform = mdbRedisClusterConfigAToTerraform;
class MdbRedisClusterConfigAOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._clientOutputBufferLimitNormal !== undefined) {
            hasAnyValues = true;
            internalValueResult.clientOutputBufferLimitNormal = this._clientOutputBufferLimitNormal;
        }
        if (this._clientOutputBufferLimitPubsub !== undefined) {
            hasAnyValues = true;
            internalValueResult.clientOutputBufferLimitPubsub = this._clientOutputBufferLimitPubsub;
        }
        if (this._databases !== undefined) {
            hasAnyValues = true;
            internalValueResult.databases = this._databases;
        }
        if (this._maxmemoryPolicy !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxmemoryPolicy = this._maxmemoryPolicy;
        }
        if (this._notifyKeyspaceEvents !== undefined) {
            hasAnyValues = true;
            internalValueResult.notifyKeyspaceEvents = this._notifyKeyspaceEvents;
        }
        if (this._password !== undefined) {
            hasAnyValues = true;
            internalValueResult.password = this._password;
        }
        if (this._slowlogLogSlowerThan !== undefined) {
            hasAnyValues = true;
            internalValueResult.slowlogLogSlowerThan = this._slowlogLogSlowerThan;
        }
        if (this._slowlogMaxLen !== undefined) {
            hasAnyValues = true;
            internalValueResult.slowlogMaxLen = this._slowlogMaxLen;
        }
        if (this._timeout !== undefined) {
            hasAnyValues = true;
            internalValueResult.timeout = this._timeout;
        }
        if (this._version !== undefined) {
            hasAnyValues = true;
            internalValueResult.version = this._version;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._clientOutputBufferLimitNormal = undefined;
            this._clientOutputBufferLimitPubsub = undefined;
            this._databases = undefined;
            this._maxmemoryPolicy = undefined;
            this._notifyKeyspaceEvents = undefined;
            this._password = undefined;
            this._slowlogLogSlowerThan = undefined;
            this._slowlogMaxLen = undefined;
            this._timeout = undefined;
            this._version = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._clientOutputBufferLimitNormal = value.clientOutputBufferLimitNormal;
            this._clientOutputBufferLimitPubsub = value.clientOutputBufferLimitPubsub;
            this._databases = value.databases;
            this._maxmemoryPolicy = value.maxmemoryPolicy;
            this._notifyKeyspaceEvents = value.notifyKeyspaceEvents;
            this._password = value.password;
            this._slowlogLogSlowerThan = value.slowlogLogSlowerThan;
            this._slowlogMaxLen = value.slowlogMaxLen;
            this._timeout = value.timeout;
            this._version = value.version;
        }
    }
    get clientOutputBufferLimitNormal() {
        return this.getStringAttribute('client_output_buffer_limit_normal');
    }
    set clientOutputBufferLimitNormal(value) {
        this._clientOutputBufferLimitNormal = value;
    }
    resetClientOutputBufferLimitNormal() {
        this._clientOutputBufferLimitNormal = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clientOutputBufferLimitNormalInput() {
        return this._clientOutputBufferLimitNormal;
    }
    get clientOutputBufferLimitPubsub() {
        return this.getStringAttribute('client_output_buffer_limit_pubsub');
    }
    set clientOutputBufferLimitPubsub(value) {
        this._clientOutputBufferLimitPubsub = value;
    }
    resetClientOutputBufferLimitPubsub() {
        this._clientOutputBufferLimitPubsub = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clientOutputBufferLimitPubsubInput() {
        return this._clientOutputBufferLimitPubsub;
    }
    get databases() {
        return this.getNumberAttribute('databases');
    }
    set databases(value) {
        this._databases = value;
    }
    resetDatabases() {
        this._databases = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get databasesInput() {
        return this._databases;
    }
    get maxmemoryPolicy() {
        return this.getStringAttribute('maxmemory_policy');
    }
    set maxmemoryPolicy(value) {
        this._maxmemoryPolicy = value;
    }
    resetMaxmemoryPolicy() {
        this._maxmemoryPolicy = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxmemoryPolicyInput() {
        return this._maxmemoryPolicy;
    }
    get notifyKeyspaceEvents() {
        return this.getStringAttribute('notify_keyspace_events');
    }
    set notifyKeyspaceEvents(value) {
        this._notifyKeyspaceEvents = value;
    }
    resetNotifyKeyspaceEvents() {
        this._notifyKeyspaceEvents = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get notifyKeyspaceEventsInput() {
        return this._notifyKeyspaceEvents;
    }
    get password() {
        return this.getStringAttribute('password');
    }
    set password(value) {
        this._password = value;
    }
    // Temporarily expose input value. Use with caution.
    get passwordInput() {
        return this._password;
    }
    get slowlogLogSlowerThan() {
        return this.getNumberAttribute('slowlog_log_slower_than');
    }
    set slowlogLogSlowerThan(value) {
        this._slowlogLogSlowerThan = value;
    }
    resetSlowlogLogSlowerThan() {
        this._slowlogLogSlowerThan = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get slowlogLogSlowerThanInput() {
        return this._slowlogLogSlowerThan;
    }
    get slowlogMaxLen() {
        return this.getNumberAttribute('slowlog_max_len');
    }
    set slowlogMaxLen(value) {
        this._slowlogMaxLen = value;
    }
    resetSlowlogMaxLen() {
        this._slowlogMaxLen = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get slowlogMaxLenInput() {
        return this._slowlogMaxLen;
    }
    get timeout() {
        return this.getNumberAttribute('timeout');
    }
    set timeout(value) {
        this._timeout = value;
    }
    resetTimeout() {
        this._timeout = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutInput() {
        return this._timeout;
    }
    get version() {
        return this.getStringAttribute('version');
    }
    set version(value) {
        this._version = value;
    }
    // Temporarily expose input value. Use with caution.
    get versionInput() {
        return this._version;
    }
}
exports.MdbRedisClusterConfigAOutputReference = MdbRedisClusterConfigAOutputReference;
_a = JSII_RTTI_SYMBOL_1;
MdbRedisClusterConfigAOutputReference[_a] = { fqn: "@cdktf/provider-yandex.MdbRedisClusterConfigAOutputReference", version: "2.0.32" };
function mdbRedisClusterHostToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        assign_public_ip: cdktf.booleanToTerraform(struct.assignPublicIp),
        replica_priority: cdktf.numberToTerraform(struct.replicaPriority),
        shard_name: cdktf.stringToTerraform(struct.shardName),
        subnet_id: cdktf.stringToTerraform(struct.subnetId),
        zone: cdktf.stringToTerraform(struct.zone),
    };
}
exports.mdbRedisClusterHostToTerraform = mdbRedisClusterHostToTerraform;
class MdbRedisClusterHostOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._assignPublicIp !== undefined) {
            hasAnyValues = true;
            internalValueResult.assignPublicIp = this._assignPublicIp;
        }
        if (this._replicaPriority !== undefined) {
            hasAnyValues = true;
            internalValueResult.replicaPriority = this._replicaPriority;
        }
        if (this._shardName !== undefined) {
            hasAnyValues = true;
            internalValueResult.shardName = this._shardName;
        }
        if (this._subnetId !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnetId = this._subnetId;
        }
        if (this._zone !== undefined) {
            hasAnyValues = true;
            internalValueResult.zone = this._zone;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._assignPublicIp = undefined;
            this._replicaPriority = undefined;
            this._shardName = undefined;
            this._subnetId = undefined;
            this._zone = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._assignPublicIp = value.assignPublicIp;
            this._replicaPriority = value.replicaPriority;
            this._shardName = value.shardName;
            this._subnetId = value.subnetId;
            this._zone = value.zone;
        }
    }
    get assignPublicIp() {
        return this.getBooleanAttribute('assign_public_ip');
    }
    set assignPublicIp(value) {
        this._assignPublicIp = value;
    }
    resetAssignPublicIp() {
        this._assignPublicIp = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get assignPublicIpInput() {
        return this._assignPublicIp;
    }
    // fqdn - computed: true, optional: false, required: false
    get fqdn() {
        return this.getStringAttribute('fqdn');
    }
    get replicaPriority() {
        return this.getNumberAttribute('replica_priority');
    }
    set replicaPriority(value) {
        this._replicaPriority = value;
    }
    resetReplicaPriority() {
        this._replicaPriority = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get replicaPriorityInput() {
        return this._replicaPriority;
    }
    get shardName() {
        return this.getStringAttribute('shard_name');
    }
    set shardName(value) {
        this._shardName = value;
    }
    resetShardName() {
        this._shardName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get shardNameInput() {
        return this._shardName;
    }
    get subnetId() {
        return this.getStringAttribute('subnet_id');
    }
    set subnetId(value) {
        this._subnetId = value;
    }
    resetSubnetId() {
        this._subnetId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get subnetIdInput() {
        return this._subnetId;
    }
    get zone() {
        return this.getStringAttribute('zone');
    }
    set zone(value) {
        this._zone = value;
    }
    // Temporarily expose input value. Use with caution.
    get zoneInput() {
        return this._zone;
    }
}
exports.MdbRedisClusterHostOutputReference = MdbRedisClusterHostOutputReference;
_b = JSII_RTTI_SYMBOL_1;
MdbRedisClusterHostOutputReference[_b] = { fqn: "@cdktf/provider-yandex.MdbRedisClusterHostOutputReference", version: "2.0.32" };
class MdbRedisClusterHostList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new MdbRedisClusterHostOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.MdbRedisClusterHostList = MdbRedisClusterHostList;
_c = JSII_RTTI_SYMBOL_1;
MdbRedisClusterHostList[_c] = { fqn: "@cdktf/provider-yandex.MdbRedisClusterHostList", version: "2.0.32" };
function mdbRedisClusterMaintenanceWindowToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        day: cdktf.stringToTerraform(struct.day),
        hour: cdktf.numberToTerraform(struct.hour),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.mdbRedisClusterMaintenanceWindowToTerraform = mdbRedisClusterMaintenanceWindowToTerraform;
class MdbRedisClusterMaintenanceWindowOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._day !== undefined) {
            hasAnyValues = true;
            internalValueResult.day = this._day;
        }
        if (this._hour !== undefined) {
            hasAnyValues = true;
            internalValueResult.hour = this._hour;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._day = undefined;
            this._hour = undefined;
            this._type = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._day = value.day;
            this._hour = value.hour;
            this._type = value.type;
        }
    }
    get day() {
        return this.getStringAttribute('day');
    }
    set day(value) {
        this._day = value;
    }
    resetDay() {
        this._day = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dayInput() {
        return this._day;
    }
    get hour() {
        return this.getNumberAttribute('hour');
    }
    set hour(value) {
        this._hour = value;
    }
    resetHour() {
        this._hour = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get hourInput() {
        return this._hour;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.MdbRedisClusterMaintenanceWindowOutputReference = MdbRedisClusterMaintenanceWindowOutputReference;
_d = JSII_RTTI_SYMBOL_1;
MdbRedisClusterMaintenanceWindowOutputReference[_d] = { fqn: "@cdktf/provider-yandex.MdbRedisClusterMaintenanceWindowOutputReference", version: "2.0.32" };
function mdbRedisClusterResourcesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        disk_size: cdktf.numberToTerraform(struct.diskSize),
        disk_type_id: cdktf.stringToTerraform(struct.diskTypeId),
        resource_preset_id: cdktf.stringToTerraform(struct.resourcePresetId),
    };
}
exports.mdbRedisClusterResourcesToTerraform = mdbRedisClusterResourcesToTerraform;
class MdbRedisClusterResourcesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._diskSize !== undefined) {
            hasAnyValues = true;
            internalValueResult.diskSize = this._diskSize;
        }
        if (this._diskTypeId !== undefined) {
            hasAnyValues = true;
            internalValueResult.diskTypeId = this._diskTypeId;
        }
        if (this._resourcePresetId !== undefined) {
            hasAnyValues = true;
            internalValueResult.resourcePresetId = this._resourcePresetId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._diskSize = undefined;
            this._diskTypeId = undefined;
            this._resourcePresetId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._diskSize = value.diskSize;
            this._diskTypeId = value.diskTypeId;
            this._resourcePresetId = value.resourcePresetId;
        }
    }
    get diskSize() {
        return this.getNumberAttribute('disk_size');
    }
    set diskSize(value) {
        this._diskSize = value;
    }
    // Temporarily expose input value. Use with caution.
    get diskSizeInput() {
        return this._diskSize;
    }
    get diskTypeId() {
        return this.getStringAttribute('disk_type_id');
    }
    set diskTypeId(value) {
        this._diskTypeId = value;
    }
    resetDiskTypeId() {
        this._diskTypeId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get diskTypeIdInput() {
        return this._diskTypeId;
    }
    get resourcePresetId() {
        return this.getStringAttribute('resource_preset_id');
    }
    set resourcePresetId(value) {
        this._resourcePresetId = value;
    }
    // Temporarily expose input value. Use with caution.
    get resourcePresetIdInput() {
        return this._resourcePresetId;
    }
}
exports.MdbRedisClusterResourcesOutputReference = MdbRedisClusterResourcesOutputReference;
_e = JSII_RTTI_SYMBOL_1;
MdbRedisClusterResourcesOutputReference[_e] = { fqn: "@cdktf/provider-yandex.MdbRedisClusterResourcesOutputReference", version: "2.0.32" };
function mdbRedisClusterTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.mdbRedisClusterTimeoutsToTerraform = mdbRedisClusterTimeoutsToTerraform;
class MdbRedisClusterTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.MdbRedisClusterTimeoutsOutputReference = MdbRedisClusterTimeoutsOutputReference;
_f = JSII_RTTI_SYMBOL_1;
MdbRedisClusterTimeoutsOutputReference[_f] = { fqn: "@cdktf/provider-yandex.MdbRedisClusterTimeoutsOutputReference", version: "2.0.32" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/r/mdb_redis_cluster yandex_mdb_redis_cluster}
*/
class MdbRedisCluster extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/r/mdb_redis_cluster yandex_mdb_redis_cluster} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options MdbRedisClusterConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'yandex_mdb_redis_cluster',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // config - computed: false, optional: false, required: true
        this._config = new MdbRedisClusterConfigAOutputReference(this, "config");
        // host - computed: false, optional: false, required: true
        this._host = new MdbRedisClusterHostList(this, "host", false);
        // maintenance_window - computed: false, optional: true, required: false
        this._maintenanceWindow = new MdbRedisClusterMaintenanceWindowOutputReference(this, "maintenance_window");
        // resources - computed: false, optional: false, required: true
        this._resources = new MdbRedisClusterResourcesOutputReference(this, "resources");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new MdbRedisClusterTimeoutsOutputReference(this, "timeouts");
        this._deletionProtection = config.deletionProtection;
        this._description = config.description;
        this._environment = config.environment;
        this._folderId = config.folderId;
        this._id = config.id;
        this._labels = config.labels;
        this._name = config.name;
        this._networkId = config.networkId;
        this._persistenceMode = config.persistenceMode;
        this._securityGroupIds = config.securityGroupIds;
        this._sharded = config.sharded;
        this._tlsEnabled = config.tlsEnabled;
        this._config.internalValue = config.config;
        this._host.internalValue = config.host;
        this._maintenanceWindow.internalValue = config.maintenanceWindow;
        this._resources.internalValue = config.resources;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // created_at - computed: true, optional: false, required: false
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    get deletionProtection() {
        return this.getBooleanAttribute('deletion_protection');
    }
    set deletionProtection(value) {
        this._deletionProtection = value;
    }
    resetDeletionProtection() {
        this._deletionProtection = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deletionProtectionInput() {
        return this._deletionProtection;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get environment() {
        return this.getStringAttribute('environment');
    }
    set environment(value) {
        this._environment = value;
    }
    // Temporarily expose input value. Use with caution.
    get environmentInput() {
        return this._environment;
    }
    get folderId() {
        return this.getStringAttribute('folder_id');
    }
    set folderId(value) {
        this._folderId = value;
    }
    resetFolderId() {
        this._folderId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get folderIdInput() {
        return this._folderId;
    }
    // health - computed: true, optional: false, required: false
    get health() {
        return this.getStringAttribute('health');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get networkId() {
        return this.getStringAttribute('network_id');
    }
    set networkId(value) {
        this._networkId = value;
    }
    // Temporarily expose input value. Use with caution.
    get networkIdInput() {
        return this._networkId;
    }
    get persistenceMode() {
        return this.getStringAttribute('persistence_mode');
    }
    set persistenceMode(value) {
        this._persistenceMode = value;
    }
    resetPersistenceMode() {
        this._persistenceMode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get persistenceModeInput() {
        return this._persistenceMode;
    }
    get securityGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('security_group_ids'));
    }
    set securityGroupIds(value) {
        this._securityGroupIds = value;
    }
    resetSecurityGroupIds() {
        this._securityGroupIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupIdsInput() {
        return this._securityGroupIds;
    }
    get sharded() {
        return this.getBooleanAttribute('sharded');
    }
    set sharded(value) {
        this._sharded = value;
    }
    resetSharded() {
        this._sharded = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get shardedInput() {
        return this._sharded;
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    get tlsEnabled() {
        return this.getBooleanAttribute('tls_enabled');
    }
    set tlsEnabled(value) {
        this._tlsEnabled = value;
    }
    resetTlsEnabled() {
        this._tlsEnabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get tlsEnabledInput() {
        return this._tlsEnabled;
    }
    get config() {
        return this._config;
    }
    putConfig(value) {
        this._config.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get configInput() {
        return this._config.internalValue;
    }
    get host() {
        return this._host;
    }
    putHost(value) {
        this._host.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get hostInput() {
        return this._host.internalValue;
    }
    get maintenanceWindow() {
        return this._maintenanceWindow;
    }
    putMaintenanceWindow(value) {
        this._maintenanceWindow.internalValue = value;
    }
    resetMaintenanceWindow() {
        this._maintenanceWindow.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maintenanceWindowInput() {
        return this._maintenanceWindow.internalValue;
    }
    get resources() {
        return this._resources;
    }
    putResources(value) {
        this._resources.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get resourcesInput() {
        return this._resources.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            deletion_protection: cdktf.booleanToTerraform(this._deletionProtection),
            description: cdktf.stringToTerraform(this._description),
            environment: cdktf.stringToTerraform(this._environment),
            folder_id: cdktf.stringToTerraform(this._folderId),
            id: cdktf.stringToTerraform(this._id),
            labels: cdktf.hashMapper(cdktf.stringToTerraform)(this._labels),
            name: cdktf.stringToTerraform(this._name),
            network_id: cdktf.stringToTerraform(this._networkId),
            persistence_mode: cdktf.stringToTerraform(this._persistenceMode),
            security_group_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(this._securityGroupIds),
            sharded: cdktf.booleanToTerraform(this._sharded),
            tls_enabled: cdktf.booleanToTerraform(this._tlsEnabled),
            config: mdbRedisClusterConfigAToTerraform(this._config.internalValue),
            host: cdktf.listMapper(mdbRedisClusterHostToTerraform, true)(this._host.internalValue),
            maintenance_window: mdbRedisClusterMaintenanceWindowToTerraform(this._maintenanceWindow.internalValue),
            resources: mdbRedisClusterResourcesToTerraform(this._resources.internalValue),
            timeouts: mdbRedisClusterTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.MdbRedisCluster = MdbRedisCluster;
_g = JSII_RTTI_SYMBOL_1;
MdbRedisCluster[_g] = { fqn: "@cdktf/provider-yandex.MdbRedisCluster", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
MdbRedisCluster.tfResourceType = "yandex_mdb_redis_cluster";
