// @cdktf/provider-yandex/lib/mdb-elasticsearch-cluster.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MdbElasticsearchCluster = exports.MdbElasticsearchClusterTimeoutsOutputReference = exports.mdbElasticsearchClusterTimeoutsToTerraform = exports.MdbElasticsearchClusterMaintenanceWindowOutputReference = exports.mdbElasticsearchClusterMaintenanceWindowToTerraform = exports.MdbElasticsearchClusterHostList = exports.MdbElasticsearchClusterHostOutputReference = exports.mdbElasticsearchClusterHostToTerraform = exports.MdbElasticsearchClusterConfigAOutputReference = exports.mdbElasticsearchClusterConfigAToTerraform = exports.MdbElasticsearchClusterConfigMasterNodeOutputReference = exports.mdbElasticsearchClusterConfigMasterNodeToTerraform = exports.MdbElasticsearchClusterConfigMasterNodeResourcesOutputReference = exports.mdbElasticsearchClusterConfigMasterNodeResourcesToTerraform = exports.MdbElasticsearchClusterConfigDataNodeOutputReference = exports.mdbElasticsearchClusterConfigDataNodeToTerraform = exports.MdbElasticsearchClusterConfigDataNodeResourcesOutputReference = exports.mdbElasticsearchClusterConfigDataNodeResourcesToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function mdbElasticsearchClusterConfigDataNodeResourcesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        disk_size: cdktf.numberToTerraform(struct.diskSize),
        disk_type_id: cdktf.stringToTerraform(struct.diskTypeId),
        resource_preset_id: cdktf.stringToTerraform(struct.resourcePresetId),
    };
}
exports.mdbElasticsearchClusterConfigDataNodeResourcesToTerraform = mdbElasticsearchClusterConfigDataNodeResourcesToTerraform;
class MdbElasticsearchClusterConfigDataNodeResourcesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._diskSize !== undefined) {
            hasAnyValues = true;
            internalValueResult.diskSize = this._diskSize;
        }
        if (this._diskTypeId !== undefined) {
            hasAnyValues = true;
            internalValueResult.diskTypeId = this._diskTypeId;
        }
        if (this._resourcePresetId !== undefined) {
            hasAnyValues = true;
            internalValueResult.resourcePresetId = this._resourcePresetId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._diskSize = undefined;
            this._diskTypeId = undefined;
            this._resourcePresetId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._diskSize = value.diskSize;
            this._diskTypeId = value.diskTypeId;
            this._resourcePresetId = value.resourcePresetId;
        }
    }
    get diskSize() {
        return this.getNumberAttribute('disk_size');
    }
    set diskSize(value) {
        this._diskSize = value;
    }
    // Temporarily expose input value. Use with caution.
    get diskSizeInput() {
        return this._diskSize;
    }
    get diskTypeId() {
        return this.getStringAttribute('disk_type_id');
    }
    set diskTypeId(value) {
        this._diskTypeId = value;
    }
    // Temporarily expose input value. Use with caution.
    get diskTypeIdInput() {
        return this._diskTypeId;
    }
    get resourcePresetId() {
        return this.getStringAttribute('resource_preset_id');
    }
    set resourcePresetId(value) {
        this._resourcePresetId = value;
    }
    // Temporarily expose input value. Use with caution.
    get resourcePresetIdInput() {
        return this._resourcePresetId;
    }
}
exports.MdbElasticsearchClusterConfigDataNodeResourcesOutputReference = MdbElasticsearchClusterConfigDataNodeResourcesOutputReference;
_a = JSII_RTTI_SYMBOL_1;
MdbElasticsearchClusterConfigDataNodeResourcesOutputReference[_a] = { fqn: "@cdktf/provider-yandex.MdbElasticsearchClusterConfigDataNodeResourcesOutputReference", version: "2.0.32" };
function mdbElasticsearchClusterConfigDataNodeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        resources: mdbElasticsearchClusterConfigDataNodeResourcesToTerraform(struct.resources),
    };
}
exports.mdbElasticsearchClusterConfigDataNodeToTerraform = mdbElasticsearchClusterConfigDataNodeToTerraform;
class MdbElasticsearchClusterConfigDataNodeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // resources - computed: false, optional: false, required: true
        this._resources = new MdbElasticsearchClusterConfigDataNodeResourcesOutputReference(this, "resources");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._resources?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.resources = this._resources?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._resources.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._resources.internalValue = value.resources;
        }
    }
    get resources() {
        return this._resources;
    }
    putResources(value) {
        this._resources.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get resourcesInput() {
        return this._resources.internalValue;
    }
}
exports.MdbElasticsearchClusterConfigDataNodeOutputReference = MdbElasticsearchClusterConfigDataNodeOutputReference;
_b = JSII_RTTI_SYMBOL_1;
MdbElasticsearchClusterConfigDataNodeOutputReference[_b] = { fqn: "@cdktf/provider-yandex.MdbElasticsearchClusterConfigDataNodeOutputReference", version: "2.0.32" };
function mdbElasticsearchClusterConfigMasterNodeResourcesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        disk_size: cdktf.numberToTerraform(struct.diskSize),
        disk_type_id: cdktf.stringToTerraform(struct.diskTypeId),
        resource_preset_id: cdktf.stringToTerraform(struct.resourcePresetId),
    };
}
exports.mdbElasticsearchClusterConfigMasterNodeResourcesToTerraform = mdbElasticsearchClusterConfigMasterNodeResourcesToTerraform;
class MdbElasticsearchClusterConfigMasterNodeResourcesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._diskSize !== undefined) {
            hasAnyValues = true;
            internalValueResult.diskSize = this._diskSize;
        }
        if (this._diskTypeId !== undefined) {
            hasAnyValues = true;
            internalValueResult.diskTypeId = this._diskTypeId;
        }
        if (this._resourcePresetId !== undefined) {
            hasAnyValues = true;
            internalValueResult.resourcePresetId = this._resourcePresetId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._diskSize = undefined;
            this._diskTypeId = undefined;
            this._resourcePresetId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._diskSize = value.diskSize;
            this._diskTypeId = value.diskTypeId;
            this._resourcePresetId = value.resourcePresetId;
        }
    }
    get diskSize() {
        return this.getNumberAttribute('disk_size');
    }
    set diskSize(value) {
        this._diskSize = value;
    }
    // Temporarily expose input value. Use with caution.
    get diskSizeInput() {
        return this._diskSize;
    }
    get diskTypeId() {
        return this.getStringAttribute('disk_type_id');
    }
    set diskTypeId(value) {
        this._diskTypeId = value;
    }
    // Temporarily expose input value. Use with caution.
    get diskTypeIdInput() {
        return this._diskTypeId;
    }
    get resourcePresetId() {
        return this.getStringAttribute('resource_preset_id');
    }
    set resourcePresetId(value) {
        this._resourcePresetId = value;
    }
    // Temporarily expose input value. Use with caution.
    get resourcePresetIdInput() {
        return this._resourcePresetId;
    }
}
exports.MdbElasticsearchClusterConfigMasterNodeResourcesOutputReference = MdbElasticsearchClusterConfigMasterNodeResourcesOutputReference;
_c = JSII_RTTI_SYMBOL_1;
MdbElasticsearchClusterConfigMasterNodeResourcesOutputReference[_c] = { fqn: "@cdktf/provider-yandex.MdbElasticsearchClusterConfigMasterNodeResourcesOutputReference", version: "2.0.32" };
function mdbElasticsearchClusterConfigMasterNodeToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        resources: mdbElasticsearchClusterConfigMasterNodeResourcesToTerraform(struct.resources),
    };
}
exports.mdbElasticsearchClusterConfigMasterNodeToTerraform = mdbElasticsearchClusterConfigMasterNodeToTerraform;
class MdbElasticsearchClusterConfigMasterNodeOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // resources - computed: false, optional: false, required: true
        this._resources = new MdbElasticsearchClusterConfigMasterNodeResourcesOutputReference(this, "resources");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._resources?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.resources = this._resources?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._resources.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._resources.internalValue = value.resources;
        }
    }
    get resources() {
        return this._resources;
    }
    putResources(value) {
        this._resources.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get resourcesInput() {
        return this._resources.internalValue;
    }
}
exports.MdbElasticsearchClusterConfigMasterNodeOutputReference = MdbElasticsearchClusterConfigMasterNodeOutputReference;
_d = JSII_RTTI_SYMBOL_1;
MdbElasticsearchClusterConfigMasterNodeOutputReference[_d] = { fqn: "@cdktf/provider-yandex.MdbElasticsearchClusterConfigMasterNodeOutputReference", version: "2.0.32" };
function mdbElasticsearchClusterConfigAToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        admin_password: cdktf.stringToTerraform(struct.adminPassword),
        edition: cdktf.stringToTerraform(struct.edition),
        plugins: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.plugins),
        version: cdktf.stringToTerraform(struct.version),
        data_node: mdbElasticsearchClusterConfigDataNodeToTerraform(struct.dataNode),
        master_node: mdbElasticsearchClusterConfigMasterNodeToTerraform(struct.masterNode),
    };
}
exports.mdbElasticsearchClusterConfigAToTerraform = mdbElasticsearchClusterConfigAToTerraform;
class MdbElasticsearchClusterConfigAOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // data_node - computed: false, optional: false, required: true
        this._dataNode = new MdbElasticsearchClusterConfigDataNodeOutputReference(this, "data_node");
        // master_node - computed: false, optional: true, required: false
        this._masterNode = new MdbElasticsearchClusterConfigMasterNodeOutputReference(this, "master_node");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._adminPassword !== undefined) {
            hasAnyValues = true;
            internalValueResult.adminPassword = this._adminPassword;
        }
        if (this._edition !== undefined) {
            hasAnyValues = true;
            internalValueResult.edition = this._edition;
        }
        if (this._plugins !== undefined) {
            hasAnyValues = true;
            internalValueResult.plugins = this._plugins;
        }
        if (this._version !== undefined) {
            hasAnyValues = true;
            internalValueResult.version = this._version;
        }
        if (this._dataNode?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.dataNode = this._dataNode?.internalValue;
        }
        if (this._masterNode?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.masterNode = this._masterNode?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._adminPassword = undefined;
            this._edition = undefined;
            this._plugins = undefined;
            this._version = undefined;
            this._dataNode.internalValue = undefined;
            this._masterNode.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._adminPassword = value.adminPassword;
            this._edition = value.edition;
            this._plugins = value.plugins;
            this._version = value.version;
            this._dataNode.internalValue = value.dataNode;
            this._masterNode.internalValue = value.masterNode;
        }
    }
    get adminPassword() {
        return this.getStringAttribute('admin_password');
    }
    set adminPassword(value) {
        this._adminPassword = value;
    }
    // Temporarily expose input value. Use with caution.
    get adminPasswordInput() {
        return this._adminPassword;
    }
    get edition() {
        return this.getStringAttribute('edition');
    }
    set edition(value) {
        this._edition = value;
    }
    resetEdition() {
        this._edition = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get editionInput() {
        return this._edition;
    }
    get plugins() {
        return cdktf.Fn.tolist(this.getListAttribute('plugins'));
    }
    set plugins(value) {
        this._plugins = value;
    }
    resetPlugins() {
        this._plugins = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get pluginsInput() {
        return this._plugins;
    }
    get version() {
        return this.getStringAttribute('version');
    }
    set version(value) {
        this._version = value;
    }
    resetVersion() {
        this._version = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get versionInput() {
        return this._version;
    }
    get dataNode() {
        return this._dataNode;
    }
    putDataNode(value) {
        this._dataNode.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get dataNodeInput() {
        return this._dataNode.internalValue;
    }
    get masterNode() {
        return this._masterNode;
    }
    putMasterNode(value) {
        this._masterNode.internalValue = value;
    }
    resetMasterNode() {
        this._masterNode.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get masterNodeInput() {
        return this._masterNode.internalValue;
    }
}
exports.MdbElasticsearchClusterConfigAOutputReference = MdbElasticsearchClusterConfigAOutputReference;
_e = JSII_RTTI_SYMBOL_1;
MdbElasticsearchClusterConfigAOutputReference[_e] = { fqn: "@cdktf/provider-yandex.MdbElasticsearchClusterConfigAOutputReference", version: "2.0.32" };
function mdbElasticsearchClusterHostToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        assign_public_ip: cdktf.booleanToTerraform(struct.assignPublicIp),
        name: cdktf.stringToTerraform(struct.name),
        subnet_id: cdktf.stringToTerraform(struct.subnetId),
        type: cdktf.stringToTerraform(struct.type),
        zone: cdktf.stringToTerraform(struct.zone),
    };
}
exports.mdbElasticsearchClusterHostToTerraform = mdbElasticsearchClusterHostToTerraform;
class MdbElasticsearchClusterHostOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._assignPublicIp !== undefined) {
            hasAnyValues = true;
            internalValueResult.assignPublicIp = this._assignPublicIp;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._subnetId !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnetId = this._subnetId;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._zone !== undefined) {
            hasAnyValues = true;
            internalValueResult.zone = this._zone;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._assignPublicIp = undefined;
            this._name = undefined;
            this._subnetId = undefined;
            this._type = undefined;
            this._zone = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._assignPublicIp = value.assignPublicIp;
            this._name = value.name;
            this._subnetId = value.subnetId;
            this._type = value.type;
            this._zone = value.zone;
        }
    }
    get assignPublicIp() {
        return this.getBooleanAttribute('assign_public_ip');
    }
    set assignPublicIp(value) {
        this._assignPublicIp = value;
    }
    resetAssignPublicIp() {
        this._assignPublicIp = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get assignPublicIpInput() {
        return this._assignPublicIp;
    }
    // fqdn - computed: true, optional: false, required: false
    get fqdn() {
        return this.getStringAttribute('fqdn');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get subnetId() {
        return this.getStringAttribute('subnet_id');
    }
    set subnetId(value) {
        this._subnetId = value;
    }
    resetSubnetId() {
        this._subnetId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get subnetIdInput() {
        return this._subnetId;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get zone() {
        return this.getStringAttribute('zone');
    }
    set zone(value) {
        this._zone = value;
    }
    // Temporarily expose input value. Use with caution.
    get zoneInput() {
        return this._zone;
    }
}
exports.MdbElasticsearchClusterHostOutputReference = MdbElasticsearchClusterHostOutputReference;
_f = JSII_RTTI_SYMBOL_1;
MdbElasticsearchClusterHostOutputReference[_f] = { fqn: "@cdktf/provider-yandex.MdbElasticsearchClusterHostOutputReference", version: "2.0.32" };
class MdbElasticsearchClusterHostList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new MdbElasticsearchClusterHostOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.MdbElasticsearchClusterHostList = MdbElasticsearchClusterHostList;
_g = JSII_RTTI_SYMBOL_1;
MdbElasticsearchClusterHostList[_g] = { fqn: "@cdktf/provider-yandex.MdbElasticsearchClusterHostList", version: "2.0.32" };
function mdbElasticsearchClusterMaintenanceWindowToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        day: cdktf.stringToTerraform(struct.day),
        hour: cdktf.numberToTerraform(struct.hour),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.mdbElasticsearchClusterMaintenanceWindowToTerraform = mdbElasticsearchClusterMaintenanceWindowToTerraform;
class MdbElasticsearchClusterMaintenanceWindowOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._day !== undefined) {
            hasAnyValues = true;
            internalValueResult.day = this._day;
        }
        if (this._hour !== undefined) {
            hasAnyValues = true;
            internalValueResult.hour = this._hour;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._day = undefined;
            this._hour = undefined;
            this._type = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._day = value.day;
            this._hour = value.hour;
            this._type = value.type;
        }
    }
    get day() {
        return this.getStringAttribute('day');
    }
    set day(value) {
        this._day = value;
    }
    resetDay() {
        this._day = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dayInput() {
        return this._day;
    }
    get hour() {
        return this.getNumberAttribute('hour');
    }
    set hour(value) {
        this._hour = value;
    }
    resetHour() {
        this._hour = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get hourInput() {
        return this._hour;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.MdbElasticsearchClusterMaintenanceWindowOutputReference = MdbElasticsearchClusterMaintenanceWindowOutputReference;
_h = JSII_RTTI_SYMBOL_1;
MdbElasticsearchClusterMaintenanceWindowOutputReference[_h] = { fqn: "@cdktf/provider-yandex.MdbElasticsearchClusterMaintenanceWindowOutputReference", version: "2.0.32" };
function mdbElasticsearchClusterTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.mdbElasticsearchClusterTimeoutsToTerraform = mdbElasticsearchClusterTimeoutsToTerraform;
class MdbElasticsearchClusterTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.MdbElasticsearchClusterTimeoutsOutputReference = MdbElasticsearchClusterTimeoutsOutputReference;
_j = JSII_RTTI_SYMBOL_1;
MdbElasticsearchClusterTimeoutsOutputReference[_j] = { fqn: "@cdktf/provider-yandex.MdbElasticsearchClusterTimeoutsOutputReference", version: "2.0.32" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/r/mdb_elasticsearch_cluster yandex_mdb_elasticsearch_cluster}
*/
class MdbElasticsearchCluster extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/r/mdb_elasticsearch_cluster yandex_mdb_elasticsearch_cluster} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options MdbElasticsearchClusterConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'yandex_mdb_elasticsearch_cluster',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // config - computed: false, optional: false, required: true
        this._config = new MdbElasticsearchClusterConfigAOutputReference(this, "config");
        // host - computed: false, optional: true, required: false
        this._host = new MdbElasticsearchClusterHostList(this, "host", true);
        // maintenance_window - computed: false, optional: true, required: false
        this._maintenanceWindow = new MdbElasticsearchClusterMaintenanceWindowOutputReference(this, "maintenance_window");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new MdbElasticsearchClusterTimeoutsOutputReference(this, "timeouts");
        this._deletionProtection = config.deletionProtection;
        this._description = config.description;
        this._environment = config.environment;
        this._folderId = config.folderId;
        this._id = config.id;
        this._labels = config.labels;
        this._name = config.name;
        this._networkId = config.networkId;
        this._securityGroupIds = config.securityGroupIds;
        this._serviceAccountId = config.serviceAccountId;
        this._config.internalValue = config.config;
        this._host.internalValue = config.host;
        this._maintenanceWindow.internalValue = config.maintenanceWindow;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // created_at - computed: true, optional: false, required: false
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    get deletionProtection() {
        return this.getBooleanAttribute('deletion_protection');
    }
    set deletionProtection(value) {
        this._deletionProtection = value;
    }
    resetDeletionProtection() {
        this._deletionProtection = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deletionProtectionInput() {
        return this._deletionProtection;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get environment() {
        return this.getStringAttribute('environment');
    }
    set environment(value) {
        this._environment = value;
    }
    // Temporarily expose input value. Use with caution.
    get environmentInput() {
        return this._environment;
    }
    get folderId() {
        return this.getStringAttribute('folder_id');
    }
    set folderId(value) {
        this._folderId = value;
    }
    resetFolderId() {
        this._folderId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get folderIdInput() {
        return this._folderId;
    }
    // health - computed: true, optional: false, required: false
    get health() {
        return this.getStringAttribute('health');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get networkId() {
        return this.getStringAttribute('network_id');
    }
    set networkId(value) {
        this._networkId = value;
    }
    // Temporarily expose input value. Use with caution.
    get networkIdInput() {
        return this._networkId;
    }
    get securityGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('security_group_ids'));
    }
    set securityGroupIds(value) {
        this._securityGroupIds = value;
    }
    resetSecurityGroupIds() {
        this._securityGroupIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupIdsInput() {
        return this._securityGroupIds;
    }
    get serviceAccountId() {
        return this.getStringAttribute('service_account_id');
    }
    set serviceAccountId(value) {
        this._serviceAccountId = value;
    }
    resetServiceAccountId() {
        this._serviceAccountId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get serviceAccountIdInput() {
        return this._serviceAccountId;
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    get config() {
        return this._config;
    }
    putConfig(value) {
        this._config.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get configInput() {
        return this._config.internalValue;
    }
    get host() {
        return this._host;
    }
    putHost(value) {
        this._host.internalValue = value;
    }
    resetHost() {
        this._host.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get hostInput() {
        return this._host.internalValue;
    }
    get maintenanceWindow() {
        return this._maintenanceWindow;
    }
    putMaintenanceWindow(value) {
        this._maintenanceWindow.internalValue = value;
    }
    resetMaintenanceWindow() {
        this._maintenanceWindow.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maintenanceWindowInput() {
        return this._maintenanceWindow.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            deletion_protection: cdktf.booleanToTerraform(this._deletionProtection),
            description: cdktf.stringToTerraform(this._description),
            environment: cdktf.stringToTerraform(this._environment),
            folder_id: cdktf.stringToTerraform(this._folderId),
            id: cdktf.stringToTerraform(this._id),
            labels: cdktf.hashMapper(cdktf.stringToTerraform)(this._labels),
            name: cdktf.stringToTerraform(this._name),
            network_id: cdktf.stringToTerraform(this._networkId),
            security_group_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(this._securityGroupIds),
            service_account_id: cdktf.stringToTerraform(this._serviceAccountId),
            config: mdbElasticsearchClusterConfigAToTerraform(this._config.internalValue),
            host: cdktf.listMapper(mdbElasticsearchClusterHostToTerraform, true)(this._host.internalValue),
            maintenance_window: mdbElasticsearchClusterMaintenanceWindowToTerraform(this._maintenanceWindow.internalValue),
            timeouts: mdbElasticsearchClusterTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.MdbElasticsearchCluster = MdbElasticsearchCluster;
_k = JSII_RTTI_SYMBOL_1;
MdbElasticsearchCluster[_k] = { fqn: "@cdktf/provider-yandex.MdbElasticsearchCluster", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
MdbElasticsearchCluster.tfResourceType = "yandex_mdb_elasticsearch_cluster";
