// @cdktf/provider-yandex/lib/mdb-mysql-user.js
"use strict";
var _a, _b, _c, _d, _e;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MdbMysqlUser = exports.MdbMysqlUserTimeoutsOutputReference = exports.mdbMysqlUserTimeoutsToTerraform = exports.MdbMysqlUserPermissionList = exports.MdbMysqlUserPermissionOutputReference = exports.mdbMysqlUserPermissionToTerraform = exports.MdbMysqlUserConnectionLimitsOutputReference = exports.mdbMysqlUserConnectionLimitsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function mdbMysqlUserConnectionLimitsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        max_connections_per_hour: cdktf.numberToTerraform(struct.maxConnectionsPerHour),
        max_questions_per_hour: cdktf.numberToTerraform(struct.maxQuestionsPerHour),
        max_updates_per_hour: cdktf.numberToTerraform(struct.maxUpdatesPerHour),
        max_user_connections: cdktf.numberToTerraform(struct.maxUserConnections),
    };
}
exports.mdbMysqlUserConnectionLimitsToTerraform = mdbMysqlUserConnectionLimitsToTerraform;
class MdbMysqlUserConnectionLimitsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._maxConnectionsPerHour !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxConnectionsPerHour = this._maxConnectionsPerHour;
        }
        if (this._maxQuestionsPerHour !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxQuestionsPerHour = this._maxQuestionsPerHour;
        }
        if (this._maxUpdatesPerHour !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxUpdatesPerHour = this._maxUpdatesPerHour;
        }
        if (this._maxUserConnections !== undefined) {
            hasAnyValues = true;
            internalValueResult.maxUserConnections = this._maxUserConnections;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._maxConnectionsPerHour = undefined;
            this._maxQuestionsPerHour = undefined;
            this._maxUpdatesPerHour = undefined;
            this._maxUserConnections = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._maxConnectionsPerHour = value.maxConnectionsPerHour;
            this._maxQuestionsPerHour = value.maxQuestionsPerHour;
            this._maxUpdatesPerHour = value.maxUpdatesPerHour;
            this._maxUserConnections = value.maxUserConnections;
        }
    }
    get maxConnectionsPerHour() {
        return this.getNumberAttribute('max_connections_per_hour');
    }
    set maxConnectionsPerHour(value) {
        this._maxConnectionsPerHour = value;
    }
    resetMaxConnectionsPerHour() {
        this._maxConnectionsPerHour = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxConnectionsPerHourInput() {
        return this._maxConnectionsPerHour;
    }
    get maxQuestionsPerHour() {
        return this.getNumberAttribute('max_questions_per_hour');
    }
    set maxQuestionsPerHour(value) {
        this._maxQuestionsPerHour = value;
    }
    resetMaxQuestionsPerHour() {
        this._maxQuestionsPerHour = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxQuestionsPerHourInput() {
        return this._maxQuestionsPerHour;
    }
    get maxUpdatesPerHour() {
        return this.getNumberAttribute('max_updates_per_hour');
    }
    set maxUpdatesPerHour(value) {
        this._maxUpdatesPerHour = value;
    }
    resetMaxUpdatesPerHour() {
        this._maxUpdatesPerHour = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxUpdatesPerHourInput() {
        return this._maxUpdatesPerHour;
    }
    get maxUserConnections() {
        return this.getNumberAttribute('max_user_connections');
    }
    set maxUserConnections(value) {
        this._maxUserConnections = value;
    }
    resetMaxUserConnections() {
        this._maxUserConnections = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maxUserConnectionsInput() {
        return this._maxUserConnections;
    }
}
exports.MdbMysqlUserConnectionLimitsOutputReference = MdbMysqlUserConnectionLimitsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
MdbMysqlUserConnectionLimitsOutputReference[_a] = { fqn: "@cdktf/provider-yandex.MdbMysqlUserConnectionLimitsOutputReference", version: "2.0.32" };
function mdbMysqlUserPermissionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        database_name: cdktf.stringToTerraform(struct.databaseName),
        roles: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.roles),
    };
}
exports.mdbMysqlUserPermissionToTerraform = mdbMysqlUserPermissionToTerraform;
class MdbMysqlUserPermissionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._databaseName !== undefined) {
            hasAnyValues = true;
            internalValueResult.databaseName = this._databaseName;
        }
        if (this._roles !== undefined) {
            hasAnyValues = true;
            internalValueResult.roles = this._roles;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._databaseName = undefined;
            this._roles = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._databaseName = value.databaseName;
            this._roles = value.roles;
        }
    }
    get databaseName() {
        return this.getStringAttribute('database_name');
    }
    set databaseName(value) {
        this._databaseName = value;
    }
    // Temporarily expose input value. Use with caution.
    get databaseNameInput() {
        return this._databaseName;
    }
    get roles() {
        return this.getListAttribute('roles');
    }
    set roles(value) {
        this._roles = value;
    }
    resetRoles() {
        this._roles = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get rolesInput() {
        return this._roles;
    }
}
exports.MdbMysqlUserPermissionOutputReference = MdbMysqlUserPermissionOutputReference;
_b = JSII_RTTI_SYMBOL_1;
MdbMysqlUserPermissionOutputReference[_b] = { fqn: "@cdktf/provider-yandex.MdbMysqlUserPermissionOutputReference", version: "2.0.32" };
class MdbMysqlUserPermissionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new MdbMysqlUserPermissionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.MdbMysqlUserPermissionList = MdbMysqlUserPermissionList;
_c = JSII_RTTI_SYMBOL_1;
MdbMysqlUserPermissionList[_c] = { fqn: "@cdktf/provider-yandex.MdbMysqlUserPermissionList", version: "2.0.32" };
function mdbMysqlUserTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        read: cdktf.stringToTerraform(struct.read),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.mdbMysqlUserTimeoutsToTerraform = mdbMysqlUserTimeoutsToTerraform;
class MdbMysqlUserTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._read !== undefined) {
            hasAnyValues = true;
            internalValueResult.read = this._read;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._read = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._read = value.read;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get read() {
        return this.getStringAttribute('read');
    }
    set read(value) {
        this._read = value;
    }
    resetRead() {
        this._read = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get readInput() {
        return this._read;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.MdbMysqlUserTimeoutsOutputReference = MdbMysqlUserTimeoutsOutputReference;
_d = JSII_RTTI_SYMBOL_1;
MdbMysqlUserTimeoutsOutputReference[_d] = { fqn: "@cdktf/provider-yandex.MdbMysqlUserTimeoutsOutputReference", version: "2.0.32" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/r/mdb_mysql_user yandex_mdb_mysql_user}
*/
class MdbMysqlUser extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/r/mdb_mysql_user yandex_mdb_mysql_user} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options MdbMysqlUserConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'yandex_mdb_mysql_user',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // connection_limits - computed: false, optional: true, required: false
        this._connectionLimits = new MdbMysqlUserConnectionLimitsOutputReference(this, "connection_limits");
        // permission - computed: false, optional: true, required: false
        this._permission = new MdbMysqlUserPermissionList(this, "permission", true);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new MdbMysqlUserTimeoutsOutputReference(this, "timeouts");
        this._authenticationPlugin = config.authenticationPlugin;
        this._clusterId = config.clusterId;
        this._globalPermissions = config.globalPermissions;
        this._id = config.id;
        this._name = config.name;
        this._password = config.password;
        this._connectionLimits.internalValue = config.connectionLimits;
        this._permission.internalValue = config.permission;
        this._timeouts.internalValue = config.timeouts;
    }
    get authenticationPlugin() {
        return this.getStringAttribute('authentication_plugin');
    }
    set authenticationPlugin(value) {
        this._authenticationPlugin = value;
    }
    resetAuthenticationPlugin() {
        this._authenticationPlugin = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get authenticationPluginInput() {
        return this._authenticationPlugin;
    }
    get clusterId() {
        return this.getStringAttribute('cluster_id');
    }
    set clusterId(value) {
        this._clusterId = value;
    }
    // Temporarily expose input value. Use with caution.
    get clusterIdInput() {
        return this._clusterId;
    }
    get globalPermissions() {
        return cdktf.Fn.tolist(this.getListAttribute('global_permissions'));
    }
    set globalPermissions(value) {
        this._globalPermissions = value;
    }
    resetGlobalPermissions() {
        this._globalPermissions = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get globalPermissionsInput() {
        return this._globalPermissions;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get password() {
        return this.getStringAttribute('password');
    }
    set password(value) {
        this._password = value;
    }
    // Temporarily expose input value. Use with caution.
    get passwordInput() {
        return this._password;
    }
    get connectionLimits() {
        return this._connectionLimits;
    }
    putConnectionLimits(value) {
        this._connectionLimits.internalValue = value;
    }
    resetConnectionLimits() {
        this._connectionLimits.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get connectionLimitsInput() {
        return this._connectionLimits.internalValue;
    }
    get permission() {
        return this._permission;
    }
    putPermission(value) {
        this._permission.internalValue = value;
    }
    resetPermission() {
        this._permission.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get permissionInput() {
        return this._permission.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            authentication_plugin: cdktf.stringToTerraform(this._authenticationPlugin),
            cluster_id: cdktf.stringToTerraform(this._clusterId),
            global_permissions: cdktf.listMapper(cdktf.stringToTerraform, false)(this._globalPermissions),
            id: cdktf.stringToTerraform(this._id),
            name: cdktf.stringToTerraform(this._name),
            password: cdktf.stringToTerraform(this._password),
            connection_limits: mdbMysqlUserConnectionLimitsToTerraform(this._connectionLimits.internalValue),
            permission: cdktf.listMapper(mdbMysqlUserPermissionToTerraform, true)(this._permission.internalValue),
            timeouts: mdbMysqlUserTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.MdbMysqlUser = MdbMysqlUser;
_e = JSII_RTTI_SYMBOL_1;
MdbMysqlUser[_e] = { fqn: "@cdktf/provider-yandex.MdbMysqlUser", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
MdbMysqlUser.tfResourceType = "yandex_mdb_mysql_user";
