// @cdktf/provider-yandex/lib/mdb-mongodb-cluster.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MdbMongodbCluster = exports.MdbMongodbClusterUserList = exports.MdbMongodbClusterUserOutputReference = exports.mdbMongodbClusterUserToTerraform = exports.MdbMongodbClusterUserPermissionList = exports.MdbMongodbClusterUserPermissionOutputReference = exports.mdbMongodbClusterUserPermissionToTerraform = exports.MdbMongodbClusterTimeoutsOutputReference = exports.mdbMongodbClusterTimeoutsToTerraform = exports.MdbMongodbClusterResourcesOutputReference = exports.mdbMongodbClusterResourcesToTerraform = exports.MdbMongodbClusterMaintenanceWindowOutputReference = exports.mdbMongodbClusterMaintenanceWindowToTerraform = exports.MdbMongodbClusterHostList = exports.MdbMongodbClusterHostOutputReference = exports.mdbMongodbClusterHostToTerraform = exports.MdbMongodbClusterDatabaseList = exports.MdbMongodbClusterDatabaseOutputReference = exports.mdbMongodbClusterDatabaseToTerraform = exports.MdbMongodbClusterClusterConfigOutputReference = exports.mdbMongodbClusterClusterConfigToTerraform = exports.MdbMongodbClusterClusterConfigMongodOutputReference = exports.mdbMongodbClusterClusterConfigMongodToTerraform = exports.MdbMongodbClusterClusterConfigMongodSetParameterOutputReference = exports.mdbMongodbClusterClusterConfigMongodSetParameterToTerraform = exports.MdbMongodbClusterClusterConfigMongodSecurityOutputReference = exports.mdbMongodbClusterClusterConfigMongodSecurityToTerraform = exports.MdbMongodbClusterClusterConfigMongodSecurityKmipOutputReference = exports.mdbMongodbClusterClusterConfigMongodSecurityKmipToTerraform = exports.MdbMongodbClusterClusterConfigMongodAuditLogOutputReference = exports.mdbMongodbClusterClusterConfigMongodAuditLogToTerraform = exports.MdbMongodbClusterClusterConfigBackupWindowStartOutputReference = exports.mdbMongodbClusterClusterConfigBackupWindowStartToTerraform = exports.MdbMongodbClusterClusterConfigAccessOutputReference = exports.mdbMongodbClusterClusterConfigAccessToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function mdbMongodbClusterClusterConfigAccessToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {};
}
exports.mdbMongodbClusterClusterConfigAccessToTerraform = mdbMongodbClusterClusterConfigAccessToTerraform;
class MdbMongodbClusterClusterConfigAccessOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
        }
    }
    // data_lens - computed: true, optional: false, required: false
    get dataLens() {
        return this.getBooleanAttribute('data_lens');
    }
}
exports.MdbMongodbClusterClusterConfigAccessOutputReference = MdbMongodbClusterClusterConfigAccessOutputReference;
_a = JSII_RTTI_SYMBOL_1;
MdbMongodbClusterClusterConfigAccessOutputReference[_a] = { fqn: "@cdktf/provider-yandex.MdbMongodbClusterClusterConfigAccessOutputReference", version: "2.0.32" };
function mdbMongodbClusterClusterConfigBackupWindowStartToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        hours: cdktf.numberToTerraform(struct.hours),
        minutes: cdktf.numberToTerraform(struct.minutes),
    };
}
exports.mdbMongodbClusterClusterConfigBackupWindowStartToTerraform = mdbMongodbClusterClusterConfigBackupWindowStartToTerraform;
class MdbMongodbClusterClusterConfigBackupWindowStartOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._hours !== undefined) {
            hasAnyValues = true;
            internalValueResult.hours = this._hours;
        }
        if (this._minutes !== undefined) {
            hasAnyValues = true;
            internalValueResult.minutes = this._minutes;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._hours = undefined;
            this._minutes = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._hours = value.hours;
            this._minutes = value.minutes;
        }
    }
    get hours() {
        return this.getNumberAttribute('hours');
    }
    set hours(value) {
        this._hours = value;
    }
    resetHours() {
        this._hours = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get hoursInput() {
        return this._hours;
    }
    get minutes() {
        return this.getNumberAttribute('minutes');
    }
    set minutes(value) {
        this._minutes = value;
    }
    resetMinutes() {
        this._minutes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get minutesInput() {
        return this._minutes;
    }
}
exports.MdbMongodbClusterClusterConfigBackupWindowStartOutputReference = MdbMongodbClusterClusterConfigBackupWindowStartOutputReference;
_b = JSII_RTTI_SYMBOL_1;
MdbMongodbClusterClusterConfigBackupWindowStartOutputReference[_b] = { fqn: "@cdktf/provider-yandex.MdbMongodbClusterClusterConfigBackupWindowStartOutputReference", version: "2.0.32" };
function mdbMongodbClusterClusterConfigMongodAuditLogToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        filter: cdktf.stringToTerraform(struct.filter),
        runtime_configuration: cdktf.booleanToTerraform(struct.runtimeConfiguration),
    };
}
exports.mdbMongodbClusterClusterConfigMongodAuditLogToTerraform = mdbMongodbClusterClusterConfigMongodAuditLogToTerraform;
class MdbMongodbClusterClusterConfigMongodAuditLogOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._filter !== undefined) {
            hasAnyValues = true;
            internalValueResult.filter = this._filter;
        }
        if (this._runtimeConfiguration !== undefined) {
            hasAnyValues = true;
            internalValueResult.runtimeConfiguration = this._runtimeConfiguration;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._filter = undefined;
            this._runtimeConfiguration = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._filter = value.filter;
            this._runtimeConfiguration = value.runtimeConfiguration;
        }
    }
    get filter() {
        return this.getStringAttribute('filter');
    }
    set filter(value) {
        this._filter = value;
    }
    resetFilter() {
        this._filter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get filterInput() {
        return this._filter;
    }
    get runtimeConfiguration() {
        return this.getBooleanAttribute('runtime_configuration');
    }
    set runtimeConfiguration(value) {
        this._runtimeConfiguration = value;
    }
    resetRuntimeConfiguration() {
        this._runtimeConfiguration = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get runtimeConfigurationInput() {
        return this._runtimeConfiguration;
    }
}
exports.MdbMongodbClusterClusterConfigMongodAuditLogOutputReference = MdbMongodbClusterClusterConfigMongodAuditLogOutputReference;
_c = JSII_RTTI_SYMBOL_1;
MdbMongodbClusterClusterConfigMongodAuditLogOutputReference[_c] = { fqn: "@cdktf/provider-yandex.MdbMongodbClusterClusterConfigMongodAuditLogOutputReference", version: "2.0.32" };
function mdbMongodbClusterClusterConfigMongodSecurityKmipToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        client_certificate: cdktf.stringToTerraform(struct.clientCertificate),
        key_identifier: cdktf.stringToTerraform(struct.keyIdentifier),
        port: cdktf.numberToTerraform(struct.port),
        server_ca: cdktf.stringToTerraform(struct.serverCa),
        server_name: cdktf.stringToTerraform(struct.serverName),
    };
}
exports.mdbMongodbClusterClusterConfigMongodSecurityKmipToTerraform = mdbMongodbClusterClusterConfigMongodSecurityKmipToTerraform;
class MdbMongodbClusterClusterConfigMongodSecurityKmipOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._clientCertificate !== undefined) {
            hasAnyValues = true;
            internalValueResult.clientCertificate = this._clientCertificate;
        }
        if (this._keyIdentifier !== undefined) {
            hasAnyValues = true;
            internalValueResult.keyIdentifier = this._keyIdentifier;
        }
        if (this._port !== undefined) {
            hasAnyValues = true;
            internalValueResult.port = this._port;
        }
        if (this._serverCa !== undefined) {
            hasAnyValues = true;
            internalValueResult.serverCa = this._serverCa;
        }
        if (this._serverName !== undefined) {
            hasAnyValues = true;
            internalValueResult.serverName = this._serverName;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._clientCertificate = undefined;
            this._keyIdentifier = undefined;
            this._port = undefined;
            this._serverCa = undefined;
            this._serverName = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._clientCertificate = value.clientCertificate;
            this._keyIdentifier = value.keyIdentifier;
            this._port = value.port;
            this._serverCa = value.serverCa;
            this._serverName = value.serverName;
        }
    }
    get clientCertificate() {
        return this.getStringAttribute('client_certificate');
    }
    set clientCertificate(value) {
        this._clientCertificate = value;
    }
    resetClientCertificate() {
        this._clientCertificate = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clientCertificateInput() {
        return this._clientCertificate;
    }
    get keyIdentifier() {
        return this.getStringAttribute('key_identifier');
    }
    set keyIdentifier(value) {
        this._keyIdentifier = value;
    }
    resetKeyIdentifier() {
        this._keyIdentifier = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get keyIdentifierInput() {
        return this._keyIdentifier;
    }
    get port() {
        return this.getNumberAttribute('port');
    }
    set port(value) {
        this._port = value;
    }
    resetPort() {
        this._port = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get portInput() {
        return this._port;
    }
    get serverCa() {
        return this.getStringAttribute('server_ca');
    }
    set serverCa(value) {
        this._serverCa = value;
    }
    resetServerCa() {
        this._serverCa = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get serverCaInput() {
        return this._serverCa;
    }
    get serverName() {
        return this.getStringAttribute('server_name');
    }
    set serverName(value) {
        this._serverName = value;
    }
    resetServerName() {
        this._serverName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get serverNameInput() {
        return this._serverName;
    }
}
exports.MdbMongodbClusterClusterConfigMongodSecurityKmipOutputReference = MdbMongodbClusterClusterConfigMongodSecurityKmipOutputReference;
_d = JSII_RTTI_SYMBOL_1;
MdbMongodbClusterClusterConfigMongodSecurityKmipOutputReference[_d] = { fqn: "@cdktf/provider-yandex.MdbMongodbClusterClusterConfigMongodSecurityKmipOutputReference", version: "2.0.32" };
function mdbMongodbClusterClusterConfigMongodSecurityToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enable_encryption: cdktf.booleanToTerraform(struct.enableEncryption),
        kmip: mdbMongodbClusterClusterConfigMongodSecurityKmipToTerraform(struct.kmip),
    };
}
exports.mdbMongodbClusterClusterConfigMongodSecurityToTerraform = mdbMongodbClusterClusterConfigMongodSecurityToTerraform;
class MdbMongodbClusterClusterConfigMongodSecurityOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // kmip - computed: false, optional: true, required: false
        this._kmip = new MdbMongodbClusterClusterConfigMongodSecurityKmipOutputReference(this, "kmip");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enableEncryption !== undefined) {
            hasAnyValues = true;
            internalValueResult.enableEncryption = this._enableEncryption;
        }
        if (this._kmip?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.kmip = this._kmip?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enableEncryption = undefined;
            this._kmip.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enableEncryption = value.enableEncryption;
            this._kmip.internalValue = value.kmip;
        }
    }
    get enableEncryption() {
        return this.getBooleanAttribute('enable_encryption');
    }
    set enableEncryption(value) {
        this._enableEncryption = value;
    }
    resetEnableEncryption() {
        this._enableEncryption = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enableEncryptionInput() {
        return this._enableEncryption;
    }
    get kmip() {
        return this._kmip;
    }
    putKmip(value) {
        this._kmip.internalValue = value;
    }
    resetKmip() {
        this._kmip.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get kmipInput() {
        return this._kmip.internalValue;
    }
}
exports.MdbMongodbClusterClusterConfigMongodSecurityOutputReference = MdbMongodbClusterClusterConfigMongodSecurityOutputReference;
_e = JSII_RTTI_SYMBOL_1;
MdbMongodbClusterClusterConfigMongodSecurityOutputReference[_e] = { fqn: "@cdktf/provider-yandex.MdbMongodbClusterClusterConfigMongodSecurityOutputReference", version: "2.0.32" };
function mdbMongodbClusterClusterConfigMongodSetParameterToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        audit_authorization_success: cdktf.booleanToTerraform(struct.auditAuthorizationSuccess),
    };
}
exports.mdbMongodbClusterClusterConfigMongodSetParameterToTerraform = mdbMongodbClusterClusterConfigMongodSetParameterToTerraform;
class MdbMongodbClusterClusterConfigMongodSetParameterOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._auditAuthorizationSuccess !== undefined) {
            hasAnyValues = true;
            internalValueResult.auditAuthorizationSuccess = this._auditAuthorizationSuccess;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._auditAuthorizationSuccess = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._auditAuthorizationSuccess = value.auditAuthorizationSuccess;
        }
    }
    get auditAuthorizationSuccess() {
        return this.getBooleanAttribute('audit_authorization_success');
    }
    set auditAuthorizationSuccess(value) {
        this._auditAuthorizationSuccess = value;
    }
    resetAuditAuthorizationSuccess() {
        this._auditAuthorizationSuccess = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get auditAuthorizationSuccessInput() {
        return this._auditAuthorizationSuccess;
    }
}
exports.MdbMongodbClusterClusterConfigMongodSetParameterOutputReference = MdbMongodbClusterClusterConfigMongodSetParameterOutputReference;
_f = JSII_RTTI_SYMBOL_1;
MdbMongodbClusterClusterConfigMongodSetParameterOutputReference[_f] = { fqn: "@cdktf/provider-yandex.MdbMongodbClusterClusterConfigMongodSetParameterOutputReference", version: "2.0.32" };
function mdbMongodbClusterClusterConfigMongodToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        audit_log: mdbMongodbClusterClusterConfigMongodAuditLogToTerraform(struct.auditLog),
        security: mdbMongodbClusterClusterConfigMongodSecurityToTerraform(struct.security),
        set_parameter: mdbMongodbClusterClusterConfigMongodSetParameterToTerraform(struct.setParameter),
    };
}
exports.mdbMongodbClusterClusterConfigMongodToTerraform = mdbMongodbClusterClusterConfigMongodToTerraform;
class MdbMongodbClusterClusterConfigMongodOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // audit_log - computed: false, optional: true, required: false
        this._auditLog = new MdbMongodbClusterClusterConfigMongodAuditLogOutputReference(this, "audit_log");
        // security - computed: false, optional: true, required: false
        this._security = new MdbMongodbClusterClusterConfigMongodSecurityOutputReference(this, "security");
        // set_parameter - computed: false, optional: true, required: false
        this._setParameter = new MdbMongodbClusterClusterConfigMongodSetParameterOutputReference(this, "set_parameter");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._auditLog?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.auditLog = this._auditLog?.internalValue;
        }
        if (this._security?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.security = this._security?.internalValue;
        }
        if (this._setParameter?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.setParameter = this._setParameter?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._auditLog.internalValue = undefined;
            this._security.internalValue = undefined;
            this._setParameter.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._auditLog.internalValue = value.auditLog;
            this._security.internalValue = value.security;
            this._setParameter.internalValue = value.setParameter;
        }
    }
    get auditLog() {
        return this._auditLog;
    }
    putAuditLog(value) {
        this._auditLog.internalValue = value;
    }
    resetAuditLog() {
        this._auditLog.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get auditLogInput() {
        return this._auditLog.internalValue;
    }
    get security() {
        return this._security;
    }
    putSecurity(value) {
        this._security.internalValue = value;
    }
    resetSecurity() {
        this._security.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityInput() {
        return this._security.internalValue;
    }
    get setParameter() {
        return this._setParameter;
    }
    putSetParameter(value) {
        this._setParameter.internalValue = value;
    }
    resetSetParameter() {
        this._setParameter.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get setParameterInput() {
        return this._setParameter.internalValue;
    }
}
exports.MdbMongodbClusterClusterConfigMongodOutputReference = MdbMongodbClusterClusterConfigMongodOutputReference;
_g = JSII_RTTI_SYMBOL_1;
MdbMongodbClusterClusterConfigMongodOutputReference[_g] = { fqn: "@cdktf/provider-yandex.MdbMongodbClusterClusterConfigMongodOutputReference", version: "2.0.32" };
function mdbMongodbClusterClusterConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        feature_compatibility_version: cdktf.stringToTerraform(struct.featureCompatibilityVersion),
        version: cdktf.stringToTerraform(struct.version),
        access: mdbMongodbClusterClusterConfigAccessToTerraform(struct.access),
        backup_window_start: mdbMongodbClusterClusterConfigBackupWindowStartToTerraform(struct.backupWindowStart),
        mongod: mdbMongodbClusterClusterConfigMongodToTerraform(struct.mongod),
    };
}
exports.mdbMongodbClusterClusterConfigToTerraform = mdbMongodbClusterClusterConfigToTerraform;
class MdbMongodbClusterClusterConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // access - computed: false, optional: true, required: false
        this._access = new MdbMongodbClusterClusterConfigAccessOutputReference(this, "access");
        // backup_window_start - computed: false, optional: true, required: false
        this._backupWindowStart = new MdbMongodbClusterClusterConfigBackupWindowStartOutputReference(this, "backup_window_start");
        // mongod - computed: false, optional: true, required: false
        this._mongod = new MdbMongodbClusterClusterConfigMongodOutputReference(this, "mongod");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._featureCompatibilityVersion !== undefined) {
            hasAnyValues = true;
            internalValueResult.featureCompatibilityVersion = this._featureCompatibilityVersion;
        }
        if (this._version !== undefined) {
            hasAnyValues = true;
            internalValueResult.version = this._version;
        }
        if (this._access?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.access = this._access?.internalValue;
        }
        if (this._backupWindowStart?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.backupWindowStart = this._backupWindowStart?.internalValue;
        }
        if (this._mongod?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.mongod = this._mongod?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._featureCompatibilityVersion = undefined;
            this._version = undefined;
            this._access.internalValue = undefined;
            this._backupWindowStart.internalValue = undefined;
            this._mongod.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._featureCompatibilityVersion = value.featureCompatibilityVersion;
            this._version = value.version;
            this._access.internalValue = value.access;
            this._backupWindowStart.internalValue = value.backupWindowStart;
            this._mongod.internalValue = value.mongod;
        }
    }
    get featureCompatibilityVersion() {
        return this.getStringAttribute('feature_compatibility_version');
    }
    set featureCompatibilityVersion(value) {
        this._featureCompatibilityVersion = value;
    }
    resetFeatureCompatibilityVersion() {
        this._featureCompatibilityVersion = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get featureCompatibilityVersionInput() {
        return this._featureCompatibilityVersion;
    }
    get version() {
        return this.getStringAttribute('version');
    }
    set version(value) {
        this._version = value;
    }
    // Temporarily expose input value. Use with caution.
    get versionInput() {
        return this._version;
    }
    get access() {
        return this._access;
    }
    putAccess(value) {
        this._access.internalValue = value;
    }
    resetAccess() {
        this._access.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get accessInput() {
        return this._access.internalValue;
    }
    get backupWindowStart() {
        return this._backupWindowStart;
    }
    putBackupWindowStart(value) {
        this._backupWindowStart.internalValue = value;
    }
    resetBackupWindowStart() {
        this._backupWindowStart.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get backupWindowStartInput() {
        return this._backupWindowStart.internalValue;
    }
    get mongod() {
        return this._mongod;
    }
    putMongod(value) {
        this._mongod.internalValue = value;
    }
    resetMongod() {
        this._mongod.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get mongodInput() {
        return this._mongod.internalValue;
    }
}
exports.MdbMongodbClusterClusterConfigOutputReference = MdbMongodbClusterClusterConfigOutputReference;
_h = JSII_RTTI_SYMBOL_1;
MdbMongodbClusterClusterConfigOutputReference[_h] = { fqn: "@cdktf/provider-yandex.MdbMongodbClusterClusterConfigOutputReference", version: "2.0.32" };
function mdbMongodbClusterDatabaseToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
    };
}
exports.mdbMongodbClusterDatabaseToTerraform = mdbMongodbClusterDatabaseToTerraform;
class MdbMongodbClusterDatabaseOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
}
exports.MdbMongodbClusterDatabaseOutputReference = MdbMongodbClusterDatabaseOutputReference;
_j = JSII_RTTI_SYMBOL_1;
MdbMongodbClusterDatabaseOutputReference[_j] = { fqn: "@cdktf/provider-yandex.MdbMongodbClusterDatabaseOutputReference", version: "2.0.32" };
class MdbMongodbClusterDatabaseList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new MdbMongodbClusterDatabaseOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.MdbMongodbClusterDatabaseList = MdbMongodbClusterDatabaseList;
_k = JSII_RTTI_SYMBOL_1;
MdbMongodbClusterDatabaseList[_k] = { fqn: "@cdktf/provider-yandex.MdbMongodbClusterDatabaseList", version: "2.0.32" };
function mdbMongodbClusterHostToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        assign_public_ip: cdktf.booleanToTerraform(struct.assignPublicIp),
        role: cdktf.stringToTerraform(struct.role),
        shard_name: cdktf.stringToTerraform(struct.shardName),
        subnet_id: cdktf.stringToTerraform(struct.subnetId),
        type: cdktf.stringToTerraform(struct.type),
        zone_id: cdktf.stringToTerraform(struct.zoneId),
    };
}
exports.mdbMongodbClusterHostToTerraform = mdbMongodbClusterHostToTerraform;
class MdbMongodbClusterHostOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._assignPublicIp !== undefined) {
            hasAnyValues = true;
            internalValueResult.assignPublicIp = this._assignPublicIp;
        }
        if (this._role !== undefined) {
            hasAnyValues = true;
            internalValueResult.role = this._role;
        }
        if (this._shardName !== undefined) {
            hasAnyValues = true;
            internalValueResult.shardName = this._shardName;
        }
        if (this._subnetId !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnetId = this._subnetId;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        if (this._zoneId !== undefined) {
            hasAnyValues = true;
            internalValueResult.zoneId = this._zoneId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._assignPublicIp = undefined;
            this._role = undefined;
            this._shardName = undefined;
            this._subnetId = undefined;
            this._type = undefined;
            this._zoneId = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._assignPublicIp = value.assignPublicIp;
            this._role = value.role;
            this._shardName = value.shardName;
            this._subnetId = value.subnetId;
            this._type = value.type;
            this._zoneId = value.zoneId;
        }
    }
    get assignPublicIp() {
        return this.getBooleanAttribute('assign_public_ip');
    }
    set assignPublicIp(value) {
        this._assignPublicIp = value;
    }
    resetAssignPublicIp() {
        this._assignPublicIp = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get assignPublicIpInput() {
        return this._assignPublicIp;
    }
    // health - computed: true, optional: false, required: false
    get health() {
        return this.getStringAttribute('health');
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get role() {
        return this.getStringAttribute('role');
    }
    set role(value) {
        this._role = value;
    }
    resetRole() {
        this._role = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get roleInput() {
        return this._role;
    }
    get shardName() {
        return this.getStringAttribute('shard_name');
    }
    set shardName(value) {
        this._shardName = value;
    }
    resetShardName() {
        this._shardName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get shardNameInput() {
        return this._shardName;
    }
    get subnetId() {
        return this.getStringAttribute('subnet_id');
    }
    set subnetId(value) {
        this._subnetId = value;
    }
    // Temporarily expose input value. Use with caution.
    get subnetIdInput() {
        return this._subnetId;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    resetType() {
        this._type = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
    get zoneId() {
        return this.getStringAttribute('zone_id');
    }
    set zoneId(value) {
        this._zoneId = value;
    }
    // Temporarily expose input value. Use with caution.
    get zoneIdInput() {
        return this._zoneId;
    }
}
exports.MdbMongodbClusterHostOutputReference = MdbMongodbClusterHostOutputReference;
_l = JSII_RTTI_SYMBOL_1;
MdbMongodbClusterHostOutputReference[_l] = { fqn: "@cdktf/provider-yandex.MdbMongodbClusterHostOutputReference", version: "2.0.32" };
class MdbMongodbClusterHostList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new MdbMongodbClusterHostOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.MdbMongodbClusterHostList = MdbMongodbClusterHostList;
_m = JSII_RTTI_SYMBOL_1;
MdbMongodbClusterHostList[_m] = { fqn: "@cdktf/provider-yandex.MdbMongodbClusterHostList", version: "2.0.32" };
function mdbMongodbClusterMaintenanceWindowToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        day: cdktf.stringToTerraform(struct.day),
        hour: cdktf.numberToTerraform(struct.hour),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.mdbMongodbClusterMaintenanceWindowToTerraform = mdbMongodbClusterMaintenanceWindowToTerraform;
class MdbMongodbClusterMaintenanceWindowOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._day !== undefined) {
            hasAnyValues = true;
            internalValueResult.day = this._day;
        }
        if (this._hour !== undefined) {
            hasAnyValues = true;
            internalValueResult.hour = this._hour;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._day = undefined;
            this._hour = undefined;
            this._type = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._day = value.day;
            this._hour = value.hour;
            this._type = value.type;
        }
    }
    get day() {
        return this.getStringAttribute('day');
    }
    set day(value) {
        this._day = value;
    }
    resetDay() {
        this._day = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dayInput() {
        return this._day;
    }
    get hour() {
        return this.getNumberAttribute('hour');
    }
    set hour(value) {
        this._hour = value;
    }
    resetHour() {
        this._hour = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get hourInput() {
        return this._hour;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.MdbMongodbClusterMaintenanceWindowOutputReference = MdbMongodbClusterMaintenanceWindowOutputReference;
_o = JSII_RTTI_SYMBOL_1;
MdbMongodbClusterMaintenanceWindowOutputReference[_o] = { fqn: "@cdktf/provider-yandex.MdbMongodbClusterMaintenanceWindowOutputReference", version: "2.0.32" };
function mdbMongodbClusterResourcesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        disk_size: cdktf.numberToTerraform(struct.diskSize),
        disk_type_id: cdktf.stringToTerraform(struct.diskTypeId),
        resource_preset_id: cdktf.stringToTerraform(struct.resourcePresetId),
    };
}
exports.mdbMongodbClusterResourcesToTerraform = mdbMongodbClusterResourcesToTerraform;
class MdbMongodbClusterResourcesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._diskSize !== undefined) {
            hasAnyValues = true;
            internalValueResult.diskSize = this._diskSize;
        }
        if (this._diskTypeId !== undefined) {
            hasAnyValues = true;
            internalValueResult.diskTypeId = this._diskTypeId;
        }
        if (this._resourcePresetId !== undefined) {
            hasAnyValues = true;
            internalValueResult.resourcePresetId = this._resourcePresetId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._diskSize = undefined;
            this._diskTypeId = undefined;
            this._resourcePresetId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._diskSize = value.diskSize;
            this._diskTypeId = value.diskTypeId;
            this._resourcePresetId = value.resourcePresetId;
        }
    }
    get diskSize() {
        return this.getNumberAttribute('disk_size');
    }
    set diskSize(value) {
        this._diskSize = value;
    }
    // Temporarily expose input value. Use with caution.
    get diskSizeInput() {
        return this._diskSize;
    }
    get diskTypeId() {
        return this.getStringAttribute('disk_type_id');
    }
    set diskTypeId(value) {
        this._diskTypeId = value;
    }
    // Temporarily expose input value. Use with caution.
    get diskTypeIdInput() {
        return this._diskTypeId;
    }
    get resourcePresetId() {
        return this.getStringAttribute('resource_preset_id');
    }
    set resourcePresetId(value) {
        this._resourcePresetId = value;
    }
    // Temporarily expose input value. Use with caution.
    get resourcePresetIdInput() {
        return this._resourcePresetId;
    }
}
exports.MdbMongodbClusterResourcesOutputReference = MdbMongodbClusterResourcesOutputReference;
_p = JSII_RTTI_SYMBOL_1;
MdbMongodbClusterResourcesOutputReference[_p] = { fqn: "@cdktf/provider-yandex.MdbMongodbClusterResourcesOutputReference", version: "2.0.32" };
function mdbMongodbClusterTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        default: cdktf.stringToTerraform(struct.default),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.mdbMongodbClusterTimeoutsToTerraform = mdbMongodbClusterTimeoutsToTerraform;
class MdbMongodbClusterTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._default !== undefined) {
            hasAnyValues = true;
            internalValueResult.default = this._default;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._default = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._default = value.default;
            this._update = value.update;
        }
    }
    get default() {
        return this.getStringAttribute('default');
    }
    set default(value) {
        this._default = value;
    }
    resetDefault() {
        this._default = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get defaultInput() {
        return this._default;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.MdbMongodbClusterTimeoutsOutputReference = MdbMongodbClusterTimeoutsOutputReference;
_q = JSII_RTTI_SYMBOL_1;
MdbMongodbClusterTimeoutsOutputReference[_q] = { fqn: "@cdktf/provider-yandex.MdbMongodbClusterTimeoutsOutputReference", version: "2.0.32" };
function mdbMongodbClusterUserPermissionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        database_name: cdktf.stringToTerraform(struct.databaseName),
        roles: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.roles),
    };
}
exports.mdbMongodbClusterUserPermissionToTerraform = mdbMongodbClusterUserPermissionToTerraform;
class MdbMongodbClusterUserPermissionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._databaseName !== undefined) {
            hasAnyValues = true;
            internalValueResult.databaseName = this._databaseName;
        }
        if (this._roles !== undefined) {
            hasAnyValues = true;
            internalValueResult.roles = this._roles;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._databaseName = undefined;
            this._roles = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._databaseName = value.databaseName;
            this._roles = value.roles;
        }
    }
    get databaseName() {
        return this.getStringAttribute('database_name');
    }
    set databaseName(value) {
        this._databaseName = value;
    }
    // Temporarily expose input value. Use with caution.
    get databaseNameInput() {
        return this._databaseName;
    }
    get roles() {
        return this.getListAttribute('roles');
    }
    set roles(value) {
        this._roles = value;
    }
    resetRoles() {
        this._roles = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get rolesInput() {
        return this._roles;
    }
}
exports.MdbMongodbClusterUserPermissionOutputReference = MdbMongodbClusterUserPermissionOutputReference;
_r = JSII_RTTI_SYMBOL_1;
MdbMongodbClusterUserPermissionOutputReference[_r] = { fqn: "@cdktf/provider-yandex.MdbMongodbClusterUserPermissionOutputReference", version: "2.0.32" };
class MdbMongodbClusterUserPermissionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new MdbMongodbClusterUserPermissionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.MdbMongodbClusterUserPermissionList = MdbMongodbClusterUserPermissionList;
_s = JSII_RTTI_SYMBOL_1;
MdbMongodbClusterUserPermissionList[_s] = { fqn: "@cdktf/provider-yandex.MdbMongodbClusterUserPermissionList", version: "2.0.32" };
function mdbMongodbClusterUserToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        password: cdktf.stringToTerraform(struct.password),
        permission: cdktf.listMapper(mdbMongodbClusterUserPermissionToTerraform, true)(struct.permission),
    };
}
exports.mdbMongodbClusterUserToTerraform = mdbMongodbClusterUserToTerraform;
class MdbMongodbClusterUserOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // permission - computed: false, optional: true, required: false
        this._permission = new MdbMongodbClusterUserPermissionList(this, "permission", true);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._password !== undefined) {
            hasAnyValues = true;
            internalValueResult.password = this._password;
        }
        if (this._permission?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.permission = this._permission?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._password = undefined;
            this._permission.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._password = value.password;
            this._permission.internalValue = value.permission;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get password() {
        return this.getStringAttribute('password');
    }
    set password(value) {
        this._password = value;
    }
    // Temporarily expose input value. Use with caution.
    get passwordInput() {
        return this._password;
    }
    get permission() {
        return this._permission;
    }
    putPermission(value) {
        this._permission.internalValue = value;
    }
    resetPermission() {
        this._permission.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get permissionInput() {
        return this._permission.internalValue;
    }
}
exports.MdbMongodbClusterUserOutputReference = MdbMongodbClusterUserOutputReference;
_t = JSII_RTTI_SYMBOL_1;
MdbMongodbClusterUserOutputReference[_t] = { fqn: "@cdktf/provider-yandex.MdbMongodbClusterUserOutputReference", version: "2.0.32" };
class MdbMongodbClusterUserList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new MdbMongodbClusterUserOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.MdbMongodbClusterUserList = MdbMongodbClusterUserList;
_u = JSII_RTTI_SYMBOL_1;
MdbMongodbClusterUserList[_u] = { fqn: "@cdktf/provider-yandex.MdbMongodbClusterUserList", version: "2.0.32" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/r/mdb_mongodb_cluster yandex_mdb_mongodb_cluster}
*/
class MdbMongodbCluster extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/r/mdb_mongodb_cluster yandex_mdb_mongodb_cluster} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options MdbMongodbClusterConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'yandex_mdb_mongodb_cluster',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // cluster_config - computed: false, optional: false, required: true
        this._clusterConfig = new MdbMongodbClusterClusterConfigOutputReference(this, "cluster_config");
        // database - computed: false, optional: false, required: true
        this._database = new MdbMongodbClusterDatabaseList(this, "database", true);
        // host - computed: false, optional: false, required: true
        this._host = new MdbMongodbClusterHostList(this, "host", false);
        // maintenance_window - computed: false, optional: true, required: false
        this._maintenanceWindow = new MdbMongodbClusterMaintenanceWindowOutputReference(this, "maintenance_window");
        // resources - computed: false, optional: false, required: true
        this._resources = new MdbMongodbClusterResourcesOutputReference(this, "resources");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new MdbMongodbClusterTimeoutsOutputReference(this, "timeouts");
        // user - computed: false, optional: false, required: true
        this._user = new MdbMongodbClusterUserList(this, "user", true);
        this._clusterId = config.clusterId;
        this._deletionProtection = config.deletionProtection;
        this._description = config.description;
        this._environment = config.environment;
        this._folderId = config.folderId;
        this._id = config.id;
        this._labels = config.labels;
        this._name = config.name;
        this._networkId = config.networkId;
        this._securityGroupIds = config.securityGroupIds;
        this._clusterConfig.internalValue = config.clusterConfig;
        this._database.internalValue = config.database;
        this._host.internalValue = config.host;
        this._maintenanceWindow.internalValue = config.maintenanceWindow;
        this._resources.internalValue = config.resources;
        this._timeouts.internalValue = config.timeouts;
        this._user.internalValue = config.user;
    }
    get clusterId() {
        return this.getStringAttribute('cluster_id');
    }
    set clusterId(value) {
        this._clusterId = value;
    }
    resetClusterId() {
        this._clusterId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get clusterIdInput() {
        return this._clusterId;
    }
    // created_at - computed: true, optional: false, required: false
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    get deletionProtection() {
        return this.getBooleanAttribute('deletion_protection');
    }
    set deletionProtection(value) {
        this._deletionProtection = value;
    }
    resetDeletionProtection() {
        this._deletionProtection = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deletionProtectionInput() {
        return this._deletionProtection;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get environment() {
        return this.getStringAttribute('environment');
    }
    set environment(value) {
        this._environment = value;
    }
    // Temporarily expose input value. Use with caution.
    get environmentInput() {
        return this._environment;
    }
    get folderId() {
        return this.getStringAttribute('folder_id');
    }
    set folderId(value) {
        this._folderId = value;
    }
    resetFolderId() {
        this._folderId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get folderIdInput() {
        return this._folderId;
    }
    // health - computed: true, optional: false, required: false
    get health() {
        return this.getStringAttribute('health');
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get networkId() {
        return this.getStringAttribute('network_id');
    }
    set networkId(value) {
        this._networkId = value;
    }
    // Temporarily expose input value. Use with caution.
    get networkIdInput() {
        return this._networkId;
    }
    get securityGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('security_group_ids'));
    }
    set securityGroupIds(value) {
        this._securityGroupIds = value;
    }
    resetSecurityGroupIds() {
        this._securityGroupIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupIdsInput() {
        return this._securityGroupIds;
    }
    // sharded - computed: true, optional: false, required: false
    get sharded() {
        return this.getBooleanAttribute('sharded');
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    get clusterConfig() {
        return this._clusterConfig;
    }
    putClusterConfig(value) {
        this._clusterConfig.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get clusterConfigInput() {
        return this._clusterConfig.internalValue;
    }
    get database() {
        return this._database;
    }
    putDatabase(value) {
        this._database.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get databaseInput() {
        return this._database.internalValue;
    }
    get host() {
        return this._host;
    }
    putHost(value) {
        this._host.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get hostInput() {
        return this._host.internalValue;
    }
    get maintenanceWindow() {
        return this._maintenanceWindow;
    }
    putMaintenanceWindow(value) {
        this._maintenanceWindow.internalValue = value;
    }
    resetMaintenanceWindow() {
        this._maintenanceWindow.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maintenanceWindowInput() {
        return this._maintenanceWindow.internalValue;
    }
    get resources() {
        return this._resources;
    }
    putResources(value) {
        this._resources.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get resourcesInput() {
        return this._resources.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    get user() {
        return this._user;
    }
    putUser(value) {
        this._user.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get userInput() {
        return this._user.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            cluster_id: cdktf.stringToTerraform(this._clusterId),
            deletion_protection: cdktf.booleanToTerraform(this._deletionProtection),
            description: cdktf.stringToTerraform(this._description),
            environment: cdktf.stringToTerraform(this._environment),
            folder_id: cdktf.stringToTerraform(this._folderId),
            id: cdktf.stringToTerraform(this._id),
            labels: cdktf.hashMapper(cdktf.stringToTerraform)(this._labels),
            name: cdktf.stringToTerraform(this._name),
            network_id: cdktf.stringToTerraform(this._networkId),
            security_group_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(this._securityGroupIds),
            cluster_config: mdbMongodbClusterClusterConfigToTerraform(this._clusterConfig.internalValue),
            database: cdktf.listMapper(mdbMongodbClusterDatabaseToTerraform, true)(this._database.internalValue),
            host: cdktf.listMapper(mdbMongodbClusterHostToTerraform, true)(this._host.internalValue),
            maintenance_window: mdbMongodbClusterMaintenanceWindowToTerraform(this._maintenanceWindow.internalValue),
            resources: mdbMongodbClusterResourcesToTerraform(this._resources.internalValue),
            timeouts: mdbMongodbClusterTimeoutsToTerraform(this._timeouts.internalValue),
            user: cdktf.listMapper(mdbMongodbClusterUserToTerraform, true)(this._user.internalValue),
        };
    }
}
exports.MdbMongodbCluster = MdbMongodbCluster;
_v = JSII_RTTI_SYMBOL_1;
MdbMongodbCluster[_v] = { fqn: "@cdktf/provider-yandex.MdbMongodbCluster", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
MdbMongodbCluster.tfResourceType = "yandex_mdb_mongodb_cluster";
