// @cdktf/provider-yandex/lib/mdb-postgresql-cluster.js
"use strict";
var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MdbPostgresqlCluster = exports.MdbPostgresqlClusterUserList = exports.MdbPostgresqlClusterUserOutputReference = exports.mdbPostgresqlClusterUserToTerraform = exports.MdbPostgresqlClusterUserPermissionList = exports.MdbPostgresqlClusterUserPermissionOutputReference = exports.mdbPostgresqlClusterUserPermissionToTerraform = exports.MdbPostgresqlClusterTimeoutsOutputReference = exports.mdbPostgresqlClusterTimeoutsToTerraform = exports.MdbPostgresqlClusterRestoreOutputReference = exports.mdbPostgresqlClusterRestoreToTerraform = exports.MdbPostgresqlClusterMaintenanceWindowOutputReference = exports.mdbPostgresqlClusterMaintenanceWindowToTerraform = exports.MdbPostgresqlClusterHostList = exports.MdbPostgresqlClusterHostOutputReference = exports.mdbPostgresqlClusterHostToTerraform = exports.MdbPostgresqlClusterDatabaseList = exports.MdbPostgresqlClusterDatabaseOutputReference = exports.mdbPostgresqlClusterDatabaseToTerraform = exports.MdbPostgresqlClusterDatabaseExtensionList = exports.MdbPostgresqlClusterDatabaseExtensionOutputReference = exports.mdbPostgresqlClusterDatabaseExtensionToTerraform = exports.MdbPostgresqlClusterConfigAOutputReference = exports.mdbPostgresqlClusterConfigAToTerraform = exports.MdbPostgresqlClusterConfigResourcesOutputReference = exports.mdbPostgresqlClusterConfigResourcesToTerraform = exports.MdbPostgresqlClusterConfigPoolerConfigOutputReference = exports.mdbPostgresqlClusterConfigPoolerConfigToTerraform = exports.MdbPostgresqlClusterConfigPerformanceDiagnosticsOutputReference = exports.mdbPostgresqlClusterConfigPerformanceDiagnosticsToTerraform = exports.MdbPostgresqlClusterConfigBackupWindowStartOutputReference = exports.mdbPostgresqlClusterConfigBackupWindowStartToTerraform = exports.MdbPostgresqlClusterConfigAccessOutputReference = exports.mdbPostgresqlClusterConfigAccessToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function mdbPostgresqlClusterConfigAccessToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        data_lens: cdktf.booleanToTerraform(struct.dataLens),
        serverless: cdktf.booleanToTerraform(struct.serverless),
        web_sql: cdktf.booleanToTerraform(struct.webSql),
    };
}
exports.mdbPostgresqlClusterConfigAccessToTerraform = mdbPostgresqlClusterConfigAccessToTerraform;
class MdbPostgresqlClusterConfigAccessOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._dataLens !== undefined) {
            hasAnyValues = true;
            internalValueResult.dataLens = this._dataLens;
        }
        if (this._serverless !== undefined) {
            hasAnyValues = true;
            internalValueResult.serverless = this._serverless;
        }
        if (this._webSql !== undefined) {
            hasAnyValues = true;
            internalValueResult.webSql = this._webSql;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._dataLens = undefined;
            this._serverless = undefined;
            this._webSql = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._dataLens = value.dataLens;
            this._serverless = value.serverless;
            this._webSql = value.webSql;
        }
    }
    get dataLens() {
        return this.getBooleanAttribute('data_lens');
    }
    set dataLens(value) {
        this._dataLens = value;
    }
    resetDataLens() {
        this._dataLens = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dataLensInput() {
        return this._dataLens;
    }
    get serverless() {
        return this.getBooleanAttribute('serverless');
    }
    set serverless(value) {
        this._serverless = value;
    }
    resetServerless() {
        this._serverless = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get serverlessInput() {
        return this._serverless;
    }
    get webSql() {
        return this.getBooleanAttribute('web_sql');
    }
    set webSql(value) {
        this._webSql = value;
    }
    resetWebSql() {
        this._webSql = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get webSqlInput() {
        return this._webSql;
    }
}
exports.MdbPostgresqlClusterConfigAccessOutputReference = MdbPostgresqlClusterConfigAccessOutputReference;
_a = JSII_RTTI_SYMBOL_1;
MdbPostgresqlClusterConfigAccessOutputReference[_a] = { fqn: "@cdktf/provider-yandex.MdbPostgresqlClusterConfigAccessOutputReference", version: "2.0.32" };
function mdbPostgresqlClusterConfigBackupWindowStartToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        hours: cdktf.numberToTerraform(struct.hours),
        minutes: cdktf.numberToTerraform(struct.minutes),
    };
}
exports.mdbPostgresqlClusterConfigBackupWindowStartToTerraform = mdbPostgresqlClusterConfigBackupWindowStartToTerraform;
class MdbPostgresqlClusterConfigBackupWindowStartOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._hours !== undefined) {
            hasAnyValues = true;
            internalValueResult.hours = this._hours;
        }
        if (this._minutes !== undefined) {
            hasAnyValues = true;
            internalValueResult.minutes = this._minutes;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._hours = undefined;
            this._minutes = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._hours = value.hours;
            this._minutes = value.minutes;
        }
    }
    get hours() {
        return this.getNumberAttribute('hours');
    }
    set hours(value) {
        this._hours = value;
    }
    resetHours() {
        this._hours = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get hoursInput() {
        return this._hours;
    }
    get minutes() {
        return this.getNumberAttribute('minutes');
    }
    set minutes(value) {
        this._minutes = value;
    }
    resetMinutes() {
        this._minutes = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get minutesInput() {
        return this._minutes;
    }
}
exports.MdbPostgresqlClusterConfigBackupWindowStartOutputReference = MdbPostgresqlClusterConfigBackupWindowStartOutputReference;
_b = JSII_RTTI_SYMBOL_1;
MdbPostgresqlClusterConfigBackupWindowStartOutputReference[_b] = { fqn: "@cdktf/provider-yandex.MdbPostgresqlClusterConfigBackupWindowStartOutputReference", version: "2.0.32" };
function mdbPostgresqlClusterConfigPerformanceDiagnosticsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        enabled: cdktf.booleanToTerraform(struct.enabled),
        sessions_sampling_interval: cdktf.numberToTerraform(struct.sessionsSamplingInterval),
        statements_sampling_interval: cdktf.numberToTerraform(struct.statementsSamplingInterval),
    };
}
exports.mdbPostgresqlClusterConfigPerformanceDiagnosticsToTerraform = mdbPostgresqlClusterConfigPerformanceDiagnosticsToTerraform;
class MdbPostgresqlClusterConfigPerformanceDiagnosticsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._enabled !== undefined) {
            hasAnyValues = true;
            internalValueResult.enabled = this._enabled;
        }
        if (this._sessionsSamplingInterval !== undefined) {
            hasAnyValues = true;
            internalValueResult.sessionsSamplingInterval = this._sessionsSamplingInterval;
        }
        if (this._statementsSamplingInterval !== undefined) {
            hasAnyValues = true;
            internalValueResult.statementsSamplingInterval = this._statementsSamplingInterval;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._enabled = undefined;
            this._sessionsSamplingInterval = undefined;
            this._statementsSamplingInterval = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._enabled = value.enabled;
            this._sessionsSamplingInterval = value.sessionsSamplingInterval;
            this._statementsSamplingInterval = value.statementsSamplingInterval;
        }
    }
    get enabled() {
        return this.getBooleanAttribute('enabled');
    }
    set enabled(value) {
        this._enabled = value;
    }
    resetEnabled() {
        this._enabled = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get enabledInput() {
        return this._enabled;
    }
    get sessionsSamplingInterval() {
        return this.getNumberAttribute('sessions_sampling_interval');
    }
    set sessionsSamplingInterval(value) {
        this._sessionsSamplingInterval = value;
    }
    // Temporarily expose input value. Use with caution.
    get sessionsSamplingIntervalInput() {
        return this._sessionsSamplingInterval;
    }
    get statementsSamplingInterval() {
        return this.getNumberAttribute('statements_sampling_interval');
    }
    set statementsSamplingInterval(value) {
        this._statementsSamplingInterval = value;
    }
    // Temporarily expose input value. Use with caution.
    get statementsSamplingIntervalInput() {
        return this._statementsSamplingInterval;
    }
}
exports.MdbPostgresqlClusterConfigPerformanceDiagnosticsOutputReference = MdbPostgresqlClusterConfigPerformanceDiagnosticsOutputReference;
_c = JSII_RTTI_SYMBOL_1;
MdbPostgresqlClusterConfigPerformanceDiagnosticsOutputReference[_c] = { fqn: "@cdktf/provider-yandex.MdbPostgresqlClusterConfigPerformanceDiagnosticsOutputReference", version: "2.0.32" };
function mdbPostgresqlClusterConfigPoolerConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        pool_discard: cdktf.booleanToTerraform(struct.poolDiscard),
        pooling_mode: cdktf.stringToTerraform(struct.poolingMode),
    };
}
exports.mdbPostgresqlClusterConfigPoolerConfigToTerraform = mdbPostgresqlClusterConfigPoolerConfigToTerraform;
class MdbPostgresqlClusterConfigPoolerConfigOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._poolDiscard !== undefined) {
            hasAnyValues = true;
            internalValueResult.poolDiscard = this._poolDiscard;
        }
        if (this._poolingMode !== undefined) {
            hasAnyValues = true;
            internalValueResult.poolingMode = this._poolingMode;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._poolDiscard = undefined;
            this._poolingMode = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._poolDiscard = value.poolDiscard;
            this._poolingMode = value.poolingMode;
        }
    }
    get poolDiscard() {
        return this.getBooleanAttribute('pool_discard');
    }
    set poolDiscard(value) {
        this._poolDiscard = value;
    }
    resetPoolDiscard() {
        this._poolDiscard = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get poolDiscardInput() {
        return this._poolDiscard;
    }
    get poolingMode() {
        return this.getStringAttribute('pooling_mode');
    }
    set poolingMode(value) {
        this._poolingMode = value;
    }
    resetPoolingMode() {
        this._poolingMode = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get poolingModeInput() {
        return this._poolingMode;
    }
}
exports.MdbPostgresqlClusterConfigPoolerConfigOutputReference = MdbPostgresqlClusterConfigPoolerConfigOutputReference;
_d = JSII_RTTI_SYMBOL_1;
MdbPostgresqlClusterConfigPoolerConfigOutputReference[_d] = { fqn: "@cdktf/provider-yandex.MdbPostgresqlClusterConfigPoolerConfigOutputReference", version: "2.0.32" };
function mdbPostgresqlClusterConfigResourcesToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        disk_size: cdktf.numberToTerraform(struct.diskSize),
        disk_type_id: cdktf.stringToTerraform(struct.diskTypeId),
        resource_preset_id: cdktf.stringToTerraform(struct.resourcePresetId),
    };
}
exports.mdbPostgresqlClusterConfigResourcesToTerraform = mdbPostgresqlClusterConfigResourcesToTerraform;
class MdbPostgresqlClusterConfigResourcesOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._diskSize !== undefined) {
            hasAnyValues = true;
            internalValueResult.diskSize = this._diskSize;
        }
        if (this._diskTypeId !== undefined) {
            hasAnyValues = true;
            internalValueResult.diskTypeId = this._diskTypeId;
        }
        if (this._resourcePresetId !== undefined) {
            hasAnyValues = true;
            internalValueResult.resourcePresetId = this._resourcePresetId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._diskSize = undefined;
            this._diskTypeId = undefined;
            this._resourcePresetId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._diskSize = value.diskSize;
            this._diskTypeId = value.diskTypeId;
            this._resourcePresetId = value.resourcePresetId;
        }
    }
    get diskSize() {
        return this.getNumberAttribute('disk_size');
    }
    set diskSize(value) {
        this._diskSize = value;
    }
    // Temporarily expose input value. Use with caution.
    get diskSizeInput() {
        return this._diskSize;
    }
    get diskTypeId() {
        return this.getStringAttribute('disk_type_id');
    }
    set diskTypeId(value) {
        this._diskTypeId = value;
    }
    resetDiskTypeId() {
        this._diskTypeId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get diskTypeIdInput() {
        return this._diskTypeId;
    }
    get resourcePresetId() {
        return this.getStringAttribute('resource_preset_id');
    }
    set resourcePresetId(value) {
        this._resourcePresetId = value;
    }
    // Temporarily expose input value. Use with caution.
    get resourcePresetIdInput() {
        return this._resourcePresetId;
    }
}
exports.MdbPostgresqlClusterConfigResourcesOutputReference = MdbPostgresqlClusterConfigResourcesOutputReference;
_e = JSII_RTTI_SYMBOL_1;
MdbPostgresqlClusterConfigResourcesOutputReference[_e] = { fqn: "@cdktf/provider-yandex.MdbPostgresqlClusterConfigResourcesOutputReference", version: "2.0.32" };
function mdbPostgresqlClusterConfigAToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        autofailover: cdktf.booleanToTerraform(struct.autofailover),
        backup_retain_period_days: cdktf.numberToTerraform(struct.backupRetainPeriodDays),
        postgresql_config: cdktf.hashMapper(cdktf.stringToTerraform)(struct.postgresqlConfig),
        version: cdktf.stringToTerraform(struct.version),
        access: mdbPostgresqlClusterConfigAccessToTerraform(struct.access),
        backup_window_start: mdbPostgresqlClusterConfigBackupWindowStartToTerraform(struct.backupWindowStart),
        performance_diagnostics: mdbPostgresqlClusterConfigPerformanceDiagnosticsToTerraform(struct.performanceDiagnostics),
        pooler_config: mdbPostgresqlClusterConfigPoolerConfigToTerraform(struct.poolerConfig),
        resources: mdbPostgresqlClusterConfigResourcesToTerraform(struct.resources),
    };
}
exports.mdbPostgresqlClusterConfigAToTerraform = mdbPostgresqlClusterConfigAToTerraform;
class MdbPostgresqlClusterConfigAOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
        // access - computed: false, optional: true, required: false
        this._access = new MdbPostgresqlClusterConfigAccessOutputReference(this, "access");
        // backup_window_start - computed: false, optional: true, required: false
        this._backupWindowStart = new MdbPostgresqlClusterConfigBackupWindowStartOutputReference(this, "backup_window_start");
        // performance_diagnostics - computed: false, optional: true, required: false
        this._performanceDiagnostics = new MdbPostgresqlClusterConfigPerformanceDiagnosticsOutputReference(this, "performance_diagnostics");
        // pooler_config - computed: false, optional: true, required: false
        this._poolerConfig = new MdbPostgresqlClusterConfigPoolerConfigOutputReference(this, "pooler_config");
        // resources - computed: false, optional: false, required: true
        this._resources = new MdbPostgresqlClusterConfigResourcesOutputReference(this, "resources");
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._autofailover !== undefined) {
            hasAnyValues = true;
            internalValueResult.autofailover = this._autofailover;
        }
        if (this._backupRetainPeriodDays !== undefined) {
            hasAnyValues = true;
            internalValueResult.backupRetainPeriodDays = this._backupRetainPeriodDays;
        }
        if (this._postgresqlConfig !== undefined) {
            hasAnyValues = true;
            internalValueResult.postgresqlConfig = this._postgresqlConfig;
        }
        if (this._version !== undefined) {
            hasAnyValues = true;
            internalValueResult.version = this._version;
        }
        if (this._access?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.access = this._access?.internalValue;
        }
        if (this._backupWindowStart?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.backupWindowStart = this._backupWindowStart?.internalValue;
        }
        if (this._performanceDiagnostics?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.performanceDiagnostics = this._performanceDiagnostics?.internalValue;
        }
        if (this._poolerConfig?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.poolerConfig = this._poolerConfig?.internalValue;
        }
        if (this._resources?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.resources = this._resources?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._autofailover = undefined;
            this._backupRetainPeriodDays = undefined;
            this._postgresqlConfig = undefined;
            this._version = undefined;
            this._access.internalValue = undefined;
            this._backupWindowStart.internalValue = undefined;
            this._performanceDiagnostics.internalValue = undefined;
            this._poolerConfig.internalValue = undefined;
            this._resources.internalValue = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._autofailover = value.autofailover;
            this._backupRetainPeriodDays = value.backupRetainPeriodDays;
            this._postgresqlConfig = value.postgresqlConfig;
            this._version = value.version;
            this._access.internalValue = value.access;
            this._backupWindowStart.internalValue = value.backupWindowStart;
            this._performanceDiagnostics.internalValue = value.performanceDiagnostics;
            this._poolerConfig.internalValue = value.poolerConfig;
            this._resources.internalValue = value.resources;
        }
    }
    get autofailover() {
        return this.getBooleanAttribute('autofailover');
    }
    set autofailover(value) {
        this._autofailover = value;
    }
    resetAutofailover() {
        this._autofailover = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get autofailoverInput() {
        return this._autofailover;
    }
    get backupRetainPeriodDays() {
        return this.getNumberAttribute('backup_retain_period_days');
    }
    set backupRetainPeriodDays(value) {
        this._backupRetainPeriodDays = value;
    }
    resetBackupRetainPeriodDays() {
        this._backupRetainPeriodDays = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get backupRetainPeriodDaysInput() {
        return this._backupRetainPeriodDays;
    }
    get postgresqlConfig() {
        return this.getStringMapAttribute('postgresql_config');
    }
    set postgresqlConfig(value) {
        this._postgresqlConfig = value;
    }
    resetPostgresqlConfig() {
        this._postgresqlConfig = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get postgresqlConfigInput() {
        return this._postgresqlConfig;
    }
    get version() {
        return this.getStringAttribute('version');
    }
    set version(value) {
        this._version = value;
    }
    // Temporarily expose input value. Use with caution.
    get versionInput() {
        return this._version;
    }
    get access() {
        return this._access;
    }
    putAccess(value) {
        this._access.internalValue = value;
    }
    resetAccess() {
        this._access.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get accessInput() {
        return this._access.internalValue;
    }
    get backupWindowStart() {
        return this._backupWindowStart;
    }
    putBackupWindowStart(value) {
        this._backupWindowStart.internalValue = value;
    }
    resetBackupWindowStart() {
        this._backupWindowStart.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get backupWindowStartInput() {
        return this._backupWindowStart.internalValue;
    }
    get performanceDiagnostics() {
        return this._performanceDiagnostics;
    }
    putPerformanceDiagnostics(value) {
        this._performanceDiagnostics.internalValue = value;
    }
    resetPerformanceDiagnostics() {
        this._performanceDiagnostics.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get performanceDiagnosticsInput() {
        return this._performanceDiagnostics.internalValue;
    }
    get poolerConfig() {
        return this._poolerConfig;
    }
    putPoolerConfig(value) {
        this._poolerConfig.internalValue = value;
    }
    resetPoolerConfig() {
        this._poolerConfig.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get poolerConfigInput() {
        return this._poolerConfig.internalValue;
    }
    get resources() {
        return this._resources;
    }
    putResources(value) {
        this._resources.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get resourcesInput() {
        return this._resources.internalValue;
    }
}
exports.MdbPostgresqlClusterConfigAOutputReference = MdbPostgresqlClusterConfigAOutputReference;
_f = JSII_RTTI_SYMBOL_1;
MdbPostgresqlClusterConfigAOutputReference[_f] = { fqn: "@cdktf/provider-yandex.MdbPostgresqlClusterConfigAOutputReference", version: "2.0.32" };
function mdbPostgresqlClusterDatabaseExtensionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        name: cdktf.stringToTerraform(struct.name),
        version: cdktf.stringToTerraform(struct.version),
    };
}
exports.mdbPostgresqlClusterDatabaseExtensionToTerraform = mdbPostgresqlClusterDatabaseExtensionToTerraform;
class MdbPostgresqlClusterDatabaseExtensionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._version !== undefined) {
            hasAnyValues = true;
            internalValueResult.version = this._version;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._name = undefined;
            this._version = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._name = value.name;
            this._version = value.version;
        }
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get version() {
        return this.getStringAttribute('version');
    }
    set version(value) {
        this._version = value;
    }
    resetVersion() {
        this._version = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get versionInput() {
        return this._version;
    }
}
exports.MdbPostgresqlClusterDatabaseExtensionOutputReference = MdbPostgresqlClusterDatabaseExtensionOutputReference;
_g = JSII_RTTI_SYMBOL_1;
MdbPostgresqlClusterDatabaseExtensionOutputReference[_g] = { fqn: "@cdktf/provider-yandex.MdbPostgresqlClusterDatabaseExtensionOutputReference", version: "2.0.32" };
class MdbPostgresqlClusterDatabaseExtensionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new MdbPostgresqlClusterDatabaseExtensionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.MdbPostgresqlClusterDatabaseExtensionList = MdbPostgresqlClusterDatabaseExtensionList;
_h = JSII_RTTI_SYMBOL_1;
MdbPostgresqlClusterDatabaseExtensionList[_h] = { fqn: "@cdktf/provider-yandex.MdbPostgresqlClusterDatabaseExtensionList", version: "2.0.32" };
function mdbPostgresqlClusterDatabaseToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        lc_collate: cdktf.stringToTerraform(struct.lcCollate),
        lc_type: cdktf.stringToTerraform(struct.lcType),
        name: cdktf.stringToTerraform(struct.name),
        owner: cdktf.stringToTerraform(struct.owner),
        extension: cdktf.listMapper(mdbPostgresqlClusterDatabaseExtensionToTerraform, true)(struct.extension),
    };
}
exports.mdbPostgresqlClusterDatabaseToTerraform = mdbPostgresqlClusterDatabaseToTerraform;
class MdbPostgresqlClusterDatabaseOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // extension - computed: false, optional: true, required: false
        this._extension = new MdbPostgresqlClusterDatabaseExtensionList(this, "extension", true);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._lcCollate !== undefined) {
            hasAnyValues = true;
            internalValueResult.lcCollate = this._lcCollate;
        }
        if (this._lcType !== undefined) {
            hasAnyValues = true;
            internalValueResult.lcType = this._lcType;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._owner !== undefined) {
            hasAnyValues = true;
            internalValueResult.owner = this._owner;
        }
        if (this._extension?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.extension = this._extension?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._lcCollate = undefined;
            this._lcType = undefined;
            this._name = undefined;
            this._owner = undefined;
            this._extension.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._lcCollate = value.lcCollate;
            this._lcType = value.lcType;
            this._name = value.name;
            this._owner = value.owner;
            this._extension.internalValue = value.extension;
        }
    }
    get lcCollate() {
        return this.getStringAttribute('lc_collate');
    }
    set lcCollate(value) {
        this._lcCollate = value;
    }
    resetLcCollate() {
        this._lcCollate = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get lcCollateInput() {
        return this._lcCollate;
    }
    get lcType() {
        return this.getStringAttribute('lc_type');
    }
    set lcType(value) {
        this._lcType = value;
    }
    resetLcType() {
        this._lcType = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get lcTypeInput() {
        return this._lcType;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get owner() {
        return this.getStringAttribute('owner');
    }
    set owner(value) {
        this._owner = value;
    }
    // Temporarily expose input value. Use with caution.
    get ownerInput() {
        return this._owner;
    }
    get extension() {
        return this._extension;
    }
    putExtension(value) {
        this._extension.internalValue = value;
    }
    resetExtension() {
        this._extension.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get extensionInput() {
        return this._extension.internalValue;
    }
}
exports.MdbPostgresqlClusterDatabaseOutputReference = MdbPostgresqlClusterDatabaseOutputReference;
_j = JSII_RTTI_SYMBOL_1;
MdbPostgresqlClusterDatabaseOutputReference[_j] = { fqn: "@cdktf/provider-yandex.MdbPostgresqlClusterDatabaseOutputReference", version: "2.0.32" };
class MdbPostgresqlClusterDatabaseList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new MdbPostgresqlClusterDatabaseOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.MdbPostgresqlClusterDatabaseList = MdbPostgresqlClusterDatabaseList;
_k = JSII_RTTI_SYMBOL_1;
MdbPostgresqlClusterDatabaseList[_k] = { fqn: "@cdktf/provider-yandex.MdbPostgresqlClusterDatabaseList", version: "2.0.32" };
function mdbPostgresqlClusterHostToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        assign_public_ip: cdktf.booleanToTerraform(struct.assignPublicIp),
        name: cdktf.stringToTerraform(struct.name),
        priority: cdktf.numberToTerraform(struct.priority),
        replication_source_name: cdktf.stringToTerraform(struct.replicationSourceName),
        subnet_id: cdktf.stringToTerraform(struct.subnetId),
        zone: cdktf.stringToTerraform(struct.zone),
    };
}
exports.mdbPostgresqlClusterHostToTerraform = mdbPostgresqlClusterHostToTerraform;
class MdbPostgresqlClusterHostOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._assignPublicIp !== undefined) {
            hasAnyValues = true;
            internalValueResult.assignPublicIp = this._assignPublicIp;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._priority !== undefined) {
            hasAnyValues = true;
            internalValueResult.priority = this._priority;
        }
        if (this._replicationSourceName !== undefined) {
            hasAnyValues = true;
            internalValueResult.replicationSourceName = this._replicationSourceName;
        }
        if (this._subnetId !== undefined) {
            hasAnyValues = true;
            internalValueResult.subnetId = this._subnetId;
        }
        if (this._zone !== undefined) {
            hasAnyValues = true;
            internalValueResult.zone = this._zone;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._assignPublicIp = undefined;
            this._name = undefined;
            this._priority = undefined;
            this._replicationSourceName = undefined;
            this._subnetId = undefined;
            this._zone = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._assignPublicIp = value.assignPublicIp;
            this._name = value.name;
            this._priority = value.priority;
            this._replicationSourceName = value.replicationSourceName;
            this._subnetId = value.subnetId;
            this._zone = value.zone;
        }
    }
    get assignPublicIp() {
        return this.getBooleanAttribute('assign_public_ip');
    }
    set assignPublicIp(value) {
        this._assignPublicIp = value;
    }
    resetAssignPublicIp() {
        this._assignPublicIp = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get assignPublicIpInput() {
        return this._assignPublicIp;
    }
    // fqdn - computed: true, optional: false, required: false
    get fqdn() {
        return this.getStringAttribute('fqdn');
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get priority() {
        return this.getNumberAttribute('priority');
    }
    set priority(value) {
        this._priority = value;
    }
    resetPriority() {
        this._priority = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get priorityInput() {
        return this._priority;
    }
    // replication_source - computed: true, optional: false, required: false
    get replicationSource() {
        return this.getStringAttribute('replication_source');
    }
    get replicationSourceName() {
        return this.getStringAttribute('replication_source_name');
    }
    set replicationSourceName(value) {
        this._replicationSourceName = value;
    }
    resetReplicationSourceName() {
        this._replicationSourceName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get replicationSourceNameInput() {
        return this._replicationSourceName;
    }
    // role - computed: true, optional: false, required: false
    get role() {
        return this.getStringAttribute('role');
    }
    get subnetId() {
        return this.getStringAttribute('subnet_id');
    }
    set subnetId(value) {
        this._subnetId = value;
    }
    resetSubnetId() {
        this._subnetId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get subnetIdInput() {
        return this._subnetId;
    }
    get zone() {
        return this.getStringAttribute('zone');
    }
    set zone(value) {
        this._zone = value;
    }
    // Temporarily expose input value. Use with caution.
    get zoneInput() {
        return this._zone;
    }
}
exports.MdbPostgresqlClusterHostOutputReference = MdbPostgresqlClusterHostOutputReference;
_l = JSII_RTTI_SYMBOL_1;
MdbPostgresqlClusterHostOutputReference[_l] = { fqn: "@cdktf/provider-yandex.MdbPostgresqlClusterHostOutputReference", version: "2.0.32" };
class MdbPostgresqlClusterHostList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new MdbPostgresqlClusterHostOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.MdbPostgresqlClusterHostList = MdbPostgresqlClusterHostList;
_m = JSII_RTTI_SYMBOL_1;
MdbPostgresqlClusterHostList[_m] = { fqn: "@cdktf/provider-yandex.MdbPostgresqlClusterHostList", version: "2.0.32" };
function mdbPostgresqlClusterMaintenanceWindowToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        day: cdktf.stringToTerraform(struct.day),
        hour: cdktf.numberToTerraform(struct.hour),
        type: cdktf.stringToTerraform(struct.type),
    };
}
exports.mdbPostgresqlClusterMaintenanceWindowToTerraform = mdbPostgresqlClusterMaintenanceWindowToTerraform;
class MdbPostgresqlClusterMaintenanceWindowOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._day !== undefined) {
            hasAnyValues = true;
            internalValueResult.day = this._day;
        }
        if (this._hour !== undefined) {
            hasAnyValues = true;
            internalValueResult.hour = this._hour;
        }
        if (this._type !== undefined) {
            hasAnyValues = true;
            internalValueResult.type = this._type;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._day = undefined;
            this._hour = undefined;
            this._type = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._day = value.day;
            this._hour = value.hour;
            this._type = value.type;
        }
    }
    get day() {
        return this.getStringAttribute('day');
    }
    set day(value) {
        this._day = value;
    }
    resetDay() {
        this._day = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get dayInput() {
        return this._day;
    }
    get hour() {
        return this.getNumberAttribute('hour');
    }
    set hour(value) {
        this._hour = value;
    }
    resetHour() {
        this._hour = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get hourInput() {
        return this._hour;
    }
    get type() {
        return this.getStringAttribute('type');
    }
    set type(value) {
        this._type = value;
    }
    // Temporarily expose input value. Use with caution.
    get typeInput() {
        return this._type;
    }
}
exports.MdbPostgresqlClusterMaintenanceWindowOutputReference = MdbPostgresqlClusterMaintenanceWindowOutputReference;
_o = JSII_RTTI_SYMBOL_1;
MdbPostgresqlClusterMaintenanceWindowOutputReference[_o] = { fqn: "@cdktf/provider-yandex.MdbPostgresqlClusterMaintenanceWindowOutputReference", version: "2.0.32" };
function mdbPostgresqlClusterRestoreToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        backup_id: cdktf.stringToTerraform(struct.backupId),
        time: cdktf.stringToTerraform(struct.time),
        time_inclusive: cdktf.booleanToTerraform(struct.timeInclusive),
    };
}
exports.mdbPostgresqlClusterRestoreToTerraform = mdbPostgresqlClusterRestoreToTerraform;
class MdbPostgresqlClusterRestoreOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._backupId !== undefined) {
            hasAnyValues = true;
            internalValueResult.backupId = this._backupId;
        }
        if (this._time !== undefined) {
            hasAnyValues = true;
            internalValueResult.time = this._time;
        }
        if (this._timeInclusive !== undefined) {
            hasAnyValues = true;
            internalValueResult.timeInclusive = this._timeInclusive;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._backupId = undefined;
            this._time = undefined;
            this._timeInclusive = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._backupId = value.backupId;
            this._time = value.time;
            this._timeInclusive = value.timeInclusive;
        }
    }
    get backupId() {
        return this.getStringAttribute('backup_id');
    }
    set backupId(value) {
        this._backupId = value;
    }
    // Temporarily expose input value. Use with caution.
    get backupIdInput() {
        return this._backupId;
    }
    get time() {
        return this.getStringAttribute('time');
    }
    set time(value) {
        this._time = value;
    }
    resetTime() {
        this._time = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeInput() {
        return this._time;
    }
    get timeInclusive() {
        return this.getBooleanAttribute('time_inclusive');
    }
    set timeInclusive(value) {
        this._timeInclusive = value;
    }
    resetTimeInclusive() {
        this._timeInclusive = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeInclusiveInput() {
        return this._timeInclusive;
    }
}
exports.MdbPostgresqlClusterRestoreOutputReference = MdbPostgresqlClusterRestoreOutputReference;
_p = JSII_RTTI_SYMBOL_1;
MdbPostgresqlClusterRestoreOutputReference[_p] = { fqn: "@cdktf/provider-yandex.MdbPostgresqlClusterRestoreOutputReference", version: "2.0.32" };
function mdbPostgresqlClusterTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.mdbPostgresqlClusterTimeoutsToTerraform = mdbPostgresqlClusterTimeoutsToTerraform;
class MdbPostgresqlClusterTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.MdbPostgresqlClusterTimeoutsOutputReference = MdbPostgresqlClusterTimeoutsOutputReference;
_q = JSII_RTTI_SYMBOL_1;
MdbPostgresqlClusterTimeoutsOutputReference[_q] = { fqn: "@cdktf/provider-yandex.MdbPostgresqlClusterTimeoutsOutputReference", version: "2.0.32" };
function mdbPostgresqlClusterUserPermissionToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        database_name: cdktf.stringToTerraform(struct.databaseName),
    };
}
exports.mdbPostgresqlClusterUserPermissionToTerraform = mdbPostgresqlClusterUserPermissionToTerraform;
class MdbPostgresqlClusterUserPermissionOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._databaseName !== undefined) {
            hasAnyValues = true;
            internalValueResult.databaseName = this._databaseName;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._databaseName = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._databaseName = value.databaseName;
        }
    }
    get databaseName() {
        return this.getStringAttribute('database_name');
    }
    set databaseName(value) {
        this._databaseName = value;
    }
    // Temporarily expose input value. Use with caution.
    get databaseNameInput() {
        return this._databaseName;
    }
}
exports.MdbPostgresqlClusterUserPermissionOutputReference = MdbPostgresqlClusterUserPermissionOutputReference;
_r = JSII_RTTI_SYMBOL_1;
MdbPostgresqlClusterUserPermissionOutputReference[_r] = { fqn: "@cdktf/provider-yandex.MdbPostgresqlClusterUserPermissionOutputReference", version: "2.0.32" };
class MdbPostgresqlClusterUserPermissionList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new MdbPostgresqlClusterUserPermissionOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.MdbPostgresqlClusterUserPermissionList = MdbPostgresqlClusterUserPermissionList;
_s = JSII_RTTI_SYMBOL_1;
MdbPostgresqlClusterUserPermissionList[_s] = { fqn: "@cdktf/provider-yandex.MdbPostgresqlClusterUserPermissionList", version: "2.0.32" };
function mdbPostgresqlClusterUserToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        conn_limit: cdktf.numberToTerraform(struct.connLimit),
        grants: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.grants),
        login: cdktf.booleanToTerraform(struct.login),
        name: cdktf.stringToTerraform(struct.name),
        password: cdktf.stringToTerraform(struct.password),
        settings: cdktf.hashMapper(cdktf.stringToTerraform)(struct.settings),
        permission: cdktf.listMapper(mdbPostgresqlClusterUserPermissionToTerraform, true)(struct.permission),
    };
}
exports.mdbPostgresqlClusterUserToTerraform = mdbPostgresqlClusterUserToTerraform;
class MdbPostgresqlClusterUserOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
        // permission - computed: false, optional: true, required: false
        this._permission = new MdbPostgresqlClusterUserPermissionList(this, "permission", true);
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._connLimit !== undefined) {
            hasAnyValues = true;
            internalValueResult.connLimit = this._connLimit;
        }
        if (this._grants !== undefined) {
            hasAnyValues = true;
            internalValueResult.grants = this._grants;
        }
        if (this._login !== undefined) {
            hasAnyValues = true;
            internalValueResult.login = this._login;
        }
        if (this._name !== undefined) {
            hasAnyValues = true;
            internalValueResult.name = this._name;
        }
        if (this._password !== undefined) {
            hasAnyValues = true;
            internalValueResult.password = this._password;
        }
        if (this._settings !== undefined) {
            hasAnyValues = true;
            internalValueResult.settings = this._settings;
        }
        if (this._permission?.internalValue !== undefined) {
            hasAnyValues = true;
            internalValueResult.permission = this._permission?.internalValue;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._connLimit = undefined;
            this._grants = undefined;
            this._login = undefined;
            this._name = undefined;
            this._password = undefined;
            this._settings = undefined;
            this._permission.internalValue = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._connLimit = value.connLimit;
            this._grants = value.grants;
            this._login = value.login;
            this._name = value.name;
            this._password = value.password;
            this._settings = value.settings;
            this._permission.internalValue = value.permission;
        }
    }
    get connLimit() {
        return this.getNumberAttribute('conn_limit');
    }
    set connLimit(value) {
        this._connLimit = value;
    }
    resetConnLimit() {
        this._connLimit = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get connLimitInput() {
        return this._connLimit;
    }
    get grants() {
        return this.getListAttribute('grants');
    }
    set grants(value) {
        this._grants = value;
    }
    resetGrants() {
        this._grants = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get grantsInput() {
        return this._grants;
    }
    get login() {
        return this.getBooleanAttribute('login');
    }
    set login(value) {
        this._login = value;
    }
    resetLogin() {
        this._login = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get loginInput() {
        return this._login;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get password() {
        return this.getStringAttribute('password');
    }
    set password(value) {
        this._password = value;
    }
    // Temporarily expose input value. Use with caution.
    get passwordInput() {
        return this._password;
    }
    get settings() {
        return this.getStringMapAttribute('settings');
    }
    set settings(value) {
        this._settings = value;
    }
    resetSettings() {
        this._settings = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get settingsInput() {
        return this._settings;
    }
    get permission() {
        return this._permission;
    }
    putPermission(value) {
        this._permission.internalValue = value;
    }
    resetPermission() {
        this._permission.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get permissionInput() {
        return this._permission.internalValue;
    }
}
exports.MdbPostgresqlClusterUserOutputReference = MdbPostgresqlClusterUserOutputReference;
_t = JSII_RTTI_SYMBOL_1;
MdbPostgresqlClusterUserOutputReference[_t] = { fqn: "@cdktf/provider-yandex.MdbPostgresqlClusterUserOutputReference", version: "2.0.32" };
class MdbPostgresqlClusterUserList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new MdbPostgresqlClusterUserOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.MdbPostgresqlClusterUserList = MdbPostgresqlClusterUserList;
_u = JSII_RTTI_SYMBOL_1;
MdbPostgresqlClusterUserList[_u] = { fqn: "@cdktf/provider-yandex.MdbPostgresqlClusterUserList", version: "2.0.32" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/r/mdb_postgresql_cluster yandex_mdb_postgresql_cluster}
*/
class MdbPostgresqlCluster extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/r/mdb_postgresql_cluster yandex_mdb_postgresql_cluster} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options MdbPostgresqlClusterConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'yandex_mdb_postgresql_cluster',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // config - computed: false, optional: false, required: true
        this._config = new MdbPostgresqlClusterConfigAOutputReference(this, "config");
        // database - computed: false, optional: true, required: false
        this._database = new MdbPostgresqlClusterDatabaseList(this, "database", false);
        // host - computed: false, optional: false, required: true
        this._host = new MdbPostgresqlClusterHostList(this, "host", false);
        // maintenance_window - computed: false, optional: true, required: false
        this._maintenanceWindow = new MdbPostgresqlClusterMaintenanceWindowOutputReference(this, "maintenance_window");
        // restore - computed: false, optional: true, required: false
        this._restore = new MdbPostgresqlClusterRestoreOutputReference(this, "restore");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new MdbPostgresqlClusterTimeoutsOutputReference(this, "timeouts");
        // user - computed: false, optional: true, required: false
        this._user = new MdbPostgresqlClusterUserList(this, "user", false);
        this._deletionProtection = config.deletionProtection;
        this._description = config.description;
        this._environment = config.environment;
        this._folderId = config.folderId;
        this._hostGroupIds = config.hostGroupIds;
        this._hostMasterName = config.hostMasterName;
        this._id = config.id;
        this._labels = config.labels;
        this._name = config.name;
        this._networkId = config.networkId;
        this._securityGroupIds = config.securityGroupIds;
        this._config.internalValue = config.config;
        this._database.internalValue = config.database;
        this._host.internalValue = config.host;
        this._maintenanceWindow.internalValue = config.maintenanceWindow;
        this._restore.internalValue = config.restore;
        this._timeouts.internalValue = config.timeouts;
        this._user.internalValue = config.user;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // created_at - computed: true, optional: false, required: false
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    get deletionProtection() {
        return this.getBooleanAttribute('deletion_protection');
    }
    set deletionProtection(value) {
        this._deletionProtection = value;
    }
    resetDeletionProtection() {
        this._deletionProtection = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deletionProtectionInput() {
        return this._deletionProtection;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get environment() {
        return this.getStringAttribute('environment');
    }
    set environment(value) {
        this._environment = value;
    }
    // Temporarily expose input value. Use with caution.
    get environmentInput() {
        return this._environment;
    }
    get folderId() {
        return this.getStringAttribute('folder_id');
    }
    set folderId(value) {
        this._folderId = value;
    }
    resetFolderId() {
        this._folderId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get folderIdInput() {
        return this._folderId;
    }
    // health - computed: true, optional: false, required: false
    get health() {
        return this.getStringAttribute('health');
    }
    get hostGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('host_group_ids'));
    }
    set hostGroupIds(value) {
        this._hostGroupIds = value;
    }
    resetHostGroupIds() {
        this._hostGroupIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get hostGroupIdsInput() {
        return this._hostGroupIds;
    }
    get hostMasterName() {
        return this.getStringAttribute('host_master_name');
    }
    set hostMasterName(value) {
        this._hostMasterName = value;
    }
    resetHostMasterName() {
        this._hostMasterName = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get hostMasterNameInput() {
        return this._hostMasterName;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get networkId() {
        return this.getStringAttribute('network_id');
    }
    set networkId(value) {
        this._networkId = value;
    }
    // Temporarily expose input value. Use with caution.
    get networkIdInput() {
        return this._networkId;
    }
    get securityGroupIds() {
        return cdktf.Fn.tolist(this.getListAttribute('security_group_ids'));
    }
    set securityGroupIds(value) {
        this._securityGroupIds = value;
    }
    resetSecurityGroupIds() {
        this._securityGroupIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupIdsInput() {
        return this._securityGroupIds;
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    get config() {
        return this._config;
    }
    putConfig(value) {
        this._config.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get configInput() {
        return this._config.internalValue;
    }
    get database() {
        return this._database;
    }
    putDatabase(value) {
        this._database.internalValue = value;
    }
    resetDatabase() {
        this._database.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get databaseInput() {
        return this._database.internalValue;
    }
    get host() {
        return this._host;
    }
    putHost(value) {
        this._host.internalValue = value;
    }
    // Temporarily expose input value. Use with caution.
    get hostInput() {
        return this._host.internalValue;
    }
    get maintenanceWindow() {
        return this._maintenanceWindow;
    }
    putMaintenanceWindow(value) {
        this._maintenanceWindow.internalValue = value;
    }
    resetMaintenanceWindow() {
        this._maintenanceWindow.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get maintenanceWindowInput() {
        return this._maintenanceWindow.internalValue;
    }
    get restore() {
        return this._restore;
    }
    putRestore(value) {
        this._restore.internalValue = value;
    }
    resetRestore() {
        this._restore.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get restoreInput() {
        return this._restore.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    get user() {
        return this._user;
    }
    putUser(value) {
        this._user.internalValue = value;
    }
    resetUser() {
        this._user.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get userInput() {
        return this._user.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            deletion_protection: cdktf.booleanToTerraform(this._deletionProtection),
            description: cdktf.stringToTerraform(this._description),
            environment: cdktf.stringToTerraform(this._environment),
            folder_id: cdktf.stringToTerraform(this._folderId),
            host_group_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(this._hostGroupIds),
            host_master_name: cdktf.stringToTerraform(this._hostMasterName),
            id: cdktf.stringToTerraform(this._id),
            labels: cdktf.hashMapper(cdktf.stringToTerraform)(this._labels),
            name: cdktf.stringToTerraform(this._name),
            network_id: cdktf.stringToTerraform(this._networkId),
            security_group_ids: cdktf.listMapper(cdktf.stringToTerraform, false)(this._securityGroupIds),
            config: mdbPostgresqlClusterConfigAToTerraform(this._config.internalValue),
            database: cdktf.listMapper(mdbPostgresqlClusterDatabaseToTerraform, true)(this._database.internalValue),
            host: cdktf.listMapper(mdbPostgresqlClusterHostToTerraform, true)(this._host.internalValue),
            maintenance_window: mdbPostgresqlClusterMaintenanceWindowToTerraform(this._maintenanceWindow.internalValue),
            restore: mdbPostgresqlClusterRestoreToTerraform(this._restore.internalValue),
            timeouts: mdbPostgresqlClusterTimeoutsToTerraform(this._timeouts.internalValue),
            user: cdktf.listMapper(mdbPostgresqlClusterUserToTerraform, true)(this._user.internalValue),
        };
    }
}
exports.MdbPostgresqlCluster = MdbPostgresqlCluster;
_v = JSII_RTTI_SYMBOL_1;
MdbPostgresqlCluster[_v] = { fqn: "@cdktf/provider-yandex.MdbPostgresqlCluster", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
MdbPostgresqlCluster.tfResourceType = "yandex_mdb_postgresql_cluster";
