// @cdktf/provider-yandex/lib/organizationmanager-saml-federation.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrganizationmanagerSamlFederation = exports.OrganizationmanagerSamlFederationTimeoutsOutputReference = exports.organizationmanagerSamlFederationTimeoutsToTerraform = exports.OrganizationmanagerSamlFederationSecuritySettingsOutputReference = exports.organizationmanagerSamlFederationSecuritySettingsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function organizationmanagerSamlFederationSecuritySettingsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        encrypted_assertions: cdktf.booleanToTerraform(struct.encryptedAssertions),
    };
}
exports.organizationmanagerSamlFederationSecuritySettingsToTerraform = organizationmanagerSamlFederationSecuritySettingsToTerraform;
class OrganizationmanagerSamlFederationSecuritySettingsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._encryptedAssertions !== undefined) {
            hasAnyValues = true;
            internalValueResult.encryptedAssertions = this._encryptedAssertions;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._encryptedAssertions = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._encryptedAssertions = value.encryptedAssertions;
        }
    }
    get encryptedAssertions() {
        return this.getBooleanAttribute('encrypted_assertions');
    }
    set encryptedAssertions(value) {
        this._encryptedAssertions = value;
    }
    // Temporarily expose input value. Use with caution.
    get encryptedAssertionsInput() {
        return this._encryptedAssertions;
    }
}
exports.OrganizationmanagerSamlFederationSecuritySettingsOutputReference = OrganizationmanagerSamlFederationSecuritySettingsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
OrganizationmanagerSamlFederationSecuritySettingsOutputReference[_a] = { fqn: "@cdktf/provider-yandex.OrganizationmanagerSamlFederationSecuritySettingsOutputReference", version: "2.0.32" };
function organizationmanagerSamlFederationTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        read: cdktf.stringToTerraform(struct.read),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.organizationmanagerSamlFederationTimeoutsToTerraform = organizationmanagerSamlFederationTimeoutsToTerraform;
class OrganizationmanagerSamlFederationTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._read !== undefined) {
            hasAnyValues = true;
            internalValueResult.read = this._read;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._read = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._read = value.read;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get read() {
        return this.getStringAttribute('read');
    }
    set read(value) {
        this._read = value;
    }
    resetRead() {
        this._read = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get readInput() {
        return this._read;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.OrganizationmanagerSamlFederationTimeoutsOutputReference = OrganizationmanagerSamlFederationTimeoutsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
OrganizationmanagerSamlFederationTimeoutsOutputReference[_b] = { fqn: "@cdktf/provider-yandex.OrganizationmanagerSamlFederationTimeoutsOutputReference", version: "2.0.32" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/r/organizationmanager_saml_federation yandex_organizationmanager_saml_federation}
*/
class OrganizationmanagerSamlFederation extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/r/organizationmanager_saml_federation yandex_organizationmanager_saml_federation} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options OrganizationmanagerSamlFederationConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'yandex_organizationmanager_saml_federation',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // security_settings - computed: false, optional: true, required: false
        this._securitySettings = new OrganizationmanagerSamlFederationSecuritySettingsOutputReference(this, "security_settings");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new OrganizationmanagerSamlFederationTimeoutsOutputReference(this, "timeouts");
        this._autoCreateAccountOnLogin = config.autoCreateAccountOnLogin;
        this._caseInsensitiveNameIds = config.caseInsensitiveNameIds;
        this._cookieMaxAge = config.cookieMaxAge;
        this._description = config.description;
        this._id = config.id;
        this._issuer = config.issuer;
        this._labels = config.labels;
        this._name = config.name;
        this._organizationId = config.organizationId;
        this._ssoBinding = config.ssoBinding;
        this._ssoUrl = config.ssoUrl;
        this._securitySettings.internalValue = config.securitySettings;
        this._timeouts.internalValue = config.timeouts;
    }
    get autoCreateAccountOnLogin() {
        return this.getBooleanAttribute('auto_create_account_on_login');
    }
    set autoCreateAccountOnLogin(value) {
        this._autoCreateAccountOnLogin = value;
    }
    resetAutoCreateAccountOnLogin() {
        this._autoCreateAccountOnLogin = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get autoCreateAccountOnLoginInput() {
        return this._autoCreateAccountOnLogin;
    }
    get caseInsensitiveNameIds() {
        return this.getBooleanAttribute('case_insensitive_name_ids');
    }
    set caseInsensitiveNameIds(value) {
        this._caseInsensitiveNameIds = value;
    }
    resetCaseInsensitiveNameIds() {
        this._caseInsensitiveNameIds = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get caseInsensitiveNameIdsInput() {
        return this._caseInsensitiveNameIds;
    }
    get cookieMaxAge() {
        return this.getStringAttribute('cookie_max_age');
    }
    set cookieMaxAge(value) {
        this._cookieMaxAge = value;
    }
    resetCookieMaxAge() {
        this._cookieMaxAge = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get cookieMaxAgeInput() {
        return this._cookieMaxAge;
    }
    // created_at - computed: true, optional: false, required: false
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get issuer() {
        return this.getStringAttribute('issuer');
    }
    set issuer(value) {
        this._issuer = value;
    }
    // Temporarily expose input value. Use with caution.
    get issuerInput() {
        return this._issuer;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    get organizationId() {
        return this.getStringAttribute('organization_id');
    }
    set organizationId(value) {
        this._organizationId = value;
    }
    // Temporarily expose input value. Use with caution.
    get organizationIdInput() {
        return this._organizationId;
    }
    get ssoBinding() {
        return this.getStringAttribute('sso_binding');
    }
    set ssoBinding(value) {
        this._ssoBinding = value;
    }
    // Temporarily expose input value. Use with caution.
    get ssoBindingInput() {
        return this._ssoBinding;
    }
    get ssoUrl() {
        return this.getStringAttribute('sso_url');
    }
    set ssoUrl(value) {
        this._ssoUrl = value;
    }
    // Temporarily expose input value. Use with caution.
    get ssoUrlInput() {
        return this._ssoUrl;
    }
    get securitySettings() {
        return this._securitySettings;
    }
    putSecuritySettings(value) {
        this._securitySettings.internalValue = value;
    }
    resetSecuritySettings() {
        this._securitySettings.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securitySettingsInput() {
        return this._securitySettings.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            auto_create_account_on_login: cdktf.booleanToTerraform(this._autoCreateAccountOnLogin),
            case_insensitive_name_ids: cdktf.booleanToTerraform(this._caseInsensitiveNameIds),
            cookie_max_age: cdktf.stringToTerraform(this._cookieMaxAge),
            description: cdktf.stringToTerraform(this._description),
            id: cdktf.stringToTerraform(this._id),
            issuer: cdktf.stringToTerraform(this._issuer),
            labels: cdktf.hashMapper(cdktf.stringToTerraform)(this._labels),
            name: cdktf.stringToTerraform(this._name),
            organization_id: cdktf.stringToTerraform(this._organizationId),
            sso_binding: cdktf.stringToTerraform(this._ssoBinding),
            sso_url: cdktf.stringToTerraform(this._ssoUrl),
            security_settings: organizationmanagerSamlFederationSecuritySettingsToTerraform(this._securitySettings.internalValue),
            timeouts: organizationmanagerSamlFederationTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.OrganizationmanagerSamlFederation = OrganizationmanagerSamlFederation;
_c = JSII_RTTI_SYMBOL_1;
OrganizationmanagerSamlFederation[_c] = { fqn: "@cdktf/provider-yandex.OrganizationmanagerSamlFederation", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
OrganizationmanagerSamlFederation.tfResourceType = "yandex_organizationmanager_saml_federation";
