// @cdktf/provider-yandex/lib/resourcemanager-cloud-iam-member.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResourcemanagerCloudIamMember = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/r/resourcemanager_cloud_iam_member yandex_resourcemanager_cloud_iam_member}
*/
class ResourcemanagerCloudIamMember extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/r/resourcemanager_cloud_iam_member yandex_resourcemanager_cloud_iam_member} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ResourcemanagerCloudIamMemberConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'yandex_resourcemanager_cloud_iam_member',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._cloudId = config.cloudId;
        this._id = config.id;
        this._member = config.member;
        this._role = config.role;
        this._sleepAfter = config.sleepAfter;
    }
    get cloudId() {
        return this.getStringAttribute('cloud_id');
    }
    set cloudId(value) {
        this._cloudId = value;
    }
    // Temporarily expose input value. Use with caution.
    get cloudIdInput() {
        return this._cloudId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get member() {
        return this.getStringAttribute('member');
    }
    set member(value) {
        this._member = value;
    }
    // Temporarily expose input value. Use with caution.
    get memberInput() {
        return this._member;
    }
    get role() {
        return this.getStringAttribute('role');
    }
    set role(value) {
        this._role = value;
    }
    // Temporarily expose input value. Use with caution.
    get roleInput() {
        return this._role;
    }
    get sleepAfter() {
        return this.getNumberAttribute('sleep_after');
    }
    set sleepAfter(value) {
        this._sleepAfter = value;
    }
    resetSleepAfter() {
        this._sleepAfter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sleepAfterInput() {
        return this._sleepAfter;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            cloud_id: cdktf.stringToTerraform(this._cloudId),
            id: cdktf.stringToTerraform(this._id),
            member: cdktf.stringToTerraform(this._member),
            role: cdktf.stringToTerraform(this._role),
            sleep_after: cdktf.numberToTerraform(this._sleepAfter),
        };
    }
}
exports.ResourcemanagerCloudIamMember = ResourcemanagerCloudIamMember;
_a = JSII_RTTI_SYMBOL_1;
ResourcemanagerCloudIamMember[_a] = { fqn: "@cdktf/provider-yandex.ResourcemanagerCloudIamMember", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
ResourcemanagerCloudIamMember.tfResourceType = "yandex_resourcemanager_cloud_iam_member";
