// @cdktf/provider-yandex/lib/resourcemanager-folder-iam-member.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResourcemanagerFolderIamMember = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/r/resourcemanager_folder_iam_member yandex_resourcemanager_folder_iam_member}
*/
class ResourcemanagerFolderIamMember extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/r/resourcemanager_folder_iam_member yandex_resourcemanager_folder_iam_member} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ResourcemanagerFolderIamMemberConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'yandex_resourcemanager_folder_iam_member',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._folderId = config.folderId;
        this._id = config.id;
        this._member = config.member;
        this._role = config.role;
        this._sleepAfter = config.sleepAfter;
    }
    get folderId() {
        return this.getStringAttribute('folder_id');
    }
    set folderId(value) {
        this._folderId = value;
    }
    // Temporarily expose input value. Use with caution.
    get folderIdInput() {
        return this._folderId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get member() {
        return this.getStringAttribute('member');
    }
    set member(value) {
        this._member = value;
    }
    // Temporarily expose input value. Use with caution.
    get memberInput() {
        return this._member;
    }
    get role() {
        return this.getStringAttribute('role');
    }
    set role(value) {
        this._role = value;
    }
    // Temporarily expose input value. Use with caution.
    get roleInput() {
        return this._role;
    }
    get sleepAfter() {
        return this.getNumberAttribute('sleep_after');
    }
    set sleepAfter(value) {
        this._sleepAfter = value;
    }
    resetSleepAfter() {
        this._sleepAfter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sleepAfterInput() {
        return this._sleepAfter;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            folder_id: cdktf.stringToTerraform(this._folderId),
            id: cdktf.stringToTerraform(this._id),
            member: cdktf.stringToTerraform(this._member),
            role: cdktf.stringToTerraform(this._role),
            sleep_after: cdktf.numberToTerraform(this._sleepAfter),
        };
    }
}
exports.ResourcemanagerFolderIamMember = ResourcemanagerFolderIamMember;
_a = JSII_RTTI_SYMBOL_1;
ResourcemanagerFolderIamMember[_a] = { fqn: "@cdktf/provider-yandex.ResourcemanagerFolderIamMember", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
ResourcemanagerFolderIamMember.tfResourceType = "yandex_resourcemanager_folder_iam_member";
