// @cdktf/provider-yandex/lib/serverless-container-iam-binding.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServerlessContainerIamBinding = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/r/serverless_container_iam_binding yandex_serverless_container_iam_binding}
*/
class ServerlessContainerIamBinding extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/r/serverless_container_iam_binding yandex_serverless_container_iam_binding} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options ServerlessContainerIamBindingConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'yandex_serverless_container_iam_binding',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        this._containerId = config.containerId;
        this._id = config.id;
        this._members = config.members;
        this._role = config.role;
        this._sleepAfter = config.sleepAfter;
    }
    get containerId() {
        return this.getStringAttribute('container_id');
    }
    set containerId(value) {
        this._containerId = value;
    }
    // Temporarily expose input value. Use with caution.
    get containerIdInput() {
        return this._containerId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get members() {
        return cdktf.Fn.tolist(this.getListAttribute('members'));
    }
    set members(value) {
        this._members = value;
    }
    // Temporarily expose input value. Use with caution.
    get membersInput() {
        return this._members;
    }
    get role() {
        return this.getStringAttribute('role');
    }
    set role(value) {
        this._role = value;
    }
    // Temporarily expose input value. Use with caution.
    get roleInput() {
        return this._role;
    }
    get sleepAfter() {
        return this.getNumberAttribute('sleep_after');
    }
    set sleepAfter(value) {
        this._sleepAfter = value;
    }
    resetSleepAfter() {
        this._sleepAfter = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get sleepAfterInput() {
        return this._sleepAfter;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            container_id: cdktf.stringToTerraform(this._containerId),
            id: cdktf.stringToTerraform(this._id),
            members: cdktf.listMapper(cdktf.stringToTerraform, false)(this._members),
            role: cdktf.stringToTerraform(this._role),
            sleep_after: cdktf.numberToTerraform(this._sleepAfter),
        };
    }
}
exports.ServerlessContainerIamBinding = ServerlessContainerIamBinding;
_a = JSII_RTTI_SYMBOL_1;
ServerlessContainerIamBinding[_a] = { fqn: "@cdktf/provider-yandex.ServerlessContainerIamBinding", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
ServerlessContainerIamBinding.tfResourceType = "yandex_serverless_container_iam_binding";
