// @cdktf/provider-yandex/lib/vpc-default-security-group.js
"use strict";
var _a, _b, _c, _d, _e, _f;
Object.defineProperty(exports, "__esModule", { value: true });
exports.VpcDefaultSecurityGroup = exports.VpcDefaultSecurityGroupTimeoutsOutputReference = exports.vpcDefaultSecurityGroupTimeoutsToTerraform = exports.VpcDefaultSecurityGroupIngressList = exports.VpcDefaultSecurityGroupIngressOutputReference = exports.vpcDefaultSecurityGroupIngressToTerraform = exports.VpcDefaultSecurityGroupEgressList = exports.VpcDefaultSecurityGroupEgressOutputReference = exports.vpcDefaultSecurityGroupEgressToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function vpcDefaultSecurityGroupEgressToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        description: cdktf.stringToTerraform(struct.description),
        from_port: cdktf.numberToTerraform(struct.fromPort),
        labels: cdktf.hashMapper(cdktf.stringToTerraform)(struct.labels),
        port: cdktf.numberToTerraform(struct.port),
        predefined_target: cdktf.stringToTerraform(struct.predefinedTarget),
        protocol: cdktf.stringToTerraform(struct.protocol),
        security_group_id: cdktf.stringToTerraform(struct.securityGroupId),
        to_port: cdktf.numberToTerraform(struct.toPort),
        v4_cidr_blocks: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.v4CidrBlocks),
        v6_cidr_blocks: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.v6CidrBlocks),
    };
}
exports.vpcDefaultSecurityGroupEgressToTerraform = vpcDefaultSecurityGroupEgressToTerraform;
class VpcDefaultSecurityGroupEgressOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._description !== undefined) {
            hasAnyValues = true;
            internalValueResult.description = this._description;
        }
        if (this._fromPort !== undefined) {
            hasAnyValues = true;
            internalValueResult.fromPort = this._fromPort;
        }
        if (this._labels !== undefined) {
            hasAnyValues = true;
            internalValueResult.labels = this._labels;
        }
        if (this._port !== undefined) {
            hasAnyValues = true;
            internalValueResult.port = this._port;
        }
        if (this._predefinedTarget !== undefined) {
            hasAnyValues = true;
            internalValueResult.predefinedTarget = this._predefinedTarget;
        }
        if (this._protocol !== undefined) {
            hasAnyValues = true;
            internalValueResult.protocol = this._protocol;
        }
        if (this._securityGroupId !== undefined) {
            hasAnyValues = true;
            internalValueResult.securityGroupId = this._securityGroupId;
        }
        if (this._toPort !== undefined) {
            hasAnyValues = true;
            internalValueResult.toPort = this._toPort;
        }
        if (this._v4CidrBlocks !== undefined) {
            hasAnyValues = true;
            internalValueResult.v4CidrBlocks = this._v4CidrBlocks;
        }
        if (this._v6CidrBlocks !== undefined) {
            hasAnyValues = true;
            internalValueResult.v6CidrBlocks = this._v6CidrBlocks;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._description = undefined;
            this._fromPort = undefined;
            this._labels = undefined;
            this._port = undefined;
            this._predefinedTarget = undefined;
            this._protocol = undefined;
            this._securityGroupId = undefined;
            this._toPort = undefined;
            this._v4CidrBlocks = undefined;
            this._v6CidrBlocks = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._description = value.description;
            this._fromPort = value.fromPort;
            this._labels = value.labels;
            this._port = value.port;
            this._predefinedTarget = value.predefinedTarget;
            this._protocol = value.protocol;
            this._securityGroupId = value.securityGroupId;
            this._toPort = value.toPort;
            this._v4CidrBlocks = value.v4CidrBlocks;
            this._v6CidrBlocks = value.v6CidrBlocks;
        }
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get fromPort() {
        return this.getNumberAttribute('from_port');
    }
    set fromPort(value) {
        this._fromPort = value;
    }
    resetFromPort() {
        this._fromPort = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get fromPortInput() {
        return this._fromPort;
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get port() {
        return this.getNumberAttribute('port');
    }
    set port(value) {
        this._port = value;
    }
    resetPort() {
        this._port = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get portInput() {
        return this._port;
    }
    get predefinedTarget() {
        return this.getStringAttribute('predefined_target');
    }
    set predefinedTarget(value) {
        this._predefinedTarget = value;
    }
    resetPredefinedTarget() {
        this._predefinedTarget = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get predefinedTargetInput() {
        return this._predefinedTarget;
    }
    get protocol() {
        return this.getStringAttribute('protocol');
    }
    set protocol(value) {
        this._protocol = value;
    }
    // Temporarily expose input value. Use with caution.
    get protocolInput() {
        return this._protocol;
    }
    get securityGroupId() {
        return this.getStringAttribute('security_group_id');
    }
    set securityGroupId(value) {
        this._securityGroupId = value;
    }
    resetSecurityGroupId() {
        this._securityGroupId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupIdInput() {
        return this._securityGroupId;
    }
    get toPort() {
        return this.getNumberAttribute('to_port');
    }
    set toPort(value) {
        this._toPort = value;
    }
    resetToPort() {
        this._toPort = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get toPortInput() {
        return this._toPort;
    }
    get v4CidrBlocks() {
        return this.getListAttribute('v4_cidr_blocks');
    }
    set v4CidrBlocks(value) {
        this._v4CidrBlocks = value;
    }
    resetV4CidrBlocks() {
        this._v4CidrBlocks = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get v4CidrBlocksInput() {
        return this._v4CidrBlocks;
    }
    get v6CidrBlocks() {
        return this.getListAttribute('v6_cidr_blocks');
    }
    set v6CidrBlocks(value) {
        this._v6CidrBlocks = value;
    }
    resetV6CidrBlocks() {
        this._v6CidrBlocks = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get v6CidrBlocksInput() {
        return this._v6CidrBlocks;
    }
}
exports.VpcDefaultSecurityGroupEgressOutputReference = VpcDefaultSecurityGroupEgressOutputReference;
_a = JSII_RTTI_SYMBOL_1;
VpcDefaultSecurityGroupEgressOutputReference[_a] = { fqn: "@cdktf/provider-yandex.VpcDefaultSecurityGroupEgressOutputReference", version: "2.0.32" };
class VpcDefaultSecurityGroupEgressList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new VpcDefaultSecurityGroupEgressOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.VpcDefaultSecurityGroupEgressList = VpcDefaultSecurityGroupEgressList;
_b = JSII_RTTI_SYMBOL_1;
VpcDefaultSecurityGroupEgressList[_b] = { fqn: "@cdktf/provider-yandex.VpcDefaultSecurityGroupEgressList", version: "2.0.32" };
function vpcDefaultSecurityGroupIngressToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        description: cdktf.stringToTerraform(struct.description),
        from_port: cdktf.numberToTerraform(struct.fromPort),
        labels: cdktf.hashMapper(cdktf.stringToTerraform)(struct.labels),
        port: cdktf.numberToTerraform(struct.port),
        predefined_target: cdktf.stringToTerraform(struct.predefinedTarget),
        protocol: cdktf.stringToTerraform(struct.protocol),
        security_group_id: cdktf.stringToTerraform(struct.securityGroupId),
        to_port: cdktf.numberToTerraform(struct.toPort),
        v4_cidr_blocks: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.v4CidrBlocks),
        v6_cidr_blocks: cdktf.listMapper(cdktf.stringToTerraform, false)(struct.v6CidrBlocks),
    };
}
exports.vpcDefaultSecurityGroupIngressToTerraform = vpcDefaultSecurityGroupIngressToTerraform;
class VpcDefaultSecurityGroupIngressOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param complexObjectIndex the index of this item in the list
    * @param complexObjectIsFromSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, complexObjectIndex, complexObjectIsFromSet) {
        super(terraformResource, terraformAttribute, complexObjectIsFromSet, complexObjectIndex);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._description !== undefined) {
            hasAnyValues = true;
            internalValueResult.description = this._description;
        }
        if (this._fromPort !== undefined) {
            hasAnyValues = true;
            internalValueResult.fromPort = this._fromPort;
        }
        if (this._labels !== undefined) {
            hasAnyValues = true;
            internalValueResult.labels = this._labels;
        }
        if (this._port !== undefined) {
            hasAnyValues = true;
            internalValueResult.port = this._port;
        }
        if (this._predefinedTarget !== undefined) {
            hasAnyValues = true;
            internalValueResult.predefinedTarget = this._predefinedTarget;
        }
        if (this._protocol !== undefined) {
            hasAnyValues = true;
            internalValueResult.protocol = this._protocol;
        }
        if (this._securityGroupId !== undefined) {
            hasAnyValues = true;
            internalValueResult.securityGroupId = this._securityGroupId;
        }
        if (this._toPort !== undefined) {
            hasAnyValues = true;
            internalValueResult.toPort = this._toPort;
        }
        if (this._v4CidrBlocks !== undefined) {
            hasAnyValues = true;
            internalValueResult.v4CidrBlocks = this._v4CidrBlocks;
        }
        if (this._v6CidrBlocks !== undefined) {
            hasAnyValues = true;
            internalValueResult.v6CidrBlocks = this._v6CidrBlocks;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._description = undefined;
            this._fromPort = undefined;
            this._labels = undefined;
            this._port = undefined;
            this._predefinedTarget = undefined;
            this._protocol = undefined;
            this._securityGroupId = undefined;
            this._toPort = undefined;
            this._v4CidrBlocks = undefined;
            this._v6CidrBlocks = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._description = value.description;
            this._fromPort = value.fromPort;
            this._labels = value.labels;
            this._port = value.port;
            this._predefinedTarget = value.predefinedTarget;
            this._protocol = value.protocol;
            this._securityGroupId = value.securityGroupId;
            this._toPort = value.toPort;
            this._v4CidrBlocks = value.v4CidrBlocks;
            this._v6CidrBlocks = value.v6CidrBlocks;
        }
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get fromPort() {
        return this.getNumberAttribute('from_port');
    }
    set fromPort(value) {
        this._fromPort = value;
    }
    resetFromPort() {
        this._fromPort = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get fromPortInput() {
        return this._fromPort;
    }
    // id - computed: true, optional: false, required: false
    get id() {
        return this.getStringAttribute('id');
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get port() {
        return this.getNumberAttribute('port');
    }
    set port(value) {
        this._port = value;
    }
    resetPort() {
        this._port = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get portInput() {
        return this._port;
    }
    get predefinedTarget() {
        return this.getStringAttribute('predefined_target');
    }
    set predefinedTarget(value) {
        this._predefinedTarget = value;
    }
    resetPredefinedTarget() {
        this._predefinedTarget = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get predefinedTargetInput() {
        return this._predefinedTarget;
    }
    get protocol() {
        return this.getStringAttribute('protocol');
    }
    set protocol(value) {
        this._protocol = value;
    }
    // Temporarily expose input value. Use with caution.
    get protocolInput() {
        return this._protocol;
    }
    get securityGroupId() {
        return this.getStringAttribute('security_group_id');
    }
    set securityGroupId(value) {
        this._securityGroupId = value;
    }
    resetSecurityGroupId() {
        this._securityGroupId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupIdInput() {
        return this._securityGroupId;
    }
    get toPort() {
        return this.getNumberAttribute('to_port');
    }
    set toPort(value) {
        this._toPort = value;
    }
    resetToPort() {
        this._toPort = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get toPortInput() {
        return this._toPort;
    }
    get v4CidrBlocks() {
        return this.getListAttribute('v4_cidr_blocks');
    }
    set v4CidrBlocks(value) {
        this._v4CidrBlocks = value;
    }
    resetV4CidrBlocks() {
        this._v4CidrBlocks = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get v4CidrBlocksInput() {
        return this._v4CidrBlocks;
    }
    get v6CidrBlocks() {
        return this.getListAttribute('v6_cidr_blocks');
    }
    set v6CidrBlocks(value) {
        this._v6CidrBlocks = value;
    }
    resetV6CidrBlocks() {
        this._v6CidrBlocks = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get v6CidrBlocksInput() {
        return this._v6CidrBlocks;
    }
}
exports.VpcDefaultSecurityGroupIngressOutputReference = VpcDefaultSecurityGroupIngressOutputReference;
_c = JSII_RTTI_SYMBOL_1;
VpcDefaultSecurityGroupIngressOutputReference[_c] = { fqn: "@cdktf/provider-yandex.VpcDefaultSecurityGroupIngressOutputReference", version: "2.0.32" };
class VpcDefaultSecurityGroupIngressList extends cdktf.ComplexList {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    * @param wrapsSet whether the list is wrapping a set (will add tolist() to be able to access an item via an index)
    */
    constructor(terraformResource, terraformAttribute, wrapsSet) {
        super(terraformResource, terraformAttribute, wrapsSet);
        this.terraformResource = terraformResource;
        this.terraformAttribute = terraformAttribute;
        this.wrapsSet = wrapsSet;
    }
    /**
    * @param index the index of the item to return
    */
    get(index) {
        return new VpcDefaultSecurityGroupIngressOutputReference(this.terraformResource, this.terraformAttribute, index, this.wrapsSet);
    }
}
exports.VpcDefaultSecurityGroupIngressList = VpcDefaultSecurityGroupIngressList;
_d = JSII_RTTI_SYMBOL_1;
VpcDefaultSecurityGroupIngressList[_d] = { fqn: "@cdktf/provider-yandex.VpcDefaultSecurityGroupIngressList", version: "2.0.32" };
function vpcDefaultSecurityGroupTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        read: cdktf.stringToTerraform(struct.read),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.vpcDefaultSecurityGroupTimeoutsToTerraform = vpcDefaultSecurityGroupTimeoutsToTerraform;
class VpcDefaultSecurityGroupTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._read !== undefined) {
            hasAnyValues = true;
            internalValueResult.read = this._read;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._read = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._read = value.read;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get read() {
        return this.getStringAttribute('read');
    }
    set read(value) {
        this._read = value;
    }
    resetRead() {
        this._read = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get readInput() {
        return this._read;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.VpcDefaultSecurityGroupTimeoutsOutputReference = VpcDefaultSecurityGroupTimeoutsOutputReference;
_e = JSII_RTTI_SYMBOL_1;
VpcDefaultSecurityGroupTimeoutsOutputReference[_e] = { fqn: "@cdktf/provider-yandex.VpcDefaultSecurityGroupTimeoutsOutputReference", version: "2.0.32" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/r/vpc_default_security_group yandex_vpc_default_security_group}
*/
class VpcDefaultSecurityGroup extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/r/vpc_default_security_group yandex_vpc_default_security_group} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options VpcDefaultSecurityGroupConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'yandex_vpc_default_security_group',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // egress - computed: false, optional: true, required: false
        this._egress = new VpcDefaultSecurityGroupEgressList(this, "egress", true);
        // ingress - computed: false, optional: true, required: false
        this._ingress = new VpcDefaultSecurityGroupIngressList(this, "ingress", true);
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new VpcDefaultSecurityGroupTimeoutsOutputReference(this, "timeouts");
        this._description = config.description;
        this._folderId = config.folderId;
        this._id = config.id;
        this._labels = config.labels;
        this._networkId = config.networkId;
        this._egress.internalValue = config.egress;
        this._ingress.internalValue = config.ingress;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // created_at - computed: true, optional: false, required: false
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get folderId() {
        return this.getStringAttribute('folder_id');
    }
    set folderId(value) {
        this._folderId = value;
    }
    resetFolderId() {
        this._folderId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get folderIdInput() {
        return this._folderId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    // name - computed: true, optional: false, required: false
    get name() {
        return this.getStringAttribute('name');
    }
    get networkId() {
        return this.getStringAttribute('network_id');
    }
    set networkId(value) {
        this._networkId = value;
    }
    // Temporarily expose input value. Use with caution.
    get networkIdInput() {
        return this._networkId;
    }
    // status - computed: true, optional: false, required: false
    get status() {
        return this.getStringAttribute('status');
    }
    get egress() {
        return this._egress;
    }
    putEgress(value) {
        this._egress.internalValue = value;
    }
    resetEgress() {
        this._egress.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get egressInput() {
        return this._egress.internalValue;
    }
    get ingress() {
        return this._ingress;
    }
    putIngress(value) {
        this._ingress.internalValue = value;
    }
    resetIngress() {
        this._ingress.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ingressInput() {
        return this._ingress.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            folder_id: cdktf.stringToTerraform(this._folderId),
            id: cdktf.stringToTerraform(this._id),
            labels: cdktf.hashMapper(cdktf.stringToTerraform)(this._labels),
            network_id: cdktf.stringToTerraform(this._networkId),
            egress: cdktf.listMapper(vpcDefaultSecurityGroupEgressToTerraform, true)(this._egress.internalValue),
            ingress: cdktf.listMapper(vpcDefaultSecurityGroupIngressToTerraform, true)(this._ingress.internalValue),
            timeouts: vpcDefaultSecurityGroupTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.VpcDefaultSecurityGroup = VpcDefaultSecurityGroup;
_f = JSII_RTTI_SYMBOL_1;
VpcDefaultSecurityGroup[_f] = { fqn: "@cdktf/provider-yandex.VpcDefaultSecurityGroup", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
VpcDefaultSecurityGroup.tfResourceType = "yandex_vpc_default_security_group";
