// @cdktf/provider-yandex/lib/vpc-address.js
"use strict";
var _a, _b, _c;
Object.defineProperty(exports, "__esModule", { value: true });
exports.VpcAddress = exports.VpcAddressTimeoutsOutputReference = exports.vpcAddressTimeoutsToTerraform = exports.VpcAddressExternalIpv4AddressOutputReference = exports.vpcAddressExternalIpv4AddressToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function vpcAddressExternalIpv4AddressToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        ddos_protection_provider: cdktf.stringToTerraform(struct.ddosProtectionProvider),
        outgoing_smtp_capability: cdktf.stringToTerraform(struct.outgoingSmtpCapability),
        zone_id: cdktf.stringToTerraform(struct.zoneId),
    };
}
exports.vpcAddressExternalIpv4AddressToTerraform = vpcAddressExternalIpv4AddressToTerraform;
class VpcAddressExternalIpv4AddressOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._ddosProtectionProvider !== undefined) {
            hasAnyValues = true;
            internalValueResult.ddosProtectionProvider = this._ddosProtectionProvider;
        }
        if (this._outgoingSmtpCapability !== undefined) {
            hasAnyValues = true;
            internalValueResult.outgoingSmtpCapability = this._outgoingSmtpCapability;
        }
        if (this._zoneId !== undefined) {
            hasAnyValues = true;
            internalValueResult.zoneId = this._zoneId;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this._ddosProtectionProvider = undefined;
            this._outgoingSmtpCapability = undefined;
            this._zoneId = undefined;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this._ddosProtectionProvider = value.ddosProtectionProvider;
            this._outgoingSmtpCapability = value.outgoingSmtpCapability;
            this._zoneId = value.zoneId;
        }
    }
    // address - computed: true, optional: false, required: false
    get address() {
        return this.getStringAttribute('address');
    }
    get ddosProtectionProvider() {
        return this.getStringAttribute('ddos_protection_provider');
    }
    set ddosProtectionProvider(value) {
        this._ddosProtectionProvider = value;
    }
    resetDdosProtectionProvider() {
        this._ddosProtectionProvider = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get ddosProtectionProviderInput() {
        return this._ddosProtectionProvider;
    }
    get outgoingSmtpCapability() {
        return this.getStringAttribute('outgoing_smtp_capability');
    }
    set outgoingSmtpCapability(value) {
        this._outgoingSmtpCapability = value;
    }
    resetOutgoingSmtpCapability() {
        this._outgoingSmtpCapability = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get outgoingSmtpCapabilityInput() {
        return this._outgoingSmtpCapability;
    }
    get zoneId() {
        return this.getStringAttribute('zone_id');
    }
    set zoneId(value) {
        this._zoneId = value;
    }
    resetZoneId() {
        this._zoneId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get zoneIdInput() {
        return this._zoneId;
    }
}
exports.VpcAddressExternalIpv4AddressOutputReference = VpcAddressExternalIpv4AddressOutputReference;
_a = JSII_RTTI_SYMBOL_1;
VpcAddressExternalIpv4AddressOutputReference[_a] = { fqn: "@cdktf/provider-yandex.VpcAddressExternalIpv4AddressOutputReference", version: "2.0.32" };
function vpcAddressTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.vpcAddressTimeoutsToTerraform = vpcAddressTimeoutsToTerraform;
class VpcAddressTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.VpcAddressTimeoutsOutputReference = VpcAddressTimeoutsOutputReference;
_b = JSII_RTTI_SYMBOL_1;
VpcAddressTimeoutsOutputReference[_b] = { fqn: "@cdktf/provider-yandex.VpcAddressTimeoutsOutputReference", version: "2.0.32" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/r/vpc_address yandex_vpc_address}
*/
class VpcAddress extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/r/vpc_address yandex_vpc_address} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options VpcAddressConfig = {}
    */
    constructor(scope, id, config = {}) {
        super(scope, id, {
            terraformResourceType: 'yandex_vpc_address',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // external_ipv4_address - computed: false, optional: true, required: false
        this._externalIpv4Address = new VpcAddressExternalIpv4AddressOutputReference(this, "external_ipv4_address");
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new VpcAddressTimeoutsOutputReference(this, "timeouts");
        this._description = config.description;
        this._folderId = config.folderId;
        this._id = config.id;
        this._labels = config.labels;
        this._name = config.name;
        this._externalIpv4Address.internalValue = config.externalIpv4Address;
        this._timeouts.internalValue = config.timeouts;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // created_at - computed: true, optional: false, required: false
    get createdAt() {
        return this.getStringAttribute('created_at');
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get folderId() {
        return this.getStringAttribute('folder_id');
    }
    set folderId(value) {
        this._folderId = value;
    }
    resetFolderId() {
        this._folderId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get folderIdInput() {
        return this._folderId;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get name() {
        return this.getStringAttribute('name');
    }
    set name(value) {
        this._name = value;
    }
    resetName() {
        this._name = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get nameInput() {
        return this._name;
    }
    // reserved - computed: true, optional: false, required: false
    get reserved() {
        return this.getBooleanAttribute('reserved');
    }
    // used - computed: true, optional: false, required: false
    get used() {
        return this.getBooleanAttribute('used');
    }
    get externalIpv4Address() {
        return this._externalIpv4Address;
    }
    putExternalIpv4Address(value) {
        this._externalIpv4Address.internalValue = value;
    }
    resetExternalIpv4Address() {
        this._externalIpv4Address.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get externalIpv4AddressInput() {
        return this._externalIpv4Address.internalValue;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            folder_id: cdktf.stringToTerraform(this._folderId),
            id: cdktf.stringToTerraform(this._id),
            labels: cdktf.hashMapper(cdktf.stringToTerraform)(this._labels),
            name: cdktf.stringToTerraform(this._name),
            external_ipv4_address: vpcAddressExternalIpv4AddressToTerraform(this._externalIpv4Address.internalValue),
            timeouts: vpcAddressTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.VpcAddress = VpcAddress;
_c = JSII_RTTI_SYMBOL_1;
VpcAddress[_c] = { fqn: "@cdktf/provider-yandex.VpcAddress", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
VpcAddress.tfResourceType = "yandex_vpc_address";
