// @cdktf/provider-yandex/lib/vpc-security-group-rule.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.VpcSecurityGroupRule = exports.VpcSecurityGroupRuleTimeoutsOutputReference = exports.vpcSecurityGroupRuleTimeoutsToTerraform = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function vpcSecurityGroupRuleTimeoutsToTerraform(struct) {
    if (!cdktf.canInspect(struct) || cdktf.Tokenization.isResolvable(struct)) {
        return struct;
    }
    if (cdktf.isComplexElement(struct)) {
        throw new Error("A complex element was used as configuration, this is not supported: https://cdk.tf/complex-object-as-configuration");
    }
    return {
        create: cdktf.stringToTerraform(struct.create),
        delete: cdktf.stringToTerraform(struct.delete),
        read: cdktf.stringToTerraform(struct.read),
        update: cdktf.stringToTerraform(struct.update),
    };
}
exports.vpcSecurityGroupRuleTimeoutsToTerraform = vpcSecurityGroupRuleTimeoutsToTerraform;
class VpcSecurityGroupRuleTimeoutsOutputReference extends cdktf.ComplexObject {
    /**
    * @param terraformResource The parent resource
    * @param terraformAttribute The attribute on the parent resource this class is referencing
    */
    constructor(terraformResource, terraformAttribute) {
        super(terraformResource, terraformAttribute, false, 0);
        this.isEmptyObject = false;
    }
    get internalValue() {
        if (this.resolvableValue) {
            return this.resolvableValue;
        }
        let hasAnyValues = this.isEmptyObject;
        const internalValueResult = {};
        if (this._create !== undefined) {
            hasAnyValues = true;
            internalValueResult.create = this._create;
        }
        if (this._delete !== undefined) {
            hasAnyValues = true;
            internalValueResult.delete = this._delete;
        }
        if (this._read !== undefined) {
            hasAnyValues = true;
            internalValueResult.read = this._read;
        }
        if (this._update !== undefined) {
            hasAnyValues = true;
            internalValueResult.update = this._update;
        }
        return hasAnyValues ? internalValueResult : undefined;
    }
    set internalValue(value) {
        if (value === undefined) {
            this.isEmptyObject = false;
            this.resolvableValue = undefined;
            this._create = undefined;
            this._delete = undefined;
            this._read = undefined;
            this._update = undefined;
        }
        else if (cdktf.Tokenization.isResolvable(value)) {
            this.isEmptyObject = false;
            this.resolvableValue = value;
        }
        else {
            this.isEmptyObject = Object.keys(value).length === 0;
            this.resolvableValue = undefined;
            this._create = value.create;
            this._delete = value.delete;
            this._read = value.read;
            this._update = value.update;
        }
    }
    get create() {
        return this.getStringAttribute('create');
    }
    set create(value) {
        this._create = value;
    }
    resetCreate() {
        this._create = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get createInput() {
        return this._create;
    }
    get delete() {
        return this.getStringAttribute('delete');
    }
    set delete(value) {
        this._delete = value;
    }
    resetDelete() {
        this._delete = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get deleteInput() {
        return this._delete;
    }
    get read() {
        return this.getStringAttribute('read');
    }
    set read(value) {
        this._read = value;
    }
    resetRead() {
        this._read = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get readInput() {
        return this._read;
    }
    get update() {
        return this.getStringAttribute('update');
    }
    set update(value) {
        this._update = value;
    }
    resetUpdate() {
        this._update = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get updateInput() {
        return this._update;
    }
}
exports.VpcSecurityGroupRuleTimeoutsOutputReference = VpcSecurityGroupRuleTimeoutsOutputReference;
_a = JSII_RTTI_SYMBOL_1;
VpcSecurityGroupRuleTimeoutsOutputReference[_a] = { fqn: "@cdktf/provider-yandex.VpcSecurityGroupRuleTimeoutsOutputReference", version: "2.0.32" };
/**
* Represents a {@link https://www.terraform.io/docs/providers/yandex/r/vpc_security_group_rule yandex_vpc_security_group_rule}
*/
class VpcSecurityGroupRule extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
    * Create a new {@link https://www.terraform.io/docs/providers/yandex/r/vpc_security_group_rule yandex_vpc_security_group_rule} Resource
    *
    * @param scope The scope in which to define this construct
    * @param id The scoped construct ID. Must be unique amongst siblings in the same scope
    * @param options VpcSecurityGroupRuleConfig
    */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'yandex_vpc_security_group_rule',
            terraformGeneratorMetadata: {
                providerName: 'yandex',
                providerVersion: '0.77.0',
                providerVersionConstraint: '~> 0.73'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle,
            provisioners: config.provisioners,
            connection: config.connection,
            forEach: config.forEach
        });
        // timeouts - computed: false, optional: true, required: false
        this._timeouts = new VpcSecurityGroupRuleTimeoutsOutputReference(this, "timeouts");
        this._description = config.description;
        this._direction = config.direction;
        this._fromPort = config.fromPort;
        this._id = config.id;
        this._labels = config.labels;
        this._port = config.port;
        this._predefinedTarget = config.predefinedTarget;
        this._protocol = config.protocol;
        this._securityGroupBinding = config.securityGroupBinding;
        this._securityGroupId = config.securityGroupId;
        this._toPort = config.toPort;
        this._v4CidrBlocks = config.v4CidrBlocks;
        this._v6CidrBlocks = config.v6CidrBlocks;
        this._timeouts.internalValue = config.timeouts;
    }
    get description() {
        return this.getStringAttribute('description');
    }
    set description(value) {
        this._description = value;
    }
    resetDescription() {
        this._description = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get descriptionInput() {
        return this._description;
    }
    get direction() {
        return this.getStringAttribute('direction');
    }
    set direction(value) {
        this._direction = value;
    }
    // Temporarily expose input value. Use with caution.
    get directionInput() {
        return this._direction;
    }
    get fromPort() {
        return this.getNumberAttribute('from_port');
    }
    set fromPort(value) {
        this._fromPort = value;
    }
    resetFromPort() {
        this._fromPort = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get fromPortInput() {
        return this._fromPort;
    }
    get id() {
        return this.getStringAttribute('id');
    }
    set id(value) {
        this._id = value;
    }
    resetId() {
        this._id = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get idInput() {
        return this._id;
    }
    get labels() {
        return this.getStringMapAttribute('labels');
    }
    set labels(value) {
        this._labels = value;
    }
    resetLabels() {
        this._labels = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get labelsInput() {
        return this._labels;
    }
    get port() {
        return this.getNumberAttribute('port');
    }
    set port(value) {
        this._port = value;
    }
    resetPort() {
        this._port = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get portInput() {
        return this._port;
    }
    get predefinedTarget() {
        return this.getStringAttribute('predefined_target');
    }
    set predefinedTarget(value) {
        this._predefinedTarget = value;
    }
    resetPredefinedTarget() {
        this._predefinedTarget = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get predefinedTargetInput() {
        return this._predefinedTarget;
    }
    get protocol() {
        return this.getStringAttribute('protocol');
    }
    set protocol(value) {
        this._protocol = value;
    }
    resetProtocol() {
        this._protocol = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get protocolInput() {
        return this._protocol;
    }
    get securityGroupBinding() {
        return this.getStringAttribute('security_group_binding');
    }
    set securityGroupBinding(value) {
        this._securityGroupBinding = value;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupBindingInput() {
        return this._securityGroupBinding;
    }
    get securityGroupId() {
        return this.getStringAttribute('security_group_id');
    }
    set securityGroupId(value) {
        this._securityGroupId = value;
    }
    resetSecurityGroupId() {
        this._securityGroupId = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get securityGroupIdInput() {
        return this._securityGroupId;
    }
    get toPort() {
        return this.getNumberAttribute('to_port');
    }
    set toPort(value) {
        this._toPort = value;
    }
    resetToPort() {
        this._toPort = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get toPortInput() {
        return this._toPort;
    }
    get v4CidrBlocks() {
        return this.getListAttribute('v4_cidr_blocks');
    }
    set v4CidrBlocks(value) {
        this._v4CidrBlocks = value;
    }
    resetV4CidrBlocks() {
        this._v4CidrBlocks = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get v4CidrBlocksInput() {
        return this._v4CidrBlocks;
    }
    get v6CidrBlocks() {
        return this.getListAttribute('v6_cidr_blocks');
    }
    set v6CidrBlocks(value) {
        this._v6CidrBlocks = value;
    }
    resetV6CidrBlocks() {
        this._v6CidrBlocks = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get v6CidrBlocksInput() {
        return this._v6CidrBlocks;
    }
    get timeouts() {
        return this._timeouts;
    }
    putTimeouts(value) {
        this._timeouts.internalValue = value;
    }
    resetTimeouts() {
        this._timeouts.internalValue = undefined;
    }
    // Temporarily expose input value. Use with caution.
    get timeoutsInput() {
        return this._timeouts.internalValue;
    }
    // =========
    // SYNTHESIS
    // =========
    synthesizeAttributes() {
        return {
            description: cdktf.stringToTerraform(this._description),
            direction: cdktf.stringToTerraform(this._direction),
            from_port: cdktf.numberToTerraform(this._fromPort),
            id: cdktf.stringToTerraform(this._id),
            labels: cdktf.hashMapper(cdktf.stringToTerraform)(this._labels),
            port: cdktf.numberToTerraform(this._port),
            predefined_target: cdktf.stringToTerraform(this._predefinedTarget),
            protocol: cdktf.stringToTerraform(this._protocol),
            security_group_binding: cdktf.stringToTerraform(this._securityGroupBinding),
            security_group_id: cdktf.stringToTerraform(this._securityGroupId),
            to_port: cdktf.numberToTerraform(this._toPort),
            v4_cidr_blocks: cdktf.listMapper(cdktf.stringToTerraform, false)(this._v4CidrBlocks),
            v6_cidr_blocks: cdktf.listMapper(cdktf.stringToTerraform, false)(this._v6CidrBlocks),
            timeouts: vpcSecurityGroupRuleTimeoutsToTerraform(this._timeouts.internalValue),
        };
    }
}
exports.VpcSecurityGroupRule = VpcSecurityGroupRule;
_b = JSII_RTTI_SYMBOL_1;
VpcSecurityGroupRule[_b] = { fqn: "@cdktf/provider-yandex.VpcSecurityGroupRule", version: "2.0.32" };
// =================
// STATIC PROPERTIES
// =================
VpcSecurityGroupRule.tfResourceType = "yandex_vpc_security_group_rule";
